<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />


<meta name="author" content="John Blischak" />

<meta name="date" content="2017-09-05" />

<title>Contributing</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-1.1/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-1.1/highlight.js"></script>
<link href="site_libs/font-awesome-4.5.0/css/font-awesome.min.css" rel="stylesheet" />

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs && document.readyState && document.readyState === "complete") {
   window.setTimeout(function() {
      hljs.initHighlighting();
   }, 0);
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>


</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
button.code-folding-btn:focus {
  outline: none;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}

.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>


<div class="container-fluid main-container">

<!-- tabsets -->
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->




<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
  padding-left: 25px;
  text-indent: 0;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>

<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">singlecell-qtl</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Overview</a>
</li>
<li>
  <a href="data-collection.html">Data</a>
</li>
<li>
  <a href="analysis.html">Analysis</a>
</li>
<li>
  <a href="contributing.html">Contributing</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/jdblischak/singlecell-qtl">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Contributing</h1>
<h4 class="author"><em>John Blischak</em></h4>
<h4 class="date"><em>2017-09-05</em></h4>

</div>


<!-- The file analysis/chunks.R contains chunks that define default settings
shared across the workflowr files. -->
<!-- Update knitr chunk options -->
<!-- Insert the date the file was last updated -->
<p><strong>Last updated:</strong> 2018-01-29</p>
<!-- Insert the code version (Git commit SHA1) if Git repository exists and R
 package git2r is installed -->
<p><strong>Code version:</strong> f6b7f76</p>
<section id="initial-setup" class="level2">
<h2>Initial setup</h2>
<p>The initial setup only has to be performed once for each computer that you use.</p>
<p>To ensure all contributors are using the same computational environment, we use <a href="https://conda.io/docs/">conda</a> to manage software dependencies (made possible by the <a href="https://bioconda.github.io">bioconda</a> and <a href="https://conda-forge.org/">conda-forge</a> projects). Please complete the following steps to replicate the computing environment. Note that this is only guaranteed to work on a Linux-64 based architecture, but in theory should be able to work on macOS as well. All commands shown below are intended to be run in a Bash shell from the root of the project directory.</p>
<ol type="1">
<li><p>Install Git and register for an account on GitHub</p></li>
<li>Download and install <a href="https://conda.io/miniconda.html">Miniconda3</a>
<ul>
<li>Run <code>bash Downloads/Miniconda3-latest-MacOSX-x86_64.sh</code></li>
<li>Press <code>ENTER</code> to start installation</li>
<li>Press <code>q</code> when finished reading the license and then type <code>yes</code></li>
<li>Type the location for installation (press <code>ENTER</code> to accept the default of <code>~/miniconda3</code>)</li>
<li>Type <code>yes</code> to add Miniconda to your <code>PATH</code> variable</li>
<li>Run <code>source ~/.bash_profile</code> on macOS or <code>source ~/.bashrc</code> on Linux to update <code>PATH</code></li>
<li>Run <code>which conda</code> to confirm that <code>conda</code> was installed</li>
</ul></li>
<li><p>Clone this repository (or your personal fork) <code>git clone https://github.com/jdblischak/singlecell-qtl.git  cd singlecell-qtl</code></p></li>
<li><p>Create the conda environment “scqtl” using <code>environment.yaml</code> <code>conda env create --file environment.yaml</code></p></li>
<li><p>To use the conda environment, you must first activate it by running <code>source activate scqtl</code>. This will override your default settings for R, Python, and various other software packages. When you are done working on this project, you can either logout of the current session or deactivate the environment. <code># Activate conda environment scqtl  source activate scqtl  # Deactivate conda environment  source deactivate</code></p></li>
<li><p>Initialize <a href="https://git-lfs.github.com/">git-lfs</a> and download the latest version of large data files <code>git lfs install  git lfs pull</code></p></li>
</ol>
<p><strong>Warning:</strong> If you are using RStudio, you need to ensure that it recognizes your conda environment. If you launch RStudio by clicking on an icon, it doesn’t use the current environment you have configured in your shell. On a Linux-based system, the solution is to launch RStudio directly from the shell with <code>rstudio</code>. On macOS, running <code>open -a rstudio .</code> from the shell causes RStudio to recognize most of the environment variables, but strangely it does not set the correct library path to the conda R packages. Suggestions for how to fix this are welcome.</p>
</section>
<section id="regular-setup" class="level2">
<h2>Regular setup</h2>
<p>Once you have performed the <a href="#initial-setup">initial setup</a> on your computer, the regular setup you have to perform each time you want to work on the project is much quicker. Navigate to the singlecell-qtl directory, activate the conda environment, and pull the latest changes from the GitHub repository.</p>
<pre><code>cd singlecell-qtl
source activate scqtl
git pull</code></pre>
<p>If there are updates to <code>environment.yaml</code>, you can update the “scqtl” conda environment. You may want to do this occasionally just to be sure you have all the software you need.</p>
<pre><code>conda env update --file environment.yaml</code></pre>
</section>
<section id="bioinformatics-pipeline" class="level2">
<h2>Bioinformatics pipeline</h2>
<p>The main steps of the bioinformatics pipeline are:</p>
<ul>
<li>Reads that do not have a valid UMI are removed</li>
<li>UMIs are extracted with <a href="https://github.com/CGATOxford/UMI-tools">UMI-tools</a> (<code>umi_tools extract</code>)</li>
<li>Reads are mapped to the genome with <a href="http://bioinf.wehi.edu.au/subjunc/">Subjunc</a></li>
<li>Reads are deduplicated based on their UMI with <a href="https://github.com/CGATOxford/UMI-tools">UMI-tools</a> (<code>umi_tools dedup</code>; output referred to as molecules)</li>
<li>Molecules are assigned to genes with <a href="http://bioinf.wehi.edu.au/featureCounts/">featureCounts</a></li>
<li>The identity of the single cells is determined with <a href="http://genome.sph.umich.edu/wiki/VerifyBamID">verifyBamID</a></li>
</ul>
<p>The pipeline is implemented with <a href="https://snakemake.readthedocs.io/en/stable/">Snakemake</a>. The <a href="https://github.com/jdblischak/singlecell-qtl/blob/master/Snakefile">Snakefile</a> controls this process, ensuring that output files are updated when upstream files are changed. Options for the pipeline are set in <a href="https://github.com/jdblischak/singlecell-qtl/blob/master/config.yaml">config.yaml</a>, options for submission to Slurm are set in <a href="https://github.com/jdblischak/singlecell-qtl/blob/master/cluster.json">cluster.json</a>, the Bash script for executing Snakemake is in <a href="https://github.com/jdblischak/singlecell-qtl/blob/master/submit-snakemake.sh">submit-snakemake.sh</a>, and the Slurm sbatch file for submitting the pipeline is <a href="https://github.com/jdblischak/singlecell-qtl/blob/master/snakemake.sbatch">snakemake.sbatch</a>.</p>
</section>
<section id="using-expressionset-objects" class="level2">
<h2>Using ExpressionSet objects</h2>
<p>The processed data for each C1 chip is stored in its own RDS file, a binary format for storing R objects. The RDS file contains an <a href="https://www.bioconductor.org/packages/release/bioc/vignettes/Biobase/inst/doc/ExpressionSetIntroduction.pdf">ExpressionSet</a> object. The ExpressionSet class stores the RNA-seq counts along with metadata on the samples, features (e.g. genes), and experimental details.</p>
<p>To import the data, use the R function <code>readRDS()</code>. Because the ExpressionSet class is defined in the Bioconductor package Biobase, it will be loaded the first time you access the loaded ExpressionSet object. To avoid the long startup message that the Biobase package prints, here it is loaded prior to importing the object.</p>
<pre class="r"><code>suppressPackageStartupMessages(library(&quot;Biobase&quot;))
eset &lt;- readRDS(&quot;../data/eset/03162017.rds&quot;)</code></pre>
<p>Executing the object in the R console provides an overall summary of what it contains.</p>
<pre class="r"><code>eset</code></pre>
<pre><code>ExpressionSet (storageMode: lockedEnvironment)
assayData: 54792 features, 96 samples 
  element names: exprs 
protocolData: none
phenoData
  sampleNames: 03162017-A01 03162017-A02 ... 03162017-H12 (96
    total)
  varLabels: experiment well ... valid_id (40 total)
  varMetadata: labelDescription
featureData
  featureNames: ENSG00000000003 ENSG00000000005 ... WBGene00235374
    (54792 total)
  fvarLabels: chr start ... source (6 total)
  fvarMetadata: labelDescription
experimentData: use &#39;experimentData(object)&#39;
Annotation:  </code></pre>
<p>But each element of the object can be extracted individually, for example:</p>
<pre class="r"><code># Dimensions
dim(eset)</code></pre>
<pre><code>Features  Samples 
   54792       96 </code></pre>
<pre class="r"><code># Sample names
head(sampleNames(eset))</code></pre>
<pre><code>[1] &quot;03162017-A01&quot; &quot;03162017-A02&quot; &quot;03162017-A03&quot; &quot;03162017-A04&quot;
[5] &quot;03162017-A05&quot; &quot;03162017-A06&quot;</code></pre>
<pre class="r"><code># Expression data
exprs(eset)[1:5, 1:5]</code></pre>
<pre><code>                03162017-A01 03162017-A02 03162017-A03 03162017-A04
ENSG00000000003            0            3            3            3
ENSG00000000005            0            0            0            0
ENSG00000000419            0            3            1            2
ENSG00000000457            1            0            1            0
ENSG00000000460            0            2            0            0
                03162017-A05
ENSG00000000003            2
ENSG00000000005            0
ENSG00000000419            2
ENSG00000000457            2
ENSG00000000460            0</code></pre>
<pre class="r"><code># Sample-level variables (&quot;phenotype data&quot;)
pData(eset)[1:5, 1:5]</code></pre>
<pre><code>             experiment well batch cell_number concentration
03162017-A01   03162017  A01    b1           1     0.6790047
03162017-A02   03162017  A02    b1           1     1.6430792
03162017-A03   03162017  A03    b1           1     0.4034764
03162017-A04   03162017  A04    b1           0     0.9992873
03162017-A05   03162017  A05    b1           1     0.3265886</code></pre>
<pre class="r"><code># Descriptions of the sample-level variables
varMetadata(eset)[1:5, , drop = FALSE]</code></pre>
<pre><code>                                                      labelDescription
experiment            ID of C1 chip (i.e. processing date in MMDDYYYY)
well                   Well of C1 chip (96 total, rows A-H, cols 1-12)
batch                 Batch the C1 chip was processed in (b1, b2, ...)
cell_number    The number of cells observed in the well via microscopy
concentration The cDNA concentration of the well prior to library prep</code></pre>
<pre class="r"><code># Gene-level variables (&quot;feature data&quot;)
head(fData(eset))</code></pre>
<pre><code>                 chr     start       end     name strand     source
ENSG00000000003  hsX  99883667  99894988   TSPAN6      - H. sapiens
ENSG00000000005  hsX  99839799  99854882     TNMD      + H. sapiens
ENSG00000000419 hs20  49551404  49575092     DPM1      - H. sapiens
ENSG00000000457  hs1 169818772 169863408    SCYL3      - H. sapiens
ENSG00000000460  hs1 169631245 169823221 C1orf112      + H. sapiens
ENSG00000000938  hs1  27938575  27961788      FGR      - H. sapiens</code></pre>
<pre class="r"><code># Descriptions of the gene-level metrics
fvarMetadata(eset)</code></pre>
<pre><code>                                               labelDescription
chr                                                  Chromosome
start  Most 5&#39; start position (GRCh37/hg19; 1-based; inclusive)
end      Most 3&#39; end position (GRCh37/hg19; 1-based; inclusive)
name                                                  Gene name
strand      Strand (+ = positive/forward; - = negative/reverse)
source                                            Source of RNA</code></pre>
<pre class="r"><code># Experimental details
experimentData(eset)</code></pre>
<pre><code>Experiment data
  Experimenter name: PoYuan Tung, Harold Pimentel, Joyce Hsiao, John Blischak 
  Laboratory: Gilad/Pritchard 
  Contact information: https://github.com/jdblischak/singlecell-qtl/issues 
  Title: singlecell-qtl 
  URL: https://jdblischak.github.io/singlecell-qtl 
  PMIDs:  

  Abstract: A 15 word abstract is available. Use &#39;abstract&#39; method.
  Information is available on: preprocessing </code></pre>
<pre class="r"><code># Abstract
abstract(eset)</code></pre>
<pre><code>[1] &quot;Mapping genetic variants associated with cell-to-cell variation in gene expression in the HapMap Yoruba population.&quot;</code></pre>
<pre class="r"><code># Proprocessing steps
preproc(eset)</code></pre>
<pre><code>$mapping
[1] &quot;Mapped to GRCh37 with Sunjunc&quot;

$umi
[1] &quot;UMIs were extracted and deduplicated with umi_tools&quot;

$counting
[1] &quot;Counts per gene (Ensembl 75) were obtained with featureCounts&quot;

$verify
[1] &quot;Individuals were identified by verifyBamID&quot;</code></pre>
<p>A nice feature of ExpressionSets is that any subsetting applied to either the sample-level or gene-level variables is automatically propagated to the expression matrix.</p>
<pre class="r"><code>eset_quality &lt;- eset[fData(eset)$source == &quot;H. sapiens&quot;,
                     eset$cell_number == 1 &amp; eset$tra1.60]
dim(eset_quality)</code></pre>
<pre><code>Features  Samples 
   20327       85 </code></pre>
<p>Multiple ExpressionSet objects can also be combined. The following code imports and combines the data for all 10 C1 chips from batch 1.</p>
<pre class="r"><code>chips &lt;- c(&quot;03162017.rds&quot;, &quot;03172017.rds&quot;, &quot;03232017.rds&quot;,
           &quot;03302017.rds&quot;, &quot;03312017.rds&quot;, &quot;04052017.rds&quot;,
           &quot;04072017.rds&quot;, &quot;04132017.rds&quot;, &quot;04142017.rds&quot;,
           &quot;04202017.rds&quot;)
fname &lt;- Map(function(x) paste0(&quot;../data/eset/&quot;, x), chips)
batch1 &lt;- Reduce(combine, Map(readRDS, fname))
dim(batch1)</code></pre>
<pre><code>Features  Samples 
   54792      960 </code></pre>
<p>To read in the entire data set, use the combined ExpressionSet object (it’s much faster to load just one serialized data file).</p>
<pre class="r"><code>eset_total &lt;- readRDS(&quot;../data/eset.rds&quot;)
dim(eset_total)</code></pre>
<pre><code>Features  Samples 
   54792     4992 </code></pre>
<p>To learn more about the data that can be stored in an ExpressionSet object and methods to access the data, please see the help pages for ExpressionSet and its parent class eSet.</p>
<pre class="r"><code>?ExpressionSet
?eSet</code></pre>
</section>
<section id="using-combined-data-files" class="level2">
<h2>Using combined data files</h2>
<p>For convenience, the combined data are available in single files.</p>
<p>The annotation for all the single cells is in <a href="https://github.com/jdblischak/singlecell-qtl/blob/master/data/scqtl-annotation.txt">data/scqtl-annotation.txt</a>. The description of the annotation columns is in <a href="https://github.com/jdblischak/singlecell-qtl/blob/master/data/scqtl-annotation-description.txt">data/scqtl-annotation-description.txt</a>.</p>
<pre class="r"><code>anno &lt;- read.delim(&quot;../data/scqtl-annotation.txt&quot;, stringsAsFactors = FALSE)
anno_descrip &lt;- read.delim(&quot;../data/scqtl-annotation-description.txt&quot;,
                           stringsAsFactors = FALSE)
head(anno_descrip, 3)</code></pre>
<pre><code>    variable                                      description
1 experiment ID of C1 chip (i.e. processing date in MMDDYYYY)
2       well  Well of C1 chip (96 total, rows A-H, cols 1-12)
3      batch Batch the C1 chip was processed in (b1, b2, ...)</code></pre>
<p>The combined counts are in <a href="https://github.com/jdblischak/singlecell-qtl/blob/master/data/scqtl-counts.txt.gz">data/scqtl-counts.txt.gz</a>. Note that because this file is large, it is stored using <a href="https://git-lfs.github.com/">Git LFS</a>. Thus you must follow the <a href="#initial-setup">setup instructions above</a> to download this file to your local computer. I recommend importing it with <code>data.table::fread</code>.</p>
<pre class="r"><code>library(&quot;data.table&quot;)
counts &lt;- data.frame(fread(&quot;zcat ../data/scqtl-counts.txt.gz&quot;), row.names = 1)</code></pre>
<pre><code>
Read 49.2% of 20327 rows
Read 98.4% of 20327 rows
Read 20327 rows and 4993 (of 4993) columns from 0.193 GB file in 00:00:07</code></pre>
<pre class="r"><code>dim(counts)</code></pre>
<pre><code>[1] 20327  4992</code></pre>
<pre class="r"><code>counts[1:5, 1:2]</code></pre>
<pre><code>                NA19098.03162017.A01 NA18505.03162017.A02
ENSG00000000003                    0                    3
ENSG00000000005                    0                    0
ENSG00000000419                    0                    3
ENSG00000000457                    1                    0
ENSG00000000460                    0                    2</code></pre>
</section>
<section id="data-analysis-workflow" class="level2">
<h2>Data analysis workflow</h2>
<p>The data analysis is organized using the R package <a href="https://github.com/jdblischak/workflowr">workflowr</a>. Here are the basics:</p>
<pre class="r"><code># Create a new analysis file using the workflowr
wflow_open(&quot;analysis/new.Rmd&quot;)
# Build the website locally
wflow_build()
# Build and commit the website files
wflow_publish(&quot;analysis/new.Rmd&quot;, &quot;Add a new analysis&quot;)</code></pre>
<p>Ideally <code>wflow_publish()</code> will be run using the conda environment scqtl so that the final results are always generated using the same versions of the software. If you are using RStudio, you may need to open R from the command line to execute <code>wflow_publish()</code> in the conda environment.</p>
<p>More information on workflowr can be found in the <a href="https://jdblischak.github.io/workflowr/">online documentation</a>.</p>
</section>
<section id="adding-large-files-to-the-git-repo" class="level2">
<h2>Adding large files to the Git repo</h2>
<p>Since we use <a href="https://git-lfs.github.com/">Git LFS</a>, we can version large binary files with Git without worrying about the size of the Git repo. Before you do anything with large files, make sure you have activated the conda environment to ensure that you have Git LFS installed.</p>
<ol type="1">
<li><p>Activate the conda environment: <code>source activate scqtl</code></p></li>
<li><p>Check the currently tracked file extensions: <code>git lfs track</code></p></li>
<li><p>If the file extension is not yet being tracked, you first need to track it. For example, if you wanted to use Git LFS with files that end in the extension <code>.xyz</code>, you would run the following: <code>git lfs track &quot;*.xyz&quot;  git add .gitattributes</code></p></li>
<li><p>Once the file extension is being tracked, add and commit the file like any other file (you need to specify the <code>-f</code> flag for “force” because the directory <code>data/</code> is ignored by default to prevent accidentally committing a large file): <code>git add -f data/filename.xyz  git commit</code></p></li>
<li><p>Before pushing to GitHub, confirm that the large file was properly committed by running <code>git log --stat</code>. The file that was committed should only be a few lines long because it is a plain text file with a pointer to the version of the large file on GitHub’s servers.</p></li>
</ol>
</section>
<section id="installation-of-r-packages-outside-of-conda-environment" class="level2">
<h2>Installation of R packages outside of conda environment</h2>
<p>There may be situations where you cannot access the R packages installed in the conda environment (e.g. running RStudio on macOS, or RStudio Server). The code below installs the necessary R packages from <a href="https://cran.r-project.org/">CRAN</a>, <a href="https://bioconductor.org">Bioconductor</a>, and <a href="https://github.com">GitHub</a>.</p>
<p>Install R packages from <a href="https://cran.r-project.org/">CRAN</a>:</p>
<pre class="r"><code>install.packages(&quot;cowplot&quot;, &quot;data.table&quot;, &quot;devtools&quot;, &quot;dplyr&quot;, &quot;ggplot2&quot;,
                 &quot;readr&quot;, &quot;rmarkdown&quot;, &quot;stringr&quot;, &quot;testit&quot;, &quot;tidyr&quot;)</code></pre>
<p>Install R packages from <a href="https://bioconductor.org">Bioconductor</a>:</p>
<pre class="r"><code>source(&quot;https://bioconductor.org/biocLite.R&quot;)
biocLite(c(&quot;Biobase&quot;, &quot;biomaRt&quot;, &quot;edgeR&quot;))</code></pre>
<p>Install R packages from <a href="https://github.com">GitHub</a>:</p>
<pre class="r"><code>devtools::install_github(&quot;jdblischak/workflowr&quot;)</code></pre>
</section>
<section id="tips-and-tricks" class="level2">
<h2>Tips and tricks</h2>
<section id="enable-bash-auto-completion-for-git-on-macos" class="level3">
<h3>Enable Bash auto-completion for Git on macOS</h3>
<p>Manually typing <code>git pull origin master</code> and other long Git commands is a pain. To enable <a href="https://git-scm.com/book/en/v2/Appendix-A:-Git-in-Other-Environments-Git-in-Bash">Bash auto-completion</a> (i.e. tab-completion) for Git on macOS, follow these <a href="http://www.codethatmatters.com/2010/01/git-autocomplete-in-mac-os-x/">steps</a>:</p>
<ol type="1">
<li><p>Download the Bash completion script: <code>curl -L https://raw.github.com/git/git/master/contrib/completion/git-completion.bash &gt; ~/git-completion.bash</code></p></li>
<li><p>Add the following line to your .bash_profile: <code>source ~/git-completion.bash</code></p></li>
<li><p>Re-run your .bash_profile: <code>source ~/.bash_profile</code></p></li>
</ol>
</section>
<section id="install-git-lfs-locally" class="level3">
<h3>Install git-lfs locally</h3>
<p>The <a href="#initial-setup">Initial setup instructions</a> install git-lfs using conda. Therefore you need to have the conda environment active (<code>source activate scqtl</code>) whenever you run a Git command, which ensures that the large files are always handled properly by Git. Since it can be easy to forget to first activate the conda environment, a safety measure would be to install git-lfs on your local machine (i.e. outside of the conda environment). This will ensure that git-lfs is always available even when you forget to activate the conda environment.</p>
<p>There are multiple <a href="https://github.com/git-lfs/git-lfs#getting-started">methods</a> to install git-lfs. Below are instructions to download a pre-built binary.</p>
<ol type="1">
<li><p>Download the software: <code>cd ~/Downloads  # 64-bit macOS:  curl -OL https://github.com/git-lfs/git-lfs/releases/download/v2.3.0/git-lfs-darwin-amd64-2.3.0.tar.gz  # 64-bit Linux:  curl -OL https://github.com/git-lfs/git-lfs/releases/download/v2.3.0/git-lfs-linux-amd64-2.3.0.tar.gz</code></p></li>
<li><p>Extract the software and copy the executable files into <code>~/bin</code> (note: you can put it anywhere you like, but this is a conventional place to put executable files): <code>mkdir -p ~/bin  tar xzf git-lfs-*.tar.gz  cd git-lfs-2.3.0/  cp git-lfs install.sh ~/bin</code></p></li>
<li><p>Add <code>~/bin</code> to your <code>PATH</code> by adding the following line to your <code>~/.bash_profile</code> (macOS) or <code>.bashrc</code> (Linux): <code>export PATH=&quot;~/bin:$PATH&quot;</code> Note that <code>~/bin</code> is already included in <code>PATH</code> by default on at least some Linux distributions.</p></li>
<li><p>Update <code>PATH</code>: <code># macOS  source ~/.bash_profile  # Linux  source ~/.bashrc</code></p></li>
<li><p>Just to be safe, run the git-lfs install script (this modifies the file <code>~/.gitconfig</code>). This is unnecessary if you have already gone through the <a href="#initial-setup">Initial setup instructions</a> on this machine, but it doesn’t hurt to run it again: <code>git lfs install</code></p></li>
<li><p>To confirm that you have git-lfs installed locally, run <code>which git-lfs</code>. This should point to <code>~/bin/git-lfs</code> and not your conda environment. Also run <code>cat ~/.gitconfig</code>. It should contain the following (it will also likely have additional lines): <code>[filter &quot;lfs&quot;]          clean = git-lfs clean -- %f          smudge = git-lfs smudge -- %f          process = git-lfs filter-process          required = true</code></p></li>
</ol>
</section>
<section id="upgrade-rstudio" class="level3">
<h3>Upgrade RStudio</h3>
<p>RStudio has had some major updates with its 1.0 release. If you are using a version less than 1.0 (check by clicking <code>Help</code> -&gt; <code>About RStudio</code>), you should strongly consider <a href="https://www.rstudio.com/products/rstudio/download/#download">upgrading</a>. The most noticeable change for our workflow is that it has built-in support for R Markdown websites. Thus if you click the <code>Knit</code> button for an R Markdown file in the workflowr project, it will recognize the settings in the file <code>analysis/_site.yml</code> and save the output HTML in <code>docs/</code>.</p>
<p>One annoying new feature you will notice is that the chunks of R Markdown documents appear inline in the document instead of being executed in the R console. To restore the previous setting, go to <code>Tools</code> -&gt; <code>Global Options...</code> -&gt; <code>R Markdown</code> and unselect “Show output inline for all R Markdown documents”.</p>
</section>
<section id="open-rstudio-from-the-terminal" class="level3">
<h3>Open RStudio from the Terminal</h3>
<p>On macOS and Linux systems you can run RStudio from a terminal and specify which working directory to startup within. Additionally, on Linux systems if you run RStudio from a terminal and specify no command line argument then RStudio will startup using the current working directory of the terminal.</p>
<p>For example, on macOS you could use the following commands to open RStudio (respectively) in the <code>~/projects/foo</code> directory or the current working directory:</p>
<pre><code>$ open -a RStudio ~/projects/foo
$ open -a RStudio .</code></pre>
<p>On Linux you would use the following commands (note that no <code>.</code> is necessary in the second invocation):</p>
<pre><code>$ rstudio ~/projects/foo
$ rstudio</code></pre>
</section>
</section>
<section id="faq" class="level2">
<h2>FAQ</h2>
<section id="why-does-git-keep-asking-for-my-password-repeatedly" class="level3">
<h3>Why does Git keep asking for my password repeatedly?</h3>
<p>We use <a href="https://git-lfs.github.com/">git-lfs</a> to version control large files. To push updated versions of these large files requires authentication of the https server where they are stored, and this must be done for each file separately! To reduce the number of times you have to manually enter your password, enable <a href="https://help.github.com/articles/caching-your-github-password-in-git/">caching of your GitHub password</a>:</p>
<pre><code>git config --global credential.helper cache</code></pre>
</section>
</section>
<section id="session-information" class="level2">
<h2>Session information</h2>
<!-- Insert the session information into the document -->
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 3.4.1 (2017-06-30)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Scientific Linux 7.2 (Nitrogen)

Matrix products: default
BLAS: /project2/gilad/jdblischak/miniconda3/envs/scqtl/lib/R/lib/libRblas.so
LAPACK: /project2/gilad/jdblischak/miniconda3/envs/scqtl/lib/R/lib/libRlapack.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] parallel  methods   stats     graphics  grDevices utils     datasets 
[8] base     

other attached packages:
[1] data.table_1.10.4   Biobase_2.38.0      BiocGenerics_0.24.0

loaded via a namespace (and not attached):
 [1] Rcpp_0.12.13    digest_0.6.12   rprojroot_1.2   backports_1.0.5
 [5] git2r_0.19.0    magrittr_1.5    evaluate_0.10.1 stringi_1.1.2  
 [9] rmarkdown_1.6   tools_3.4.1     stringr_1.2.0   yaml_2.1.14    
[13] compiler_3.4.1  htmltools_0.3.6 knitr_1.16     </code></pre>
</section>

<hr>
<p>
    This <a href="http://rmarkdown.rstudio.com">R Markdown</a> site was created with <a href="https://github.com/jdblischak/workflowr">workflowr</a>
</p>
<hr>

<!-- To enable disqus, uncomment the section below and provide your disqus_shortname -->

<!-- disqus
  <div id="disqus_thread"></div>
    <script type="text/javascript">
        /* * * CONFIGURATION VARIABLES: EDIT BEFORE PASTING INTO YOUR WEBPAGE * * */
        var disqus_shortname = 'rmarkdown'; // required: replace example with your forum shortname

        /* * * DON'T EDIT BELOW THIS LINE * * */
        (function() {
            var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
            dsq.src = '//' + disqus_shortname + '.disqus.com/embed.js';
            (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
        })();
    </script>
    <noscript>Please enable JavaScript to view the <a href="http://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
    <a href="http://disqus.com" class="dsq-brlink">comments powered by <span class="logo-disqus">Disqus</span></a>
-->


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
