<!DOCTYPE html>
<html lang="en">
<head>
<!-- 2017-12-19 Tue 11:36 -->
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
<title>Hurdle model estimation</title>
<meta name="generator" content="Org mode">
<meta name="author" content="Abhishek Sarkar">
<style type="text/css">
 <!--/*--><![CDATA[/*><!--*/
  .title  { text-align: center;
             margin-bottom: .2em; }
  .subtitle { text-align: center;
              font-size: medium;
              font-weight: bold;
              margin-top:0; }
  .todo   { font-family: monospace; color: red; }
  .done   { font-family: monospace; color: green; }
  .priority { font-family: monospace; color: orange; }
  .tag    { background-color: #eee; font-family: monospace;
            padding: 2px; font-size: 80%; font-weight: normal; }
  .timestamp { color: #bebebe; }
  .timestamp-kwd { color: #5f9ea0; }
  .org-right  { margin-left: auto; margin-right: 0px;  text-align: right; }
  .org-left   { margin-left: 0px;  margin-right: auto; text-align: left; }
  .org-center { margin-left: auto; margin-right: auto; text-align: center; }
  .underline { text-decoration: underline; }
  #postamble p, #preamble p { font-size: 90%; margin: .2em; }
  p.verse { margin-left: 3%; }
  pre {
    border: 1px solid #ccc;
    box-shadow: 3px 3px 3px #eee;
    padding: 8pt;
    font-family: monospace;
    overflow: auto;
    margin: 1.2em;
  }
  pre.src {
    position: relative;
    overflow: visible;
    padding-top: 1.2em;
  }
  pre.src:before {
    display: none;
    position: absolute;
    background-color: white;
    top: -10px;
    right: 10px;
    padding: 3px;
    border: 1px solid black;
  }
  pre.src:hover:before { display: inline;}
  /* Languages per Org manual */
  pre.src-asymptote:before { content: 'Asymptote'; }
  pre.src-awk:before { content: 'Awk'; }
  pre.src-C:before { content: 'C'; }
  /* pre.src-C++ doesn't work in CSS */
  pre.src-clojure:before { content: 'Clojure'; }
  pre.src-css:before { content: 'CSS'; }
  pre.src-D:before { content: 'D'; }
  pre.src-ditaa:before { content: 'ditaa'; }
  pre.src-dot:before { content: 'Graphviz'; }
  pre.src-calc:before { content: 'Emacs Calc'; }
  pre.src-emacs-lisp:before { content: 'Emacs Lisp'; }
  pre.src-fortran:before { content: 'Fortran'; }
  pre.src-gnuplot:before { content: 'gnuplot'; }
  pre.src-haskell:before { content: 'Haskell'; }
  pre.src-hledger:before { content: 'hledger'; }
  pre.src-java:before { content: 'Java'; }
  pre.src-js:before { content: 'Javascript'; }
  pre.src-latex:before { content: 'LaTeX'; }
  pre.src-ledger:before { content: 'Ledger'; }
  pre.src-lisp:before { content: 'Lisp'; }
  pre.src-lilypond:before { content: 'Lilypond'; }
  pre.src-lua:before { content: 'Lua'; }
  pre.src-matlab:before { content: 'MATLAB'; }
  pre.src-mscgen:before { content: 'Mscgen'; }
  pre.src-ocaml:before { content: 'Objective Caml'; }
  pre.src-octave:before { content: 'Octave'; }
  pre.src-org:before { content: 'Org mode'; }
  pre.src-oz:before { content: 'OZ'; }
  pre.src-plantuml:before { content: 'Plantuml'; }
  pre.src-processing:before { content: 'Processing.js'; }
  pre.src-python:before { content: 'Python'; }
  pre.src-R:before { content: 'R'; }
  pre.src-ruby:before { content: 'Ruby'; }
  pre.src-sass:before { content: 'Sass'; }
  pre.src-scheme:before { content: 'Scheme'; }
  pre.src-screen:before { content: 'Gnu Screen'; }
  pre.src-sed:before { content: 'Sed'; }
  pre.src-sh:before { content: 'shell'; }
  pre.src-sql:before { content: 'SQL'; }
  pre.src-sqlite:before { content: 'SQLite'; }
  /* additional languages in org.el's org-babel-load-languages alist */
  pre.src-forth:before { content: 'Forth'; }
  pre.src-io:before { content: 'IO'; }
  pre.src-J:before { content: 'J'; }
  pre.src-makefile:before { content: 'Makefile'; }
  pre.src-maxima:before { content: 'Maxima'; }
  pre.src-perl:before { content: 'Perl'; }
  pre.src-picolisp:before { content: 'Pico Lisp'; }
  pre.src-scala:before { content: 'Scala'; }
  pre.src-shell:before { content: 'Shell Script'; }
  pre.src-ebnf2ps:before { content: 'ebfn2ps'; }
  /* additional language identifiers per "defun org-babel-execute"
       in ob-*.el */
  pre.src-cpp:before  { content: 'C++'; }
  pre.src-abc:before  { content: 'ABC'; }
  pre.src-coq:before  { content: 'Coq'; }
  pre.src-groovy:before  { content: 'Groovy'; }
  /* additional language identifiers from org-babel-shell-names in
     ob-shell.el: ob-shell is the only babel language using a lambda to put
     the execution function name together. */
  pre.src-bash:before  { content: 'bash'; }
  pre.src-csh:before  { content: 'csh'; }
  pre.src-ash:before  { content: 'ash'; }
  pre.src-dash:before  { content: 'dash'; }
  pre.src-ksh:before  { content: 'ksh'; }
  pre.src-mksh:before  { content: 'mksh'; }
  pre.src-posh:before  { content: 'posh'; }
  /* Additional Emacs modes also supported by the LaTeX listings package */
  pre.src-ada:before { content: 'Ada'; }
  pre.src-asm:before { content: 'Assembler'; }
  pre.src-caml:before { content: 'Caml'; }
  pre.src-delphi:before { content: 'Delphi'; }
  pre.src-html:before { content: 'HTML'; }
  pre.src-idl:before { content: 'IDL'; }
  pre.src-mercury:before { content: 'Mercury'; }
  pre.src-metapost:before { content: 'MetaPost'; }
  pre.src-modula-2:before { content: 'Modula-2'; }
  pre.src-pascal:before { content: 'Pascal'; }
  pre.src-ps:before { content: 'PostScript'; }
  pre.src-prolog:before { content: 'Prolog'; }
  pre.src-simula:before { content: 'Simula'; }
  pre.src-tcl:before { content: 'tcl'; }
  pre.src-tex:before { content: 'TeX'; }
  pre.src-plain-tex:before { content: 'Plain TeX'; }
  pre.src-verilog:before { content: 'Verilog'; }
  pre.src-vhdl:before { content: 'VHDL'; }
  pre.src-xml:before { content: 'XML'; }
  pre.src-nxml:before { content: 'XML'; }
  /* add a generic configuration mode; LaTeX export needs an additional
     (add-to-list 'org-latex-listings-langs '(conf " ")) in .emacs */
  pre.src-conf:before { content: 'Configuration File'; }

  table { border-collapse:collapse; }
  caption.t-above { caption-side: top; }
  caption.t-bottom { caption-side: bottom; }
  td, th { vertical-align:top;  }
  th.org-right  { text-align: center;  }
  th.org-left   { text-align: center;   }
  th.org-center { text-align: center; }
  td.org-right  { text-align: right;  }
  td.org-left   { text-align: left;   }
  td.org-center { text-align: center; }
  dt { font-weight: bold; }
  .footpara { display: inline; }
  .footdef  { margin-bottom: 1em; }
  .figure { padding: 1em; }
  .figure p { text-align: center; }
  .inlinetask {
    padding: 10px;
    border: 2px solid gray;
    margin: 10px;
    background: #ffffcc;
  }
  #org-div-home-and-up
   { text-align: right; font-size: 70%; white-space: nowrap; }
  textarea { overflow-x: auto; }
  .linenr { font-size: smaller }
  .code-highlighted { background-color: #ffff00; }
  .org-info-js_info-navigation { border-style: none; }
  #org-info-js_console-label
    { font-size: 10px; font-weight: bold; white-space: nowrap; }
  .org-info-js_search-highlight
    { background-color: #ffff00; color: #000000; font-weight: bold; }
  .org-svg { width: 90%; }
  /*]]>*/-->
</style>
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<link rel="stylesheet" type="text/css" href="https://www.pirilampo.org/styles/readtheorg/css/htmlize.css"/>
<style type="text/css">body {width: 60em; margin:auto} pre.src {overflow:auto}</style>
<script type="text/javascript">
/*
@licstart  The following is the entire license notice for the
JavaScript code in this tag.

Copyright (C) 2012-2017 Free Software Foundation, Inc.

The JavaScript code in this tag is free software: you can
redistribute it and/or modify it under the terms of the GNU
General Public License (GNU GPL) as published by the Free Software
Foundation, either version 3 of the License, or (at your option)
any later version.  The code is distributed WITHOUT ANY WARRANTY;
without even the implied warranty of MERCHANTABILITY or FITNESS
FOR A PARTICULAR PURPOSE.  See the GNU GPL for more details.

As additional permission under GNU GPL version 3 section 7, you
may distribute non-source (e.g., minimized or compacted) forms of
that code without the copy of the GNU GPL normally required by
section 4, provided you include this license notice and a URL
through which recipients can access the Corresponding Source.


@licend  The above is the entire license notice
for the JavaScript code in this tag.
*/
<!--/*--><![CDATA[/*><!--*/
 function CodeHighlightOn(elem, id)
 {
   var target = document.getElementById(id);
   if(null != target) {
     elem.cacheClassElem = elem.className;
     elem.cacheClassTarget = target.className;
     target.className = "code-highlighted";
     elem.className   = "code-highlighted";
   }
 }
 function CodeHighlightOff(elem, id)
 {
   var target = document.getElementById(id);
   if(elem.cacheClassElem)
     elem.className = elem.cacheClassElem;
   if(elem.cacheClassTarget)
     target.className = elem.cacheClassTarget;
 }
/*]]>*///-->
</script>
<script type="text/x-mathjax-config">
    MathJax.Hub.Config({
        displayAlign: "center",
        displayIndent: "0em",

        "HTML-CSS": { scale: 100,
                        linebreaks: { automatic: "false" },
                        webFont: "TeX"
                       },
        SVG: {scale: 100,
              linebreaks: { automatic: "false" },
              font: "TeX"},
        NativeMML: {scale: 100},
        TeX: { equationNumbers: {autoNumber: "AMS"},
               MultLineWidth: "85%",
               TagSide: "right",
               TagIndent: ".8em"
             }
});
</script>
<script type="text/javascript"
        src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-AMS_HTML"></script>
</head>
<body>
<div id="content">
<h1 class="title">Hurdle model estimation</h1>
<div id="table-of-contents">
<h2>Table of Contents</h2>
<div id="text-table-of-contents">
<ul>
<li><a href="#orga5b8a16">Introduction</a></li>
<li><a href="#orgb2384c1">Test case</a></li>
<li><a href="#orgef8e65e">Model specification and inference</a></li>
<li><a href="#orged40312">Next steps</a></li>
</ul>
</div>
</div>

<div id="outline-container-orga5b8a16" class="outline-2">
<h2 id="orga5b8a16">Introduction</h2>
<div class="outline-text-2" id="text-orga5b8a16">
<p>
The key idea of <code>voom</code> (<a href="http://genomebiology.com/2014/15/2/R29">Law et al 2014</a>) is that the distribution of
log-transformed counts per million is approximately Gaussian. Consider the
expression of a single gene in individual \(i\), cell \(j\).
</p>

<p>
\[ r_{ij} = \mbox{number of reads} \]
</p>

<p>
Assuming \(r_{ij}\) follows the NB2 negative binomial model (Hilbe 2012):
</p>

<p>
\[ E[r_{ij}] = \lambda_{ij} \]
\[ V[r_{ij}] = \lambda_{ij} + \phi_{ij} \lambda_{ij}^2 \]
</p>

<p>
Let \(L_j\) be the library size of cell \(j\). Then, by the definition of CPM:
</p>

<p>
\[ y_{ij} = \log_2(10^6 \times \frac{r_{ij}}{L_j} + 1) \]
</p>

<p>
\[ = \log_2\left(10^6 \times \left(r_{ij} + \frac{L_j}{10^6}\right)\right) - \log_2(L_j) \]
</p>

<p>
Assuming that \(L_j \ll 10^6\):
</p>

<p>
\[ y_{ij} \approx \log_2(r_{ij}) - \log_2(L_j) + 6 \log_2(10) \]
</p>

<p>
By first order Taylor expansion:
</p>

<p>
\[ E[y_{ij}] = \mu_{ij} \approx \log_2 \lambda_{ij} + \mathrm{const} \]
</p>

<p>
By the delta method:
</p>

<p>
\[ V[y_{ij}] = \sigma_{ij}^2 \approx V[r_{ij}] / \lambda_{ij}^2 = 1/\lambda_{ij} + \phi_{ij} \]
</p>

<p>
We assume:
</p>

<p>
\[ y_{ij} \approx N(\mu_{ij}, \sigma^2_{ij}) \]
</p>

<p>
Now, we seek to write a hierarchical model for \(y\) in terms of the genotype
of individual \(i\) to call mean and variance QTLs.
</p>

<p>
Our idea is to start from generalized linear models for the underlying rate
and dispersion:
</p>

<p>
\[ \log \lambda_{ij} = X_i \beta_\lambda + u_i^\lambda + \epsilon^\lambda_{ij} \]
</p>

<p>
\[ \log \phi_{ij} = X_i \beta_\phi + u_i^\phi + \epsilon^\phi_{ij} \]
</p>

<p>
We assume errors are uncorrelated with \(X, \beta, u\) so we can write
\(\sigma^2 = V[\epsilon_{ij}^\lambda] + V[\epsilon_{ij}^\phi]\)
</p>

<p>
Then, the likelihood of each data point is given by:
</p>

<p>
\[ y_{ij} \sim N(X_i \beta_\lambda + u_i^\lambda, \exp(-(X_i \beta_\lambda +
  u_i^\lambda)) + \exp(X_i \beta_\phi + u_i^\phi) + \sigma^2) \]
</p>

<p>
In this model, between individual variance is explained by <i>cis</i>-genotype
(\(X_i \beta_\lambda)\), unobserved factors (\(u_i^\lambda)\), and sampling
(\(\epsilon_{ij}^\lambda\)).
</p>

<p>
Within individual variance is explained by the rate of expression (as derived
above), <i>cis</i>-genotype (\(X_i \beta_\phi\)), unobserved factors \(u_i^\phi\),
and sampling (\(\epsilon_{ij}^\phi\)).
</p>

<p>
Unobserved factors are needed to account for the fact that two individuals
with the same genotype could still have different underlying rate and
dispersion of expression.
</p>

<p>
We can estimate the posterior \(p(\beta, u \mid Y, \cdot)\) this model using
a combination of black-box variational inference and variational EM.
</p>

<p>
<b>TODO:</b> Marginalize out \(u\)
</p>

<p>
<b>TODO:</b> Specify priors on the correct scales for \(\beta\)
</p>
</div>
</div>

<div id="outline-container-orgb2384c1" class="outline-2">
<h2 id="orgb2384c1">Test case</h2>
<div class="outline-text-2" id="text-orgb2384c1">
<div class="org-src-container">
<pre class="src src-ipython">gene_info.query(<span class="org-string">'name == "ZSWIM7"'</span>)
</pre>
</div>

<pre class="example">
                chr     start       end    name strand      source
gene                                                                
ENSG00000214941  hs17  15879874  15903031  ZSWIM7      -  H. sapiens
</pre>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">with</span> gzip.<span class="org-builtin">open</span>(<span class="org-string">'/project2/gilad/singlecell-qtl/bulk/genotypes.vcf.gz'</span>, <span class="org-string">'rt'</span>) <span class="org-keyword">as</span> f:
  <span class="org-variable-name">header</span> = <span class="org-builtin">next</span>(f).split()
<span class="org-variable-name">vcf</span> = tabix.<span class="org-builtin">open</span>(<span class="org-string">'/project2/gilad/singlecell-qtl/bulk/genotypes.vcf.gz'</span>)
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">def</span> <span class="org-function-name">_extract_genotypes</span>(row, vcf, header, window):
  <span class="org-keyword">if</span> row[<span class="org-string">'strand'</span>] == <span class="org-string">'+'</span>:
    <span class="org-variable-name">start</span> = row[<span class="org-string">'start'</span>] - window
    <span class="org-variable-name">end</span> = row[<span class="org-string">'start'</span>]
  <span class="org-keyword">else</span>:
    <span class="org-variable-name">start</span> = row[<span class="org-string">'end'</span>]
    <span class="org-variable-name">end</span> = row[<span class="org-string">'end'</span>] + window
  <span class="org-variable-name">result</span> = pd.DataFrame(<span class="org-builtin">list</span>(vcf.query(<span class="org-string">'chr{}'</span>.<span class="org-builtin">format</span>(row[<span class="org-string">'chr'</span>][2:]), start, end)), columns=header).iloc[:,9:].astype(np.float32)
  <span class="org-keyword">return</span> result
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">genotypes</span> = _extract_genotypes(gene_info.loc[<span class="org-string">'ENSG00000214941'</span>], vcf=vcf, header=header, window=<span class="org-builtin">int</span>(1e6)).rename(columns=<span class="org-keyword">lambda</span> x: <span class="org-string">'NA{}'</span>.<span class="org-builtin">format</span>(x))[<span class="org-builtin">sorted</span>(annotations_qc[<span class="org-string">'chip_id'</span>].unique())]
<span class="org-variable-name">genotypes</span> = genotypes.T.transform(<span class="org-keyword">lambda</span> x: x - x.mean()).fillna(0)
genotypes.shape
</pre>
</div>

<pre class="example">
(21, 2506)

</pre>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">normalizers</span> = (6 * np.log(10) - np.log(umi_qc.agg(np.<span class="org-builtin">sum</span>))) / np.log(2)
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">data</span> = {<span class="org-string">'onehot'</span>: onehot,
        <span class="org-string">'genotypes'</span>: genotypes.values,
        <span class="org-string">'counts'</span>: umi_qc.loc[<span class="org-string">'ENSG00000214941'</span>].values.astype(np.float32).reshape(-1, 1),
        <span class="org-string">'log_cpm'</span>: cpm(umi_qc, log2=<span class="org-constant">True</span>).loc[<span class="org-string">'ENSG00000214941'</span>].values.astype(np.float32).reshape(-1, 1),
        <span class="org-string">'normalizers'</span>: normalizers.values.reshape(-1, 1)}
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">with</span> <span class="org-builtin">open</span>(<span class="org-string">'test_data.pkl'</span>, <span class="org-string">'wb'</span>) <span class="org-keyword">as</span> f:
  pickle.dump(data, f)
</pre>
</div>

<p>
Plot the data, its mean, and twice its standard deviation per individual:
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">onehot_cpm</span> = np.ma.masked_equal(data[<span class="org-string">'log_cpm'</span>] * data[<span class="org-string">'onehot'</span>], 0)

plt.clf()
plt.gcf().set_size_inches(8, 6)
plt.scatter(x=np.where(data[<span class="org-string">'onehot'</span>] == 1)[1] + np.random.normal(scale=0.1, size=m), y=data[<span class="org-string">'log_cpm'</span>], s=2, alpha=0.5)
plt.errorbar(x=np.arange(n), y=onehot_cpm.mean(axis=0), yerr=2 * onehot_cpm.std(axis=0), fmt=<span class="org-string">'o'</span>, c=<span class="org-string">'red'</span>)
plt.xlabel(<span class="org-string">'Individual'</span>)
plt.ylabel(<span class="org-string">'$\log_2(CPM + 1)$'</span>)
plt.gca().set_xticks([])
plt.gcf()
</pre>
</div>


<div class="figure">
<p><img src="figure/hurdle.org/data.png" alt="data.png">
</p>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">with</span> <span class="org-builtin">open</span>(<span class="org-string">'test_data.pkl'</span>, <span class="org-string">'rb'</span>) <span class="org-keyword">as</span> f:
  <span class="org-variable-name">data</span> = pickle.load(f)
</pre>
</div>
</div>
</div>

<div id="outline-container-orgef8e65e" class="outline-2">
<h2 id="orgef8e65e">Model specification and inference</h2>
<div class="outline-text-2" id="text-orgef8e65e">
<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">m</span>, <span class="org-variable-name">n</span> = data[<span class="org-string">'onehot'</span>].shape
<span class="org-variable-name">_</span>, <span class="org-variable-name">p</span> = data[<span class="org-string">'genotypes'</span>].shape

<span class="org-variable-name">onehot</span> = tf.placeholder(tf.float32, [m, n])
<span class="org-variable-name">genotypes</span> = tf.placeholder(tf.float32, [n, 1])
<span class="org-variable-name">cell_bias</span> = tf.placeholder(tf.float32, [m, 1])

<span class="org-variable-name">ind_bias_scale</span> = tf.Variable(tf.ones([n, 1]))
<span class="org-variable-name">ind_bias</span> = ed.models.NormalWithSoftplusScale(loc=tf.zeros([n, 1]), scale=ind_bias_scale)
<span class="org-variable-name">rate_effect_scale</span> = tf.Variable(tf.ones([1, 1]))
<span class="org-variable-name">rate_effect</span> = ed.models.NormalWithSoftplusScale(loc=tf.zeros([1, 1]), scale=rate_effect_scale)
<span class="org-variable-name">log_rate</span> = tf.matmul(onehot, tf.matmul(genotypes, rate_effect) + ind_bias)
<span class="org-variable-name">mean</span> = log_rate + cell_bias

<span class="org-variable-name">disp_effect_scale</span> = tf.Variable(tf.ones([1, 1]))
<span class="org-variable-name">disp_bias</span> = ed.models.NormalWithSoftplusScale(loc=tf.zeros([n, 1]), scale=ind_bias_scale)
<span class="org-variable-name">disp_effect</span> = ed.models.Normal(loc=tf.zeros([1, 1]), scale=disp_effect_scale)
<span class="org-variable-name">disp</span> = tf.matmul(onehot, tf.matmul(genotypes, disp_effect) + disp_bias)
<span class="org-variable-name">resid_var_scale</span> = tf.Variable(tf.zeros([1]))
<span class="org-variable-name">var</span> = tf.exp(-log_rate) + tf.exp(disp) + tf.exp(resid_var_scale)

<span class="org-variable-name">log_cpm</span> = ed.models.Normal(loc=mean, scale=tf.sqrt(var))

<span class="org-variable-name">q_ind_bias</span> = ed.models.Normal(loc=tf.Variable(tf.zeros([n, 1])), scale=tf.Variable(tf.ones([n, 1])))
<span class="org-variable-name">q_disp_bias</span> = ed.models.Normal(loc=tf.Variable(tf.zeros([n, 1])), scale=tf.Variable(tf.ones([n, 1])))
<span class="org-variable-name">q_rate_effect</span> = ed.models.Normal(loc=tf.Variable(tf.random_normal([1, 1])), scale=tf.Variable(tf.random_normal([1, 1])))
<span class="org-variable-name">q_disp_effect</span> = ed.models.Normal(loc=tf.Variable(tf.random_normal([1, 1])), scale=tf.Variable(tf.random_normal([1, 1])))

<span class="org-variable-name">inf</span> = ed.ReparameterizationKLKLqp(
  latent_vars={
    rate_effect: q_rate_effect,
    ind_bias: q_ind_bias,
    disp_effect: q_disp_effect,
    disp_bias: q_disp_bias,
  },
  data={
    onehot: data[<span class="org-string">'onehot'</span>],
    genotypes: data[<span class="org-string">'genotypes'</span>][:,0:1],
    cell_bias: data[<span class="org-string">'normalizers'</span>],
    log_cpm: data[<span class="org-string">'log_cpm'</span>],
  })
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython">inf.run(n_samples=10, optimizer=tf.train.AdamOptimizer(learning_rate=5e-2))
</pre>
</div>

<pre class="example">
1000/1000 [100%] ██████████████████████████████ Elapsed: 7s | Loss: 4225.791

</pre>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">res</span> = pd.DataFrame(
  np.hstack(ed.get_session().run([q_rate_effect.mean(), q_rate_effect.variance(), q_disp_effect.mean(), q_disp_effect.variance()])),
  columns=[<span class="org-string">'rate_effect_mean'</span>, <span class="org-string">'rate_effect_var'</span>, <span class="org-string">'disp_effect_mean'</span>, <span class="org-string">'disp_effect_var'</span>])
res
</pre>
</div>

<pre class="example">
 rate_effect_mean  rate_effect_var  disp_effect_mean  disp_effect_var
0         -0.005486         0.000523          0.282421         0.053938
</pre>

<p>
Plot a posterior predictive draw, and the real data means and twice standard
deviations.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">post_pred</span> = ed.get_session().run(
  ed.copy(log_cpm, inf.latent_vars),
  {
    onehot: data[<span class="org-string">'onehot'</span>],
    genotypes: data[<span class="org-string">'genotypes'</span>][:,0:1],
    cell_bias: data[<span class="org-string">'normalizers'</span>],
    log_cpm: data[<span class="org-string">'log_cpm'</span>],
  })

plt.clf()
plt.gcf().set_size_inches(8, 6)
plt.scatter(x=np.where(data[<span class="org-string">'onehot'</span>] == 1)[1] + np.random.normal(scale=0.1, size=m), y=post_pred, s=2)
plt.errorbar(x=np.arange(n), y=onehot_cpm.mean(axis=0), yerr=2 * onehot_cpm.std(axis=0), fmt=<span class="org-string">'o'</span>, c=<span class="org-string">'red'</span>)
plt.xlabel(<span class="org-string">'Individual'</span>)
plt.ylabel(<span class="org-string">'$\log_2(CPM + 1)$'</span>)
plt.gca().set_xticks([])
plt.gcf()
</pre>
</div>


<div class="figure">
<p><img src="figure/hurdle.org/post-pred.png" alt="post-pred.png">
</p>
</div>
</div>
</div>

<div id="outline-container-orged40312" class="outline-2">
<h2 id="orged40312">Next steps</h2>
<div class="outline-text-2" id="text-orged40312">
<p>
The key idea of <code>mast</code> (<a href="https://dx.doi.org/10.1186/s13059-015-0844-5">Finak et al 2015</a>) is to model non-zero \(R\) using a
Gaussian distribution, and model zero \(R\) using logistic regression. The
key distinction between this approach and the <a href="zinb.html">zero-inflated
negative binomial model</a> is that zeroes are assumed to arise from only one
process.
</p>

<p>
This naturally suggests a Bayesian model which simultaneously calls mean and
variance QTLs based on the likelihood:
</p>

<p>
\[ Y \mid Y = 0 \sim \mathrm{Bernoulli}(1 - \mathrm{sigmoid}(AX\theta_m + \delta)) \]
</p>

<p>
\[ Y \mid Y > 0 \sim N(AX\theta_m + \delta, \exp(-AX\theta_m - \delta) +
  \exp(AX\theta_v) + \sigma^2) \]
</p>

<p>
where \(\delta\) is a known constant that depends on library size, \(A\) maps
from cells to individuals, and \(X\) is the genotype matrix.
</p>

<p>
The key issues are:
</p>

<ol class="org-ol">
<li><p>
<b>Do we need to actually fit the dropout model?</b> <code>mast</code> conditions on the
observed \(Y = 0\), not on a latent \(Z = 0\), which means we could simply
do the same and ignore zeros.
</p>

<p>
If we did so, then mean/variance QTL effect size estimation will be less
robust for genes with high dropout. This might not be a problem depending
on the stringency of gene filtering.
</p></li>

<li><p>
<b>For a single gene, do we need to worry about mean QTLs in LD with
variance QTLs?</b> We previously <a href="https://github.com/YPARK/fqtl">built multivariate mean/variance QTL models</a>
which could account for LD, and could share information between the mean
and variance models.
</p>

<p>
The fundamental problem is that if we assume that the mean and dispersion
both have genetic components, then the mean is no longer independent of
the dispersion.
</p>

<p>
This actually could be derived without using the fact that both depend on
the same genotypes if we use second-order Taylor expansion:
</p>

<p>
\[ \mu \approx \log_2 \lambda + \frac{V[R]}{2 \lambda^2} \]
</p></li>

<li><p>
<b>Do we need to share parameters between genes?</b> <code>mast</code> assumes genes are
conditionally independent. But this is no longer true when nearby genes
can be driven by overlapping (or correlated) <i>cis</i>-genotypes.
</p>

<p>
We previously developed multiresponse QTL models which learned the target
genes of causal variants, allowing the true target gene to explain away
nearby correlated genes (<a href="https://www.biorxiv.org/content/early/2017/11/14/219428">Park et al 2017</a>).
</p></li>
</ol>
</div>
</div>
</div>
<div id="postamble" class="status">
<p class="author">Author: Abhishek Sarkar</p>
<p class="date">Created: 2017-12-19 Tue 11:36</p>
<p class="validation"><a href="http://validator.w3.org/check?uri=referer">Validate</a></p>
</div>
</body>
</html>
