<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />


<meta name="author" content="John Blischak" />

<meta name="date" content="2017-08-23" />

<title>Identifying single cells</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-4.5.0/css/font-awesome.min.css" rel="stylesheet" />

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>


</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
button.code-folding-btn:focus {
  outline: none;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}

.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>


<div class="container-fluid main-container">

<!-- tabsets -->
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->




<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
  padding-left: 25px;
  text-indent: 0;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>

<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">singlecell-qtl</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Overview</a>
</li>
<li>
  <a href="data-collection.html">Data</a>
</li>
<li>
  <a href="analysis.html">Analysis</a>
</li>
<li>
  <a href="contributing.html">Contributing</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/jdblischak/singlecell-qtl">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Identifying single cells</h1>
<h4 class="author"><em>John Blischak</em></h4>
<h4 class="date"><em>2017-08-23</em></h4>

</div>


<!-- The file analysis/chunks.R contains chunks that define default settings
shared across the workflowr files. -->
<!-- Update knitr chunk options -->
<!-- Insert the date the file was last updated -->
<p><strong>Last updated:</strong> 2018-06-01</p>
<!-- Insert the code version (Git commit SHA1) if Git repository exists and R
 package git2r is installed -->
<p><strong>Code version:</strong> 6b81757</p>
<p><a href="https://catalog.coriell.org/0/Sections/Collections/NHGRI/1000IndSamples.aspx?PgId=665&amp;did=MGENOME27">Link to Yoruba samples on Coriell</a></p>
<div id="setup" class="section level2">
<h2>Setup</h2>
<pre class="r"><code>library(&quot;cowplot&quot;)
library(&quot;dplyr&quot;)
library(&quot;ggplot2&quot;)
library(&quot;stringr&quot;)
library(&quot;tidyr&quot;)
theme_set(theme_cowplot())
library(&quot;Biobase&quot;)</code></pre>
<p>Import data.</p>
<pre class="r"><code>eset &lt;- readRDS(&quot;../data/eset.rds&quot;)
(chips &lt;- unique(eset$experiment))</code></pre>
<pre><code> [1] &quot;02192018&quot; &quot;02202018&quot; &quot;02212018&quot; &quot;02222018&quot; &quot;02242018&quot; &quot;02262018&quot;
 [7] &quot;02272018&quot; &quot;02282018&quot; &quot;03012018&quot; &quot;03052018&quot; &quot;03062018&quot; &quot;03072018&quot;
[13] &quot;03162017&quot; &quot;03172017&quot; &quot;03232017&quot; &quot;03302017&quot; &quot;03312017&quot; &quot;04052017&quot;
[19] &quot;04072017&quot; &quot;04132017&quot; &quot;04142017&quot; &quot;04202017&quot; &quot;08102017&quot; &quot;08112017&quot;
[25] &quot;08142017&quot; &quot;08152017&quot; &quot;08162017&quot; &quot;08182017&quot; &quot;08212017&quot; &quot;08222017&quot;
[31] &quot;08232017&quot; &quot;08242017&quot; &quot;08282017&quot; &quot;08292017&quot; &quot;08302017&quot; &quot;08312017&quot;
[37] &quot;09252017&quot; &quot;09262017&quot; &quot;09272017&quot; &quot;10022017&quot; &quot;10042017&quot; &quot;10052017&quot;
[43] &quot;10062017&quot; &quot;10092017&quot; &quot;10102017&quot; &quot;10112017&quot; &quot;10122017&quot; &quot;10132017&quot;
[49] &quot;10162017&quot; &quot;10172017&quot; &quot;10302017&quot; &quot;11022017&quot; &quot;11032017&quot; &quot;11062017&quot;
[55] &quot;11072017&quot; &quot;11082017&quot; &quot;11092017&quot; &quot;11102017&quot; &quot;11132017&quot; &quot;11142017&quot;
[61] &quot;11152017&quot; &quot;11162017&quot; &quot;11172017&quot; &quot;11202017&quot; &quot;11212017&quot; &quot;11272017&quot;
[67] &quot;11282017&quot; &quot;11292017&quot; &quot;11302017&quot; &quot;12012017&quot; &quot;12042017&quot; &quot;12052017&quot;
[73] &quot;12062017&quot; &quot;12072017&quot; &quot;12082017&quot; &quot;12112017&quot; &quot;12122017&quot; &quot;12132017&quot;
[79] &quot;12142017&quot;</code></pre>
<pre class="r"><code>eset</code></pre>
<pre><code>ExpressionSet (storageMode: lockedEnvironment)
assayData: 54792 features, 7584 samples 
  element names: exprs 
protocolData: none
phenoData
  sampleNames: 02192018-A01 02192018-A02 ... 12142017-H12 (7584
    total)
  varLabels: experiment well ... valid_id (40 total)
  varMetadata: labelDescription
featureData
  featureNames: ENSG00000000003 ENSG00000000005 ... WBGene00235374
    (54792 total)
  fvarLabels: chr start ... source (6 total)
  fvarMetadata: labelDescription
experimentData: use &#39;experimentData(object)&#39;
Annotation:  </code></pre>
</div>
<div id="how-many-of-the-identities-are-plausible" class="section level2">
<h2>How many of the identities are plausible?</h2>
<pre class="r"><code>table(eset$valid_id)</code></pre>
<pre><code>
FALSE  TRUE 
  130  7454 </code></pre>
<pre class="r"><code>mean(eset$valid_id)</code></pre>
<pre><code>[1] 0.9828586</code></pre>
</div>
<div id="plot-per-c1-chip" class="section level2">
<h2>Plot per C1 chip</h2>
<pre class="r"><code>anno &lt;- pData(eset)
df_e &lt;- anno %&gt;%
  group_by(experiment) %&gt;%
  summarize(i1 = unique(individual.1),
            i2 = unique(individual.2),
            i3 = unique(individual.3),
            i4 = unique(individual.4))
op &lt;- par(cex = 2, las = 2, mfrow = c(ceiling(length(chips) / 5), 5))
for (i in 1:nrow(df_e)) {
  e &lt;- df_e$experiment[i]
  n1 &lt;- sum(anno$chip_id[anno$experiment == e] == df_e$i1[i])
  n2 &lt;- sum(anno$chip_id[anno$experiment == e] == df_e$i2[i])
  n3 &lt;- sum(anno$chip_id[anno$experiment == e] == df_e$i3[i])
  n4 &lt;- sum(anno$chip_id[anno$experiment == e] == df_e$i4[i])
  n_other &lt;- 96 - sum(anno$chip_id[anno$experiment == e] %in%
                      as.character(df_e[i, ]))
  barplot(c(n1, n2, n3, n4, n_other), main = sprintf(&quot;C1 chip %s&quot;, e),
          names.arg = c(df_e$i1[i], df_e$i2[i], df_e$i3[i],df_e$i4[i], &quot;Other&quot;),
          ylab = &quot;Number of single cells&quot;)
}
par(op)</code></pre>
<p><img src="figure/identity.Rmd/ind-per-chip-1.png" width="768" style="display: block; margin: auto;" /></p>
<p>Colored by individual.</p>
<pre class="r"><code>anno$valid_id_display &lt;- factor(anno$valid_id, levels = c(TRUE, FALSE),
                                labels = c(&quot;Valid&quot;, &quot;Invalid&quot;))
ggplot(anno, aes(x = valid_id_display, fill = chip_id)) +
  geom_bar() +
  facet_wrap(~experiment) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1),
        legend.position = &quot;bottom&quot;) +
  labs(x = &quot;&quot;, y = &quot;Number of single cells&quot;,
       title = &quot;Valid vs. invalid assignments colored by individual&quot;)</code></pre>
<p><img src="figure/identity.Rmd/ind-per-chip-color-1.png" width="768" style="display: block; margin: auto;" /></p>
</div>
<div id="expected-versus-observed-cells" class="section level2">
<h2>Expected versus observed cells</h2>
<p>Each individual is expected to get 25% of the wells (i.e. 24) for each C1 chip it is on.</p>
<pre class="r"><code>expected &lt;- df_e %&gt;% select(starts_with(&quot;i&quot;)) %&gt;% unlist %&gt;% table * 96 / 4
expected &lt;- as.data.frame(expected)
colnames(expected) &lt;- c(&quot;ind&quot;, &quot;count&quot;)
observed &lt;- table(anno$chip_id)
observed &lt;- as.data.frame(observed)
colnames(observed) &lt;- c(&quot;ind&quot;, &quot;count&quot;)
df_obs &lt;- merge(expected, observed, by = &quot;ind&quot;, suffixes = c(&quot;.exp&quot;, &quot;.obs&quot;),
                all.x = TRUE, sort = TRUE)
# Some individuals were assigned zero cells
df_obs$count.obs[is.na(df_obs$count.obs)] &lt;- 0
df_obs</code></pre>
<pre><code>       ind count.exp count.obs
1  NA18489       120       110
2  NA18498        72        21
3  NA18499       216       166
4  NA18501       216       280
5  NA18502       144       227
6  NA18505       216       165
7  NA18507       360       419
8  NA18508       144       210
9  NA18511       120       149
10 NA18516        96       123
11 NA18517       192       137
12 NA18519       168       124
13 NA18520       168       169
14 NA18522       144       141
15 NA18852       144       102
16 NA18853        96        90
17 NA18855       120       108
18 NA18856        72        93
19 NA18858       192       188
20 NA18859       168       168
21 NA18862        96       137
22 NA18870        96        96
23 NA18907       120       115
24 NA18912        96       121
25 NA18913       216       191
26 NA19092        48        54
27 NA19093       120       123
28 NA19098       168       145
29 NA19099       120       123
30 NA19101        96       113
31 NA19102        96        99
32 NA19108        96       125
33 NA19114       120       100
34 NA19116       120       161
35 NA19127        96        85
36 NA19128        96       144
37 NA19130       144       133
38 NA19140       120       108
39 NA19143       144       112
40 NA19144       144       180
41 NA19152       192       155
42 NA19153        96       136
43 NA19159       168       187
44 NA19160       120       124
45 NA19190       120       115
46 NA19193       192       168
47 NA19203       168       156
48 NA19204       192       136
49 NA19206       120        85
50 NA19207       192       190
51 NA19209        96       102
52 NA19210       144       122
53 NA19225        96       113
54 NA19257       168       101</code></pre>
<pre class="r"><code># The max observed number of cells for any indvidual
max_cells &lt;- max(df_obs$count.exp, df_obs$count.obs)
ggplot(df_obs, aes(x = count.exp, y = count.obs)) +
  geom_text(aes(label = ind)) +
  geom_abline(intercept = 0, slope = 1, color = &quot;red&quot;, linetype = &quot;dashed&quot;) +
  xlim(0, max_cells) + ylim(0, max_cells) +
  labs(x = &quot;Expected number of cells per individual&quot;,
       y = &quot;Observed number of cells per individual&quot;,
       title = &quot;Observed versus expected number of cells&quot;)</code></pre>
<p><img src="figure/identity.Rmd/obs-v-exp-1.png" width="768" style="display: block; margin: auto;" /></p>
</div>
<div id="unexpectedly-high-counts" class="section level2">
<h2>Unexpectedly high counts</h2>
<pre class="r"><code>observed %&gt;% filter(count &gt; 10, !(ind %in% expected$ind)) %&gt;%
  arrange(desc(count))</code></pre>
<pre><code>      ind count
1 NA19119    17</code></pre>
</div>
<div id="top-individuals-per-c1-chip" class="section level2">
<h2>Top individuals per C1 chip</h2>
<pre class="r"><code># Return all individuals with more than cutoff cells on the C1 chip
table_to_string &lt;- function(x, cutoff = 5) {
  tab &lt;- sort(table(x), decreasing = TRUE)
  tab &lt;- tab[tab &gt; cutoff]
  string &lt;- mapply(paste, names(tab), tab, MoreArgs = list(sep = &quot; - &quot;))
  return(paste(string, collapse = &quot;; &quot;))
}
top &lt;- anno %&gt;%
  group_by(experiment, individual.1, individual.2, individual.3, individual.4) %&gt;%
  summarize(Observed = table_to_string(chip_id))
knitr::kable(top, col.names = c(&quot;C1 chip&quot;, &quot;Expected 1&quot;, &quot;Expected 2&quot;,
                               &quot;Expected 3&quot;, &quot;Expected 4&quot;, &quot;Observed (individual - number of cells)&quot;))</code></pre>
<table>
<thead>
<tr class="header">
<th align="left">C1 chip</th>
<th align="left">Expected 1</th>
<th align="left">Expected 2</th>
<th align="left">Expected 3</th>
<th align="left">Expected 4</th>
<th align="left">Observed (individual - number of cells)</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">02192018</td>
<td align="left">NA18517</td>
<td align="left">NA18913</td>
<td align="left">NA19193</td>
<td align="left">NA19210</td>
<td align="left">NA19193 - 32; NA18517 - 25; NA19210 - 20; NA18913 - 19</td>
</tr>
<tr class="even">
<td align="left">02202018</td>
<td align="left">NA18913</td>
<td align="left">NA19159</td>
<td align="left">NA19203</td>
<td align="left">NA19204</td>
<td align="left">NA19159 - 52; NA18913 - 16; NA19204 - 15; NA19203 - 13</td>
</tr>
<tr class="odd">
<td align="left">02212018</td>
<td align="left">NA18505</td>
<td align="left">NA18913</td>
<td align="left">NA19098</td>
<td align="left">NA19143</td>
<td align="left">NA18505 - 35; NA18913 - 24; NA19098 - 21; NA19143 - 15</td>
</tr>
<tr class="even">
<td align="left">02222018</td>
<td align="left">NA19130</td>
<td align="left">NA19143</td>
<td align="left">NA19159</td>
<td align="left">NA19193</td>
<td align="left">NA19193 - 29; NA19130 - 24; NA19159 - 24; NA19143 - 16</td>
</tr>
<tr class="odd">
<td align="left">02242018</td>
<td align="left">NA18505</td>
<td align="left">NA19130</td>
<td align="left">NA19204</td>
<td align="left">NA19210</td>
<td align="left">NA19130 - 35; NA19210 - 29; NA19204 - 23; NA18505 - 8</td>
</tr>
<tr class="even">
<td align="left">02262018</td>
<td align="left">NA18517</td>
<td align="left">NA18913</td>
<td align="left">NA19098</td>
<td align="left">NA19203</td>
<td align="left">NA19203 - 43; NA19098 - 24; NA18913 - 19; NA18517 - 9</td>
</tr>
<tr class="odd">
<td align="left">02272018</td>
<td align="left">NA18505</td>
<td align="left">NA18859</td>
<td align="left">NA19193</td>
<td align="left">NA19204</td>
<td align="left">NA18859 - 32; NA19204 - 28; NA18505 - 21; NA19193 - 15</td>
</tr>
<tr class="even">
<td align="left">02282018</td>
<td align="left">NA18519</td>
<td align="left">NA18913</td>
<td align="left">NA19098</td>
<td align="left">NA19204</td>
<td align="left">NA18913 - 41; NA19098 - 23; NA18519 - 17; NA19204 - 15</td>
</tr>
<tr class="odd">
<td align="left">03012018</td>
<td align="left">NA18505</td>
<td align="left">NA19159</td>
<td align="left">NA19193</td>
<td align="left">NA19203</td>
<td align="left">NA19159 - 32; NA19203 - 29; NA19193 - 22; NA18505 - 13</td>
</tr>
<tr class="even">
<td align="left">03052018</td>
<td align="left">NA18517</td>
<td align="left">NA18859</td>
<td align="left">NA19159</td>
<td align="left">NA19203</td>
<td align="left">NA18859 - 31; NA19203 - 26; NA18517 - 24; NA19159 - 15</td>
</tr>
<tr class="odd">
<td align="left">03062018</td>
<td align="left">NA18505</td>
<td align="left">NA18519</td>
<td align="left">NA18522</td>
<td align="left">NA19159</td>
<td align="left">NA18522 - 36; NA18505 - 24; NA19159 - 23; NA18519 - 12</td>
</tr>
<tr class="even">
<td align="left">03072018</td>
<td align="left">NA18519</td>
<td align="left">NA18522</td>
<td align="left">NA18859</td>
<td align="left">NA19203</td>
<td align="left">NA18859 - 30; NA18522 - 24; NA18519 - 23; NA19203 - 18</td>
</tr>
<tr class="odd">
<td align="left">03162017</td>
<td align="left">NA19098</td>
<td align="left">NA18852</td>
<td align="left">NA18505</td>
<td align="left">NA18520</td>
<td align="left">NA18520 - 34; NA19098 - 24; NA18505 - 22; NA18852 - 14</td>
</tr>
<tr class="even">
<td align="left">03172017</td>
<td align="left">NA18502</td>
<td align="left">NA18505</td>
<td align="left">NA18852</td>
<td align="left">NA19098</td>
<td align="left">NA19098 - 29; NA18505 - 26; NA18502 - 22; NA18852 - 18</td>
</tr>
<tr class="odd">
<td align="left">03232017</td>
<td align="left">NA18520</td>
<td align="left">NA18502</td>
<td align="left">NA18505</td>
<td align="left">NA18852</td>
<td align="left">NA18502 - 36; NA18520 - 29; NA18852 - 18; NA18505 - 13</td>
</tr>
<tr class="even">
<td align="left">03302017</td>
<td align="left">NA18502</td>
<td align="left">NA18520</td>
<td align="left">NA18852</td>
<td align="left">NA19098</td>
<td align="left">NA18502 - 49; NA18852 - 21; NA18520 - 13; NA19098 - 13</td>
</tr>
<tr class="odd">
<td align="left">03312017</td>
<td align="left">NA18502</td>
<td align="left">NA18520</td>
<td align="left">NA19092</td>
<td align="left">NA18856</td>
<td align="left">NA18856 - 28; NA19092 - 25; NA18502 - 22; NA18520 - 20</td>
</tr>
<tr class="even">
<td align="left">04052017</td>
<td align="left">NA18502</td>
<td align="left">NA18505</td>
<td align="left">NA18520</td>
<td align="left">NA18852</td>
<td align="left">NA18502 - 66; NA18852 - 14; NA18520 - 13</td>
</tr>
<tr class="odd">
<td align="left">04072017</td>
<td align="left">NA18502</td>
<td align="left">NA18856</td>
<td align="left">NA19098</td>
<td align="left">NA19092</td>
<td align="left">NA18502 - 32; NA19092 - 29; NA18856 - 27; NA19098 - 7</td>
</tr>
<tr class="even">
<td align="left">04132017</td>
<td align="left">NA18498</td>
<td align="left">NA18520</td>
<td align="left">NA18852</td>
<td align="left">NA19203</td>
<td align="left">NA18862 - 30; NA18520 - 20; NA19203 - 18; NA18852 - 16; NA18498 - 7</td>
</tr>
<tr class="odd">
<td align="left">04142017</td>
<td align="left">NA18498</td>
<td align="left">NA18507</td>
<td align="left">NA18520</td>
<td align="left">NA19203</td>
<td align="left">NA18520 - 40; NA18507 - 23; NA18862 - 14; NA19203 - 9; NA18498 - 7</td>
</tr>
<tr class="even">
<td align="left">04202017</td>
<td align="left">NA18498</td>
<td align="left">NA18508</td>
<td align="left">NA18856</td>
<td align="left">NA19190</td>
<td align="left">NA18856 - 36; NA18508 - 29; NA19190 - 16; NA18862 - 8; NA18498 - 6</td>
</tr>
<tr class="odd">
<td align="left">08102017</td>
<td align="left">NA18501</td>
<td align="left">NA18507</td>
<td align="left">NA18499</td>
<td align="left">NA19257</td>
<td align="left">NA18501 - 30; NA18507 - 28; NA18499 - 16; NA19257 - 16</td>
</tr>
<tr class="even">
<td align="left">08112017</td>
<td align="left">NA18853</td>
<td align="left">NA19210</td>
<td align="left">NA18508</td>
<td align="left">NA19257</td>
<td align="left">NA18508 - 36; NA18853 - 26; NA19210 - 19; NA19257 - 13</td>
</tr>
<tr class="odd">
<td align="left">08142017</td>
<td align="left">NA18507</td>
<td align="left">NA18508</td>
<td align="left">NA18499</td>
<td align="left">NA19257</td>
<td align="left">NA18508 - 46; NA18507 - 23; NA19257 - 14; NA18499 - 11</td>
</tr>
<tr class="even">
<td align="left">08152017</td>
<td align="left">NA18862</td>
<td align="left">NA19159</td>
<td align="left">NA18507</td>
<td align="left">NA19257</td>
<td align="left">NA18507 - 39; NA19159 - 28; NA18862 - 18; NA19257 - 9</td>
</tr>
<tr class="odd">
<td align="left">08162017</td>
<td align="left">NA19159</td>
<td align="left">NA19190</td>
<td align="left">NA18508</td>
<td align="left">NA18499</td>
<td align="left">NA18508 - 43; NA18499 - 30; NA19159 - 13; NA19190 - 10</td>
</tr>
<tr class="even">
<td align="left">08182017</td>
<td align="left">NA18853</td>
<td align="left">NA18870</td>
<td align="left">NA18507</td>
<td align="left">NA18499</td>
<td align="left">NA18507 - 37; NA18499 - 21; NA18870 - 21; NA18853 - 15</td>
</tr>
<tr class="odd">
<td align="left">08212017</td>
<td align="left">NA19128</td>
<td align="left">NA18507</td>
<td align="left">NA18508</td>
<td align="left">NA18507</td>
<td align="left">NA19128 - 38; NA18507 - 34; NA18508 - 23</td>
</tr>
<tr class="even">
<td align="left">08222017</td>
<td align="left">NA18870</td>
<td align="left">NA19128</td>
<td align="left">NA19190</td>
<td align="left">NA19257</td>
<td align="left">NA19128 - 29; NA19190 - 28; NA18870 - 24; NA19257 - 13</td>
</tr>
<tr class="odd">
<td align="left">08232017</td>
<td align="left">NA18862</td>
<td align="left">NA19128</td>
<td align="left">NA19210</td>
<td align="left">NA18499</td>
<td align="left">NA19128 - 57; NA18862 - 14; NA18499 - 13; NA19210 - 11</td>
</tr>
<tr class="even">
<td align="left">08242017</td>
<td align="left">NA18501</td>
<td align="left">NA18870</td>
<td align="left">NA18862</td>
<td align="left">NA18508</td>
<td align="left">NA18508 - 33; NA18862 - 29; NA18501 - 18; NA18870 - 16</td>
</tr>
<tr class="odd">
<td align="left">08282017</td>
<td align="left">NA18853</td>
<td align="left">NA18862</td>
<td align="left">NA18507</td>
<td align="left">NA19190</td>
<td align="left">NA19190 - 32; NA18862 - 24; NA18507 - 22; NA18853 - 18</td>
</tr>
<tr class="even">
<td align="left">08292017</td>
<td align="left">NA18501</td>
<td align="left">NA18507</td>
<td align="left">NA19190</td>
<td align="left">NA19210</td>
<td align="left">NA19190 - 29; NA18507 - 26; NA19210 - 22; NA18501 - 18</td>
</tr>
<tr class="odd">
<td align="left">08302017</td>
<td align="left">NA18870</td>
<td align="left">NA18507</td>
<td align="left">NA19257</td>
<td align="left">NA19210</td>
<td align="left">NA19257 - 29; NA18870 - 21; NA19210 - 21; NA18507 - 20</td>
</tr>
<tr class="even">
<td align="left">08312017</td>
<td align="left">NA18501</td>
<td align="left">NA18853</td>
<td align="left">NA19128</td>
<td align="left">NA19257</td>
<td align="left">NA18501 - 40; NA18853 - 31; NA19128 - 17; NA19257 - 7</td>
</tr>
<tr class="odd">
<td align="left">09252017</td>
<td align="left">NA18858</td>
<td align="left">NA18858</td>
<td align="left">NA19114</td>
<td align="left">NA18499</td>
<td align="left">NA18858 - 52; NA18499 - 25; NA19114 - 19</td>
</tr>
<tr class="even">
<td align="left">09262017</td>
<td align="left">NA18522</td>
<td align="left">NA19114</td>
<td align="left">NA19153</td>
<td align="left">NA19207</td>
<td align="left">NA19207 - 38; NA18522 - 22; NA19153 - 22; NA19114 - 14</td>
</tr>
<tr class="odd">
<td align="left">09272017</td>
<td align="left">NA18489</td>
<td align="left">NA19093</td>
<td align="left">NA19114</td>
<td align="left">NA19116</td>
<td align="left">NA19116 - 39; NA19093 - 32; NA19114 - 17; NA18489 - 8</td>
</tr>
<tr class="even">
<td align="left">10022017</td>
<td align="left">NA18522</td>
<td align="left">NA18858</td>
<td align="left">NA19116</td>
<td align="left">NA19193</td>
<td align="left">NA19116 - 32; NA18522 - 23; NA19193 - 21; NA18858 - 20</td>
</tr>
<tr class="odd">
<td align="left">10042017</td>
<td align="left">NA18489</td>
<td align="left">NA18858</td>
<td align="left">NA19153</td>
<td align="left">NA19193</td>
<td align="left">NA18858 - 40; NA18489 - 26; NA19153 - 23; NA19193 - 7</td>
</tr>
<tr class="even">
<td align="left">10052017</td>
<td align="left">NA19093</td>
<td align="left">NA19207</td>
<td align="left">NA19193</td>
<td align="left">NA18499</td>
<td align="left">NA19093 - 42; NA19207 - 28; NA19193 - 18; NA18499 - 8</td>
</tr>
<tr class="odd">
<td align="left">10062017</td>
<td align="left">NA18489</td>
<td align="left">NA19093</td>
<td align="left">NA19114</td>
<td align="left">NA19116</td>
<td align="left">NA19114 - 34; NA18489 - 26; NA19116 - 26; NA19093 - 10</td>
</tr>
<tr class="even">
<td align="left">10092017</td>
<td align="left">NA18519</td>
<td align="left">NA19101</td>
<td align="left">NA19114</td>
<td align="left">NA19193</td>
<td align="left">NA19101 - 42; NA19193 - 24; NA19114 - 16; NA18519 - 13</td>
</tr>
<tr class="odd">
<td align="left">10102017</td>
<td align="left">NA19101</td>
<td align="left">NA19116</td>
<td align="left">NA19153</td>
<td align="left">NA18499</td>
<td align="left">NA19153 - 30; NA19116 - 27; NA19101 - 26; NA18499 - 13</td>
</tr>
<tr class="even">
<td align="left">10112017</td>
<td align="left">NA18519</td>
<td align="left">NA18858</td>
<td align="left">NA19116</td>
<td align="left">NA19207</td>
<td align="left">NA19116 - 37; NA18519 - 22; NA19207 - 20; NA18858 - 17</td>
</tr>
<tr class="odd">
<td align="left">10122017</td>
<td align="left">NA18489</td>
<td align="left">NA18519</td>
<td align="left">NA18522</td>
<td align="left">NA18499</td>
<td align="left">NA18499 - 29; NA18519 - 27; NA18489 - 20; NA18522 - 20</td>
</tr>
<tr class="even">
<td align="left">10132017</td>
<td align="left">NA18489</td>
<td align="left">NA18858</td>
<td align="left">NA19101</td>
<td align="left">NA19207</td>
<td align="left">NA18489 - 30; NA19207 - 27; NA18858 - 20; NA19101 - 19</td>
</tr>
<tr class="odd">
<td align="left">10162017</td>
<td align="left">NA18519</td>
<td align="left">NA18858</td>
<td align="left">NA19093</td>
<td align="left">NA19153</td>
<td align="left">NA19153 - 60; NA19093 - 15; NA18858 - 11; NA18519 - 10</td>
</tr>
<tr class="even">
<td align="left">10172017</td>
<td align="left">NA18522</td>
<td align="left">NA18858</td>
<td align="left">NA19093</td>
<td align="left">NA19101</td>
<td align="left">NA18858 - 28; NA19101 - 26; NA19093 - 24; NA18522 - 14</td>
</tr>
<tr class="odd">
<td align="left">10302017</td>
<td align="left">NA18855</td>
<td align="left">NA19099</td>
<td align="left">NA19144</td>
<td align="left">NA19160</td>
<td align="left">NA19160 - 43; NA18855 - 17; NA18870 - 14; NA19099 - 11; NA19144 - 11</td>
</tr>
<tr class="even">
<td align="left">11022017</td>
<td align="left">NA19152</td>
<td align="left">NA19160</td>
<td align="left">NA19206</td>
<td align="left">NA19207</td>
<td align="left">NA19160 - 32; NA19206 - 25; NA19152 - 23; NA19207 - 16</td>
</tr>
<tr class="odd">
<td align="left">11032017</td>
<td align="left">NA18507</td>
<td align="left">NA18516</td>
<td align="left">NA18855</td>
<td align="left">NA19207</td>
<td align="left">NA18516 - 35; NA18855 - 29; NA18507 - 15; NA19207 - 15</td>
</tr>
<tr class="even">
<td align="left">11062017</td>
<td align="left">NA18507</td>
<td align="left">NA18913</td>
<td align="left">NA19099</td>
<td align="left">NA19152</td>
<td align="left">NA19099 - 39; NA18507 - 24; NA18913 - 17; NA19152 - 16</td>
</tr>
<tr class="odd">
<td align="left">11072017</td>
<td align="left">NA18855</td>
<td align="left">NA19099</td>
<td align="left">NA19144</td>
<td align="left">NA19160</td>
<td align="left">NA19144 - 58; NA19099 - 18; NA18855 - 10; NA19160 - 10</td>
</tr>
<tr class="even">
<td align="left">11082017</td>
<td align="left">NA18507</td>
<td align="left">NA19160</td>
<td align="left">NA19152</td>
<td align="left">NA19209</td>
<td align="left">NA18507 - 41; NA19209 - 24; NA19152 - 19; NA19160 - 12</td>
</tr>
<tr class="odd">
<td align="left">11092017</td>
<td align="left">NA18907</td>
<td align="left">NA18516</td>
<td align="left">NA18913</td>
<td align="left">NA19160</td>
<td align="left">NA18516 - 42; NA19160 - 27; NA18907 - 17; NA18913 - 10</td>
</tr>
<tr class="even">
<td align="left">11102017</td>
<td align="left">NA18855</td>
<td align="left">NA18913</td>
<td align="left">NA19206</td>
<td align="left">NA19209</td>
<td align="left">NA19209 - 32; NA18913 - 25; NA18855 - 22; NA19206 - 17</td>
</tr>
<tr class="odd">
<td align="left">11132017</td>
<td align="left">NA18907</td>
<td align="left">NA19099</td>
<td align="left">NA19207</td>
<td align="left">NA19209</td>
<td align="left">NA18907 - 32; NA19207 - 31; NA19209 - 18; NA19099 - 15</td>
</tr>
<tr class="even">
<td align="left">11142017</td>
<td align="left">NA18516</td>
<td align="left">NA19144</td>
<td align="left">NA19152</td>
<td align="left">NA19209</td>
<td align="left">NA19144 - 36; NA19209 - 27; NA18516 - 22; NA19152 - 11</td>
</tr>
<tr class="odd">
<td align="left">11152017</td>
<td align="left">NA18507</td>
<td align="left">NA18907</td>
<td align="left">NA19206</td>
<td align="left">NA19144</td>
<td align="left">NA18507 - 46; NA19144 - 21; NA18907 - 16; NA19206 - 13</td>
</tr>
<tr class="even">
<td align="left">11162017</td>
<td align="left">NA18516</td>
<td align="left">NA19099</td>
<td align="left">NA19152</td>
<td align="left">NA19206</td>
<td align="left">NA19099 - 39; NA18516 - 23; NA19152 - 18; NA19206 - 16</td>
</tr>
<tr class="odd">
<td align="left">11172017</td>
<td align="left">NA18913</td>
<td align="left">NA19144</td>
<td align="left">NA19152</td>
<td align="left">NA19207</td>
<td align="left">NA19144 - 32; NA19152 - 31; NA18913 - 17; NA19207 - 15</td>
</tr>
<tr class="even">
<td align="left">11202017</td>
<td align="left">NA18907</td>
<td align="left">NA18855</td>
<td align="left">NA19152</td>
<td align="left">NA19152</td>
<td align="left">NA19152 - 37; NA18855 - 30; NA18907 - 29</td>
</tr>
<tr class="odd">
<td align="left">11212017</td>
<td align="left">NA18507</td>
<td align="left">NA18907</td>
<td align="left">NA19144</td>
<td align="left">NA19206</td>
<td align="left">NA18507 - 40; NA18907 - 21; NA19144 - 21; NA19206 - 14</td>
</tr>
<tr class="even">
<td align="left">11272017</td>
<td align="left">NA18511</td>
<td align="left">NA18859</td>
<td align="left">NA19127</td>
<td align="left">NA19130</td>
<td align="left">NA18511 - 31; NA18859 - 31; NA19130 - 20; NA19127 - 14</td>
</tr>
<tr class="odd">
<td align="left">11282017</td>
<td align="left">NA18511</td>
<td align="left">NA18912</td>
<td align="left">NA19102</td>
<td align="left">NA19225</td>
<td align="left">NA18912 - 34; NA19225 - 29; NA19102 - 21; NA18511 - 12</td>
</tr>
<tr class="even">
<td align="left">11292017</td>
<td align="left">NA18912</td>
<td align="left">NA19130</td>
<td align="left">NA19140</td>
<td align="left">NA19204</td>
<td align="left">NA18912 - 49; NA19130 - 20; NA19140 - 19; NA19204 - 7</td>
</tr>
<tr class="odd">
<td align="left">11302017</td>
<td align="left">NA18501</td>
<td align="left">NA18859</td>
<td align="left">NA19102</td>
<td align="left">NA19204</td>
<td align="left">NA18501 - 56; NA19102 - 16; NA18859 - 15; NA19204 - 9</td>
</tr>
<tr class="even">
<td align="left">12012017</td>
<td align="left">NA18501</td>
<td align="left">NA18511</td>
<td align="left">NA18517</td>
<td align="left">NA19140</td>
<td align="left">NA18501 - 41; NA18511 - 25; NA19140 - 17; NA18517 - 13</td>
</tr>
<tr class="odd">
<td align="left">12042017</td>
<td align="left">NA18517</td>
<td align="left">NA19127</td>
<td align="left">NA19204</td>
<td align="left">NA19225</td>
<td align="left">NA19225 - 38; NA19204 - 23; NA18517 - 18; NA19127 - 16</td>
</tr>
<tr class="even">
<td align="left">12052017</td>
<td align="left">NA18501</td>
<td align="left">NA18511</td>
<td align="left">NA18517</td>
<td align="left">NA19140</td>
<td align="left">NA18511 - 40; NA18501 - 28; NA19140 - 16; NA18517 - 12</td>
</tr>
<tr class="odd">
<td align="left">12062017</td>
<td align="left">NA19102</td>
<td align="left">NA19108</td>
<td align="left">NA19127</td>
<td align="left">NA19140</td>
<td align="left">NA19127 - 31; NA19108 - 23; NA19102 - 22; NA19140 - 19</td>
</tr>
<tr class="even">
<td align="left">12072017</td>
<td align="left">NA18517</td>
<td align="left">NA18859</td>
<td align="left">NA18912</td>
<td align="left">NA19108</td>
<td align="left">NA19108 - 47; NA18859 - 18; NA18517 - 17; NA18912 - 11</td>
</tr>
<tr class="odd">
<td align="left">12082017</td>
<td align="left">NA18501</td>
<td align="left">NA19108</td>
<td align="left">NA19130</td>
<td align="left">NA19225</td>
<td align="left">NA18501 - 28; NA19108 - 25; NA19225 - 25; NA19130 - 17</td>
</tr>
<tr class="even">
<td align="left">12112017</td>
<td align="left">NA18511</td>
<td align="left">NA19108</td>
<td align="left">NA19143</td>
<td align="left">NA19204</td>
<td align="left">NA18511 - 40; NA19108 - 30; NA19204 - 16; NA19143 - 10</td>
</tr>
<tr class="odd">
<td align="left">12122017</td>
<td align="left">NA18859</td>
<td align="left">NA19140</td>
<td align="left">NA19143</td>
<td align="left">NA19225</td>
<td align="left">NA19140 - 36; NA19143 - 28; NA19225 - 21; NA18859 - 11</td>
</tr>
<tr class="even">
<td align="left">12132017</td>
<td align="left">NA18501</td>
<td align="left">NA18912</td>
<td align="left">NA19127</td>
<td align="left">NA19143</td>
<td align="left">NA18912 - 27; NA19127 - 24; NA19143 - 22; NA18501 - 21</td>
</tr>
<tr class="odd">
<td align="left">12142017</td>
<td align="left">NA18517</td>
<td align="left">NA19102</td>
<td align="left">NA19130</td>
<td align="left">NA19143</td>
<td align="left">NA19102 - 40; NA19143 - 20; NA18517 - 19; NA19130 - 17</td>
</tr>
</tbody>
</table>
</div>
<div id="potential-reasons-for-failure" class="section level2">
<h2>Potential reasons for failure</h2>
<p>Sequencing depth</p>
<pre class="r"><code>reason_raw &lt;- ggplot(anno, aes(x = valid_id, y = raw)) +
  geom_boxplot() +
  scale_x_discrete(labels = c(&quot;Wrong&quot;, &quot;Correct&quot;)) +
  labs(x = &quot;Predicted individual of origin&quot;,
       y = &quot;Raw number of reads&quot;,
       title = &quot;Raw number of reads&quot;)
reason_raw</code></pre>
<p><img src="figure/identity.Rmd/reason-raw-1.png" width="768" style="display: block; margin: auto;" /></p>
<p>Percent mapped</p>
<pre class="r"><code>reason_raw %+% aes(y = mapped / raw * 100) +
  labs(y = &quot;Percent of reads mapped&quot;,
       title = &quot;Percent of reads mapped&quot;)</code></pre>
<p><img src="figure/identity.Rmd/reason-mapped-1.png" width="768" style="display: block; margin: auto;" /></p>
<p>Cell number</p>
<pre class="r"><code>barplot(with(anno, table(valid_id, cell_number)), beside = TRUE,
        col = c(&quot;red&quot;, &quot;black&quot;),
        xlab = &quot;Number of cells in well&quot;, main = &quot;Number of cells in well&quot;,
        ylab = &quot;Number of cells correct (black) versus wrong (red)&quot;)</code></pre>
<p><img src="figure/identity.Rmd/reason-cell-number-1.png" width="768" style="display: block; margin: auto;" /></p>
<p>Concentration</p>
<pre class="r"><code>reason_raw %+% aes(y = concentration) +
  labs(y = &quot;Concentration&quot;,
       title = &quot;Concentration&quot;)</code></pre>
<p><img src="figure/identity.Rmd/reason-concentation-1.png" width="768" style="display: block; margin: auto;" /></p>
<p>TRA-1-60</p>
<pre class="r"><code>barplot(with(anno, table(valid_id, tra1.60)), beside = TRUE,
        col = c(&quot;red&quot;, &quot;black&quot;),
        xlab = &quot;TRA-1-60&quot;, main = &quot;TRA-1-60&quot;,
        ylab = &quot;Number of cells correct (black) versus wrong (red)&quot;)</code></pre>
<p><img src="figure/identity.Rmd/reason-tra-1-60-1.png" width="768" style="display: block; margin: auto;" /></p>
<p>SNPs with at least one overlapping read</p>
<pre class="r"><code>reason_raw %+% aes(y = snps_w_min) +
  labs(y = &quot;SNPs with at least one overlapping read&quot;,
       title = &quot;SNPs with at least one overlapping read&quot;)</code></pre>
<p><img src="figure/identity.Rmd/reason-snps-w-min-1.png" width="768" style="display: block; margin: auto;" /></p>
<p>chipmix</p>
<pre class="r"><code>reason_raw %+% aes(y = chipmix) +
  labs(y = &quot;chipmix&quot;, title = &quot;chipmix&quot;)</code></pre>
<p><img src="figure/identity.Rmd/reason-chipmix-1.png" width="768" style="display: block; margin: auto;" /></p>
<p>freemix</p>
<pre class="r"><code>reason_raw %+% aes(y = freemix) +
  labs(y = &quot;freemix&quot;, title = &quot;freemix&quot;)</code></pre>
<p><img src="figure/identity.Rmd/reason-freemix-1.png" width="768" style="display: block; margin: auto;" /></p>
<p>chipmix versus freemix</p>
<pre class="r"><code>ggplot(anno, aes(x = freemix, y = chipmix, color = valid_id)) +
  geom_point() +
  facet_wrap(~valid_id) +
  labs(title = &quot;chipmix versus freemix&quot;)</code></pre>
<p><img src="figure/identity.Rmd/chipmix-v-freemix-1.png" width="768" style="display: block; margin: auto;" /></p>
<p>chipmix versus SNPs with overlapping reads</p>
<pre class="r"><code>ggplot(anno, aes(x = snps_w_min, y = chipmix, color = valid_id)) +
  geom_point() +
  facet_wrap(~valid_id) +
  labs(title = &quot;chipmix versus SNPs with at least one overlapping read&quot;,
       x = &quot;SNPs with at least one overlapping read&quot;)</code></pre>
<p><img src="figure/identity.Rmd/chipmix-v-snps-w-min-1.png" width="768" style="display: block; margin: auto;" /></p>
<p>chipmix versus raw number of reads</p>
<pre class="r"><code>ggplot(anno, aes(x = raw, y = chipmix, color = valid_id)) +
  geom_point() +
  facet_wrap(~valid_id) +
  labs(title = &quot;chipmix versus number of raw reads&quot;,
       x = &quot;Number of raw reads&quot;)</code></pre>
<p><img src="figure/identity.Rmd/chipmix-v-raw-1.png" width="768" style="display: block; margin: auto;" /></p>
</div>
<div id="session-information" class="section level2">
<h2>Session information</h2>
<!-- Insert the session information into the document -->
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 3.4.1 (2017-06-30)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Scientific Linux 7.4 (Nitrogen)

Matrix products: default
BLAS: /project2/gilad/jdblischak/miniconda3/envs/scqtl/lib/R/lib/libRblas.so
LAPACK: /project2/gilad/jdblischak/miniconda3/envs/scqtl/lib/R/lib/libRlapack.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] parallel  methods   stats     graphics  grDevices utils     datasets 
[8] base     

other attached packages:
[1] bindrcpp_0.2        Biobase_2.38.0      BiocGenerics_0.24.0
[4] tidyr_0.7.1         stringr_1.2.0       dplyr_0.7.4        
[7] cowplot_0.9.1       ggplot2_2.2.1      

loaded via a namespace (and not attached):
 [1] Rcpp_0.12.13     bindr_0.1        knitr_1.20       magrittr_1.5    
 [5] munsell_0.4.3    colorspace_1.3-2 R6_2.2.0         rlang_0.1.2     
 [9] highr_0.6        plyr_1.8.4       tools_3.4.1      grid_3.4.1      
[13] gtable_0.2.0     git2r_0.19.0     htmltools_0.3.6  assertthat_0.1  
[17] yaml_2.1.14      lazyeval_0.2.0   rprojroot_1.2    digest_0.6.12   
[21] tibble_1.3.3     purrr_0.2.2      glue_1.1.1       evaluate_0.10.1 
[25] rmarkdown_1.8    labeling_0.3     stringi_1.1.2    compiler_3.4.1  
[29] scales_0.5.0     backports_1.0.5  pkgconfig_2.0.1 </code></pre>
</div>

<hr>
<p>
    This <a href="http://rmarkdown.rstudio.com">R Markdown</a> site was created with <a href="https://github.com/jdblischak/workflowr">workflowr</a>
</p>
<hr>

<!-- To enable disqus, uncomment the section below and provide your disqus_shortname -->

<!-- disqus
  <div id="disqus_thread"></div>
    <script type="text/javascript">
        /* * * CONFIGURATION VARIABLES: EDIT BEFORE PASTING INTO YOUR WEBPAGE * * */
        var disqus_shortname = 'rmarkdown'; // required: replace example with your forum shortname

        /* * * DON'T EDIT BELOW THIS LINE * * */
        (function() {
            var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
            dsq.src = '//' + disqus_shortname + '.disqus.com/embed.js';
            (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
        })();
    </script>
    <noscript>Please enable JavaScript to view the <a href="http://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
    <a href="http://disqus.com" class="dsq-brlink">comments powered by <span class="logo-disqus">Disqus</span></a>
-->


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
