<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />


<meta name="author" content="John Blischak" />

<meta name="date" content="2017-08-23" />

<title>Identifying single cells</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-1.1/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-1.1/highlight.js"></script>
<link href="site_libs/font-awesome-4.5.0/css/font-awesome.min.css" rel="stylesheet" />

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs && document.readyState && document.readyState === "complete") {
   window.setTimeout(function() {
      hljs.initHighlighting();
   }, 0);
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>


</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
button.code-folding-btn:focus {
  outline: none;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}

.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>


<div class="container-fluid main-container">

<!-- tabsets -->
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->




<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
  padding-left: 25px;
  text-indent: 0;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>

<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">singlecell-qtl</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/jdblischak/singlecell-qtl">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Identifying single cells</h1>
<h4 class="author"><em>John Blischak</em></h4>
<h4 class="date"><em>2017-08-23</em></h4>

</div>


<!-- The file analysis/chunks.R contains chunks that define default settings
shared across the workflowr files. -->
<!-- Update knitr chunk options -->
<!-- Insert the date the file was last updated -->
<p><strong>Last updated:</strong> 2017-08-25</p>
<!-- Insert the code version (Git commit SHA1) if Git repository exists and R
 package git2r is installed -->
<p><strong>Code version:</strong> 06d517b</p>
<p><a href="https://catalog.coriell.org/0/Sections/Collections/NHGRI/1000IndSamples.aspx?PgId=665&amp;did=MGENOME27">Link to Yoruba samples on Coriell</a></p>
<div id="setup" class="section level2">
<h2>Setup</h2>
<pre class="r"><code>library(&quot;cowplot&quot;)
library(&quot;dplyr&quot;)
library(&quot;ggplot2&quot;)
library(&quot;readr&quot;)
library(&quot;stringr&quot;)
library(&quot;tidyr&quot;)
theme_set(theme_cowplot())
source(&quot;../code/functions.R&quot;)</code></pre>
<p>Import totals counts.</p>
<pre class="r"><code>totals &lt;- read_tsv(&quot;../data/totals.txt&quot;)
dim(totals)</code></pre>
<pre><code>[1] 959  17</code></pre>
<pre class="r"><code>totals[1:5, 1:10]</code></pre>
<pre><code># A tibble: 5 Ã— 10
             sample experiment  well    id    raw    umi mapped unmapped
              &lt;chr&gt;      &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;  &lt;int&gt;  &lt;int&gt;  &lt;int&gt;    &lt;int&gt;
1 03302017-D10_S814   03302017   D10  S814  33968  18296  16111     2185
2 03172017-A11_S395   03172017   A11  S395 919300 587356 460302   127054
3 04132017-B03_S879   04132017   B03  S879 786849 479560 393533    86027
4 03172017-H12_S480   03172017   H12  S480 829154 536921 431628   105293
5 04132017-G07_S943   04132017   G07  S943 478981 299414 236353    63061
# ... with 2 more variables: reads_ce &lt;int&gt;, reads_dm &lt;int&gt;</code></pre>
<pre class="r"><code>colnames(totals)</code></pre>
<pre><code> [1] &quot;sample&quot;     &quot;experiment&quot; &quot;well&quot;       &quot;id&quot;         &quot;raw&quot;       
 [6] &quot;umi&quot;        &quot;mapped&quot;     &quot;unmapped&quot;   &quot;reads_ce&quot;   &quot;reads_dm&quot;  
[11] &quot;reads_ercc&quot; &quot;reads_hs&quot;   &quot;molecules&quot;  &quot;mol_ce&quot;     &quot;mol_dm&quot;    
[16] &quot;mol_ercc&quot;   &quot;mol_hs&quot;    </code></pre>
<p>Import annotation.</p>
<pre class="r"><code>anno &lt;- read_tsv(&quot;../data/batch1_qc.txt&quot;,
                 col_types = cols(ERCC = col_character()))
# Cleanup ERCC column
anno$ERCC &lt;- ifelse(is.na(anno$ERCC), &quot;Not added&quot;, anno$ERCC)
anno$ERCC &lt;- factor(anno$ERCC, levels = c(&quot;Not added&quot;, &quot;1:100000&quot;, &quot;1:50000&quot;),
                    labels = c(&quot;Not added&quot;, &quot;100x dilution&quot;, &quot;50x dilution&quot;))
# Prepend &quot;NA&quot; to individual IDs
anno &lt;- anno %&gt;% mutate(individual.1 = paste0(&quot;NA&quot;, individual.1),
                        individual.2 = paste0(&quot;NA&quot;, individual.2),
                        individual.3 = paste0(&quot;NA&quot;, individual.3),
                        individual.4 = paste0(&quot;NA&quot;, individual.4))
# Change individual NA18522 to NA 18852
anno &lt;- anno %&gt;% mutate(individual.1 = ifelse(individual.1 == &quot;NA18522&quot;,
                                              &quot;NA18852&quot;, individual.1),
                        individual.2 = ifelse(individual.2 == &quot;NA18522&quot;,
                                              &quot;NA18852&quot;, individual.2),
                        individual.3 = ifelse(individual.3 == &quot;NA18522&quot;,
                                              &quot;NA18852&quot;, individual.3),
                        individual.4 = ifelse(individual.4 == &quot;NA18522&quot;,
                                              &quot;NA18852&quot;, individual.4))
dim(anno)</code></pre>
<pre><code>[1] 960  13</code></pre>
<pre class="r"><code>colnames(anno)</code></pre>
<pre><code> [1] &quot;experiment&quot;    &quot;well&quot;          &quot;cell_number&quot;   &quot;concentration&quot;
 [5] &quot;tra1.60&quot;       &quot;individual.1&quot;  &quot;individual.2&quot;  &quot;individual.3&quot; 
 [9] &quot;individual.4&quot;  &quot;fly&quot;           &quot;worm&quot;          &quot;ERCC&quot;         
[13] &quot;index&quot;        </code></pre>
<p>Merge the data sets.</p>
<pre class="r"><code>totals_anno &lt;- merge(totals, anno, by = c(&quot;experiment&quot;, &quot;well&quot;))
stopifnot(nrow(totals_anno) == nrow(totals))</code></pre>
<p>Import identity information from verifyBamID.</p>
<pre class="r"><code>verify &lt;- read_tsv(&quot;../data/verify.txt&quot;)
dim(verify)</code></pre>
<pre><code>[1] 959   9</code></pre>
<pre class="r"><code># Extract sample ID into experiment and well
verify &lt;- verify %&gt;%
  extract(col = sample, into = c(&quot;experiment&quot;, &quot;well&quot;),
          regex = &quot;([[:digit:]]+)-([[:alnum:]]+)&quot;)
head(verify)</code></pre>
<pre><code># A tibble: 6 Ã— 10
  experiment  well chip_id chipmix freemix   snps reads avg_dp min_dp
       &lt;chr&gt; &lt;chr&gt;   &lt;chr&gt;   &lt;dbl&gt;   &lt;dbl&gt;  &lt;int&gt; &lt;int&gt;  &lt;dbl&gt;  &lt;int&gt;
1   03162017   A01 NA19098 0.29581 0.16808 311848   718   0.00      1
2   03162017   A02 NA18505 0.13881 0.05019 311848  3274   0.01      1
3   03162017   A03 NA18856 0.70950 0.00000 311848   290   0.00      1
4   03162017   A04 NA19098 0.23550 0.24901 311848   789   0.00      1
5   03162017   A05 NA18520 0.59094 0.29020 311848   536   0.00      1
6   03162017   A06 NA18520 0.18003 0.12005 311848  2984   0.01      1
# ... with 1 more variables: snps_w_min &lt;int&gt;</code></pre>
<p>Merge in identity.</p>
<pre class="r"><code>verify &lt;- merge(verify, totals_anno, by = c(&quot;experiment&quot;, &quot;well&quot;))</code></pre>
</div>
<div id="how-many-of-the-identities-are-plausible" class="section level2">
<h2>How many of the identities are plausible?</h2>
<pre class="r"><code>verify &lt;- verify %&gt;% mutate(plausible = chip_id %in% c(individual.1, individual.2,
                                                       individual.3, individual.4))
table(verify$plausible)</code></pre>
<pre><code>
FALSE  TRUE 
  174   785 </code></pre>
<pre class="r"><code>mean(verify$plausible)</code></pre>
<pre><code>[1] 0.818561</code></pre>
</div>
<div id="plot-per-c1-chip" class="section level2">
<h2>Plot per C1 chip</h2>
<pre class="r"><code>df_e &lt;- anno %&gt;%
  group_by(experiment) %&gt;%
  summarize(i1 = unique(individual.1),
            i2 = unique(individual.2),
            i3 = unique(individual.3),
            i4 = unique(individual.4))
op &lt;- par(cex = 2, las = 2, mfrow = c(2, 5))
for (i in 1:nrow(df_e)) {
  e &lt;- df_e$experiment[i]
  n1 &lt;- sum(verify$chip_id[verify$experiment == e] == df_e$i1[i])
  n2 &lt;- sum(verify$chip_id[verify$experiment == e] == df_e$i2[i])
  n3 &lt;- sum(verify$chip_id[verify$experiment == e] == df_e$i3[i])
  n4 &lt;- sum(verify$chip_id[verify$experiment == e] == df_e$i4[i])
  n_other &lt;- 96 - n1 - n2 -n3 - n4
  barplot(c(n1, n2, n3, n4, n_other), main = sprintf(&quot;C1 chip %s&quot;, e),
          names.arg = c(df_e$i1[i], df_e$i2[i], df_e$i3[i],df_e$i4[i], &quot;Other&quot;),
          ylab = &quot;Number of single cells&quot;)
}</code></pre>
<p><img src="figure/identity.Rmd/ind-per-chip-1.png" width="768" style="display: block; margin: auto;" /></p>
<pre class="r"><code>par(op)</code></pre>
</div>
<div id="expected-versus-observed-cells" class="section level2">
<h2>Expected versus observed cells</h2>
<p>Each individual is expected to get 25% of the wells (i.e.Â 24) for each C1 chip it is on.</p>
<pre class="r"><code>expected &lt;- df_e %&gt;% select(starts_with(&quot;i&quot;)) %&gt;% unlist %&gt;% table * 96 / 4
expected &lt;- as.data.frame(expected)
colnames(expected) &lt;- c(&quot;ind&quot;, &quot;count&quot;)
observed &lt;- table(verify$chip_id)
observed &lt;- as.data.frame(observed)
colnames(observed) &lt;- c(&quot;ind&quot;, &quot;count&quot;)
df_obs &lt;- merge(expected, observed, by = &quot;ind&quot;, suffixes = c(&quot;.exp&quot;, &quot;.obs&quot;),
                all.x = TRUE, sort = TRUE)
# Some individuals were assigned zero cells
df_obs$count.obs[is.na(df_obs$count.obs)] &lt;- 0
df_obs</code></pre>
<pre><code>       ind count.exp count.obs
1  NA18498        72        20
2  NA18502       144       207
3  NA18505        96        62
4  NA18508        24        29
5  NA18510        24         1
6  NA18520       168       153
7  NA18852       144       104
8  NA18856        72        77
9  NA19098        96        83
10 NA19190        24        18
11 NA19201        48         0
12 NA19203        48        31</code></pre>
<pre class="r"><code># The max observed number of cells for any indvidual
max_cells &lt;- max(df_obs$count.exp, df_obs$count.obs)
ggplot(df_obs, aes(x = count.exp, y = count.obs)) +
  geom_text(aes(label = ind)) +
  geom_abline(intercept = 0, slope = 1, color = &quot;red&quot;, linetype = &quot;dashed&quot;) +
  xlim(0, max_cells) + ylim(0, max_cells) +
  labs(x = &quot;Expected number of cells per individual&quot;,
       y = &quot;Observed number of cells per individual&quot;,
       title = &quot;Observed versus expected number of cells&quot;)</code></pre>
<p><img src="figure/identity.Rmd/obs-v-exp-1.png" width="768" style="display: block; margin: auto;" /></p>
</div>
<div id="unexpectedly-high-counts" class="section level2">
<h2>Unexpectedly high counts</h2>
<pre class="r"><code>observed %&gt;% filter(count &gt; 10, !(ind %in% expected$ind)) %&gt;%
  arrange(desc(count))</code></pre>
<pre><code>      ind count
1 NA19092    55
2 NA18862    53
3 NA18507    19</code></pre>
</div>
<div id="potential-reasons-for-failure" class="section level2">
<h2>Potential reasons for failure</h2>
<p>Sequencing depth</p>
<pre class="r"><code>reason_raw &lt;- ggplot(verify, aes(x = plausible, y = raw)) +
  geom_boxplot() +
  scale_x_discrete(labels = c(&quot;Wrong&quot;, &quot;Correct&quot;)) +
  labs(x = &quot;Predicted individual of origin&quot;,
       y = &quot;Raw number of reads&quot;,
       title = &quot;Raw number of reads&quot;)
reason_raw</code></pre>
<p><img src="figure/identity.Rmd/reason-raw-1.png" width="768" style="display: block; margin: auto;" /></p>
<p>Percent mapped</p>
<pre class="r"><code>reason_raw %+% aes(y = mapped / raw * 100) +
  labs(y = &quot;Percent of reads mapped&quot;,
       title = &quot;Percent of reads mapped&quot;)</code></pre>
<p><img src="figure/identity.Rmd/reason-mapped-1.png" width="768" style="display: block; margin: auto;" /></p>
<p>Cell number</p>
<pre class="r"><code>barplot(with(verify, table(plausible, cell_number)), beside = TRUE,
        col = c(&quot;red&quot;, &quot;black&quot;),
        xlab = &quot;Number of cells in well&quot;, main = &quot;Number of cells in well&quot;,
        ylab = &quot;Number of cells correct (black) versus wrong (red)&quot;)</code></pre>
<p><img src="figure/identity.Rmd/reason-cell-number-1.png" width="768" style="display: block; margin: auto;" /></p>
<p>Concentration</p>
<pre class="r"><code>reason_raw %+% aes(y = concentration) +
  labs(y = &quot;Concentration&quot;,
       title = &quot;Concentration&quot;)</code></pre>
<pre><code>Warning: Removed 192 rows containing non-finite values (stat_boxplot).</code></pre>
<p><img src="figure/identity.Rmd/reason-concentation-1.png" width="768" style="display: block; margin: auto;" /></p>
<p>TRA-1-60</p>
<pre class="r"><code>barplot(with(verify, table(plausible, tra1.60)), beside = TRUE,
        col = c(&quot;red&quot;, &quot;black&quot;),
        xlab = &quot;TRA-1-60&quot;, main = &quot;TRA-1-60&quot;,
        ylab = &quot;Number of cells correct (black) versus wrong (red)&quot;)</code></pre>
<p><img src="figure/identity.Rmd/reason-tra-1-60-1.png" width="768" style="display: block; margin: auto;" /></p>
<p>SNPs with at least one overlapping read</p>
<pre class="r"><code>reason_raw %+% aes(y = snps_w_min) +
  labs(y = &quot;SNPs with at least one overlapping read&quot;,
       title = &quot;SNPs with at least one overlapping read&quot;)</code></pre>
<p><img src="figure/identity.Rmd/reason-snps-w-min-1.png" width="768" style="display: block; margin: auto;" /></p>
<p>chipmix</p>
<pre class="r"><code>reason_raw %+% aes(y = chipmix) +
  labs(y = &quot;chipmix&quot;, title = &quot;chipmix&quot;)</code></pre>
<p><img src="figure/identity.Rmd/reason-chipmix-1.png" width="768" style="display: block; margin: auto;" /></p>
<p>freemix</p>
<pre class="r"><code>reason_raw %+% aes(y = freemix) +
  labs(y = &quot;freemix&quot;, title = &quot;freemix&quot;)</code></pre>
<p><img src="figure/identity.Rmd/reason-freemix-1.png" width="768" style="display: block; margin: auto;" /></p>
<p>chipmix versus freemix</p>
<pre class="r"><code>ggplot(verify, aes(x = freemix, y = chipmix, color = plausible)) +
  geom_point() +
  facet_wrap(~plausible) +
  labs(title = &quot;chipmix versus freemix&quot;)</code></pre>
<p><img src="figure/identity.Rmd/chipmix-v-freemix-1.png" width="768" style="display: block; margin: auto;" /></p>
<p>chipmix versus SNPs with overlapping reads</p>
<pre class="r"><code>ggplot(verify, aes(x = snps_w_min, y = chipmix, color = plausible)) +
  geom_point() +
  facet_wrap(~plausible) +
  labs(title = &quot;chipmix versus SNPs with at least one overlapping read&quot;,
       x = &quot;SNPs with at least one overlapping read&quot;)</code></pre>
<p><img src="figure/identity.Rmd/chipmix-v-snps-w-min-1.png" width="768" style="display: block; margin: auto;" /></p>
<p>chipmix versus raw number of reads</p>
<pre class="r"><code>ggplot(verify, aes(x = raw, y = chipmix, color = plausible)) +
  geom_point() +
  facet_wrap(~plausible) +
  labs(title = &quot;chipmix versus number of raw reads&quot;,
       x = &quot;Number of raw reads&quot;)</code></pre>
<p><img src="figure/identity.Rmd/chipmix-v-raw-1.png" width="768" style="display: block; margin: auto;" /></p>
</div>
<div id="session-information" class="section level2">
<h2>Session information</h2>
<!-- Insert the session information into the document -->
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 3.3.2 (2016-10-31)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Scientific Linux 7.2 (Nitrogen)

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] tidyr_0.6.1   stringr_1.2.0 readr_0.2.2   dplyr_0.5.0   cowplot_0.7.0
[6] ggplot2_2.2.1

loaded via a namespace (and not attached):
 [1] Rcpp_0.12.10     knitr_1.16       magrittr_1.5     munsell_0.4.3   
 [5] colorspace_1.3-2 R6_2.2.0         plyr_1.8.4       tools_3.3.2     
 [9] grid_3.3.2       gtable_0.2.0     DBI_0.6-1        git2r_0.19.0    
[13] htmltools_0.3.6  yaml_2.1.13      lazyeval_0.2.0   rprojroot_1.2   
[17] digest_0.6.12    assertthat_0.1   tibble_1.2       evaluate_0.10   
[21] rmarkdown_1.6    labeling_0.3     stringi_1.1.5    scales_0.4.1    
[25] backports_1.0.5 </code></pre>
</div>

<hr>
<p>
    This <a href="http://rmarkdown.rstudio.com">R Markdown</a> site was created with <a href="https://github.com/jdblischak/workflowr">workflowr</a>
</p>
<hr>

<!-- To enable disqus, uncomment the section below and provide your disqus_shortname -->

<!-- disqus
  <div id="disqus_thread"></div>
    <script type="text/javascript">
        /* * * CONFIGURATION VARIABLES: EDIT BEFORE PASTING INTO YOUR WEBPAGE * * */
        var disqus_shortname = 'rmarkdown'; // required: replace example with your forum shortname

        /* * * DON'T EDIT BELOW THIS LINE * * */
        (function() {
            var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
            dsq.src = '//' + disqus_shortname + '.disqus.com/embed.js';
            (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
        })();
    </script>
    <noscript>Please enable JavaScript to view the <a href="http://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
    <a href="http://disqus.com" class="dsq-brlink">comments powered by <span class="logo-disqus">Disqus</span></a>
-->


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
