<!DOCTYPE html>
<html lang="en">
<head>
<!-- 2018-01-10 Wed 18:03 -->
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
<title>NB estimation</title>
<meta name="generator" content="Org mode">
<meta name="author" content="Abhishek Sarkar">
<style type="text/css">
 <!--/*--><![CDATA[/*><!--*/
  .title  { text-align: center;
             margin-bottom: .2em; }
  .subtitle { text-align: center;
              font-size: medium;
              font-weight: bold;
              margin-top:0; }
  .todo   { font-family: monospace; color: red; }
  .done   { font-family: monospace; color: green; }
  .priority { font-family: monospace; color: orange; }
  .tag    { background-color: #eee; font-family: monospace;
            padding: 2px; font-size: 80%; font-weight: normal; }
  .timestamp { color: #bebebe; }
  .timestamp-kwd { color: #5f9ea0; }
  .org-right  { margin-left: auto; margin-right: 0px;  text-align: right; }
  .org-left   { margin-left: 0px;  margin-right: auto; text-align: left; }
  .org-center { margin-left: auto; margin-right: auto; text-align: center; }
  .underline { text-decoration: underline; }
  #postamble p, #preamble p { font-size: 90%; margin: .2em; }
  p.verse { margin-left: 3%; }
  pre {
    border: 1px solid #ccc;
    box-shadow: 3px 3px 3px #eee;
    padding: 8pt;
    font-family: monospace;
    overflow: auto;
    margin: 1.2em;
  }
  pre.src {
    position: relative;
    overflow: visible;
    padding-top: 1.2em;
  }
  pre.src:before {
    display: none;
    position: absolute;
    background-color: white;
    top: -10px;
    right: 10px;
    padding: 3px;
    border: 1px solid black;
  }
  pre.src:hover:before { display: inline;}
  /* Languages per Org manual */
  pre.src-asymptote:before { content: 'Asymptote'; }
  pre.src-awk:before { content: 'Awk'; }
  pre.src-C:before { content: 'C'; }
  /* pre.src-C++ doesn't work in CSS */
  pre.src-clojure:before { content: 'Clojure'; }
  pre.src-css:before { content: 'CSS'; }
  pre.src-D:before { content: 'D'; }
  pre.src-ditaa:before { content: 'ditaa'; }
  pre.src-dot:before { content: 'Graphviz'; }
  pre.src-calc:before { content: 'Emacs Calc'; }
  pre.src-emacs-lisp:before { content: 'Emacs Lisp'; }
  pre.src-fortran:before { content: 'Fortran'; }
  pre.src-gnuplot:before { content: 'gnuplot'; }
  pre.src-haskell:before { content: 'Haskell'; }
  pre.src-hledger:before { content: 'hledger'; }
  pre.src-java:before { content: 'Java'; }
  pre.src-js:before { content: 'Javascript'; }
  pre.src-latex:before { content: 'LaTeX'; }
  pre.src-ledger:before { content: 'Ledger'; }
  pre.src-lisp:before { content: 'Lisp'; }
  pre.src-lilypond:before { content: 'Lilypond'; }
  pre.src-lua:before { content: 'Lua'; }
  pre.src-matlab:before { content: 'MATLAB'; }
  pre.src-mscgen:before { content: 'Mscgen'; }
  pre.src-ocaml:before { content: 'Objective Caml'; }
  pre.src-octave:before { content: 'Octave'; }
  pre.src-org:before { content: 'Org mode'; }
  pre.src-oz:before { content: 'OZ'; }
  pre.src-plantuml:before { content: 'Plantuml'; }
  pre.src-processing:before { content: 'Processing.js'; }
  pre.src-python:before { content: 'Python'; }
  pre.src-R:before { content: 'R'; }
  pre.src-ruby:before { content: 'Ruby'; }
  pre.src-sass:before { content: 'Sass'; }
  pre.src-scheme:before { content: 'Scheme'; }
  pre.src-screen:before { content: 'Gnu Screen'; }
  pre.src-sed:before { content: 'Sed'; }
  pre.src-sh:before { content: 'shell'; }
  pre.src-sql:before { content: 'SQL'; }
  pre.src-sqlite:before { content: 'SQLite'; }
  /* additional languages in org.el's org-babel-load-languages alist */
  pre.src-forth:before { content: 'Forth'; }
  pre.src-io:before { content: 'IO'; }
  pre.src-J:before { content: 'J'; }
  pre.src-makefile:before { content: 'Makefile'; }
  pre.src-maxima:before { content: 'Maxima'; }
  pre.src-perl:before { content: 'Perl'; }
  pre.src-picolisp:before { content: 'Pico Lisp'; }
  pre.src-scala:before { content: 'Scala'; }
  pre.src-shell:before { content: 'Shell Script'; }
  pre.src-ebnf2ps:before { content: 'ebfn2ps'; }
  /* additional language identifiers per "defun org-babel-execute"
       in ob-*.el */
  pre.src-cpp:before  { content: 'C++'; }
  pre.src-abc:before  { content: 'ABC'; }
  pre.src-coq:before  { content: 'Coq'; }
  pre.src-groovy:before  { content: 'Groovy'; }
  /* additional language identifiers from org-babel-shell-names in
     ob-shell.el: ob-shell is the only babel language using a lambda to put
     the execution function name together. */
  pre.src-bash:before  { content: 'bash'; }
  pre.src-csh:before  { content: 'csh'; }
  pre.src-ash:before  { content: 'ash'; }
  pre.src-dash:before  { content: 'dash'; }
  pre.src-ksh:before  { content: 'ksh'; }
  pre.src-mksh:before  { content: 'mksh'; }
  pre.src-posh:before  { content: 'posh'; }
  /* Additional Emacs modes also supported by the LaTeX listings package */
  pre.src-ada:before { content: 'Ada'; }
  pre.src-asm:before { content: 'Assembler'; }
  pre.src-caml:before { content: 'Caml'; }
  pre.src-delphi:before { content: 'Delphi'; }
  pre.src-html:before { content: 'HTML'; }
  pre.src-idl:before { content: 'IDL'; }
  pre.src-mercury:before { content: 'Mercury'; }
  pre.src-metapost:before { content: 'MetaPost'; }
  pre.src-modula-2:before { content: 'Modula-2'; }
  pre.src-pascal:before { content: 'Pascal'; }
  pre.src-ps:before { content: 'PostScript'; }
  pre.src-prolog:before { content: 'Prolog'; }
  pre.src-simula:before { content: 'Simula'; }
  pre.src-tcl:before { content: 'tcl'; }
  pre.src-tex:before { content: 'TeX'; }
  pre.src-plain-tex:before { content: 'Plain TeX'; }
  pre.src-verilog:before { content: 'Verilog'; }
  pre.src-vhdl:before { content: 'VHDL'; }
  pre.src-xml:before { content: 'XML'; }
  pre.src-nxml:before { content: 'XML'; }
  /* add a generic configuration mode; LaTeX export needs an additional
     (add-to-list 'org-latex-listings-langs '(conf " ")) in .emacs */
  pre.src-conf:before { content: 'Configuration File'; }

  table { border-collapse:collapse; }
  caption.t-above { caption-side: top; }
  caption.t-bottom { caption-side: bottom; }
  td, th { vertical-align:top;  }
  th.org-right  { text-align: center;  }
  th.org-left   { text-align: center;   }
  th.org-center { text-align: center; }
  td.org-right  { text-align: right;  }
  td.org-left   { text-align: left;   }
  td.org-center { text-align: center; }
  dt { font-weight: bold; }
  .footpara { display: inline; }
  .footdef  { margin-bottom: 1em; }
  .figure { padding: 1em; }
  .figure p { text-align: center; }
  .inlinetask {
    padding: 10px;
    border: 2px solid gray;
    margin: 10px;
    background: #ffffcc;
  }
  #org-div-home-and-up
   { text-align: right; font-size: 70%; white-space: nowrap; }
  textarea { overflow-x: auto; }
  .linenr { font-size: smaller }
  .code-highlighted { background-color: #ffff00; }
  .org-info-js_info-navigation { border-style: none; }
  #org-info-js_console-label
    { font-size: 10px; font-weight: bold; white-space: nowrap; }
  .org-info-js_search-highlight
    { background-color: #ffff00; color: #000000; font-weight: bold; }
  .org-svg { width: 90%; }
  /*]]>*/-->
</style>
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<link rel="stylesheet" type="text/css" href="https://www.pirilampo.org/styles/readtheorg/css/htmlize.css"/>
<style type="text/css">body {width: 60em; margin:auto} pre.src {overflow:auto}</style>
<script type="text/javascript">
/*
@licstart  The following is the entire license notice for the
JavaScript code in this tag.

Copyright (C) 2012-2017 Free Software Foundation, Inc.

The JavaScript code in this tag is free software: you can
redistribute it and/or modify it under the terms of the GNU
General Public License (GNU GPL) as published by the Free Software
Foundation, either version 3 of the License, or (at your option)
any later version.  The code is distributed WITHOUT ANY WARRANTY;
without even the implied warranty of MERCHANTABILITY or FITNESS
FOR A PARTICULAR PURPOSE.  See the GNU GPL for more details.

As additional permission under GNU GPL version 3 section 7, you
may distribute non-source (e.g., minimized or compacted) forms of
that code without the copy of the GNU GPL normally required by
section 4, provided you include this license notice and a URL
through which recipients can access the Corresponding Source.


@licend  The above is the entire license notice
for the JavaScript code in this tag.
*/
<!--/*--><![CDATA[/*><!--*/
 function CodeHighlightOn(elem, id)
 {
   var target = document.getElementById(id);
   if(null != target) {
     elem.cacheClassElem = elem.className;
     elem.cacheClassTarget = target.className;
     target.className = "code-highlighted";
     elem.className   = "code-highlighted";
   }
 }
 function CodeHighlightOff(elem, id)
 {
   var target = document.getElementById(id);
   if(elem.cacheClassElem)
     elem.className = elem.cacheClassElem;
   if(elem.cacheClassTarget)
     target.className = elem.cacheClassTarget;
 }
/*]]>*///-->
</script>
</head>
<body>
<div id="content">
<h1 class="title">NB estimation</h1>
<div id="table-of-contents">
<h2>Table of Contents</h2>
<div id="text-table-of-contents">
<ul>
<li><a href="#orgc4ca55f">Introduction</a></li>
<li><a href="#org2006bde">Data</a></li>
<li><a href="#org47fd6f0">Per-individual mean, no overdispersion</a></li>
<li><a href="#org1b53395">Per-individual mean, global overdispersion</a></li>
<li><a href="#org477b2bb">Per-individual mean, per-individual overdispersion</a></li>
</ul>
</div>
</div>

<div id="outline-container-orgc4ca55f" class="outline-2">
<h2 id="orgc4ca55f">Introduction</h2>
<div class="outline-text-2" id="text-orgc4ca55f">
<p>
Here we investigate how mean and dispersion estimates can explain away each
other using single cell RNA-Seq data for the gene <i>ZSWIM7</i> (the top eQTL gene
from the iPSC bulk RNA-Seq study).
</p>

<p>
For simplicity, we ignore genotype for now.
</p>
</div>
</div>

<div id="outline-container-org2006bde" class="outline-2">
<h2 id="org2006bde">Data</h2>
<div class="outline-text-2" id="text-org2006bde">
<p>
We <a href="hurdle.html#org010e249">previously prepared the relevant data</a>:
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">with</span> <span class="org-builtin">open</span>(<span class="org-string">'test_data.pkl'</span>, <span class="org-string">'rb'</span>) <span class="org-keyword">as</span> f:
  <span class="org-variable-name">data</span> = pickle.load(f)
data.keys()
</pre>
</div>

<pre class="example">
dict_keys(['onehot', 'genotypes', 'counts', 'normalizers'])

</pre>

<p>
Plot the distribution of the data:
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">n</span> = data[<span class="org-string">'counts'</span>].shape[0]
plt.clf()
plt.gcf().set_size_inches(8, 6)
plt.scatter(x=np.where(data[<span class="org-string">'onehot'</span>] == 1)[1] + np.random.normal(scale=0.1, size=n), y=data[<span class="org-string">'counts'</span>], s=2)
plt.xlabel(<span class="org-string">'Individual'</span>)
plt.ylabel(<span class="org-string">'UMI count'</span>)
plt.gca().set_xticks([])
</pre>
</div>

<pre class="example">
[]

</pre>

<div class="figure">
<p><img src="figure/nb.org/data.png" alt="data.png">
</p>
</div>
</div>
</div>

<div id="outline-container-org47fd6f0" class="outline-2">
<h2 id="org47fd6f0">Per-individual mean, no overdispersion</h2>
<div class="outline-text-2" id="text-org47fd6f0">
<div class="org-src-container">
<pre class="src src-ipython">tf.reset_default_graph()
ed.get_session().close()

<span class="org-variable-name">n</span>, <span class="org-variable-name">p</span> = data[<span class="org-string">'genotypes'</span>].shape
<span class="org-variable-name">m</span>, <span class="org-variable-name">_</span> = data[<span class="org-string">'onehot'</span>].shape

<span class="org-variable-name">onehot</span> = tf.placeholder(tf.float32, [m, n])
<span class="org-variable-name">mean_bias</span> = ed.models.Normal(loc=tf.zeros([n, 1]), scale=tf.ones([n, 1]))
<span class="org-variable-name">counts</span> = ed.models.Poisson(rate=tf.exp(tf.matmul(onehot, mean_bias)))

<span class="org-variable-name">q_mean_bias</span> = ed.models.NormalWithSoftplusScale(
  loc=tf.Variable(tf.random_normal([n, 1])),
  scale=tf.Variable(tf.random_normal([n, 1])))
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">inf</span> = ed.ReparameterizationKLqp(
  latent_vars={mean_bias: q_mean_bias},
  data={onehot: data[<span class="org-string">'onehot'</span>], counts: data[<span class="org-string">'counts'</span>]})
inf.run(n_samples=10, optimizer=tf.train.AdamOptimizer(learning_rate=5e-3))
</pre>
</div>

<p>
1000/1000 <code>[100%]</code> ██████████████████████████████ Elapsed: 3s | Loss: 5215.281
</p>

<p>
Plot posterior 95% credible intervals for the latent mean of each individual.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">res</span> = pd.DataFrame(np.hstack(ed.get_session().run([q_mean_bias.mean(), q_mean_bias.variance()])))
plt.clf()
plt.errorbar(x=np.arange(n), y=res[0], yerr=1.96 * np.sqrt(res[1]), fmt=<span class="org-string">'o'</span>)
plt.xlabel(<span class="org-string">'Individual'</span>)
plt.ylabel(<span class="org-string">'Latent mean'</span>)
plt.gca().set_xticks([])
</pre>
</div>

<pre class="example">
[]

</pre>

<div class="figure">
<p><img src="figure/nb.org/no-overdispersion.png" alt="no-overdispersion.png">
</p>
</div>
</div>
</div>

<div id="outline-container-org1b53395" class="outline-2">
<h2 id="org1b53395">Per-individual mean, global overdispersion</h2>
<div class="outline-text-2" id="text-org1b53395">
<div class="org-src-container">
<pre class="src src-ipython">tf.reset_default_graph()
ed.get_session().close()

<span class="org-variable-name">n</span>, <span class="org-variable-name">p</span> = data[<span class="org-string">'genotypes'</span>].shape
<span class="org-variable-name">m</span>, <span class="org-variable-name">_</span> = data[<span class="org-string">'onehot'</span>].shape

<span class="org-variable-name">onehot</span> = tf.placeholder(tf.float32, [m, n])
<span class="org-variable-name">mean_bias</span> = ed.models.Normal(loc=tf.zeros([n, 1]), scale=tf.ones([n, 1]))
<span class="org-variable-name">disp_bias</span> = ed.models.Normal(loc=tf.zeros([1, 1]), scale=tf.ones([1, 1]))
<span class="org-variable-name">rate</span> = ed.models.Gamma(
  concentration=tf.exp(tf.matmul(onehot, mean_bias)),
  rate=tf.exp(disp_bias))
<span class="org-variable-name">counts</span> = ed.models.Poisson(rate=rate)

<span class="org-variable-name">q_mean_bias</span> = ed.models.NormalWithSoftplusScale(
  loc=tf.Variable(tf.random_normal([n, 1])),
  scale=tf.Variable(tf.random_normal([n, 1])))
<span class="org-variable-name">q_disp_bias</span> = ed.models.NormalWithSoftplusScale(
  loc=tf.Variable(tf.random_normal([1, 1])),
  scale=tf.Variable(tf.random_normal([1, 1])))
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">inf</span> = ed.ReparameterizationKLqp(
  latent_vars={mean_bias: q_mean_bias, disp_bias: q_disp_bias},
  data={onehot: data[<span class="org-string">'onehot'</span>], counts: data[<span class="org-string">'counts'</span>]})
inf.run(n_samples=10, optimizer=tf.train.AdamOptimizer(learning_rate=5e-3))
</pre>
</div>

<p>
1000/1000 <code>[100%]</code> ██████████████████████████████ Elapsed: 10s | Loss: 15706.714
</p>

<p>
Plot the posterior 95% credible interval for the latent means.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">res</span> = pd.DataFrame(np.hstack(ed.get_session().run([q_mean_bias.mean(), q_mean_bias.variance()])))
plt.clf()
plt.errorbar(x=np.arange(n), y=res[0], yerr=1.96 * np.sqrt(res[1]), fmt=<span class="org-string">'o'</span>)
plt.xlabel(<span class="org-string">'Individual'</span>)
plt.ylabel(<span class="org-string">'Latent mean'</span>)
plt.gca().set_xticks([])
</pre>
</div>

<pre class="example">
[]

</pre>

<div class="figure">
<p><img src="figure/nb.org/global-overdispersion.png" alt="global-overdispersion.png">
</p>
</div>

<p>
Recover the estimated posterior mean and variance of the global dispersion
parameter.
</p>

<div class="org-src-container">
<pre class="src src-ipython">pd.DataFrame(np.hstack(ed.get_session().run([q_disp_bias.mean(), q_disp_bias.variance()])))  
</pre>
</div>

<pre class="example">
        0         1
0 -0.709709  0.002269
</pre>
</div>
</div>

<div id="outline-container-org477b2bb" class="outline-2">
<h2 id="org477b2bb">Per-individual mean, per-individual overdispersion</h2>
<div class="outline-text-2" id="text-org477b2bb">
<div class="org-src-container">
<pre class="src src-ipython">tf.reset_default_graph()
ed.get_session().close()

<span class="org-variable-name">n</span>, <span class="org-variable-name">p</span> = data[<span class="org-string">'genotypes'</span>].shape
<span class="org-variable-name">m</span>, <span class="org-variable-name">_</span> = data[<span class="org-string">'onehot'</span>].shape

<span class="org-variable-name">onehot</span> = tf.placeholder(tf.float32, [m, n])
<span class="org-variable-name">genotypes</span> = tf.placeholder(tf.float32, [n, p])
<span class="org-variable-name">normalizers</span> = tf.placeholder(tf.float32, [m, 1])

<span class="org-variable-name">mean_bias</span> = ed.models.Normal(loc=tf.zeros([n, 1]), scale=tf.ones([n, 1]))
<span class="org-variable-name">disp_bias</span> = ed.models.Normal(loc=tf.zeros([n, 1]), scale=tf.ones([n, 1]))
<span class="org-variable-name">rate</span> = ed.models.Gamma(
  concentration=tf.exp(tf.matmul(onehot, mean_bias)),
  rate=tf.exp(tf.matmul(onehot, disp_bias)))
<span class="org-variable-name">counts</span> = ed.models.Poisson(rate=rate)

<span class="org-variable-name">q_mean_bias</span> = ed.models.NormalWithSoftplusScale(
  loc=tf.Variable(tf.random_normal([n, 1])),
  scale=tf.Variable(tf.random_normal([n, 1])))
<span class="org-variable-name">q_disp_bias</span> = ed.models.NormalWithSoftplusScale(
  loc=tf.Variable(tf.random_normal([n, 1])),
  scale=tf.Variable(tf.random_normal([n, 1])))

<span class="org-variable-name">inf</span> = ed.ReparameterizationKLqp(
  latent_vars={mean_bias: q_mean_bias, disp_bias: q_disp_bias},
  data={<span class="org-builtin">globals</span>()[k]: v <span class="org-keyword">for</span> k, v <span class="org-keyword">in</span> data.items()})
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython">inf.run(n_samples=10, optimizer=tf.train.AdamOptimizer(learning_rate=5e-3))
</pre>
</div>

<p>
Plot the posterior 95% credible intervals for the latent means.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">res</span> = pd.DataFrame(np.hstack(ed.get_session().run([q_mean_bias.mean(), q_mean_bias.variance()])))
plt.clf()
plt.errorbar(x=np.arange(n), y=res[0], yerr=1.96 * np.sqrt(res[1]), fmt=<span class="org-string">'o'</span>)
plt.xlabel(<span class="org-string">'Individual'</span>)
plt.ylabel(<span class="org-string">'Latent mean'</span>)
plt.gca().set_xticks([])
</pre>
</div>

<pre class="example">
[]

</pre>

<div class="figure">
<p><img src="mean-per-individual-dispersion.png" alt="mean-per-individual-dispersion.png">
</p>
</div>

<p>
Plot the posterior 95% credible intervals for the latent dispersions.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">res</span> = pd.DataFrame(np.hstack(ed.get_session().run([q_disp_bias.mean(), q_disp_bias.variance()])))
plt.clf()
plt.errorbar(x=np.arange(n), y=res[0], yerr=1.96 * np.sqrt(res[1]), fmt=<span class="org-string">'o'</span>)
plt.xlabel(<span class="org-string">'Individual'</span>)
plt.ylabel(<span class="org-string">'Latent dispersion'</span>)
plt.gca().set_xticks([])
</pre>
</div>

<pre class="example">
[]

</pre>

<div class="figure">
<p><img src="disp-per-individual-dispersion.png" alt="disp-per-individual-dispersion.png">
</p>
</div>
</div>
</div>
</div>
<div id="postamble" class="status">
<p class="author">Author: Abhishek Sarkar</p>
<p class="date">Created: 2018-01-10 Wed 18:03</p>
<p class="validation"><a href="http://validator.w3.org/check?uri=referer">Validate</a></p>
</div>
</body>
</html>
