<!DOCTYPE html>
<html lang="en">
<head>
<!-- 2018-09-05 Wed 14:22 -->
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
<title>Power to detect QTLs in single cell data</title>
<meta name="generator" content="Org mode">
<meta name="author" content="Abhishek Sarkar">
<style type="text/css">
 <!--/*--><![CDATA[/*><!--*/
  .title  { text-align: center;
             margin-bottom: .2em; }
  .subtitle { text-align: center;
              font-size: medium;
              font-weight: bold;
              margin-top:0; }
  .todo   { font-family: monospace; color: red; }
  .done   { font-family: monospace; color: green; }
  .priority { font-family: monospace; color: orange; }
  .tag    { background-color: #eee; font-family: monospace;
            padding: 2px; font-size: 80%; font-weight: normal; }
  .timestamp { color: #bebebe; }
  .timestamp-kwd { color: #5f9ea0; }
  .org-right  { margin-left: auto; margin-right: 0px;  text-align: right; }
  .org-left   { margin-left: 0px;  margin-right: auto; text-align: left; }
  .org-center { margin-left: auto; margin-right: auto; text-align: center; }
  .underline { text-decoration: underline; }
  #postamble p, #preamble p { font-size: 90%; margin: .2em; }
  p.verse { margin-left: 3%; }
  pre {
    border: 1px solid #ccc;
    box-shadow: 3px 3px 3px #eee;
    padding: 8pt;
    font-family: monospace;
    overflow: auto;
    margin: 1.2em;
  }
  pre.src {
    position: relative;
    overflow: visible;
    padding-top: 1.2em;
  }
  pre.src:before {
    display: none;
    position: absolute;
    background-color: white;
    top: -10px;
    right: 10px;
    padding: 3px;
    border: 1px solid black;
  }
  pre.src:hover:before { display: inline;}
  /* Languages per Org manual */
  pre.src-asymptote:before { content: 'Asymptote'; }
  pre.src-awk:before { content: 'Awk'; }
  pre.src-C:before { content: 'C'; }
  /* pre.src-C++ doesn't work in CSS */
  pre.src-clojure:before { content: 'Clojure'; }
  pre.src-css:before { content: 'CSS'; }
  pre.src-D:before { content: 'D'; }
  pre.src-ditaa:before { content: 'ditaa'; }
  pre.src-dot:before { content: 'Graphviz'; }
  pre.src-calc:before { content: 'Emacs Calc'; }
  pre.src-emacs-lisp:before { content: 'Emacs Lisp'; }
  pre.src-fortran:before { content: 'Fortran'; }
  pre.src-gnuplot:before { content: 'gnuplot'; }
  pre.src-haskell:before { content: 'Haskell'; }
  pre.src-hledger:before { content: 'hledger'; }
  pre.src-java:before { content: 'Java'; }
  pre.src-js:before { content: 'Javascript'; }
  pre.src-latex:before { content: 'LaTeX'; }
  pre.src-ledger:before { content: 'Ledger'; }
  pre.src-lisp:before { content: 'Lisp'; }
  pre.src-lilypond:before { content: 'Lilypond'; }
  pre.src-lua:before { content: 'Lua'; }
  pre.src-matlab:before { content: 'MATLAB'; }
  pre.src-mscgen:before { content: 'Mscgen'; }
  pre.src-ocaml:before { content: 'Objective Caml'; }
  pre.src-octave:before { content: 'Octave'; }
  pre.src-org:before { content: 'Org mode'; }
  pre.src-oz:before { content: 'OZ'; }
  pre.src-plantuml:before { content: 'Plantuml'; }
  pre.src-processing:before { content: 'Processing.js'; }
  pre.src-python:before { content: 'Python'; }
  pre.src-R:before { content: 'R'; }
  pre.src-ruby:before { content: 'Ruby'; }
  pre.src-sass:before { content: 'Sass'; }
  pre.src-scheme:before { content: 'Scheme'; }
  pre.src-screen:before { content: 'Gnu Screen'; }
  pre.src-sed:before { content: 'Sed'; }
  pre.src-sh:before { content: 'shell'; }
  pre.src-sql:before { content: 'SQL'; }
  pre.src-sqlite:before { content: 'SQLite'; }
  /* additional languages in org.el's org-babel-load-languages alist */
  pre.src-forth:before { content: 'Forth'; }
  pre.src-io:before { content: 'IO'; }
  pre.src-J:before { content: 'J'; }
  pre.src-makefile:before { content: 'Makefile'; }
  pre.src-maxima:before { content: 'Maxima'; }
  pre.src-perl:before { content: 'Perl'; }
  pre.src-picolisp:before { content: 'Pico Lisp'; }
  pre.src-scala:before { content: 'Scala'; }
  pre.src-shell:before { content: 'Shell Script'; }
  pre.src-ebnf2ps:before { content: 'ebfn2ps'; }
  /* additional language identifiers per "defun org-babel-execute"
       in ob-*.el */
  pre.src-cpp:before  { content: 'C++'; }
  pre.src-abc:before  { content: 'ABC'; }
  pre.src-coq:before  { content: 'Coq'; }
  pre.src-groovy:before  { content: 'Groovy'; }
  /* additional language identifiers from org-babel-shell-names in
     ob-shell.el: ob-shell is the only babel language using a lambda to put
     the execution function name together. */
  pre.src-bash:before  { content: 'bash'; }
  pre.src-csh:before  { content: 'csh'; }
  pre.src-ash:before  { content: 'ash'; }
  pre.src-dash:before  { content: 'dash'; }
  pre.src-ksh:before  { content: 'ksh'; }
  pre.src-mksh:before  { content: 'mksh'; }
  pre.src-posh:before  { content: 'posh'; }
  /* Additional Emacs modes also supported by the LaTeX listings package */
  pre.src-ada:before { content: 'Ada'; }
  pre.src-asm:before { content: 'Assembler'; }
  pre.src-caml:before { content: 'Caml'; }
  pre.src-delphi:before { content: 'Delphi'; }
  pre.src-html:before { content: 'HTML'; }
  pre.src-idl:before { content: 'IDL'; }
  pre.src-mercury:before { content: 'Mercury'; }
  pre.src-metapost:before { content: 'MetaPost'; }
  pre.src-modula-2:before { content: 'Modula-2'; }
  pre.src-pascal:before { content: 'Pascal'; }
  pre.src-ps:before { content: 'PostScript'; }
  pre.src-prolog:before { content: 'Prolog'; }
  pre.src-simula:before { content: 'Simula'; }
  pre.src-tcl:before { content: 'tcl'; }
  pre.src-tex:before { content: 'TeX'; }
  pre.src-plain-tex:before { content: 'Plain TeX'; }
  pre.src-verilog:before { content: 'Verilog'; }
  pre.src-vhdl:before { content: 'VHDL'; }
  pre.src-xml:before { content: 'XML'; }
  pre.src-nxml:before { content: 'XML'; }
  /* add a generic configuration mode; LaTeX export needs an additional
     (add-to-list 'org-latex-listings-langs '(conf " ")) in .emacs */
  pre.src-conf:before { content: 'Configuration File'; }

  table { border-collapse:collapse; }
  caption.t-above { caption-side: top; }
  caption.t-bottom { caption-side: bottom; }
  td, th { vertical-align:top;  }
  th.org-right  { text-align: center;  }
  th.org-left   { text-align: center;   }
  th.org-center { text-align: center; }
  td.org-right  { text-align: right;  }
  td.org-left   { text-align: left;   }
  td.org-center { text-align: center; }
  dt { font-weight: bold; }
  .footpara { display: inline; }
  .footdef  { margin-bottom: 1em; }
  .figure { padding: 1em; }
  .figure p { text-align: center; }
  .inlinetask {
    padding: 10px;
    border: 2px solid gray;
    margin: 10px;
    background: #ffffcc;
  }
  #org-div-home-and-up
   { text-align: right; font-size: 70%; white-space: nowrap; }
  textarea { overflow-x: auto; }
  .linenr { font-size: smaller }
  .code-highlighted { background-color: #ffff00; }
  .org-info-js_info-navigation { border-style: none; }
  #org-info-js_console-label
    { font-size: 10px; font-weight: bold; white-space: nowrap; }
  .org-info-js_search-highlight
    { background-color: #ffff00; color: #000000; font-weight: bold; }
  .org-svg { width: 90%; }
  /*]]>*/-->
</style>
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<link rel="stylesheet" type="text/css" href="https://www.pirilampo.org/styles/readtheorg/css/htmlize.css"/>
<style type="text/css">body {width: 60em; margin:auto} pre.src {overflow:auto}</style>
<script type="text/javascript">
/*
@licstart  The following is the entire license notice for the
JavaScript code in this tag.

Copyright (C) 2012-2017 Free Software Foundation, Inc.

The JavaScript code in this tag is free software: you can
redistribute it and/or modify it under the terms of the GNU
General Public License (GNU GPL) as published by the Free Software
Foundation, either version 3 of the License, or (at your option)
any later version.  The code is distributed WITHOUT ANY WARRANTY;
without even the implied warranty of MERCHANTABILITY or FITNESS
FOR A PARTICULAR PURPOSE.  See the GNU GPL for more details.

As additional permission under GNU GPL version 3 section 7, you
may distribute non-source (e.g., minimized or compacted) forms of
that code without the copy of the GNU GPL normally required by
section 4, provided you include this license notice and a URL
through which recipients can access the Corresponding Source.


@licend  The above is the entire license notice
for the JavaScript code in this tag.
*/
<!--/*--><![CDATA[/*><!--*/
 function CodeHighlightOn(elem, id)
 {
   var target = document.getElementById(id);
   if(null != target) {
     elem.cacheClassElem = elem.className;
     elem.cacheClassTarget = target.className;
     target.className = "code-highlighted";
     elem.className   = "code-highlighted";
   }
 }
 function CodeHighlightOff(elem, id)
 {
   var target = document.getElementById(id);
   if(elem.cacheClassElem)
     elem.className = elem.cacheClassElem;
   if(elem.cacheClassTarget)
     target.className = elem.cacheClassTarget;
 }
/*]]>*///-->
</script>
<script type="text/x-mathjax-config">
    MathJax.Hub.Config({
        displayAlign: "center",
        displayIndent: "0em",

        "HTML-CSS": { scale: 100,
                        linebreaks: { automatic: "false" },
                        webFont: "TeX"
                       },
        SVG: {scale: 100,
              linebreaks: { automatic: "false" },
              font: "TeX"},
        NativeMML: {scale: 100},
        TeX: { equationNumbers: {autoNumber: "AMS"},
               MultLineWidth: "85%",
               TagSide: "right",
               TagIndent: ".8em"
             }
});
</script>
<script type="text/javascript"
        src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-AMS_HTML"></script>
</head>
<body>
<div id="content">
<h1 class="title">Power to detect QTLs in single cell data</h1>
<div id="table-of-contents">
<h2>Table of Contents</h2>
<div id="text-table-of-contents">
<ul>
<li><a href="#org9f6cd72">Compare experimental variance to biological variance</a></li>
</ul>
</div>
</div>
<div id="outline-container-org9f6cd72" class="outline-2">
<h2 id="org9f6cd72">Compare experimental variance to biological variance</h2>
<div class="outline-text-2" id="text-org9f6cd72">
<p>
Let \hat{\theta}<sub>i</sub> be an estimate of some phenotype \(\theta_i\). Assume:
</p>

<p>
\[ \hat{\theta}_i = \theta_i + u_i \]
</p>

<p>
where \(u_i\) is a random effect capturing experimental noise, finite
sampling variance, and estimator sampling variance.
</p>

<p>
Above, we estimated the scale of the finite sampling variance and estimator
sampling variance. Here, we relate it to the genetic variance.
</p>

<p>
Extract the genotypes and UMI counts.
</p>

<div class="org-src-container">
<pre class="src src-ipython" id="org7f1be20"><span class="org-keyword">def</span> <span class="org-function-name">parse_vcf_dosage</span>(record):
  <span class="org-variable-name">geno</span> = [<span class="org-builtin">float</span>(g) <span class="org-keyword">for</span> g <span class="org-keyword">in</span> record[9:]]
  <span class="org-keyword">return</span> pd.Series(geno)

<span class="org-keyword">def</span> <span class="org-function-name">extract_geno</span>(qtls, dosages):
  <span class="org-variable-name">header</span> = pd.read_table(dosages, skiprows=2, nrows=1, header=0).columns[9:]
  <span class="org-variable-name">genotypes</span> = tabix.<span class="org-builtin">open</span>(dosages)
  <span class="org-variable-name">X</span> = np.<span class="org-builtin">round</span>(
    qtls
    .<span class="org-builtin">apply</span>(<span class="org-keyword">lambda</span> x: parse_vcf_dosage(<span class="org-builtin">next</span>(genotypes.query(x[<span class="org-string">'chr'</span>], <span class="org-builtin">int</span>(x[<span class="org-string">'var_start'</span>]) - 1, <span class="org-builtin">int</span>(x[<span class="org-string">'var_start'</span>])))), axis=1)
    .rename(columns={i: ind <span class="org-keyword">for</span> i, ind <span class="org-keyword">in</span> <span class="org-builtin">enumerate</span>(header)}))
  <span class="org-keyword">return</span> X

<span class="org-keyword">def</span> <span class="org-function-name">extract_counts</span>(qtls, counts):
  <span class="org-variable-name">X</span> = pd.concat(
    [chunk.align(qtls, axis=<span class="org-string">'index'</span>, join=<span class="org-string">'inner'</span>)[0]
     <span class="org-keyword">for</span> chunk <span class="org-keyword">in</span> pd.read_table(counts, index_col=0, chunksize=1000)])
  <span class="org-keyword">return</span> X
</pre>
</div>

<p>
Read the data.
</p>

<div class="org-src-container">
<pre class="src src-ipython" id="org4ebe18a"><span class="org-variable-name">qtls</span> = (pd.read_table(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-qtl/data/scqtl-mapping/mean.txt.gz'</span>, sep=<span class="org-string">' '</span>, index_col=0)
        .sample(n=1000, random_state=0))
<span class="org-variable-name">annotations</span> = pd.read_table(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-qtl/data/scqtl-annotation.txt'</span>)
<span class="org-variable-name">keep_samples</span> = pd.read_table(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-qtl/data/quality-single-cells.txt'</span>, index_col=0, header=<span class="org-constant">None</span>)
</pre>
</div>

<p>
Extract the subset we want.
</p>

<div class="org-src-container">
<pre class="src src-ipython" id="orgdccb38b"><span class="org-variable-name">counts</span> = extract_counts(qtls, <span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-qtl/data/scqtl-counts.txt.gz'</span>)
<span class="org-variable-name">annotations</span> = annotations.loc[keep_samples.values.ravel()].reset_index(drop=<span class="org-constant">True</span>)
<span class="org-variable-name">counts</span> = counts.loc[:,keep_samples.values.ravel()]
<span class="org-variable-name">geno</span> = extract_geno(qtls, <span class="org-string">'/scratch/midway2/aksarkar/singlecell/scqtl-mapping/yri-120-dosages.vcf.gz'</span>)
<span class="org-variable-name">geno</span> = geno.loc[:,<span class="org-builtin">list</span>(<span class="org-builtin">set</span>(annotations[<span class="org-string">'chip_id'</span>]))]
</pre>
</div>

<p>
Bootstrap the subset.
</p>

<div class="org-src-container">
<pre class="src src-ipython" id="org1ea132a"><span class="org-keyword">def</span> <span class="org-function-name">bootstrap</span>(counts, annotations):
  <span class="org-variable-name">idx</span> = np.hstack({k: np.random.choice(g, size=<span class="org-builtin">len</span>(g))
                   <span class="org-keyword">for</span> k, g <span class="org-keyword">in</span> annotations.groupby(<span class="org-string">'chip_id'</span>).groups.items()}.values())
  <span class="org-keyword">return</span> counts.iloc[:,idx], annotations.iloc[idx]
</pre>
</div>

<p>
Recover the derived mean and variance.
</p>

<p>
Compare the phenotypic variance of mean against dispersion. Use squared
coefficient of variation to get dimensionless quantities.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">scv</span> = pd.Series(np.square(st.variation(mean_by_ind, axis=1)) / np.square(st.variation(log_phi, axis=1)))
scv.describe()
</pre>
</div>

<pre class="example">
count    9957.000000
mean        0.388954
std         0.537197
min         0.000067
25%         0.164379
50%         0.267103
75%         0.459509
max        25.724706
dtype: float64
</pre>

<div class="org-src-container">
<pre class="src src-ipython">plt.clf()
plt.gcf().set_size_inches(3, 3)
plt.semilogx()
plt.semilogy()
<span class="org-variable-name">all_qtls</span> = (pd.read_table(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-qtl/data/scqtl-mapping/mean.txt.gz'</span>, sep=<span class="org-string">' '</span>, index_col=0)
            .sort_values(<span class="org-string">'p_beta'</span>)
            .head(n=241))
plt.scatter(np.square(st.variation(mean_by_ind, axis=1)),
            np.square(st.variation(log_phi, axis=1)),
            s=2, c=<span class="org-string">'.75'</span>, label=<span class="org-string">'Not eQTL'</span>)
plt.scatter(np.square(st.variation(mean_by_ind.loc[all_qtls.index], axis=1)),
            np.square(st.variation(log_phi.loc[all_qtls.index], axis=1)),
            s=2, c=<span class="org-string">'k'</span>, label=<span class="org-string">'eQTL'</span>)
plt.plot(plt.xlim(), plt.xlim(), c=<span class="org-string">'r'</span>, ls=<span class="org-string">':'</span>)
plt.legend(frameon=<span class="org-constant">False</span>, markerscale=2, handletextpad=0)
plt.xlabel(<span class="org-string">'SCV of mean'</span>)
<span class="org-variable-name">_</span> = plt.ylabel(<span class="org-string">'SCV of $\ln(\phi)$'</span>)
</pre>
</div>


<div class="figure">
<p><img src="figure/power.org/mean-var-vs-phi-var.png" alt="mean-var-vs-phi-var.png">
</p>
</div>

<p>
Partition the sum of squares.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">def</span> <span class="org-function-name">ss</span>(y):
  <span class="org-keyword">return</span> np.square(y - y.mean()).<span class="org-builtin">sum</span>()

<span class="org-keyword">def</span> <span class="org-function-name">partition</span>(x, y):
  <span class="org-variable-name">total</span> = ss(y)
  <span class="org-variable-name">within</span> = y.groupby(x.values).<span class="org-builtin">apply</span>(ss).<span class="org-builtin">sum</span>()
  <span class="org-variable-name">between</span> = total - within
  <span class="org-keyword">return</span> total, between, within

<span class="org-variable-name">X</span>, <span class="org-variable-name">Y</span> = geno.align(mean_by_ind, join=<span class="org-string">'inner'</span>)
<span class="org-variable-name">var_comps</span> = pd.DataFrame([partition(X.loc[i], Y.loc[i]) <span class="org-keyword">for</span> i <span class="org-keyword">in</span> Y.index])
<span class="org-variable-name">var_comps.index</span> = X.index
<span class="org-variable-name">var_comps.columns</span> = [<span class="org-string">'total'</span>, <span class="org-string">'between'</span>, <span class="org-string">'within'</span>]
</pre>
</div>

<p>
Look at the ANOVA for weak eQTLs.
</p>

<div class="org-src-container">
<pre class="src src-ipython">var_comps.tail()
</pre>
</div>

<pre class="example">
total       between        within
gene
ENSG00000163961  3.681769e-10  9.298483e-11  2.751920e-10
ENSG00000151806  1.530030e-10  5.271346e-11  1.002895e-10
ENSG00000116260  3.854593e-10  6.561917e-11  3.198401e-10
ENSG00000109775  3.234854e-10  1.171794e-10  2.063060e-10
ENSG00000164758  2.571498e-10  7.245402e-11  1.846958e-10
</pre>

<p>
Compare against the estimated sampling variance from the simulation.
</p>

<div class="org-src-container">
<pre class="src src-ipython">data[np.logical_and(data[<span class="org-string">'num_samples'</span>] == 95, data[<span class="org-string">'num_mols'</span>] == 114026)][<span class="org-string">'latent_mean_hat'</span>].describe()
</pre>
</div>

<pre class="example">
count    8.000000e+01
mean     9.264424e-09
std      2.667586e-08
min      1.169715e-11
25%      1.685388e-10
50%      9.926134e-10
75%      4.742833e-09
max      1.403023e-07
Name: latent_mean_hat, dtype: float64
</pre>

<p>
Estimate bootstrap SEs on a GPU.
</p>

<div class="org-src-container">
<pre class="src src-ipython" id="orgb7518ff">&lt;&lt;zinb-imports&gt;&gt;
&lt;&lt;tf-imports&gt;&gt;
<span class="org-keyword">import</span> tabix
<span class="org-keyword">import</span> scipy.linalg <span class="org-keyword">as</span> sl

&lt;&lt;tf-zinb-impl&gt;&gt;
&lt;&lt;recode-impl&gt;&gt;
&lt;&lt;extract-impl&gt;&gt;
&lt;&lt;bootstrap-impl&gt;&gt;

&lt;&lt;read-qtls&gt;&gt;
&lt;&lt;extract-counts&gt;&gt;

<span class="org-variable-name">log_mu</span> = pd.read_table(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-qtl/data/density-estimation/design1/zi2-log-mu.txt.gz'</span>, index_col=0, sep=<span class="org-string">' '</span>)
<span class="org-variable-name">log_phi</span> = pd.read_table(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-qtl/data/density-estimation/design1/zi2-log-phi.txt.gz'</span>, index_col=0, sep=<span class="org-string">' '</span>)
<span class="org-variable-name">logodds</span> = pd.read_table(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-qtl/data/density-estimation/design1/zi2-log-mu.txt.gz'</span>, index_col=0, sep=<span class="org-string">' '</span>)

<span class="org-variable-name">results</span> = []
<span class="org-keyword">for</span> trial <span class="org-keyword">in</span> <span class="org-builtin">range</span>(100):
  <span class="org-variable-name">C</span>, <span class="org-variable-name">A</span> = bootstrap(counts, annotations)
  <span class="org-variable-name">onehot</span> = recode(annotations, <span class="org-string">'chip_id'</span>)
  <span class="org-variable-name">chip</span> = recode(annotations, <span class="org-string">'experiment'</span>)
  <span class="org-variable-name">chip</span> -= chip.mean(axis=0)

  results.append(fit(
    umi=C.values.T.astype(np.float32),
    onehot=onehot.astype(np.float32),
    design=chip.astype(np.float32),
    size_factor=annotations[<span class="org-string">'mol_hs'</span>].astype(np.float32).values.reshape(-1, 1),
    learning_rate=5e-2,
    max_epochs=1000))

<span class="org-variable-name">estimates</span> = {
  <span class="org-string">'mean'</span>: np.array([np.exp(r[0] - np.log1p(np.exp(r[2]))) <span class="org-keyword">for</span> r <span class="org-keyword">in</span> results]),
  <span class="org-string">'log_mean'</span>: np.array([r[0] - np.log1p(np.exp(r[2])) <span class="org-keyword">for</span> r <span class="org-keyword">in</span> results]),
  <span class="org-string">'disp'</span>: np.array([r[1] <span class="org-keyword">for</span> r <span class="org-keyword">in</span> results])
}

<span class="org-keyword">for</span> k <span class="org-keyword">in</span> estimates:
  (pd.DataFrame(estimates[k].var(axis=0).T, columns=<span class="org-builtin">sorted</span>(<span class="org-builtin">set</span>(annotations[<span class="org-string">'chip_id'</span>])), index=counts.index)
   .to_csv(<span class="org-string">'/scratch/midway2/aksarkar/singlecell/power/{}-se.txt.gz'</span>.<span class="org-builtin">format</span>(k), sep=<span class="org-string">'\t'</span>, compression=<span class="org-string">'gzip'</span>))
</pre>
</div>

<div class="org-src-container">
<pre class="src src-sh">sbatch --partition=gpu2 --gres=gpu:1 --mem=16G --job-name=tf-zinb-se --out=tf-zinb-se.out
<span class="org-comment-delimiter">#</span><span class="org-comment">!/bin/bash</span>
<span class="org-builtin">source</span> activate scqtl
python /project2/mstephens/aksarkar/projects/singlecell-qtl/code/tf-zinb-se.py
</pre>
</div>

<pre class="example">
Submitted batch job 49245803

</pre>

<p>
Read the results.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">mean_sampling_var</span> = pd.read_table(<span class="org-string">'/scratch/midway2/aksarkar/singlecell/power/mean-se.txt.gz'</span>, index_col=0)
<span class="org-variable-name">disp_sampling_var</span> = pd.read_table(<span class="org-string">'/scratch/midway2/aksarkar/singlecell/power/disp-se.txt.gz'</span>, index_col=0)
</pre>
</div>

<p>
Estimate the reduction in effective PVE for eQTLs.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">S</span>, <span class="org-variable-name">M</span> = mean_sampling_var.align(mean_by_ind, join=<span class="org-string">'inner'</span>, axis=<span class="org-string">'index'</span>)
pd.DataFrame({<span class="org-string">'prop'</span>: S.mean(axis=1) / M.var(axis=1), <span class="org-string">'factor'</span>: M.var(axis=1) / np.ma.masked_less((M.var(axis=1) - S.mean(axis=1)).values, 0)}).describe()
</pre>
</div>

<pre class="example">
factor        prop
count  75.000000  100.000000
mean    3.365234    0.698366
std     4.746555    0.407146
min     1.109361    0.098580
25%     1.458076    0.380033
50%     2.043664    0.626091
75%     3.229525    0.978929
max    34.946118    2.004026
</pre>

<p>
Estimate how many more cells would be required so the median factor is 1.1.
</p>

<p>
\[ (\sigma^2_y + \sigma^2_u) / \sigma^2_y \leq 1.1 \]
</p>

<p>
\[ \sigma^2_u \leq .1 \sigma^2 y \]
</p>

<div class="org-src-container">
<pre class="src src-ipython">(S.mean(axis=1) / (.1 * np.ma.masked_less((M.var(axis=1) - S.mean(axis=1)).values, 0))).describe()
</pre>
</div>

<pre class="example">
count     75.000000
mean      23.652338
std       47.465547
min        1.093611
25%        4.580756
50%       10.436642
75%       22.295255
max      339.461176
dtype: float64
</pre>

<div class="org-src-container">
<pre class="src src-ipython">(M.var(axis=1) / np.ma.masked_less((M.var(axis=1) - S.mean(axis=1) / 2.82).values, 0)).describe()
</pre>
</div>

<pre class="example">
count    100.000000
mean       1.398247
std        0.388211
min        1.036224
25%        1.155754
50%        1.285377
75%        1.531795
max        3.455993
dtype: float64
</pre>

<p>
Do the same for dQTLs.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">S</span>, <span class="org-variable-name">M</span> = disp_sampling_var.align(log_phi, join=<span class="org-string">'inner'</span>, axis=<span class="org-string">'index'</span>)
pd.DataFrame({<span class="org-string">'prop'</span>: S.mean(axis=1) / M.var(axis=1), <span class="org-string">'factor'</span>: M.var(axis=1) / np.ma.masked_less((M.var(axis=1) - S.mean(axis=1)).values, 0)}).describe()
</pre>
</div>

<pre class="example">
factor        prop
count  92.000000  100.000000
mean    2.346648    0.383063
std     3.382984    0.342855
min     1.041327    0.039686
25%     1.117332    0.108174
50%     1.281986    0.228490
75%     1.914243    0.636540
max    23.984483    1.541178
</pre>

<div class="org-src-container">
<pre class="src src-ipython">(S.mean(axis=1) / (.1 * np.ma.masked_less((M.var(axis=1) - S.mean(axis=1)).values, 0))).describe()
</pre>
</div>

<pre class="example">
count     92.000000
mean      13.466479
std       33.829841
min        0.413266
25%        1.173316
50%        2.819856
75%        9.142426
max      229.844830
dtype: float64
</pre>
</div>
</div>
</div>
<div id="postamble" class="status">
<p class="author">Author: Abhishek Sarkar</p>
<p class="date">Created: 2018-09-05 Wed 14:22</p>
<p class="validation"><a href="http://validator.w3.org/check?uri=referer">Validate</a></p>
</div>
</body>
</html>
