<!DOCTYPE html>
<html lang="en">
<head>
<!-- 2018-06-04 Mon 14:11 -->
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
<title>Power to detect QTLs in single cell data</title>
<meta name="generator" content="Org mode">
<meta name="author" content="Abhishek Sarkar">
<style type="text/css">
 <!--/*--><![CDATA[/*><!--*/
  .title  { text-align: center;
             margin-bottom: .2em; }
  .subtitle { text-align: center;
              font-size: medium;
              font-weight: bold;
              margin-top:0; }
  .todo   { font-family: monospace; color: red; }
  .done   { font-family: monospace; color: green; }
  .priority { font-family: monospace; color: orange; }
  .tag    { background-color: #eee; font-family: monospace;
            padding: 2px; font-size: 80%; font-weight: normal; }
  .timestamp { color: #bebebe; }
  .timestamp-kwd { color: #5f9ea0; }
  .org-right  { margin-left: auto; margin-right: 0px;  text-align: right; }
  .org-left   { margin-left: 0px;  margin-right: auto; text-align: left; }
  .org-center { margin-left: auto; margin-right: auto; text-align: center; }
  .underline { text-decoration: underline; }
  #postamble p, #preamble p { font-size: 90%; margin: .2em; }
  p.verse { margin-left: 3%; }
  pre {
    border: 1px solid #ccc;
    box-shadow: 3px 3px 3px #eee;
    padding: 8pt;
    font-family: monospace;
    overflow: auto;
    margin: 1.2em;
  }
  pre.src {
    position: relative;
    overflow: visible;
    padding-top: 1.2em;
  }
  pre.src:before {
    display: none;
    position: absolute;
    background-color: white;
    top: -10px;
    right: 10px;
    padding: 3px;
    border: 1px solid black;
  }
  pre.src:hover:before { display: inline;}
  /* Languages per Org manual */
  pre.src-asymptote:before { content: 'Asymptote'; }
  pre.src-awk:before { content: 'Awk'; }
  pre.src-C:before { content: 'C'; }
  /* pre.src-C++ doesn't work in CSS */
  pre.src-clojure:before { content: 'Clojure'; }
  pre.src-css:before { content: 'CSS'; }
  pre.src-D:before { content: 'D'; }
  pre.src-ditaa:before { content: 'ditaa'; }
  pre.src-dot:before { content: 'Graphviz'; }
  pre.src-calc:before { content: 'Emacs Calc'; }
  pre.src-emacs-lisp:before { content: 'Emacs Lisp'; }
  pre.src-fortran:before { content: 'Fortran'; }
  pre.src-gnuplot:before { content: 'gnuplot'; }
  pre.src-haskell:before { content: 'Haskell'; }
  pre.src-hledger:before { content: 'hledger'; }
  pre.src-java:before { content: 'Java'; }
  pre.src-js:before { content: 'Javascript'; }
  pre.src-latex:before { content: 'LaTeX'; }
  pre.src-ledger:before { content: 'Ledger'; }
  pre.src-lisp:before { content: 'Lisp'; }
  pre.src-lilypond:before { content: 'Lilypond'; }
  pre.src-lua:before { content: 'Lua'; }
  pre.src-matlab:before { content: 'MATLAB'; }
  pre.src-mscgen:before { content: 'Mscgen'; }
  pre.src-ocaml:before { content: 'Objective Caml'; }
  pre.src-octave:before { content: 'Octave'; }
  pre.src-org:before { content: 'Org mode'; }
  pre.src-oz:before { content: 'OZ'; }
  pre.src-plantuml:before { content: 'Plantuml'; }
  pre.src-processing:before { content: 'Processing.js'; }
  pre.src-python:before { content: 'Python'; }
  pre.src-R:before { content: 'R'; }
  pre.src-ruby:before { content: 'Ruby'; }
  pre.src-sass:before { content: 'Sass'; }
  pre.src-scheme:before { content: 'Scheme'; }
  pre.src-screen:before { content: 'Gnu Screen'; }
  pre.src-sed:before { content: 'Sed'; }
  pre.src-sh:before { content: 'shell'; }
  pre.src-sql:before { content: 'SQL'; }
  pre.src-sqlite:before { content: 'SQLite'; }
  /* additional languages in org.el's org-babel-load-languages alist */
  pre.src-forth:before { content: 'Forth'; }
  pre.src-io:before { content: 'IO'; }
  pre.src-J:before { content: 'J'; }
  pre.src-makefile:before { content: 'Makefile'; }
  pre.src-maxima:before { content: 'Maxima'; }
  pre.src-perl:before { content: 'Perl'; }
  pre.src-picolisp:before { content: 'Pico Lisp'; }
  pre.src-scala:before { content: 'Scala'; }
  pre.src-shell:before { content: 'Shell Script'; }
  pre.src-ebnf2ps:before { content: 'ebfn2ps'; }
  /* additional language identifiers per "defun org-babel-execute"
       in ob-*.el */
  pre.src-cpp:before  { content: 'C++'; }
  pre.src-abc:before  { content: 'ABC'; }
  pre.src-coq:before  { content: 'Coq'; }
  pre.src-groovy:before  { content: 'Groovy'; }
  /* additional language identifiers from org-babel-shell-names in
     ob-shell.el: ob-shell is the only babel language using a lambda to put
     the execution function name together. */
  pre.src-bash:before  { content: 'bash'; }
  pre.src-csh:before  { content: 'csh'; }
  pre.src-ash:before  { content: 'ash'; }
  pre.src-dash:before  { content: 'dash'; }
  pre.src-ksh:before  { content: 'ksh'; }
  pre.src-mksh:before  { content: 'mksh'; }
  pre.src-posh:before  { content: 'posh'; }
  /* Additional Emacs modes also supported by the LaTeX listings package */
  pre.src-ada:before { content: 'Ada'; }
  pre.src-asm:before { content: 'Assembler'; }
  pre.src-caml:before { content: 'Caml'; }
  pre.src-delphi:before { content: 'Delphi'; }
  pre.src-html:before { content: 'HTML'; }
  pre.src-idl:before { content: 'IDL'; }
  pre.src-mercury:before { content: 'Mercury'; }
  pre.src-metapost:before { content: 'MetaPost'; }
  pre.src-modula-2:before { content: 'Modula-2'; }
  pre.src-pascal:before { content: 'Pascal'; }
  pre.src-ps:before { content: 'PostScript'; }
  pre.src-prolog:before { content: 'Prolog'; }
  pre.src-simula:before { content: 'Simula'; }
  pre.src-tcl:before { content: 'tcl'; }
  pre.src-tex:before { content: 'TeX'; }
  pre.src-plain-tex:before { content: 'Plain TeX'; }
  pre.src-verilog:before { content: 'Verilog'; }
  pre.src-vhdl:before { content: 'VHDL'; }
  pre.src-xml:before { content: 'XML'; }
  pre.src-nxml:before { content: 'XML'; }
  /* add a generic configuration mode; LaTeX export needs an additional
     (add-to-list 'org-latex-listings-langs '(conf " ")) in .emacs */
  pre.src-conf:before { content: 'Configuration File'; }

  table { border-collapse:collapse; }
  caption.t-above { caption-side: top; }
  caption.t-bottom { caption-side: bottom; }
  td, th { vertical-align:top;  }
  th.org-right  { text-align: center;  }
  th.org-left   { text-align: center;   }
  th.org-center { text-align: center; }
  td.org-right  { text-align: right;  }
  td.org-left   { text-align: left;   }
  td.org-center { text-align: center; }
  dt { font-weight: bold; }
  .footpara { display: inline; }
  .footdef  { margin-bottom: 1em; }
  .figure { padding: 1em; }
  .figure p { text-align: center; }
  .inlinetask {
    padding: 10px;
    border: 2px solid gray;
    margin: 10px;
    background: #ffffcc;
  }
  #org-div-home-and-up
   { text-align: right; font-size: 70%; white-space: nowrap; }
  textarea { overflow-x: auto; }
  .linenr { font-size: smaller }
  .code-highlighted { background-color: #ffff00; }
  .org-info-js_info-navigation { border-style: none; }
  #org-info-js_console-label
    { font-size: 10px; font-weight: bold; white-space: nowrap; }
  .org-info-js_search-highlight
    { background-color: #ffff00; color: #000000; font-weight: bold; }
  .org-svg { width: 90%; }
  /*]]>*/-->
</style>
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<link rel="stylesheet" type="text/css" href="https://www.pirilampo.org/styles/readtheorg/css/htmlize.css"/>
<style type="text/css">body {width: 60em; margin:auto} pre.src {overflow:auto}</style>
<script type="text/javascript">
/*
@licstart  The following is the entire license notice for the
JavaScript code in this tag.

Copyright (C) 2012-2017 Free Software Foundation, Inc.

The JavaScript code in this tag is free software: you can
redistribute it and/or modify it under the terms of the GNU
General Public License (GNU GPL) as published by the Free Software
Foundation, either version 3 of the License, or (at your option)
any later version.  The code is distributed WITHOUT ANY WARRANTY;
without even the implied warranty of MERCHANTABILITY or FITNESS
FOR A PARTICULAR PURPOSE.  See the GNU GPL for more details.

As additional permission under GNU GPL version 3 section 7, you
may distribute non-source (e.g., minimized or compacted) forms of
that code without the copy of the GNU GPL normally required by
section 4, provided you include this license notice and a URL
through which recipients can access the Corresponding Source.


@licend  The above is the entire license notice
for the JavaScript code in this tag.
*/
<!--/*--><![CDATA[/*><!--*/
 function CodeHighlightOn(elem, id)
 {
   var target = document.getElementById(id);
   if(null != target) {
     elem.cacheClassElem = elem.className;
     elem.cacheClassTarget = target.className;
     target.className = "code-highlighted";
     elem.className   = "code-highlighted";
   }
 }
 function CodeHighlightOff(elem, id)
 {
   var target = document.getElementById(id);
   if(elem.cacheClassElem)
     elem.className = elem.cacheClassElem;
   if(elem.cacheClassTarget)
     target.className = elem.cacheClassTarget;
 }
/*]]>*///-->
</script>
<script type="text/x-mathjax-config">
    MathJax.Hub.Config({
        displayAlign: "center",
        displayIndent: "0em",

        "HTML-CSS": { scale: 100,
                        linebreaks: { automatic: "false" },
                        webFont: "TeX"
                       },
        SVG: {scale: 100,
              linebreaks: { automatic: "false" },
              font: "TeX"},
        NativeMML: {scale: 100},
        TeX: { equationNumbers: {autoNumber: "AMS"},
               MultLineWidth: "85%",
               TagSide: "right",
               TagIndent: ".8em"
             }
});
</script>
<script type="text/javascript"
        src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-AMS_HTML"></script>
</head>
<body>
<div id="content">
<h1 class="title">Power to detect QTLs in single cell data</h1>

<div id="outline-container-orgb40523f" class="outline-2">
<h2 id="orgb40523f">Introduction</h2>
<div class="outline-text-2" id="text-orgb40523f">
<p>
We <a href="qtl-mapping.html">previously found</a> that our study lost power to detect eQTLs, and was
underpowered to directly detect dispersion-QTLs.
</p>

<p>
Here, we estimate power to detect eQTLs and dispersion-QTLs as a function of:
</p>

<ul class="org-ul">
<li>number of cells per individual</li>
<li>number of molecules per cell</li>
<li>number of individuals</li>
</ul>
</div>
</div>

<div id="outline-container-orgb01a870" class="outline-2">
<h2 id="orgb01a870">Differential dispersion</h2>
<div class="outline-text-2" id="text-orgb01a870">
<p>
Perform a nested model comparison for each gene \(k\), comparing the model:
</p>

<p>
\[ r_{ijk} \sim ZINB(\pi_{ik}, \mu_{ik}, \phi_{ik}) \]
</p>

<p>
against the model:
</p>

<p>
\[ r_{ijk} \sim ZINB(\pi_{ik}, \mu_{ik}, \phi_{i}) \]
</p>

<div class="org-src-container">
<pre class="src src-ipython" id="orga175848"><span class="org-keyword">def</span> <span class="org-function-name">lrt</span>(umi, onehot, design, size_factor):
  <span class="org-variable-name">_</span>, <span class="org-variable-name">_</span>, <span class="org-variable-name">_</span>, <span class="org-variable-name">llik0</span> = fit(
    umi=umi.values.T.astype(np.float32),
    onehot=onehot.astype(np.float32),
    design=design.astype(np.float32),
    size_factor=size_factor.astype(np.float32),
    fit_null=<span class="org-constant">True</span>,
    return_llik=<span class="org-constant">True</span>,
    learning_rate=5e-2,
    max_epochs=4000)
  <span class="org-variable-name">_</span>, <span class="org-variable-name">_</span>, <span class="org-variable-name">_</span>, <span class="org-variable-name">llik1</span> = fit(
    umi=umi.values.T.astype(np.float32),
    onehot=onehot.astype(np.float32),
    design=design.astype(np.float32),
    size_factor=size_factor.astype(np.float32),
    return_llik=<span class="org-constant">True</span>,
    learning_rate=5e-2,
    max_epochs=4000)
  <span class="org-variable-name">T</span> = 2 * (llik1 - llik0)
  <span class="org-keyword">return</span> T, st.chi2(1).logsf(T)
</pre>
</div>
</div>

<div id="outline-container-org9a3e6cd" class="outline-3">
<h3 id="org9a3e6cd">Null calibration via parametric bootstrap</h3>
<div class="outline-text-3" id="text-org9a3e6cd">
<p>
Sample null data from the model, using empirical estimates of the parameters
in the observed data.
</p>

<div class="org-src-container">
<pre class="src src-ipython">&lt;&lt;zinb-imports&gt;&gt;
&lt;&lt;tf-imports&gt;&gt;
&lt;&lt;tf-zinb-impl&gt;&gt;
&lt;&lt;lrt-impl&gt;&gt;
&lt;&lt;sim-impl&gt;&gt;

<span class="org-variable-name">log_mu</span> = pd.read_table(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-qtl/data/density-estimation/design1/zi2-log-mu.txt.gz'</span>, index_col=0, sep=<span class="org-string">' '</span>)
<span class="org-variable-name">log_phi</span> = pd.read_table(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-qtl/data/density-estimation/design1/zi2-log-phi.txt.gz'</span>, index_col=0, sep=<span class="org-string">' '</span>)
<span class="org-variable-name">logodds</span> = pd.read_table(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-qtl/data/density-estimation/design1/zi2-log-mu.txt.gz'</span>, index_col=0, sep=<span class="org-string">' '</span>)
<span class="org-variable-name">params</span> = pd.DataFrame({<span class="org-string">'log_mu'</span>: log_mu[<span class="org-string">'NA18507'</span>],
                       <span class="org-string">'log_phi'</span>: log_phi[<span class="org-string">'NA18507'</span>],
                       <span class="org-string">'logodds'</span>: logodds[<span class="org-string">'NA18507'</span>]},
                      index=log_mu.index)
<span class="org-variable-name">params</span> = params[params[<span class="org-string">'log_mu'</span>] &gt; -10]
<span class="org-variable-name">n</span> = 100
<span class="org-variable-name">umi</span> = pd.DataFrame([simulate(2 * n, size=1e5, log_mu=log_mu, log_phi=log_phi, logodds=logodds)[0][:,0]
                    <span class="org-keyword">for</span> _, (log_mu, log_phi, logodds) <span class="org-keyword">in</span> params.iterrows() <span class="org-keyword">if</span> log_mu &gt; -10], index=params.index)
<span class="org-variable-name">onehot</span> = np.zeros((2 * n, 2))
<span class="org-variable-name">onehot</span>[:n,0] = 1
<span class="org-variable-name">onehot</span>[n:,1] = 1
<span class="org-variable-name">design</span> = np.zeros((2 * n, 1))
<span class="org-variable-name">size_factor</span> = 1e5 * np.ones((2 * n, 1))
<span class="org-variable-name">T</span>, <span class="org-variable-name">P</span> = lrt(umi, onehot, design, size_factor)
pd.DataFrame({<span class="org-string">'chi2'</span>: T, <span class="org-string">'logp'</span>: P}, index=umi.index).to_csv(<span class="org-string">'null-calibration-p.txt.gz'</span>, sep=<span class="org-string">'\t'</span>, compression=<span class="org-string">'gzip'</span>)
</pre>
</div>

<div class="org-src-container">
<pre class="src src-sh">sbatch --partition=gpu --gres=gpu:1 --mem=16G --job-name=tf-lrt-null --output=tf-lrt-null-parametric.out
<span class="org-comment-delimiter">#</span><span class="org-comment">!/bin/bash</span>
<span class="org-builtin">source</span> activate scqtl
python /project2/mstephens/aksarkar/projects/singlecell-qtl/code/tf-lrt-null-parametric.py
</pre>
</div>

<pre class="example">
Submitted batch job 46322399

</pre>

<p>
Read the results.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">null_lrt</span> = pd.read_table(<span class="org-string">'/scratch/midway2/aksarkar/singlecell/power/null-calibration-p.txt.gz'</span>)
<span class="org-variable-name">N</span> = null_lrt.shape[0]
</pre>
</div>

<p>
Report how many genes were simulated.
</p>

<div class="org-src-container">
<pre class="src src-ipython">N
</pre>
</div>

<pre class="example">
1832

</pre>

<p>
Estimate bootstrap CIs for the quantiles.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">B</span> = np.sort(st.chi2(1).rvs((N, 100)), axis=0)
<span class="org-variable-name">ci</span> = np.percentile(B, [2.5, 97.5], axis=1)
</pre>
</div>

<p>
Plot the QQ-plot.
</p>

<div class="org-src-container">
<pre class="src src-ipython">plt.clf()
plt.gcf().set_size_inches(4, 4)
<span class="org-variable-name">grid</span> = st.chi2(1).ppf(np.linspace(0, 1 - 1 / N, N))
plt.scatter(grid, null_lrt[<span class="org-string">'chi2'</span>].sort_values(), c=<span class="org-string">'k'</span>, s=2)
plt.fill_between(grid, ci[0], ci[1], color=<span class="org-string">'k'</span>, alpha=0.1)
<span class="org-variable-name">lim</span> = [0, 1.01 * grid.<span class="org-builtin">max</span>()]
plt.plot(lim, lim, c=<span class="org-string">'r'</span>, lw=1)
plt.xlim(lim)
plt.xlabel(<span class="org-string">'Expected $\chi^2$ statistic'</span>)
<span class="org-variable-name">_</span> = plt.ylabel(<span class="org-string">'Observed $\chi^2$ statistic'</span>)
</pre>
</div>


<div class="figure">
<p><img src="figure/power.org/null-calibration-p.png" alt="null-calibration-p.png">
</p>
</div>
</div>
</div>

<div id="outline-container-org56b8159" class="outline-3">
<h3 id="org56b8159">Power</h3>
<div class="outline-text-3" id="text-org56b8159">
<p>
Sample from the assumed model.
</p>

<div class="org-src-container">
<pre class="src src-ipython">&lt;&lt;zinb-imports&gt;&gt;
&lt;&lt;tf-imports&gt;&gt;
&lt;&lt;tf-zinb-impl&gt;&gt;
&lt;&lt;lrt-impl&gt;&gt;
&lt;&lt;sim-impl&gt;&gt;

<span class="org-variable-name">log_mu</span> = pd.read_table(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-qtl/data/density-estimation/design1/zi2-log-mu.txt.gz'</span>, index_col=0, sep=<span class="org-string">' '</span>)
<span class="org-variable-name">log_phi</span> = pd.read_table(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-qtl/data/density-estimation/design1/zi2-log-phi.txt.gz'</span>, index_col=0, sep=<span class="org-string">' '</span>)
<span class="org-variable-name">logodds</span> = pd.read_table(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-qtl/data/density-estimation/design1/zi2-log-mu.txt.gz'</span>, index_col=0, sep=<span class="org-string">' '</span>)
<span class="org-variable-name">params</span> = pd.DataFrame({<span class="org-string">'log_mu'</span>: log_mu[<span class="org-string">'NA18507'</span>],
                       <span class="org-string">'log_phi'</span>: log_phi[<span class="org-string">'NA18507'</span>],
                       <span class="org-string">'logodds'</span>: logodds[<span class="org-string">'NA18507'</span>]},
                      index=log_mu.index)
<span class="org-variable-name">params</span> = params[params[<span class="org-string">'log_mu'</span>] &gt; -10].sample(n=50)

<span class="org-variable-name">sample_sizes</span> = np.geomspace(1e2, 1e5, 5).astype(<span class="org-builtin">int</span>)
<span class="org-variable-name">log_fold_changes</span> = np.log(np.geomspace(1.1, 2, 5))
<span class="org-variable-name">depths</span> = np.geomspace(1e4, 1e6, 5)

<span class="org-variable-name">result</span> = []
<span class="org-keyword">for</span> num_mols <span class="org-keyword">in</span> depths:
  <span class="org-keyword">for</span> log_fc <span class="org-keyword">in</span> log_fold_changes:
    <span class="org-keyword">for</span> num_samples <span class="org-keyword">in</span> sample_sizes:
      <span class="org-variable-name">umi</span> = []
      <span class="org-keyword">for</span> _, (log_mu, log_phi, logodds) <span class="org-keyword">in</span> params.iterrows():
        umi.append(np.hstack([
          simulate(num_samples, size=num_mols, log_mu=log_mu, log_phi=log_phi, logodds=logodds)[0][:,0],
          simulate(num_samples, size=num_mols, log_mu=log_mu, log_phi=log_phi + log_fc, logodds=logodds)[0][:,0]
        ]))
      <span class="org-variable-name">umi</span> = pd.DataFrame(umi, index=params.index)
      <span class="org-variable-name">onehot</span> = np.zeros((umi.shape[1], 2))
      <span class="org-variable-name">onehot</span>[:num_samples,0] = 1
      <span class="org-variable-name">onehot</span>[num_samples:,1] = 1
      <span class="org-variable-name">design</span> = np.zeros((umi.shape[1], 1))
      <span class="org-variable-name">size_factor</span> = num_mols * np.ones((umi.shape[1], 1))
      <span class="org-variable-name">T</span>, <span class="org-variable-name">P</span> = lrt(umi, onehot, design, size_factor)
      result.append(pd.DataFrame({
        <span class="org-string">'num_mols'</span>: num_mols,
        <span class="org-string">'num_samples'</span>: num_samples,
        <span class="org-string">'log_fold_change'</span>: log_fc,
        <span class="org-string">'chi2'</span>: T,
        <span class="org-string">'logp'</span>: P}))
pd.concat(result).to_csv(<span class="org-string">'lrt-power.txt.gz'</span>, sep=<span class="org-string">'\t'</span>, compression=<span class="org-string">'gzip'</span>) 
</pre>
</div>

<div class="org-src-container">
<pre class="src src-sh">sbatch --partition=gpu --gres=gpu:1 --mem=16G --job-name=tf-lrt-power --output=tf-lrt-power.out
<span class="org-comment-delimiter">#</span><span class="org-comment">!/bin/bash</span>
<span class="org-builtin">source</span> activate scqtl
python /project2/mstephens/aksarkar/projects/singlecell-qtl/code/tf-lrt-power.py
</pre>
</div>

<pre class="example">
Submitted batch job 46367157

</pre>

<div class="org-src-container">
<pre class="src src-sh">sacct -j 46367157 -o Elapsed
</pre>
</div>

<pre class="example">
   Elapsed 
---------- 
  06:59:58 
  06:59:58 
  06:59:58 

</pre>

<p>
Move the results to permanent storage.
</p>

<div class="org-src-container">
<pre class="src src-sh">rsync -FFau /scratch/midway2/aksarkar/singlecell/power /project2/mstephens/aksarkar/projects/singlecell-qtl/data/power/
</pre>
</div>

<p>
Read the results.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">lrt_results</span> = (pd.read_table(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-qtl/data/power/lrt-power.txt.gz'</span>, index_col=0)
             .reindex())
<span class="org-variable-name">features</span> = [<span class="org-string">'num_mols'</span>, <span class="org-string">'num_samples'</span>, <span class="org-string">'log_fold_change'</span>]
<span class="org-variable-name">lrt_power</span> = (lrt_results
             .groupby(features)
             .<span class="org-builtin">apply</span>(<span class="org-keyword">lambda</span> x: (np.exp(x[<span class="org-string">'logp'</span>]) &lt; 0.05).<span class="org-builtin">sum</span>() / x.shape[0])
             .to_frame()
             .reset_index())
</pre>
</div>

<p>
Plot the results.
</p>

<div class="org-src-container">
<pre class="src src-ipython">plt.clf()
plt.gcf().set_size_inches(4, 4)
<span class="org-variable-name">num_mols</span> = 1e5
<span class="org-variable-name">grid</span> = np.log(np.linspace(1.1, 2, 100))
<span class="org-variable-name">groups</span> = <span class="org-builtin">sorted</span>(<span class="org-builtin">set</span>(lrt_power[<span class="org-string">'num_samples'</span>]))
<span class="org-keyword">for</span> i, n <span class="org-keyword">in</span> <span class="org-builtin">enumerate</span>(groups):
  <span class="org-variable-name">color</span> = colorcet.cm[<span class="org-string">'inferno'</span>]((i + 1 / (2 * <span class="org-builtin">len</span>(groups))) / <span class="org-builtin">len</span>(groups))
  <span class="org-variable-name">subset</span> = lrt_power[np.logical_and(lrt_power[<span class="org-string">'num_mols'</span>] == num_mols, lrt_power[<span class="org-string">'num_samples'</span>] == n)]
  plt.plot(np.exp(subset[<span class="org-string">'log_fold_change'</span>]), subset[0], lw=1, marker=<span class="org-string">'.'</span>, ms=8, c=color,
           label=<span class="org-string">'$10^{{{:.1f}}}$'</span>.<span class="org-builtin">format</span>(np.log(n) / np.log(10)))
plt.legend(title=<span class="org-string">'# samples'</span>, frameon=<span class="org-constant">False</span>, loc=<span class="org-string">'center left'</span>, bbox_to_anchor=(1, .5))
plt.xlabel(<span class="org-string">'Fold change in dispersion'</span>)
plt.ylabel(<span class="org-string">'Power at level 0.05'</span>)
</pre>
</div>

<pre class="example">
Text(0,0.5,'Power at level 0.05')

</pre>

<div class="figure">
<p><img src="figure/power.org/lrt-power.png" alt="lrt-power.png">
</p>
</div>

<div class="org-src-container">
<pre class="src src-ipython">plt.clf()
plt.gcf().set_size_inches(4, 4)
<span class="org-variable-name">num_samples</span> = 100
<span class="org-variable-name">groups</span> = <span class="org-builtin">sorted</span>(<span class="org-builtin">set</span>(lrt_power[<span class="org-string">'num_mols'</span>]))
<span class="org-keyword">for</span> i, n <span class="org-keyword">in</span> <span class="org-builtin">enumerate</span>(groups):
  <span class="org-variable-name">color</span> = colorcet.cm[<span class="org-string">'inferno'</span>]((i + 1 / (2 * <span class="org-builtin">len</span>(groups))) / <span class="org-builtin">len</span>(groups))
  <span class="org-variable-name">subset</span> = lrt_power[np.logical_and(lrt_power[<span class="org-string">'num_mols'</span>] == n, lrt_power[<span class="org-string">'num_samples'</span>] == num_samples)]
  plt.plot(np.exp(subset[<span class="org-string">'log_fold_change'</span>]), subset[0], lw=1, marker=<span class="org-string">'.'</span>, ms=8, c=color, label=<span class="org-string">'$10^{{{:.1f}}}$'</span>.<span class="org-builtin">format</span>(np.log(n) / np.log(10)))
plt.legend(title=<span class="org-string">'# molecules'</span>, frameon=<span class="org-constant">False</span>, loc=<span class="org-string">'center left'</span>, bbox_to_anchor=(1, .5))
plt.xlabel(<span class="org-string">'Fold change in dispersion'</span>)
plt.ylabel(<span class="org-string">'Power at level 0.05'</span>)
</pre>
</div>

<pre class="example">
Text(0,0.5,'Power at level 0.05')

</pre>

<div class="figure">
<p><img src="figure/power.org/lrt-power-fixed-samples.png" alt="lrt-power-fixed-samples.png">
</p>
</div>
</div>
</div>
</div>

<div id="outline-container-org875d420" class="outline-2">
<h2 id="org875d420">QTL discovery</h2>
<div class="outline-text-2" id="text-org875d420">
<p>
We assume the phenotype is generated from \(m\) causal effects (out of \(p\)
variants) following a linear model:
</p>

<p>
\[ \theta_i = \sum_j X_{ij} \beta_j + \epsilon_i \]
</p>

<p>
\[ \mathbb{E}[x_i] = 0, \mathbb{V}[x_i] = 1 \]
</p>

<p>
\[ \beta_j \sim N(0, 1) \text{ if \(j\) causal}\]
</p>

<p>
\[ \epsilon_i \sim N(0, \mathbb{V[X \beta]} (1 / h^2 - 1)) \]
</p>

<div class="org-src-container">
<pre class="src src-ipython" id="org2dcef1f"><span class="org-keyword">def</span> <span class="org-function-name">read_dosage</span>(row, dose_file, window=100000):
  <span class="org-variable-name">x</span> = np.array([record[9:] <span class="org-keyword">for</span> record <span class="org-keyword">in</span> dose_file.query(row[<span class="org-string">'chr'</span>], row[<span class="org-string">'start'</span>] - window, row[<span class="org-string">'start'</span>] + window)]).astype(np.<span class="org-builtin">float</span>).T
  <span class="org-variable-name">x</span> -= x.mean(axis=0)
  <span class="org-variable-name">x</span> /= x.std(axis=0)
  <span class="org-keyword">return</span> x

<span class="org-keyword">def</span> <span class="org-function-name">generate_pheno</span>(x, pve, m=<span class="org-constant">None</span>):
  <span class="org-variable-name">n</span>, <span class="org-variable-name">p</span> = x.shape
  <span class="org-keyword">if</span> m <span class="org-keyword">is</span> <span class="org-constant">None</span>:
    <span class="org-variable-name">theta</span> = np.random.normal(size=p)
  <span class="org-keyword">else</span>:
    <span class="org-variable-name">theta</span> = np.zeros(p)
    <span class="org-comment-delimiter"># </span><span class="org-comment">Assume x standardized</span>
    <span class="org-variable-name">theta</span>[np.random.choice(p, m, replace=<span class="org-constant">False</span>)] = np.random.normal(size=m)
  <span class="org-variable-name">y</span> = x.dot(theta)
  <span class="org-variable-name">y</span> += np.random.normal(scale=np.sqrt(y.var() * (1 / pve - 1)), size=n)
  <span class="org-variable-name">y</span> -= y.mean()
  <span class="org-variable-name">y</span> /= y.std()
  <span class="org-keyword">return</span> y.reshape(-1, 1)

<span class="org-variable-name">_sf</span> = st.chi2(1).sf

<span class="org-keyword">def</span> <span class="org-function-name">nominal_test</span>(x, y):
  <span class="org-variable-name">n</span> = y.shape[0]
  <span class="org-comment-delimiter"># </span><span class="org-comment">Assume x standardized, so diag(X' X) = n I</span>
  <span class="org-variable-name">beta</span> = x.T.dot(y) / n
  <span class="org-variable-name">df</span> = n - 1
  <span class="org-variable-name">rss</span> = ((y ** 2).<span class="org-builtin">sum</span>() - beta ** 2 * n)
  <span class="org-variable-name">sigma2</span> = rss / df
  <span class="org-keyword">return</span> _sf(beta ** 2 / sigma2)

<span class="org-keyword">def</span> <span class="org-function-name">beta_llik</span>(theta, x):
  <span class="org-keyword">return</span> -st.beta.logpdf(x, *theta).mean()

<span class="org-keyword">def</span> <span class="org-function-name">permutation_test</span>(x, y, num_trials=100):
  <span class="org-variable-name">pval</span> = nominal_test(x, y).<span class="org-builtin">min</span>()
  <span class="org-variable-name">null_pheno</span> = y.copy()
  <span class="org-variable-name">null_pvals</span> = []
  <span class="org-keyword">for</span> _ <span class="org-keyword">in</span> <span class="org-builtin">range</span>(num_trials):
    np.random.shuffle(null_pheno)
    null_pvals.append(nominal_test(x, null_pheno).<span class="org-builtin">min</span>())
  <span class="org-variable-name">null_pvals</span> = np.array(null_pvals)
  <span class="org-variable-name">theta</span> = np.ones(2)
  <span class="org-variable-name">opt</span> = so.minimize(beta_llik, x0=theta, args=(null_pvals,))
  <span class="org-keyword">if</span> opt.success:
    <span class="org-variable-name">theta</span> = opt.x
  <span class="org-keyword">else</span>:
    <span class="org-comment-delimiter"># </span><span class="org-comment">Method of moments</span>
    <span class="org-variable-name">theta</span> = np.array([1, (1 / null_pvals.mean() - 1)])
    <span class="org-variable-name">theta</span> *= np.square(null_pvals.mean()) * ((1 - null_pvals.mean()) / null_pvals.var() - 1)
  <span class="org-keyword">return</span> st.beta.cdf(pval, *theta)

<span class="org-comment-delimiter"># </span><span class="org-comment">Important: tabix read doesn't appear to be thread-safe</span>
<span class="org-variable-name">read_lock</span> = mp.Lock()

<span class="org-keyword">def</span> <span class="org-function-name">evaluate</span>(num_individuals, num_causal, pve, num_genes=100):
  <span class="org-variable-name">query</span> = eqtls.sample(n=num_genes)
  <span class="org-variable-name">result</span> = []
  <span class="org-keyword">for</span> _, record <span class="org-keyword">in</span> query.iterrows():
    read_lock.acquire()
    <span class="org-keyword">try</span>:
      <span class="org-variable-name">x</span> = read_dosage(record, yri)
    <span class="org-keyword">finally</span>:
      read_lock.release()
    <span class="org-variable-name">keep</span> = np.random.choice(x.shape[0], num_individuals, replace=<span class="org-constant">False</span>)
    <span class="org-variable-name">x</span> = x[keep]
    <span class="org-variable-name">y</span> = generate_pheno(x, pve=pve, m=num_causal)
    <span class="org-variable-name">pval</span> = permutation_test(x, y)
    result.append({
      <span class="org-string">'gene'</span>: record.name,
      <span class="org-string">'num_individuals'</span>: x.shape[0],
      <span class="org-string">'num_snps'</span>: x.shape[1],
      <span class="org-string">'num_causal'</span>: num_causal,
      <span class="org-string">'pve'</span>: pve,
      <span class="org-string">'pval'</span>: pval})
  <span class="org-keyword">return</span> pd.DataFrame.from_dict(result)
</pre>
</div>

<p>
Re-QC dosages for 120 YRI individuals.
</p>

<div class="org-src-container">
<pre class="src src-sh">sbatch --partition=broadwl --job-name=reprocess-dosage --out=reprocess-dosage.out
<span class="org-comment-delimiter">#</span><span class="org-comment">!/bin/bash</span>
<span class="org-builtin">set</span> -e
<span class="org-builtin">source</span> activate scqtl
python /project2/mstephens/aksarkar/projects/singlecell-qtl/code/reprocess-dosage.py
bgzip dosages.vcf
tabix dosages.vcf.gz
</pre>
</div>

<pre class="example">
Submitted batch job 46369493

</pre>

<p>
Run the power calculation on 28 CPUs.
</p>

<div class="org-src-container">
<pre class="src src-ipython">&lt;&lt;power-imports&gt;&gt;
&lt;&lt;eqtl-sim-impl&gt;&gt;

<span class="org-variable-name">yri</span> = tabix.<span class="org-builtin">open</span>(<span class="org-string">'/scratch/midway2/aksarkar/singlecell/scqtl-mapping/dosages.vcf.gz'</span>)
<span class="org-comment-delimiter"># </span><span class="org-comment">Restrict to genes where we previously successfully mapped eQTLs</span>
<span class="org-variable-name">eqtls</span> = pd.read_table(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-qtl/data/scqtl-mapping/pooled.txt.gz'</span>, sep=<span class="org-string">' '</span>, index_col=0).dropna()

<span class="org-variable-name">args</span> = [(n, m, pve)
        <span class="org-keyword">for</span> n <span class="org-keyword">in</span> (53, 75, 100, 120)
        <span class="org-keyword">for</span> m <span class="org-keyword">in</span> (1, <span class="org-constant">None</span>)
        <span class="org-keyword">for</span> pve <span class="org-keyword">in</span> np.geomspace(.01, 0.5, 5)]

np.random.seed(0)
<span class="org-keyword">with</span> mp.Pool() <span class="org-keyword">as</span> pool:
  <span class="org-variable-name">result</span> = pool.starmap(evaluate, args)
pd.concat(result).to_csv(<span class="org-string">'eqtl-power.txt.gz'</span>, compression=<span class="org-string">'gzip'</span>, sep=<span class="org-string">'\t'</span>)
</pre>
</div>

<div class="org-src-container">
<pre class="src src-sh">sbatch --partition=broadwl -n1 -c28 --exclusive --mem=16G --job-name=eqtl-power --output=eqtl-power.out
<span class="org-comment-delimiter">#</span><span class="org-comment">!/bin/bash</span>
<span class="org-builtin">source</span> activate scqtl
python /project2/mstephens/aksarkar/projects/singlecell-qtl/code/eqtl-power.py
rsync -FFau ./ /project2/mstephens/aksarkar/projects/singlecell-qtl/data/power/
</pre>
</div>

<pre class="example">
Submitted batch job 46381725

</pre>

<div class="org-src-container">
<pre class="src src-sh">sacct -j 46372142 -o Elapsed
</pre>
</div>

<pre class="example">
   Elapsed 
---------- 
  00:21:47 
  00:21:47 
  00:21:47 

</pre>

<p>
Read the results.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">qtl_results</span> = pd.read_table(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-qtl/data/power/eqtl-power.txt.gz'</span>, index_col=0).fillna(-1)
<span class="org-variable-name">qtl_power</span> = qtl_results.groupby([<span class="org-string">'num_individuals'</span>, <span class="org-string">'num_causal'</span>, <span class="org-string">'pve'</span>]).<span class="org-builtin">apply</span>(<span class="org-keyword">lambda</span> x: (x[<span class="org-string">'pval'</span>] &lt; 0.05).<span class="org-builtin">sum</span>() / x.shape[0]).reset_index().rename(columns={0: <span class="org-string">'power'</span>})
</pre>
</div>

<p>
Plot the results.
</p>

<div class="org-src-container">
<pre class="src src-ipython">plt.clf()
<span class="org-variable-name">fig</span>, <span class="org-variable-name">ax</span> = plt.subplots(1, 2, sharey=<span class="org-constant">True</span>)
fig.set_size_inches(6, 3)
<span class="org-keyword">for</span> i, m <span class="org-keyword">in</span> <span class="org-builtin">enumerate</span>([1, -1]):
  <span class="org-keyword">for</span> n <span class="org-keyword">in</span> [53, 75, 100, 120]:
    <span class="org-variable-name">subset</span> = np.logical_and(qtl_power[<span class="org-string">'num_individuals'</span>] == n, qtl_power[<span class="org-string">'num_causal'</span>] == m)
    <span class="org-variable-name">color</span> = colorcet.cm[<span class="org-string">'inferno'</span>](.9 * (n - 53) / (120 - 53))
    ax[i].plot(qtl_power.loc[subset, <span class="org-string">'pve'</span>], qtl_power.loc[subset, <span class="org-string">'power'</span>], marker=<span class="org-string">'.'</span>, c=color, lw=1, ms=8, label=n)
ax[0].set_title(<span class="org-string">'1 causal'</span>)
ax[1].set_title(<span class="org-string">'Infinitesimal'</span>)
ax[1].set_xlabel(<span class="org-string">'Proportion of variance explained'</span>)
ax[0].set_ylabel(<span class="org-string">'Power at level 0.05'</span>)
ax[-1].legend(title=<span class="org-string">'# individuals'</span>, frameon=<span class="org-constant">False</span>, loc=<span class="org-string">'center left'</span>, bbox_to_anchor=(1, .5))
fig.tight_layout()
</pre>
</div>


<div class="figure">
<p><img src="figure/power.org/eqtl-power.png" alt="eqtl-power.png">
</p>
</div>

<p>
Look at the empirical CDF of <i>cis</i>-heritability of gene expression (<a href="http://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.1006423">Wheeler
et al 2016</a>). The results are <a href="https://github.com/WheelerLab/GenArchDB">available on github</a>.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">import</span> sqlite3
<span class="org-keyword">with</span> sqlite3.connect(<span class="org-string">'/project2/mstephens/aksarkar/.local/src/GenArchDB/genarch.db'</span>) <span class="org-keyword">as</span> conn:
  <span class="org-variable-name">cis_h2g</span> = pd.read_sql(<span class="org-string">'select * from results where tissue == "DGN-WB";'</span>, conn)
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython">plt.clf()
plt.gcf().set_size_inches(3, 3)
<span class="org-variable-name">n</span>, <span class="org-variable-name">bins</span>, <span class="org-variable-name">_</span> = plt.hist(cis_h2g[<span class="org-string">'pve'</span>], bins=100, histtype=<span class="org-string">'step'</span>, color=<span class="org-string">'k'</span>, cumulative=<span class="org-constant">True</span>)
plt.axvline(x=0.4, c=<span class="org-string">'r'</span>, ls=<span class="org-string">':'</span>, lw=1)
plt.axhline(y=n[np.where(bins &gt;= 0.4)[0].<span class="org-builtin">min</span>()], c=<span class="org-string">'r'</span>, ls=<span class="org-string">':'</span>, lw=1)
plt.xlabel(<span class="org-string">'Proportion of variance explained'</span>)
plt.ylabel(<span class="org-string">'Cumulative number of genes'</span>)
</pre>
</div>

<pre class="example">
Text(0,0.5,'Cumulative number of genes')

</pre>

<div class="figure">
<p><img src="figure/power.org/eqtl-heritability.png" alt="eqtl-heritability.png">
</p>
</div>

<p>
Estimate how many genes we had 80% power to detect.
</p>

<div class="org-src-container">
<pre class="src src-ipython">cis_h2g.shape[0] - n[np.where(bins &gt;= 0.4)[0].<span class="org-builtin">min</span>()]
</pre>
</div>

<pre class="example">
921.0

</pre>

<p>
The key insight is that the single cell experiment size determines the
standard error of the estimator, which can be thought of as measurement
error.
</p>

<p>
\[ \hat\theta \sim N(\theta, \sigma^2) \]
</p>

<p>
This is the same as changing effective PVE:
</p>

<p>
\[ h^2_{\mathrm{eff}} = \frac{h^2}{1 + \sigma^2} \]
</p>

<p>
We estimate the standard error as a function of the experiment size from the
simulation.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">sim_results</span> = pd.read_table(<span class="org-string">'/scratch/midway2/aksarkar/singlecell/density-estimation/simulation.txt.gz'</span>, index_col=0)
<span class="org-variable-name">mu_pass</span> = sim_results[<span class="org-string">'log_mu'</span>] &gt; -10
<span class="org-variable-name">pi_pass</span> = sim_results[<span class="org-string">'logodds'</span>] &lt;= 0
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython">sim_results[pi_pass].groupby([<span class="org-string">'num_samples'</span>, <span class="org-string">'num_mols'</span>, <span class="org-string">'log_mu'</span>]).<span class="org-builtin">apply</span>(<span class="org-keyword">lambda</span> x: pd.Series(1 / (1 + np.var(x[<span class="org-string">'log_mu_hat'</span>])))).reset_index()
</pre>
</div>

<pre class="example">
num_samples  num_mols  log_mu         0
0            95    114026     -12  0.943735
1            95    114026     -11  0.975032
2            95    114026     -10  0.984130
3            95    114026      -9  0.988944
4            95    114026      -8  0.990610
5            95    114026      -7  0.995300
6            95    114026      -6  0.992922
7         10000    114026     -12  0.970102
8         10000    114026     -11  0.980160
9         10000    114026     -10  0.992239
10        10000    114026      -9  0.995632
11        10000    114026      -8  0.998242
12        10000    114026      -7  0.999954
13        10000    114026      -6  0.999969
</pre>

<div class="org-src-container">
<pre class="src src-ipython">sim_results[np.logical_and(mu_pass, pi_pass)].groupby([<span class="org-string">'num_samples'</span>, <span class="org-string">'num_mols'</span>, <span class="org-string">'log_phi'</span>]).<span class="org-builtin">apply</span>(<span class="org-keyword">lambda</span> x: 1 / (1 + np.var(x[<span class="org-string">'log_phi_hat'</span>]))).reset_index()
</pre>
</div>

<pre class="example">
num_samples  num_mols  log_phi         0
0           95    114026       -4  0.867270
1           95    114026       -3  0.916724
2           95    114026       -2  0.950310
3           95    114026       -1  0.969675
4           95    114026        0  0.904745
5        10000    114026       -4  0.996827
6        10000    114026       -3  0.999260
7        10000    114026       -2  0.999752
8        10000    114026       -1  0.999300
9        10000    114026        0  0.933499
</pre>
</div>
</div>

<div id="outline-container-org79a18cf" class="outline-2">
<h2 id="org79a18cf">True discovery proportion</h2>
<div class="outline-text-2" id="text-org79a18cf">
<p>
Power characterizes a single test; however, in a typical study we are
performing multiple hypothesis tests and controlling the FDR.
</p>

<p>
Read the QTLs and normalized expression matrices.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">prefix</span> = <span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-qtl/data/scqtl-mapping/'</span>
<span class="org-variable-name">qtls</span> = {pheno: [pd.read_table(<span class="org-string">'{}/{}.txt.gz'</span>.<span class="org-builtin">format</span>(prefix, pheno), sep=<span class="org-string">' '</span>),
                pd.read_table(<span class="org-string">'{}/{}.bed.gz'</span>.<span class="org-builtin">format</span>(prefix, pheno)).set_index(<span class="org-string">'pid'</span>).<span class="org-builtin">filter</span>(like=<span class="org-string">'NA'</span>, axis=<span class="org-string">'columns'</span>)]
        <span class="org-keyword">for</span> pheno <span class="org-keyword">in</span> [<span class="org-string">'bulk'</span>, <span class="org-string">'pooled'</span>, <span class="org-string">'log_mu'</span>, <span class="org-string">'log_phi'</span>, <span class="org-string">'mean'</span>, <span class="org-string">'variance'</span>, <span class="org-string">'cv'</span>, <span class="org-string">'fano'</span>]}
</pre>
</div>

<p>
Estimate \(\pi_0\) via Storey's procedure.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">import</span> rpy2.robjects.packages
<span class="org-keyword">import</span> rpy2.robjects.pandas2ri

<span class="org-variable-name">qvalue</span> = rpy2.robjects.packages.importr(<span class="org-string">'qvalue'</span>)

rpy2.robjects.pandas2ri.activate()

pd.DataFrame.from_dict({k: np.array(qvalue.pi0est(qtls[k][0][<span class="org-string">'p_beta'</span>])[0]) <span class="org-keyword">for</span> k <span class="org-keyword">in</span> qtls}, orient=<span class="org-string">'index'</span>).sort_values(0)
</pre>
</div>

<pre class="example">
0
bulk      0.534070
log_mu    0.728419
pooled    0.739354
mean      0.741717
variance  0.747747
fano      0.871923
cv        0.895591
log_phi   0.971077
</pre>

<p>
Plot the p-value distribution.
</p>

<div class="org-src-container">
<pre class="src src-ipython">plt.clf()
plt.gcf().set_size_inches(3, 3)
<span class="org-variable-name">grid</span> = np.linspace(0, 1, 100)
plt.hist(qtls[<span class="org-string">'bulk'</span>][0][<span class="org-string">'p_beta'</span>].dropna(), bins=100, histtype=<span class="org-string">'step'</span>, color=<span class="org-string">'k'</span>, lw=1, label=<span class="org-string">'Bulk'</span>)
plt.hist(qtls[<span class="org-string">'pooled'</span>][0][<span class="org-string">'p_beta'</span>].dropna(), bins=100, histtype=<span class="org-string">'step'</span>, color=<span class="org-string">'r'</span>, lw=1, label=<span class="org-string">'Single cell'</span>)
plt.legend(frameon=<span class="org-constant">False</span>)
plt.xlabel(<span class="org-string">'$p$-value'</span>)
plt.ylabel(<span class="org-string">'Number of genes'</span>)
</pre>
</div>

<pre class="example">
Text(0,0.5,'Number of genes')

</pre>

<div class="figure">
<p><img src="figure/power.org/p-value-dist.png" alt="p-value-dist.png">
</p>
</div>
</div>
</div>
</div>
<div id="postamble" class="status">
<p class="author">Author: Abhishek Sarkar</p>
<p class="date">Created: 2018-06-04 Mon 14:11</p>
<p class="validation"><a href="http://validator.w3.org/check?uri=referer">Validate</a></p>
</div>
</body>
</html>
