<!DOCTYPE html>
<html lang="en">
<head>
<!-- 2018-01-22 Mon 11:04 -->
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
<title>QTL mapping pipeline</title>
<meta name="generator" content="Org mode">
<meta name="author" content="Abhishek Sarkar">
<style type="text/css">
 <!--/*--><![CDATA[/*><!--*/
  .title  { text-align: center;
             margin-bottom: .2em; }
  .subtitle { text-align: center;
              font-size: medium;
              font-weight: bold;
              margin-top:0; }
  .todo   { font-family: monospace; color: red; }
  .done   { font-family: monospace; color: green; }
  .priority { font-family: monospace; color: orange; }
  .tag    { background-color: #eee; font-family: monospace;
            padding: 2px; font-size: 80%; font-weight: normal; }
  .timestamp { color: #bebebe; }
  .timestamp-kwd { color: #5f9ea0; }
  .org-right  { margin-left: auto; margin-right: 0px;  text-align: right; }
  .org-left   { margin-left: 0px;  margin-right: auto; text-align: left; }
  .org-center { margin-left: auto; margin-right: auto; text-align: center; }
  .underline { text-decoration: underline; }
  #postamble p, #preamble p { font-size: 90%; margin: .2em; }
  p.verse { margin-left: 3%; }
  pre {
    border: 1px solid #ccc;
    box-shadow: 3px 3px 3px #eee;
    padding: 8pt;
    font-family: monospace;
    overflow: auto;
    margin: 1.2em;
  }
  pre.src {
    position: relative;
    overflow: visible;
    padding-top: 1.2em;
  }
  pre.src:before {
    display: none;
    position: absolute;
    background-color: white;
    top: -10px;
    right: 10px;
    padding: 3px;
    border: 1px solid black;
  }
  pre.src:hover:before { display: inline;}
  /* Languages per Org manual */
  pre.src-asymptote:before { content: 'Asymptote'; }
  pre.src-awk:before { content: 'Awk'; }
  pre.src-C:before { content: 'C'; }
  /* pre.src-C++ doesn't work in CSS */
  pre.src-clojure:before { content: 'Clojure'; }
  pre.src-css:before { content: 'CSS'; }
  pre.src-D:before { content: 'D'; }
  pre.src-ditaa:before { content: 'ditaa'; }
  pre.src-dot:before { content: 'Graphviz'; }
  pre.src-calc:before { content: 'Emacs Calc'; }
  pre.src-emacs-lisp:before { content: 'Emacs Lisp'; }
  pre.src-fortran:before { content: 'Fortran'; }
  pre.src-gnuplot:before { content: 'gnuplot'; }
  pre.src-haskell:before { content: 'Haskell'; }
  pre.src-hledger:before { content: 'hledger'; }
  pre.src-java:before { content: 'Java'; }
  pre.src-js:before { content: 'Javascript'; }
  pre.src-latex:before { content: 'LaTeX'; }
  pre.src-ledger:before { content: 'Ledger'; }
  pre.src-lisp:before { content: 'Lisp'; }
  pre.src-lilypond:before { content: 'Lilypond'; }
  pre.src-lua:before { content: 'Lua'; }
  pre.src-matlab:before { content: 'MATLAB'; }
  pre.src-mscgen:before { content: 'Mscgen'; }
  pre.src-ocaml:before { content: 'Objective Caml'; }
  pre.src-octave:before { content: 'Octave'; }
  pre.src-org:before { content: 'Org mode'; }
  pre.src-oz:before { content: 'OZ'; }
  pre.src-plantuml:before { content: 'Plantuml'; }
  pre.src-processing:before { content: 'Processing.js'; }
  pre.src-python:before { content: 'Python'; }
  pre.src-R:before { content: 'R'; }
  pre.src-ruby:before { content: 'Ruby'; }
  pre.src-sass:before { content: 'Sass'; }
  pre.src-scheme:before { content: 'Scheme'; }
  pre.src-screen:before { content: 'Gnu Screen'; }
  pre.src-sed:before { content: 'Sed'; }
  pre.src-sh:before { content: 'shell'; }
  pre.src-sql:before { content: 'SQL'; }
  pre.src-sqlite:before { content: 'SQLite'; }
  /* additional languages in org.el's org-babel-load-languages alist */
  pre.src-forth:before { content: 'Forth'; }
  pre.src-io:before { content: 'IO'; }
  pre.src-J:before { content: 'J'; }
  pre.src-makefile:before { content: 'Makefile'; }
  pre.src-maxima:before { content: 'Maxima'; }
  pre.src-perl:before { content: 'Perl'; }
  pre.src-picolisp:before { content: 'Pico Lisp'; }
  pre.src-scala:before { content: 'Scala'; }
  pre.src-shell:before { content: 'Shell Script'; }
  pre.src-ebnf2ps:before { content: 'ebfn2ps'; }
  /* additional language identifiers per "defun org-babel-execute"
       in ob-*.el */
  pre.src-cpp:before  { content: 'C++'; }
  pre.src-abc:before  { content: 'ABC'; }
  pre.src-coq:before  { content: 'Coq'; }
  pre.src-groovy:before  { content: 'Groovy'; }
  /* additional language identifiers from org-babel-shell-names in
     ob-shell.el: ob-shell is the only babel language using a lambda to put
     the execution function name together. */
  pre.src-bash:before  { content: 'bash'; }
  pre.src-csh:before  { content: 'csh'; }
  pre.src-ash:before  { content: 'ash'; }
  pre.src-dash:before  { content: 'dash'; }
  pre.src-ksh:before  { content: 'ksh'; }
  pre.src-mksh:before  { content: 'mksh'; }
  pre.src-posh:before  { content: 'posh'; }
  /* Additional Emacs modes also supported by the LaTeX listings package */
  pre.src-ada:before { content: 'Ada'; }
  pre.src-asm:before { content: 'Assembler'; }
  pre.src-caml:before { content: 'Caml'; }
  pre.src-delphi:before { content: 'Delphi'; }
  pre.src-html:before { content: 'HTML'; }
  pre.src-idl:before { content: 'IDL'; }
  pre.src-mercury:before { content: 'Mercury'; }
  pre.src-metapost:before { content: 'MetaPost'; }
  pre.src-modula-2:before { content: 'Modula-2'; }
  pre.src-pascal:before { content: 'Pascal'; }
  pre.src-ps:before { content: 'PostScript'; }
  pre.src-prolog:before { content: 'Prolog'; }
  pre.src-simula:before { content: 'Simula'; }
  pre.src-tcl:before { content: 'tcl'; }
  pre.src-tex:before { content: 'TeX'; }
  pre.src-plain-tex:before { content: 'Plain TeX'; }
  pre.src-verilog:before { content: 'Verilog'; }
  pre.src-vhdl:before { content: 'VHDL'; }
  pre.src-xml:before { content: 'XML'; }
  pre.src-nxml:before { content: 'XML'; }
  /* add a generic configuration mode; LaTeX export needs an additional
     (add-to-list 'org-latex-listings-langs '(conf " ")) in .emacs */
  pre.src-conf:before { content: 'Configuration File'; }

  table { border-collapse:collapse; }
  caption.t-above { caption-side: top; }
  caption.t-bottom { caption-side: bottom; }
  td, th { vertical-align:top;  }
  th.org-right  { text-align: center;  }
  th.org-left   { text-align: center;   }
  th.org-center { text-align: center; }
  td.org-right  { text-align: right;  }
  td.org-left   { text-align: left;   }
  td.org-center { text-align: center; }
  dt { font-weight: bold; }
  .footpara { display: inline; }
  .footdef  { margin-bottom: 1em; }
  .figure { padding: 1em; }
  .figure p { text-align: center; }
  .inlinetask {
    padding: 10px;
    border: 2px solid gray;
    margin: 10px;
    background: #ffffcc;
  }
  #org-div-home-and-up
   { text-align: right; font-size: 70%; white-space: nowrap; }
  textarea { overflow-x: auto; }
  .linenr { font-size: smaller }
  .code-highlighted { background-color: #ffff00; }
  .org-info-js_info-navigation { border-style: none; }
  #org-info-js_console-label
    { font-size: 10px; font-weight: bold; white-space: nowrap; }
  .org-info-js_search-highlight
    { background-color: #ffff00; color: #000000; font-weight: bold; }
  .org-svg { width: 90%; }
  /*]]>*/-->
</style>
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<link rel="stylesheet" type="text/css" href="https://www.pirilampo.org/styles/readtheorg/css/htmlize.css"/>
<style type="text/css">body {width: 60em; margin:auto} pre.src {overflow:auto}</style>
<script type="text/javascript">
/*
@licstart  The following is the entire license notice for the
JavaScript code in this tag.

Copyright (C) 2012-2017 Free Software Foundation, Inc.

The JavaScript code in this tag is free software: you can
redistribute it and/or modify it under the terms of the GNU
General Public License (GNU GPL) as published by the Free Software
Foundation, either version 3 of the License, or (at your option)
any later version.  The code is distributed WITHOUT ANY WARRANTY;
without even the implied warranty of MERCHANTABILITY or FITNESS
FOR A PARTICULAR PURPOSE.  See the GNU GPL for more details.

As additional permission under GNU GPL version 3 section 7, you
may distribute non-source (e.g., minimized or compacted) forms of
that code without the copy of the GNU GPL normally required by
section 4, provided you include this license notice and a URL
through which recipients can access the Corresponding Source.


@licend  The above is the entire license notice
for the JavaScript code in this tag.
*/
<!--/*--><![CDATA[/*><!--*/
 function CodeHighlightOn(elem, id)
 {
   var target = document.getElementById(id);
   if(null != target) {
     elem.cacheClassElem = elem.className;
     elem.cacheClassTarget = target.className;
     target.className = "code-highlighted";
     elem.className   = "code-highlighted";
   }
 }
 function CodeHighlightOff(elem, id)
 {
   var target = document.getElementById(id);
   if(elem.cacheClassElem)
     elem.className = elem.cacheClassElem;
   if(elem.cacheClassTarget)
     target.className = elem.cacheClassTarget;
 }
/*]]>*///-->
</script>
<script type="text/x-mathjax-config">
    MathJax.Hub.Config({
        displayAlign: "center",
        displayIndent: "0em",

        "HTML-CSS": { scale: 100,
                        linebreaks: { automatic: "false" },
                        webFont: "TeX"
                       },
        SVG: {scale: 100,
              linebreaks: { automatic: "false" },
              font: "TeX"},
        NativeMML: {scale: 100},
        TeX: { equationNumbers: {autoNumber: "AMS"},
               MultLineWidth: "85%",
               TagSide: "right",
               TagIndent: ".8em"
             }
});
</script>
<script type="text/javascript"
        src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-AMS_HTML"></script>
</head>
<body>
<div id="content">
<h1 class="title">QTL mapping pipeline</h1>
<div id="table-of-contents">
<h2>Table of Contents</h2>
<div id="text-table-of-contents">
<ul>
<li><a href="#org5ed2a04">Introduction</a></li>
<li><a href="#org94b2817">Genotype processing</a></li>
<li><a href="#org051ecfa">Mean-QTL calling</a></li>
<li><a href="#org018b84e">Dispersion-QTL calling</a></li>
</ul>
</div>
</div>

<div id="outline-container-org5ed2a04" class="outline-2">
<h2 id="org5ed2a04">Introduction</h2>
<div class="outline-text-2" id="text-org5ed2a04">
<p>
Here, we set up the infrastructure to call mean and variance QTLs.
</p>

<p>
We previously <a href="zinb.html">estimated means and dispersions per individual, per gene</a>. We
now fit log linear models for each:
</p>

<p>
\[ \ln \mu_{ik} = X \beta_\mu + \epsilon_\mu \]
</p>

<p>
\[ \ln \phi_{ik} = X \beta_\phi + \epsilon_\phi \]
</p>

<p>
where \(\mu_{ik}, \phi_{ik}\) are the mean and dispersion for individual
\(i\), gene \(k\), \(X\) is the genotype matrix (\(n \times 1)\), and
\(\beta_\mu, \beta_\phi\) are the effect sizes (scalar).
</p>
</div>
</div>

<div id="outline-container-org94b2817" class="outline-2">
<h2 id="org94b2817">Genotype processing</h2>
<div class="outline-text-2" id="text-org94b2817">
<p>
<code>QTLtools</code> reported 0 variants in cis for an unexpected number of genes when
using the provided genotype file, so reprocess the genotypes to fix it.
</p>

<div class="org-src-container">
<pre class="src src-sh">sbatch --partition=broadwl --job-name process-geno
<span class="org-comment-delimiter">#</span><span class="org-comment">!/bin/bash</span>
<span class="org-builtin">set</span> -e
<span class="org-builtin">source</span> activate scqtl
zcat /project/compbio/jointLCLs/genotype/hg19/YRI/vcf/chr{?,??}.hg19.vcf.gz | awk <span class="org-string">'BEGIN {print "##fileformat=VCFv4.1"} NR == 1 || ! /CHROM/ {print}'</span> | bgzip &gt;genotypes.vcf.gz
tabix genotypes.vcf.gz
</pre>
</div>
</div>
</div>

<div id="outline-container-org051ecfa" class="outline-2">
<h2 id="org051ecfa">Mean-QTL calling</h2>
<div class="outline-text-2" id="text-org051ecfa">
<p>
Read the estimated parameters (phenotypes) then center them.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">mean</span> = (pd.read_table(<span class="org-string">'/scratch/midway2/aksarkar/singlecell/mean2.txt.gz'</span>, sep=<span class="org-string">' '</span>, index_col=0)
        .transform(<span class="org-keyword">lambda</span> x: x - x.mean(), axis=1))
</pre>
</div>

<p>
Write out the <a href="https://qtltools.github.io/qtltools/pages/input_files.html">phenotype file</a> for <code>qtltools</code>.
</p>

<div class="org-src-container">
<pre class="src src-ipython" id="org6264062"><span class="org-variable-name">gene_info</span> = (pd.read_table(<span class="org-string">'/home/aksarkar/projects/singlecell-qtl/data/scqtl-genes.txt.gz'</span>)
             .set_index(<span class="org-string">'gene'</span>)
             .query(<span class="org-string">'source == "H. sapiens"'</span>)
             .query(<span class="org-string">'chr != "hsX"'</span>)
             .query(<span class="org-string">'chr != "hsY"'</span>)
             .query(<span class="org-string">'chr != "hsMT"'</span>))
gene_info.head()
</pre>
</div>

<pre class="example">
chr      start        end      name strand      source
gene
ENSG00000000419  hs20   49551404   49575092      DPM1      -  H. sapiens
ENSG00000000457   hs1  169818772  169863408     SCYL3      -  H. sapiens
ENSG00000000460   hs1  169631245  169823221  C1orf112      +  H. sapiens
ENSG00000000938   hs1   27938575   27961788       FGR      -  H. sapiens
ENSG00000000971   hs1  196621008  196716634       CFH      +  H. sapiens
</pre>

<div class="org-src-container">
<pre class="src src-ipython" id="org042c79c"><span class="org-keyword">def</span> <span class="org-function-name">qtltools_format</span>(row):
  <span class="org-variable-name">row</span>[<span class="org-string">'#Chr'</span>] = <span class="org-string">'chr{}'</span>.<span class="org-builtin">format</span>(row[<span class="org-string">'chr'</span>][2:])
  <span class="org-variable-name">row</span>[<span class="org-string">'gid'</span>] = row.name
  <span class="org-variable-name">row</span>[<span class="org-string">'pid'</span>] = row.name
  <span class="org-keyword">return</span> row

<span class="org-keyword">def</span> <span class="org-function-name">write_pheno_file</span>(pheno, gene_info, output_file):
  (gene_info
   .<span class="org-builtin">apply</span>(qtltools_format, axis=1)
   .merge(pheno, left_index=<span class="org-constant">True</span>, right_index=<span class="org-constant">True</span>)
   .to_csv(output_file,
           sep=<span class="org-string">'\t'</span>,
           columns=[<span class="org-string">'#Chr'</span>, <span class="org-string">'start'</span>, <span class="org-string">'end'</span>, <span class="org-string">'pid'</span>, <span class="org-string">'gid'</span>, <span class="org-string">'strand'</span>] + <span class="org-builtin">list</span>(pheno.columns),
           header=<span class="org-constant">True</span>,
           index=<span class="org-constant">False</span>,
           index_label=<span class="org-constant">False</span>)
  )
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython">write_pheno_file(mean, gene_info, <span class="org-string">'/scratch/midway2/aksarkar/singlecell/mean.bed'</span>)
</pre>
</div>

<p>
Index the phenotype file.
</p>

<div class="org-src-container">
<pre class="src src-sh" id="orge80d324"><span class="org-builtin">export</span> <span class="org-variable-name">input</span>=$<span class="org-variable-name">input</span>
sbatch --partition=$<span class="org-variable-name">partition</span> --wait
<span class="org-comment-delimiter">#</span><span class="org-comment">!/bin/bash</span>
sort -k1,1 -k2,2n -k3,3n $<span class="org-variable-name">input</span> | bgzip &gt;$<span class="org-variable-name">input</span>.gz
tabix -p bed $<span class="org-variable-name">input</span>.gz
</pre>
</div>

<p>
Run the QTL mapping, using the approximate permutation test implemented in
<code>QTLtools</code>.
</p>

<div class="org-src-container">
<pre class="src src-sh" id="org5a3e21d"><span class="org-builtin">export</span> <span class="org-variable-name">pheno</span>=$<span class="org-variable-name">pheno</span>
sbatch --partition=$<span class="org-variable-name">partition</span> -N1 -c16 -J $<span class="org-variable-name">pheno</span>-qtl -o $<span class="org-variable-name">pheno</span>-qtl.log
<span class="org-comment-delimiter">#</span><span class="org-comment">!/bin/bash</span>
<span class="org-builtin">source</span> activate scqtl
module load parallel
parallel -j16 qtltools cis --vcf /project2/gilad/singlecell-qtl/bulk/genotypes.vcf.gz --bed $<span class="org-variable-name">pheno</span>.bed.gz --permute=100 --chunk {#} 100 --out $<span class="org-variable-name">pheno</span>-qtl.{#}.txt ::: $(<span class="org-sh-quoted-exec">seq</span> 1 100)
</pre>
</div>

<p>
Read the results.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">def</span> <span class="org-function-name">read_qtltools_output</span>(pheno):
  <span class="org-variable-name">file_names</span> = [<span class="org-string">'{}-qtl.{}.txt'</span>.<span class="org-builtin">format</span>(pheno, i) <span class="org-keyword">for</span> i <span class="org-keyword">in</span> <span class="org-builtin">range</span>(1, 101)]
  <span class="org-keyword">return</span> (pd.concat([pd.read_table(f, header=<span class="org-constant">None</span>, sep=<span class="org-string">' '</span>)
                     <span class="org-keyword">for</span> f <span class="org-keyword">in</span> file_names <span class="org-keyword">if</span> os.path.exists(f) <span class="org-keyword">and</span>
                     os.path.getsize(f) &gt; 0])
          .rename(columns={i: x <span class="org-keyword">for</span> i, x <span class="org-keyword">in</span> <span class="org-builtin">enumerate</span>(
            [<span class="org-string">'gene'</span>, <span class="org-string">'chr'</span>, <span class="org-string">'start'</span>, <span class="org-string">'end'</span>, <span class="org-string">'strand'</span>, <span class="org-string">'num_vars'</span>,
             <span class="org-string">'distance'</span>, <span class="org-string">'id'</span>, <span class="org-string">'var_chr'</span>, <span class="org-string">'var_start'</span>, <span class="org-string">'var_end'</span>, <span class="org-string">'df'</span>,
             <span class="org-string">'dummy'</span>, <span class="org-string">'a'</span>, <span class="org-string">'b'</span>, <span class="org-string">'p_nominal'</span>, <span class="org-string">'beta'</span>, <span class="org-string">'p_empirical'</span>, <span class="org-string">'p_beta'</span>]
          )})
          .sort_values(<span class="org-string">'p_beta'</span>))
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">mean_qtls</span> = read_qtltools_output(<span class="org-string">'mean'</span>)
</pre>
</div>

<p>
Apply the Benjamini-Hochberg procedure:
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">def</span> <span class="org-function-name">bh</span>(df):
  <span class="org-keyword">return</span> df[<span class="org-string">'p_beta'</span>] &lt; .05 * np.arange(1, df.shape[0] + 1) / df.shape[0]
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">keep_mean_qtls</span> = bh(mean_qtls)
keep_mean_qtls.<span class="org-builtin">sum</span>()
</pre>
</div>

<pre class="example">
35

</pre>

<div class="org-src-container">
<pre class="src src-ipython">mean_qtls[keep_mean_qtls].merge(gene_info, left_on=<span class="org-string">'gene'</span>, right_index=<span class="org-constant">True</span>)[[<span class="org-string">'gene'</span>, <span class="org-string">'name'</span>, <span class="org-string">'id'</span>, <span class="org-string">'beta'</span>, <span class="org-string">'p_beta'</span>]]
</pre>
</div>

<pre class="example">
gene      name           id      beta        p_beta
6   ENSG00000163719    MTMR14    rs1438641 -0.178056  1.113220e-85
37  ENSG00000162623      TYW3   rs28602706 -0.469924  3.853500e-64
2   ENSG00000165323      FAT3   rs11019873  0.304741  5.130510e-28
19  ENSG00000013375      PGM3   rs76411140 -0.413721  1.935500e-26
25  ENSG00000136261      BZW2  rs140111609  0.229461  2.138450e-22
60  ENSG00000133884      DPF2  rs116180923 -0.892610  1.357080e-12
31  ENSG00000101003     GINS1  rs117553461 -0.155267  1.357910e-11
23  ENSG00000145604      SKP2   rs10063618 -0.663646  1.424240e-10
7   ENSG00000166557     TMED3   rs10519274 -0.257261  9.518380e-10
3   ENSG00000258289    CHURC1       rs8943 -0.184327  1.353480e-09
12  ENSG00000113558      SKP1    rs2110585 -0.254785  1.218800e-08
3   ENSG00000132507     EIF5A   rs28636077 -0.347747  1.630960e-08
24  ENSG00000197345    MRPL21    rs2444765 -0.306477  3.740190e-08
27  ENSG00000125450     NUP85    rs7216538 -0.196893  5.696300e-08
19  ENSG00000114098     ARMC8  rs137911898 -0.387751  1.566310e-07
54  ENSG00000149636      DSN1   rs75741568 -1.160820  5.967950e-07
14  ENSG00000131469     RPL27     rs691997 -0.256838  6.878420e-07
19  ENSG00000188846     RPL14    rs2276868 -0.294612  6.934550e-07
45  ENSG00000153006  SREK1IP1    rs3756739 -0.228774  7.929660e-07
23  ENSG00000163743     RCHY1   rs77289963 -0.149948  8.327720e-07
31  ENSG00000167196    FBXO22   rs58829645 -0.455050  2.465320e-06
20  ENSG00000214941    ZSWIM7   rs61374610  0.405601  3.606540e-06
22  ENSG00000008324    SS18L2    rs3774396  0.130177  3.976200e-06
12  ENSG00000138160     KIF11  rs111760303 -0.414358  5.566760e-06
51  ENSG00000171202  TMEM126A   rs78417512  0.214181  7.301260e-06
1   ENSG00000170291      ELP5    rs2654190 -0.281837  2.730180e-05
7   ENSG00000183648    NDUFB1    rs7151027 -0.133751  3.359160e-05
33  ENSG00000151725     CENPU     rs873687  0.214583  3.977640e-05
22  ENSG00000008988     RPS20   rs16920295 -0.155873  4.080270e-05
28  ENSG00000136908      DPM2    rs3780666 -0.112947  7.454780e-05
4   ENSG00000106853     PTGR1    rs7030900 -0.350006  1.564970e-04
24  ENSG00000138757     G3BP2    rs6830063 -0.160755  2.030550e-04
21  ENSG00000171311    EXOSC1  rs150672414 -0.137090  2.531660e-04
20  ENSG00000073712    FERMT2   rs79469134  0.150080  3.481460e-04
33  ENSG00000145725   PPIP5K2      rs32840  0.168351  4.349820e-04
</pre>
</div>
</div>

<div id="outline-container-org018b84e" class="outline-2">
<h2 id="org018b84e">Dispersion-QTL calling</h2>
<div class="outline-text-2" id="text-org018b84e">
<p>
Read the estimated parameters and center them.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">disp</span> = (pd.read_table(<span class="org-string">'/scratch/midway2/aksarkar/singlecell/dispersion2.txt.gz'</span>, sep=<span class="org-string">' '</span>, index_col=0)
        .transform(<span class="org-keyword">lambda</span> x: x - x.mean(), axis=1))
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython">write_pheno_file(disp, gene_info, <span class="org-string">'/scratch/midway2/aksarkar/singlecell/disp.bed'</span>)
</pre>
</div>

<p>
Index the phenotype file.
</p>

<pre class="example">
Submitted batch job 42306096

</pre>

<p>
Run <code>qtltools</code>.
</p>

<pre class="example">
Submitted batch job 42306172

</pre>

<p>
Read the output.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">disp_qtls</span> = read_qtltools_output(<span class="org-string">'disp'</span>)
<span class="org-variable-name">keep_disp_qtls</span> = bh(disp_qtls)
keep_disp_qtls.<span class="org-builtin">sum</span>()
</pre>
</div>

<pre class="example">
71

</pre>

<div class="org-src-container">
<pre class="src src-ipython">disp_qtls[keep_disp_qtls].merge(gene_info, left_on=<span class="org-string">'gene'</span>, right_index=<span class="org-constant">True</span>)[[<span class="org-string">'gene'</span>, <span class="org-string">'name'</span>, <span class="org-string">'id'</span>, <span class="org-string">'beta'</span>, <span class="org-string">'p_beta'</span>]]
</pre>
</div>

<pre class="example">
gene     name           id      beta        p_beta
3   ENSG00000172500     FIBP    rs2904964  3.269360  5.982850e-59
30  ENSG00000173812     EIF1  rs141050869  2.236620  4.159740e-36
30  ENSG00000170558     CDH2   rs61062435 -3.124730  5.966380e-21
19  ENSG00000136603     SKIL    rs9818860 -3.706300  1.497410e-20
14  ENSG00000109919    MTCH2  rs112598354 -3.852610  8.516750e-20
11  ENSG00000138385      SSB   rs74671773 -1.864580  7.731550e-16
9   ENSG00000174721   FGFBP3  rs187269100 -3.340610  1.541350e-14
4   ENSG00000184640    SEPT9  rs191371465 -1.613220  5.537410e-13
6   ENSG00000005955   GGNBP2  rs115379718 -1.710560  6.033150e-13
17  ENSG00000163430    FSTL1  rs116138205 -3.701820  6.574070e-13
47  ENSG00000159596   TMEM69   rs80159866 -2.123730  1.515510e-10
3   ENSG00000077782    FGFR1  rs114441110 -2.509560  4.331400e-10
28  ENSG00000145592    RPL37  rs187287478 -1.854770  5.122680e-10
6   ENSG00000122483   CCDC18  rs114147805 -3.600390  5.673670e-09
35  ENSG00000164111    ANXA5  rs115667801 -1.147140  1.413170e-08
38  ENSG00000125977   EIF2S2   rs11696530 -0.964617  1.653160e-08
20  ENSG00000113658    SMAD5   rs76508606 -3.901540  3.115330e-08
29  ENSG00000115947     ORC4  rs149751455 -3.032940  5.041500e-08
49  ENSG00000205302     SNX2  rs148079226 -2.528050  1.294880e-07
16  ENSG00000155099  TMEM55A  rs143924177 -2.212820  1.463260e-07
24  ENSG00000099995    SF3A1  rs183441901 -2.124210  1.815250e-07
20  ENSG00000138778    CENPE  rs115865776 -2.262560  4.304580e-07
14  ENSG00000163535    SGOL2  rs186627329 -0.768620  4.388750e-07
30  ENSG00000166794     PPIB  rs138296314 -2.078340  5.791490e-07
19  ENSG00000104765   BNIP3L   rs17055905 -1.117420  6.768970e-07
10  ENSG00000181192   DHTKD1  rs186223298 -1.268940  1.822810e-06
23  ENSG00000049449     RCN1   rs61891674 -3.869610  2.008800e-06
0   ENSG00000110955    ATP5B   rs11172053 -4.315500  2.393400e-06
25  ENSG00000166483     WEE1  rs141629226 -3.647780  2.537660e-06
27  ENSG00000076003     MCM6   rs77680868 -2.576420  8.176020e-06
..              ...      ...          ...       ...           ...
24  ENSG00000167088   SNRPD1     rs599489 -3.443700  3.313140e-05
40  ENSG00000132676     DAP3  rs113643555 -3.610080  3.764670e-05
48  ENSG00000171858    RPS21  rs113006457 -1.777300  3.930040e-05
21  ENSG00000188994   ZNF292  rs190797589 -2.552210  4.493770e-05
38  ENSG00000118363    SPCS2  rs111846212 -2.637530  5.210370e-05
14  ENSG00000104738     MCM4   rs67441342  1.010130  6.201410e-05
4   ENSG00000166012    TAF1D  rs114156459 -2.464560  6.227290e-05
20  ENSG00000111300    NAA25  rs183463906 -1.903450  7.658170e-05
13  ENSG00000162694    EXTL2  rs114515441 -2.430710  8.293420e-05
22  ENSG00000143543      JTB  rs145970450 -2.513580  8.404170e-05
36  ENSG00000123737   EXOSC9   rs72684801 -4.150150  8.928880e-05
24  ENSG00000160049     DFFA  rs191456265 -1.834670  9.046540e-05
21  ENSG00000103353    UBFD1  rs187686990 -3.084110  1.050590e-04
26  ENSG00000069849   ATP1B3  rs181098445 -2.219060  1.351460e-04
9   ENSG00000112305    SMAP1    rs3805977 -1.484410  1.355490e-04
26  ENSG00000133027     PEMT  rs149534292 -3.130120  1.362530e-04
25  ENSG00000172586   CHCHD1  rs150231882 -1.491500  1.570080e-04
41  ENSG00000152904    GGPS1   rs77412086 -3.326680  1.738940e-04
11  ENSG00000112695   COX7A2  rs116524695 -1.969520  2.009070e-04
28  ENSG00000143575     HAX1  rs191047822 -3.161070  2.198430e-04
28  ENSG00000116489   CAPZA1  rs112268692 -1.440870  2.665790e-04
13  ENSG00000137807    KIF23  rs146358835 -3.455250  2.677390e-04
17  ENSG00000113328    CCNG1   rs61114041 -1.741340  3.823830e-04
30  ENSG00000148773    MKI67  rs114309437 -1.111450  4.412160e-04
23  ENSG00000184076   UQCR10  rs140062784 -2.846960  4.843060e-04
38  ENSG00000112335     SNX3  rs146420004 -2.824480  5.119260e-04
5   ENSG00000145907    G3BP1   rs78692171 -0.997360  6.671770e-04
22  ENSG00000162923    WDR26   rs56887412 -1.836620  7.486450e-04
47  ENSG00000065361    ERBB3   rs12314191 -1.623870  7.577620e-04
14  ENSG00000173145    NOC3L   rs74151073 -2.267200  8.821250e-04

[71 rows x 5 columns]
</pre>
</div>
</div>
</div>
<div id="postamble" class="status">
<p class="author">Author: Abhishek Sarkar</p>
<p class="date">Created: 2018-01-22 Mon 11:04</p>
<p class="validation"><a href="http://validator.w3.org/check?uri=referer">Validate</a></p>
</div>
</body>
</html>
