<!DOCTYPE html>
<html lang="en">
<head>
<!-- 2018-03-05 Mon 14:04 -->
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
<title>QTL mapping pipeline</title>
<meta name="generator" content="Org mode">
<meta name="author" content="Abhishek Sarkar">
<style type="text/css">
 <!--/*--><![CDATA[/*><!--*/
  .title  { text-align: center;
             margin-bottom: .2em; }
  .subtitle { text-align: center;
              font-size: medium;
              font-weight: bold;
              margin-top:0; }
  .todo   { font-family: monospace; color: red; }
  .done   { font-family: monospace; color: green; }
  .priority { font-family: monospace; color: orange; }
  .tag    { background-color: #eee; font-family: monospace;
            padding: 2px; font-size: 80%; font-weight: normal; }
  .timestamp { color: #bebebe; }
  .timestamp-kwd { color: #5f9ea0; }
  .org-right  { margin-left: auto; margin-right: 0px;  text-align: right; }
  .org-left   { margin-left: 0px;  margin-right: auto; text-align: left; }
  .org-center { margin-left: auto; margin-right: auto; text-align: center; }
  .underline { text-decoration: underline; }
  #postamble p, #preamble p { font-size: 90%; margin: .2em; }
  p.verse { margin-left: 3%; }
  pre {
    border: 1px solid #ccc;
    box-shadow: 3px 3px 3px #eee;
    padding: 8pt;
    font-family: monospace;
    overflow: auto;
    margin: 1.2em;
  }
  pre.src {
    position: relative;
    overflow: visible;
    padding-top: 1.2em;
  }
  pre.src:before {
    display: none;
    position: absolute;
    background-color: white;
    top: -10px;
    right: 10px;
    padding: 3px;
    border: 1px solid black;
  }
  pre.src:hover:before { display: inline;}
  /* Languages per Org manual */
  pre.src-asymptote:before { content: 'Asymptote'; }
  pre.src-awk:before { content: 'Awk'; }
  pre.src-C:before { content: 'C'; }
  /* pre.src-C++ doesn't work in CSS */
  pre.src-clojure:before { content: 'Clojure'; }
  pre.src-css:before { content: 'CSS'; }
  pre.src-D:before { content: 'D'; }
  pre.src-ditaa:before { content: 'ditaa'; }
  pre.src-dot:before { content: 'Graphviz'; }
  pre.src-calc:before { content: 'Emacs Calc'; }
  pre.src-emacs-lisp:before { content: 'Emacs Lisp'; }
  pre.src-fortran:before { content: 'Fortran'; }
  pre.src-gnuplot:before { content: 'gnuplot'; }
  pre.src-haskell:before { content: 'Haskell'; }
  pre.src-hledger:before { content: 'hledger'; }
  pre.src-java:before { content: 'Java'; }
  pre.src-js:before { content: 'Javascript'; }
  pre.src-latex:before { content: 'LaTeX'; }
  pre.src-ledger:before { content: 'Ledger'; }
  pre.src-lisp:before { content: 'Lisp'; }
  pre.src-lilypond:before { content: 'Lilypond'; }
  pre.src-lua:before { content: 'Lua'; }
  pre.src-matlab:before { content: 'MATLAB'; }
  pre.src-mscgen:before { content: 'Mscgen'; }
  pre.src-ocaml:before { content: 'Objective Caml'; }
  pre.src-octave:before { content: 'Octave'; }
  pre.src-org:before { content: 'Org mode'; }
  pre.src-oz:before { content: 'OZ'; }
  pre.src-plantuml:before { content: 'Plantuml'; }
  pre.src-processing:before { content: 'Processing.js'; }
  pre.src-python:before { content: 'Python'; }
  pre.src-R:before { content: 'R'; }
  pre.src-ruby:before { content: 'Ruby'; }
  pre.src-sass:before { content: 'Sass'; }
  pre.src-scheme:before { content: 'Scheme'; }
  pre.src-screen:before { content: 'Gnu Screen'; }
  pre.src-sed:before { content: 'Sed'; }
  pre.src-sh:before { content: 'shell'; }
  pre.src-sql:before { content: 'SQL'; }
  pre.src-sqlite:before { content: 'SQLite'; }
  /* additional languages in org.el's org-babel-load-languages alist */
  pre.src-forth:before { content: 'Forth'; }
  pre.src-io:before { content: 'IO'; }
  pre.src-J:before { content: 'J'; }
  pre.src-makefile:before { content: 'Makefile'; }
  pre.src-maxima:before { content: 'Maxima'; }
  pre.src-perl:before { content: 'Perl'; }
  pre.src-picolisp:before { content: 'Pico Lisp'; }
  pre.src-scala:before { content: 'Scala'; }
  pre.src-shell:before { content: 'Shell Script'; }
  pre.src-ebnf2ps:before { content: 'ebfn2ps'; }
  /* additional language identifiers per "defun org-babel-execute"
       in ob-*.el */
  pre.src-cpp:before  { content: 'C++'; }
  pre.src-abc:before  { content: 'ABC'; }
  pre.src-coq:before  { content: 'Coq'; }
  pre.src-groovy:before  { content: 'Groovy'; }
  /* additional language identifiers from org-babel-shell-names in
     ob-shell.el: ob-shell is the only babel language using a lambda to put
     the execution function name together. */
  pre.src-bash:before  { content: 'bash'; }
  pre.src-csh:before  { content: 'csh'; }
  pre.src-ash:before  { content: 'ash'; }
  pre.src-dash:before  { content: 'dash'; }
  pre.src-ksh:before  { content: 'ksh'; }
  pre.src-mksh:before  { content: 'mksh'; }
  pre.src-posh:before  { content: 'posh'; }
  /* Additional Emacs modes also supported by the LaTeX listings package */
  pre.src-ada:before { content: 'Ada'; }
  pre.src-asm:before { content: 'Assembler'; }
  pre.src-caml:before { content: 'Caml'; }
  pre.src-delphi:before { content: 'Delphi'; }
  pre.src-html:before { content: 'HTML'; }
  pre.src-idl:before { content: 'IDL'; }
  pre.src-mercury:before { content: 'Mercury'; }
  pre.src-metapost:before { content: 'MetaPost'; }
  pre.src-modula-2:before { content: 'Modula-2'; }
  pre.src-pascal:before { content: 'Pascal'; }
  pre.src-ps:before { content: 'PostScript'; }
  pre.src-prolog:before { content: 'Prolog'; }
  pre.src-simula:before { content: 'Simula'; }
  pre.src-tcl:before { content: 'tcl'; }
  pre.src-tex:before { content: 'TeX'; }
  pre.src-plain-tex:before { content: 'Plain TeX'; }
  pre.src-verilog:before { content: 'Verilog'; }
  pre.src-vhdl:before { content: 'VHDL'; }
  pre.src-xml:before { content: 'XML'; }
  pre.src-nxml:before { content: 'XML'; }
  /* add a generic configuration mode; LaTeX export needs an additional
     (add-to-list 'org-latex-listings-langs '(conf " ")) in .emacs */
  pre.src-conf:before { content: 'Configuration File'; }

  table { border-collapse:collapse; }
  caption.t-above { caption-side: top; }
  caption.t-bottom { caption-side: bottom; }
  td, th { vertical-align:top;  }
  th.org-right  { text-align: center;  }
  th.org-left   { text-align: center;   }
  th.org-center { text-align: center; }
  td.org-right  { text-align: right;  }
  td.org-left   { text-align: left;   }
  td.org-center { text-align: center; }
  dt { font-weight: bold; }
  .footpara { display: inline; }
  .footdef  { margin-bottom: 1em; }
  .figure { padding: 1em; }
  .figure p { text-align: center; }
  .inlinetask {
    padding: 10px;
    border: 2px solid gray;
    margin: 10px;
    background: #ffffcc;
  }
  #org-div-home-and-up
   { text-align: right; font-size: 70%; white-space: nowrap; }
  textarea { overflow-x: auto; }
  .linenr { font-size: smaller }
  .code-highlighted { background-color: #ffff00; }
  .org-info-js_info-navigation { border-style: none; }
  #org-info-js_console-label
    { font-size: 10px; font-weight: bold; white-space: nowrap; }
  .org-info-js_search-highlight
    { background-color: #ffff00; color: #000000; font-weight: bold; }
  .org-svg { width: 90%; }
  /*]]>*/-->
</style>
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<link rel="stylesheet" type="text/css" href="https://www.pirilampo.org/styles/readtheorg/css/htmlize.css"/>
<style type="text/css">body {width: 60em; margin:auto} pre.src {overflow:auto}</style>
<script type="text/javascript">
/*
@licstart  The following is the entire license notice for the
JavaScript code in this tag.

Copyright (C) 2012-2017 Free Software Foundation, Inc.

The JavaScript code in this tag is free software: you can
redistribute it and/or modify it under the terms of the GNU
General Public License (GNU GPL) as published by the Free Software
Foundation, either version 3 of the License, or (at your option)
any later version.  The code is distributed WITHOUT ANY WARRANTY;
without even the implied warranty of MERCHANTABILITY or FITNESS
FOR A PARTICULAR PURPOSE.  See the GNU GPL for more details.

As additional permission under GNU GPL version 3 section 7, you
may distribute non-source (e.g., minimized or compacted) forms of
that code without the copy of the GNU GPL normally required by
section 4, provided you include this license notice and a URL
through which recipients can access the Corresponding Source.


@licend  The above is the entire license notice
for the JavaScript code in this tag.
*/
<!--/*--><![CDATA[/*><!--*/
 function CodeHighlightOn(elem, id)
 {
   var target = document.getElementById(id);
   if(null != target) {
     elem.cacheClassElem = elem.className;
     elem.cacheClassTarget = target.className;
     target.className = "code-highlighted";
     elem.className   = "code-highlighted";
   }
 }
 function CodeHighlightOff(elem, id)
 {
   var target = document.getElementById(id);
   if(elem.cacheClassElem)
     elem.className = elem.cacheClassElem;
   if(elem.cacheClassTarget)
     target.className = elem.cacheClassTarget;
 }
/*]]>*///-->
</script>
<script type="text/x-mathjax-config">
    MathJax.Hub.Config({
        displayAlign: "center",
        displayIndent: "0em",

        "HTML-CSS": { scale: 100,
                        linebreaks: { automatic: "false" },
                        webFont: "TeX"
                       },
        SVG: {scale: 100,
              linebreaks: { automatic: "false" },
              font: "TeX"},
        NativeMML: {scale: 100},
        TeX: { equationNumbers: {autoNumber: "AMS"},
               MultLineWidth: "85%",
               TagSide: "right",
               TagIndent: ".8em"
             }
});
</script>
<script type="text/javascript"
        src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-AMS_HTML"></script>
</head>
<body>
<div id="content">
<h1 class="title">QTL mapping pipeline</h1>

<div id="outline-container-orga163d45" class="outline-2">
<h2 id="orga163d45">Introduction</h2>
<div class="outline-text-2" id="text-orga163d45">
<p>
We previously <a href="zinb.html">estimated means and dispersions per individual, per gene</a>. In
our modular approach, we now fit log linear models for each:
</p>

<p>
\[ \ln \mu_{ik} = X \beta_\mu + \epsilon_\mu \]
</p>

<p>
\[ \ln \phi_{ik} = X \beta_\phi + \epsilon_\phi \]
</p>

<p>
where \(\mu_{ik}, \phi_{ik}\) are the mean and dispersion for individual
\(i\), gene \(k\), \(X\) is the genotype matrix (\(n \times 1)\), and
\(\beta_\mu, \beta_\phi\) are the effect sizes (scalar).
</p>

<p>
Here, we present the following analyses:
</p>

<ol class="org-ol">
<li><a href="#org592eadf">We replicate bulk eQTLs</a> in the scRNA-Seq data</li>
<li><a href="#org7540462">We show that approximating a permutation test</a> is not appropriate</li>
<li><a href="#orgaa637e2">We call mean QTLs</a> in the scRNA-Seq data</li>
<li><a href="#orgf02711e">We call robustness QTLs</a> in the scRNA-Seq data</li>
</ol>
</div>
</div>

<div id="outline-container-org9d05232" class="outline-2">
<h2 id="org9d05232">Genotype processing</h2>
<div class="outline-text-2" id="text-org9d05232">
<p>
<code>QTLtools</code> reported 0 variants in cis for an unexpected number of genes when
using the provided genotype file, so reprocess the genotypes to fix it.
</p>

<div class="org-src-container">
<pre class="src src-sh">sbatch --partition=broadwl --job-name process-geno
<span class="org-comment-delimiter">#</span><span class="org-comment">!/bin/bash</span>
<span class="org-builtin">set</span> -e
<span class="org-builtin">source</span> activate scqtl
zcat /project/compbio/jointLCLs/genotype/hg19/YRI/vcf/chr{?,??}.hg19.vcf.gz | awk <span class="org-string">'BEGIN {print "##fileformat=VCFv4.1"} NR == 1 || ! /CHROM/ {print}'</span> | bgzip &gt;genotypes.vcf.gz
tabix genotypes.vcf.gz
</pre>
</div>

<div class="org-src-container">
<pre class="src src-sh">sbatch --partition=broadwl
<span class="org-comment-delimiter">#</span><span class="org-comment">!/bin/bash</span>
<span class="org-builtin">source</span> activate scqtl
plink --memory 2000 --vcf genotypes.vcf.gz --geno 0.01 --maf 0.01 --recode vcf-iid --out genotypes-qc
bgzip genotypes-qc.vcf
tabix -f -p vcf genotypes-qc.vcf.gz
</pre>
</div>
</div>
</div>

<div id="outline-container-org592eadf" class="outline-2">
<h2 id="org592eadf">Replicate bulk eQTLs in the single cell data</h2>
<div class="outline-text-2" id="text-org592eadf">
<p>
Read the bulk eQTLs (nominally significant).
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">def</span> <span class="org-function-name">parse_bulk_qtl_record</span>(row):
  <span class="org-variable-name">gene</span>, <span class="org-variable-name">_</span> = row[0].split(<span class="org-string">'.'</span>)
  id_, *<span class="org-variable-name">_</span> = row[5].split(<span class="org-string">'.'</span>)
  <span class="org-keyword">return</span> pd.Series({<span class="org-string">'gene'</span>: gene, <span class="org-string">'id'</span>: id_, <span class="org-string">'beta'</span>: row.iloc[-3], <span class="org-string">'p_bulk'</span>: row.iloc[-1]})
<span class="org-variable-name">bulk_qtls</span> = (pd.read_table(<span class="org-string">'/project2/gilad/singlecell-qtl/bulk/permutations.all.RNAseq_run.fixed.txt.gz'</span>, sep=<span class="org-string">' '</span>, header=<span class="org-constant">None</span>)
             .fillna(<span class="org-string">''</span>)
             .<span class="org-builtin">apply</span>(parse_bulk_qtl_record, axis=1))
</pre>
</div>

<p>
Read the bulk eQTLs (significant by permutation test).
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">def</span> <span class="org-function-name">parse_perm_bulk_qtl_record</span>(row):
  gene, *<span class="org-variable-name">_</span> = row[0].split(<span class="org-string">'.'</span>)
  <span class="org-variable-name">id_</span>, <span class="org-variable-name">chr_</span>, <span class="org-variable-name">pos</span> = row[1].split(<span class="org-string">'.'</span>)
  <span class="org-keyword">return</span> pd.Series({<span class="org-string">'gene'</span>: gene, <span class="org-string">'id'</span>: id_, <span class="org-string">'chr'</span>: chr_, <span class="org-string">'pos'</span>: pos, <span class="org-string">'p_bulk'</span>: row.iloc[-1]})
<span class="org-variable-name">perm_bulk_qtls</span> = (pd.read_table(<span class="org-string">'/project2/gilad/singlecell-qtl/bulk/permutations.all.RNAseq_run.fixed.txt.gz.bh.txt'</span>, sep=<span class="org-string">' '</span>, header=0)
                  .<span class="org-builtin">apply</span>(parse_perm_bulk_qtl_record, axis=1))
</pre>
</div>

<p>
Test each bulk QTL-gene pair in the single cell data. Mean-impute and
center genotypes.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">def</span> <span class="org-function-name">parse_vcf</span>(record):
  <span class="org-variable-name">geno</span> = [<span class="org-builtin">sum</span>(<span class="org-builtin">float</span>(h) <span class="org-keyword">for</span> h <span class="org-keyword">in</span> g.split(<span class="org-string">'/'</span>)) <span class="org-keyword">if</span> <span class="org-string">'.'</span> <span class="org-keyword">not</span> <span class="org-keyword">in</span> g <span class="org-keyword">else</span> -1 <span class="org-keyword">for</span> g <span class="org-keyword">in</span> record[9:]]
  <span class="org-keyword">return</span> pd.Series(geno)

<span class="org-keyword">def</span> <span class="org-function-name">standardize</span>(geno):
  <span class="org-comment-delimiter"># </span><span class="org-comment">Important: this has to be done after filtering to common individuals</span>
  <span class="org-variable-name">orig_geno</span> = geno.copy()
  <span class="org-comment-delimiter"># </span><span class="org-comment">Important: match the qtltools implementation</span>
  <span class="org-variable-name">geno</span> = np.ma.masked_less(geno.values, 0)
  <span class="org-variable-name">geno</span> = geno.filled(geno.mean())
  <span class="org-variable-name">geno</span> -= geno.mean()
  <span class="org-keyword">return</span> geno

<span class="org-keyword">def</span> <span class="org-function-name">extract_qtl_gene_pair</span>(bulk_qtls, means):
  <span class="org-variable-name">common_means</span>, <span class="org-variable-name">common_qtls</span> = means.align(bulk_qtls.set_index(<span class="org-string">'gene'</span>), join=<span class="org-string">'inner'</span>, axis=0)
  <span class="org-variable-name">header</span> = pd.read_table(<span class="org-string">'/project2/gilad/singlecell-qtl/bulk/genotypes.vcf.gz'</span>, skiprows=1, nrows=1, header=0)[9:]
  <span class="org-variable-name">genotypes</span> = tabix.<span class="org-builtin">open</span>(<span class="org-string">'/project2/gilad/singlecell-qtl/bulk/genotypes.vcf.gz'</span>)
  <span class="org-variable-name">X</span>, <span class="org-variable-name">Y</span> = (common_qtls
          .<span class="org-builtin">apply</span>(<span class="org-keyword">lambda</span> x: parse_vcf(<span class="org-builtin">next</span>(genotypes.query(x[<span class="org-string">'chr'</span>], <span class="org-builtin">int</span>(x[<span class="org-string">'pos'</span>]) - 1, <span class="org-builtin">int</span>(x[<span class="org-string">'pos'</span>])))), axis=1)
          .rename(columns={i: ind <span class="org-keyword">for</span> i, ind <span class="org-keyword">in</span> <span class="org-builtin">enumerate</span>(header.columns[9:])})
          .align(common_means, join=<span class="org-string">'inner'</span>, axis=<span class="org-constant">None</span>))
  <span class="org-variable-name">X</span> = X.<span class="org-builtin">apply</span>(standardize, axis=1)
  <span class="org-keyword">return</span> X, Y
</pre>
</div>

<p>
Define replication as significant at FDR 5% in both bulk and single cell.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">def</span> <span class="org-function-name">bh</span>(df, column, fdr=0.05):
  <span class="org-variable-name">num_tests</span> = df.shape[0]
  <span class="org-keyword">return</span> np.logical_and.accumulate(df.sort_values(column)[column] &lt; fdr * np.arange(1, num_tests + 1) / num_tests)
</pre>
</div>
</div>

<div id="outline-container-org1b9c87d" class="outline-3">
<h3 id="org1b9c87d">NB model</h3>
<div class="outline-text-3" id="text-org1b9c87d">
<p>
Estimate a lower bound on the replication rate using the following procedure:
</p>

<ol class="org-ol">
<li>For each individual, for each gene, <a href="zinb.html#orga46d8c0">estimate a mean and dispersion in the
single cell data ignoring zero inflation</a> (i.e., just fit a negative
binomial model)</li>
<li>For each bulk RNA-Seq eQTL-gene pair, test the SNP against the estimated
means</li>
</ol>

<p>
Read the estimated parameters (phenotypes) then center them.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">with</span> sqlite3.connect(<span class="org-string">'/scratch/midway2/aksarkar/singlecell/browser.db'</span>) <span class="org-keyword">as</span> conn:
  <span class="org-variable-name">mean</span> = (pd.read_sql(<span class="org-string">'select gene, ind, nb_log_mean from params'</span>, conn)
          .pivot(index=<span class="org-string">'gene'</span>, columns=<span class="org-string">'ind'</span>, values=<span class="org-string">'nb_log_mean'</span>)
          .transform(<span class="org-keyword">lambda</span> x: x - x.mean(), axis=1))
</pre>
</div>

<p>
Perform the eQTL tests.
</p>

<div class="org-src-container">
<pre class="src src-ipython" id="org5c67bff"><span class="org-variable-name">X</span>, <span class="org-variable-name">Y</span> = extract_qtl_gene_pair(perm_bulk_qtls, mean)
<span class="org-variable-name">result</span> = []
<span class="org-variable-name">_sf</span> = st.chi2(1).sf
<span class="org-keyword">for</span> (_, x), (name, y) <span class="org-keyword">in</span> <span class="org-builtin">zip</span>(X.iterrows(), Y.iterrows()):
  <span class="org-keyword">if</span> np.isclose(x.std(), 0):
    <span class="org-keyword">print</span>(<span class="org-string">'Skipping {}'</span>.<span class="org-builtin">format</span>(name))
    <span class="org-keyword">continue</span>
  beta, rss, *<span class="org-variable-name">_</span> = np.linalg.lstsq(x.values.reshape(-1, 1), y.values.ravel(), rcond=-1)
  <span class="org-variable-name">sigma2</span> = rss / y.shape[0]
  <span class="org-variable-name">se</span> = sigma2 / np.inner(x.values, x.values)
  <span class="org-variable-name">pval</span> = _sf(np.square(beta / se))
  result.append({<span class="org-string">'gene'</span>: name, <span class="org-string">'beta'</span>: beta[0], <span class="org-string">'p_sc'</span>: pval.ravel()[0]})
<span class="org-variable-name">merged</span> = (bulk_qtls.merge(pd.DataFrame.from_dict(result), on=<span class="org-string">'gene'</span>, suffixes=[<span class="org-string">'_bulk'</span>, <span class="org-string">'_sc'</span>])
          .set_index(<span class="org-string">'gene'</span>)
          .sort_values(<span class="org-string">'p_bulk'</span>)
          .reset_index()[[<span class="org-string">'gene'</span>, <span class="org-string">'id'</span>, <span class="org-string">'p_bulk'</span>, <span class="org-string">'beta_bulk'</span>, <span class="org-string">'p_sc'</span>, <span class="org-string">'beta_sc'</span>]])
</pre>
</div>

<p>
Skipping ENSG00000141867
Skipping ENSG00000164823
</p>

<p>
Estimate the replication rate.
</p>

<div class="org-src-container">
<pre class="src src-ipython" id="org848d3e7"><span class="org-variable-name">bulk_fdr_pass</span> = bh(merged, <span class="org-string">'p_bulk'</span>)
<span class="org-variable-name">sc_fdr_pass</span> = bh(merged, <span class="org-string">'p_sc'</span>)
<span class="org-variable-name">replicated</span> = merged[np.logical_and(*bulk_fdr_pass.align(sc_fdr_pass))]
replicated.shape[0], replicated.shape[0] / merged.shape[0]
</pre>
</div>

<pre class="example">
(892, 0.9529914529914529)

</pre>

<p>
Estimate the concordance.
</p>

<div class="org-src-container">
<pre class="src src-ipython" id="org8bb546b"><span class="org-variable-name">sign_pass</span> = (replicated[<span class="org-string">'beta_bulk'</span>] * replicated[<span class="org-string">'beta_sc'</span>]).values &gt; 0
replicated[sign_pass].shape[0], sign_pass.<span class="org-builtin">sum</span>() / replicated.shape[0]
</pre>
</div>

<pre class="example">
(779, 0.8733183856502242)

</pre>

<p>
Write out tables for the interactive browser.
</p>

<div class="org-src-container">
<pre class="src src-ipython" id="org1d58682"><span class="org-keyword">with</span> sqlite3.connect(<span class="org-string">'/scratch/midway2/aksarkar/singlecell/browser.db'</span>) <span class="org-keyword">as</span> conn:
  merged.to_sql(<span class="org-string">'qtls'</span>, con=conn, if_exists=<span class="org-string">'replace'</span>)
  (X
   .reset_index()
   .melt(id_vars=<span class="org-string">'gene'</span>, var_name=<span class="org-string">'ind'</span>)
   .to_sql(name=<span class="org-string">'genotype'</span>, con=conn, index=<span class="org-constant">False</span>, if_exists=<span class="org-string">'replace'</span>))

  conn.execute(<span class="org-string">'create index ix_qtls on qtls(gene);'</span>)
  conn.execute(<span class="org-string">'create index ix_genotype on genotype(gene, ind);'</span>)
</pre>
</div>
</div>
</div>

<div id="outline-container-orga1d54a9" class="outline-3">
<h3 id="orga1d54a9">ZINB model</h3>
<div class="outline-text-3" id="text-orga1d54a9">
<p>
For each individual, for each gene, take either the negative binomial or
zero-inflated negative binomial estimate, depending on which minimized the
negative log-likelihood of the data.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">with</span> sqlite3.connect(<span class="org-string">'/scratch/midway2/aksarkar/singlecell/browser.db'</span>) <span class="org-keyword">as</span> conn:
  <span class="org-variable-name">mean</span> = (pd.read_sql(<span class="org-string">'select gene, ind, case when nb_nll &lt; zinb_nll then nb_log_mean else zinb2_log_mean end as log_mean from params;'</span>, conn)
          .pivot(index=<span class="org-string">'gene'</span>, columns=<span class="org-string">'ind'</span>, values=<span class="org-string">'log_mean'</span>)
          .transform(<span class="org-keyword">lambda</span> x: x - x.mean(), axis=1))
</pre>
</div>

<p>
Skipping ENSG00000141867
Skipping ENSG00000164823
</p>

<p>
Estimate the replication rate.
</p>

<pre class="example">
(894, 0.9551282051282052)

</pre>

<p>
Estimate the concordance.
</p>

<pre class="example">
(779, 0.8713646532438478)

</pre>

<p>
Write out tables for the interactive browser.
</p>
</div>
</div>
</div>

<div id="outline-container-org7540462" class="outline-2">
<h2 id="org7540462">Test validity of approximate permutation test</h2>
<div class="outline-text-2" id="text-org7540462">
<p>
<code>qtltools</code> tries to calibrate false discovery rates using the following
procedure:
</p>

<ol class="org-ol">
<li>For each hypothesis, permute the genotype data to estimate the null
distribution of the p-values</li>
<li>Fit a beta distribution to the permuted p-values via ML</li>
<li>Compute the lower tail probability of the observed p-value, assuming it
was generated from the fitted beta distribution</li>
<li>Apply the Benjamini-Hochberg procedure on the set of lower tail
probabilities (across all hypotheses)</li>
</ol>

<p>
Test whether the beta approximation is appropriate for our sample size by
subsetting GEUVADIS. Take all genes on chromosome 1.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">geuvadis</span> = []
<span class="org-keyword">for</span> chunk <span class="org-keyword">in</span> pd.read_table(<span class="org-string">'/project/compbio/geuvadis/analysis_results/GD462.GeneQuantRPKM.50FN.samplename.resk10.txt.gz'</span>, chunksize=100):
  geuvadis.append(chunk.query(<span class="org-string">'Chr == "1"'</span>))
<span class="org-variable-name">geuvadis</span> = pd.concat(geuvadis)
</pre>
</div>

<p>
Pick 54 random individuals from GEUVADIS.
</p>

<div class="org-src-container">
<pre class="src src-ipython">np.random.seed(0)
<span class="org-variable-name">subset</span> = np.random.choice(geuvadis.columns[5:], size=54)
</pre>
</div>

<p>
Write out the phenotype file for <code>qtltools</code>. Important: GEUVADIS VCFs code
chromosome without <code>chr</code>.
</p>

<div class="org-src-container">
<pre class="src src-ipython">write_pheno_file(geuvadis.set_index(geuvadis[<span class="org-string">'Gene_Symbol'</span>].<span class="org-builtin">apply</span>(<span class="org-keyword">lambda</span> x: x.split(<span class="org-string">'.'</span>)[0]))[subset], gene_info, <span class="org-string">'/scratch/midway2/aksarkar/singlecell/test.bed'</span>)
</pre>
</div>

<p>
Index the phenotype file. Important: <code>#</code> sorts before <code>c</code>, but after <code>1</code>.
</p>

<div class="org-src-container">
<pre class="src src-sh" id="org9ba22bd"><span class="org-builtin">export</span> <span class="org-variable-name">input</span>=$<span class="org-variable-name">input</span>
sbatch --partition=$<span class="org-variable-name">partition</span> --wait
<span class="org-comment-delimiter">#</span><span class="org-comment">!/bin/bash</span>
sort -k1,1g -k2,2n -k3,3n $<span class="org-variable-name">input</span> | bgzip &gt;$<span class="org-variable-name">input</span>.gz
tabix -p bed $<span class="org-variable-name">input</span>.gz
</pre>
</div>

<p>
<code>qtltools</code> segfaults on the VCF file, so convert to <code>plink</code> format.
</p>

<div class="org-src-container">
<pre class="src src-sh">sbatch --partition=broadwl --mem=2G
<span class="org-comment-delimiter">#</span><span class="org-comment">!/bin/bash</span>
plink --memory 2000 --geno 0.01 --maf 0.01 --vcf /project/compbio/geuvadis/genotypes/GEUVADIS.chr1.PH1PH2_465.IMPFRQFILT_BIALLELIC_PH.annotv2.genotypes.vcf.gz --recode vcf-iid --out geuvadis-chr1
bgzip geuvadis-chr1.vcf
tabix -f -p vcf geuvadis-chr1.vcf.gz
</pre>
</div>

<p>
Run <code>qtltools</code>.
</p>

<div class="org-src-container">
<pre class="src src-sh" id="org61bc797"><span class="org-builtin">export</span> <span class="org-variable-name">pheno</span>=$<span class="org-variable-name">pheno</span>
<span class="org-builtin">export</span> <span class="org-variable-name">geno</span>=$<span class="org-variable-name">geno</span>
<span class="org-builtin">export</span> <span class="org-variable-name">op</span>=$<span class="org-variable-name">op</span>
sbatch --partition=$<span class="org-variable-name">partition</span> -n1 -c28 --exclusive -J $<span class="org-variable-name">pheno</span>-qtl -o $<span class="org-variable-name">pheno</span>-qtl.log
<span class="org-comment-delimiter">#</span><span class="org-comment">!/bin/bash</span>
<span class="org-builtin">source</span> activate scqtl
module load parallel
parallel -j28 qtltools cis --vcf $<span class="org-variable-name">geno</span> --bed $<span class="org-variable-name">pheno</span>.bed.gz $<span class="org-variable-name">op</span> --chunk {#} 100 --out $<span class="org-variable-name">pheno</span>-qtl.{#}.txt ::: $(<span class="org-sh-quoted-exec">seq</span> 1 100)
</pre>
</div>

<p>
Read the results.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">def</span> <span class="org-function-name">read_qtltools_output</span>(pheno):
  <span class="org-variable-name">file_names</span> = [<span class="org-string">'{}-qtl.{}.txt'</span>.<span class="org-builtin">format</span>(pheno, i) <span class="org-keyword">for</span> i <span class="org-keyword">in</span> <span class="org-builtin">range</span>(1, 101)]
  <span class="org-keyword">return</span> (pd.concat([pd.read_table(f, header=<span class="org-constant">None</span>, sep=<span class="org-string">' '</span>)
                     <span class="org-keyword">for</span> f <span class="org-keyword">in</span> file_names <span class="org-keyword">if</span> os.path.exists(f) <span class="org-keyword">and</span>
                     os.path.getsize(f) &gt; 0])
          .rename(columns={i: x <span class="org-keyword">for</span> i, x <span class="org-keyword">in</span> <span class="org-builtin">enumerate</span>(
            [<span class="org-string">'gene'</span>, <span class="org-string">'chr'</span>, <span class="org-string">'start'</span>, <span class="org-string">'end'</span>, <span class="org-string">'strand'</span>, <span class="org-string">'num_vars'</span>,
             <span class="org-string">'distance'</span>, <span class="org-string">'id'</span>, <span class="org-string">'var_chr'</span>, <span class="org-string">'var_start'</span>, <span class="org-string">'var_end'</span>, <span class="org-string">'df'</span>,
             <span class="org-string">'dummy'</span>, <span class="org-string">'a'</span>, <span class="org-string">'b'</span>, <span class="org-string">'p_nominal'</span>, <span class="org-string">'beta'</span>, <span class="org-string">'p_empirical'</span>, <span class="org-string">'p_beta'</span>]
          )})
          .sort_values(<span class="org-string">'p_beta'</span>))
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">geuvadis_qtls</span> = read_qtltools_output(<span class="org-string">'test'</span>)
</pre>
</div>

<p>
Check the beta approximation to the permutation p-values.
</p>

<div class="org-src-container">
<pre class="src src-ipython">plt.clf()
plt.gcf().set_size_inches(6, 6)
plt.scatter(geuvadis_qtls[<span class="org-string">'p_empirical'</span>], geuvadis_qtls[<span class="org-string">'p_beta'</span>], s=1, c=<span class="org-string">'k'</span>)
plt.plot([0, 1], [0, 1], c=<span class="org-string">'r'</span>, ls=<span class="org-string">'--'</span>)
plt.xlabel(<span class="org-string">'Empirical p-value'</span>)
<span class="org-variable-name">_</span> = plt.ylabel(<span class="org-string">'Approximate p-value'</span>)
</pre>
</div>


<div class="figure">
<p><img src="figure/qtl-mapping.org/geuvadis-qtl-beta.png" alt="geuvadis-qtl-beta.png">
</p>
</div>
</div>
</div>

<div id="outline-container-orgaa637e2" class="outline-2">
<h2 id="orgaa637e2">Call mean-QTLs in the single cell data</h2>
<div class="outline-text-2" id="text-orgaa637e2">
<p>
For each individual, for each gene, take either the negative binomial or
zero-inflated negative binomial estimate, depending on which minimized the
negative log-likelihood of the data.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">with</span> sqlite3.connect(<span class="org-string">'/scratch/midway2/aksarkar/singlecell/browser.db'</span>) <span class="org-keyword">as</span> conn:
  <span class="org-variable-name">mean</span> = (pd.read_sql(<span class="org-string">'select gene, ind, case when nb_nll &lt; zinb_nll then nb_log_mean else zinb2_log_mean end as log_mean from params;'</span>, conn)
          .pivot(index=<span class="org-string">'gene'</span>, columns=<span class="org-string">'ind'</span>, values=<span class="org-string">'log_mean'</span>))
</pre>
</div>

<p>
Write out the <a href="https://qtltools.github.io/qtltools/pages/input_files.html">phenotype file</a> for <code>qtltools</code>.
</p>

<div class="org-src-container">
<pre class="src src-ipython" id="org5b69fcd"><span class="org-variable-name">gene_info</span> = (pd.read_table(<span class="org-string">'/home/aksarkar/projects/singlecell-qtl/data/scqtl-genes.txt.gz'</span>)
             .set_index(<span class="org-string">'gene'</span>)
             .query(<span class="org-string">'source == "H. sapiens"'</span>)
             .query(<span class="org-string">'chr != "hsX"'</span>)
             .query(<span class="org-string">'chr != "hsY"'</span>)
             .query(<span class="org-string">'chr != "hsMT"'</span>))
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython" id="org38e255d"><span class="org-keyword">def</span> <span class="org-function-name">qtltools_format</span>(row):
  <span class="org-variable-name">row</span>[<span class="org-string">'#Chr'</span>] = <span class="org-string">'{}'</span>.<span class="org-builtin">format</span>(row[<span class="org-string">'chr'</span>][2:])
  <span class="org-variable-name">row</span>[<span class="org-string">'gid'</span>] = row.name
  <span class="org-variable-name">row</span>[<span class="org-string">'pid'</span>] = row.name
  <span class="org-keyword">return</span> row

<span class="org-keyword">def</span> <span class="org-function-name">write_pheno_file</span>(pheno, gene_info, output_file, holdout=<span class="org-constant">True</span>, **kwargs):
  <span class="org-keyword">if</span> holdout:
    <span class="org-variable-name">genes</span> = gene_info.loc[gene_info[<span class="org-string">'chr'</span>].<span class="org-builtin">apply</span>(<span class="org-keyword">lambda</span> x: <span class="org-builtin">bool</span>(<span class="org-builtin">int</span>(x[2:]) % 2)).values]
  <span class="org-keyword">else</span>:
    <span class="org-variable-name">genes</span> = gene_info
  (genes
   .<span class="org-builtin">apply</span>(qtltools_format, **kwargs, axis=1)
   .merge(pheno, left_index=<span class="org-constant">True</span>, right_index=<span class="org-constant">True</span>)
   .to_csv(output_file,
           sep=<span class="org-string">'\t'</span>,
           columns=[<span class="org-string">'#Chr'</span>, <span class="org-string">'start'</span>, <span class="org-string">'end'</span>, <span class="org-string">'pid'</span>, <span class="org-string">'gid'</span>, <span class="org-string">'strand'</span>] + <span class="org-builtin">list</span>(pheno.columns),
           header=<span class="org-constant">True</span>,
           index=<span class="org-constant">False</span>,
           index_label=<span class="org-constant">False</span>))
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython">write_pheno_file(mean, gene_info, <span class="org-string">'/scratch/midway2/aksarkar/singlecell/mean.bed'</span>)
</pre>
</div>

<p>
Index the phenotype file.
</p>

<pre class="example">
Submitted batch job 44113019

</pre>

<pre class="example">
Submitted batch job 44113021

</pre>
</div>
</div>

<div id="outline-container-orgf02711e" class="outline-2">
<h2 id="orgf02711e">Dispersion-QTL calling</h2>
<div class="outline-text-2" id="text-orgf02711e">
<p>
Read the estimated parameters and center them.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">disp</span> = (pd.read_table(<span class="org-string">'/scratch/midway2/aksarkar/singlecell/zi2-dispersion.txt.gz'</span>, sep=<span class="org-string">' '</span>, index_col=0)
        .transform(<span class="org-keyword">lambda</span> x: x - x.mean(), axis=1))
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython">write_pheno_file(disp, gene_info, <span class="org-string">'/scratch/midway2/aksarkar/singlecell/disp.bed'</span>)
</pre>
</div>

<p>
Index the phenotype file.
</p>

<pre class="example">
Submitted batch job 43380178

</pre>

<p>
Run <code>qtltools</code>.
</p>

<pre class="example">
Submitted batch job 43380210

</pre>

<p>
Read the output.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">disp_qtls</span> = read_qtltools_output(<span class="org-string">'disp'</span>)
<span class="org-variable-name">keep_disp_qtls</span> = bh(disp_qtls)
</pre>
</div>

<pre class="example">
98

</pre>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">print</span>(disp_qtls[keep_disp_qtls].merge(gene_info, left_on=<span class="org-string">'gene'</span>, right_index=<span class="org-constant">True</span>)[[<span class="org-string">'gene'</span>, <span class="org-string">'name'</span>, <span class="org-string">'id'</span>, <span class="org-string">'beta'</span>, <span class="org-string">'p_beta'</span>]].reset_index(drop=<span class="org-constant">True</span>).to_html(classes=[<span class="org-string">'table'</span>]))
</pre>
</div>

<table border="1" class="dataframe table">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>gene</th>
      <th>name</th>
      <th>id</th>
      <th>beta</th>
      <th>p_beta</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>ENSG00000165283</td>
      <td>STOML2</td>
      <td>rs151013351</td>
      <td>-3.309280</td>
      <td>0.000000e+00</td>
    </tr>
    <tr>
      <th>1</th>
      <td>ENSG00000170085</td>
      <td>SIMC1</td>
      <td>rs115194011</td>
      <td>-4.063380</td>
      <td>0.000000e+00</td>
    </tr>
    <tr>
      <th>2</th>
      <td>ENSG00000197724</td>
      <td>PHF2</td>
      <td>rs113363990</td>
      <td>-4.412720</td>
      <td>7.294860e-228</td>
    </tr>
    <tr>
      <th>3</th>
      <td>ENSG00000174405</td>
      <td>LIG4</td>
      <td>rs115154048</td>
      <td>-5.077620</td>
      <td>2.507990e-182</td>
    </tr>
    <tr>
      <th>4</th>
      <td>ENSG00000176697</td>
      <td>BDNF</td>
      <td>rs143827698</td>
      <td>-4.773690</td>
      <td>2.225750e-170</td>
    </tr>
    <tr>
      <th>5</th>
      <td>ENSG00000123810</td>
      <td>B9D2</td>
      <td>rs79026999</td>
      <td>-4.370730</td>
      <td>1.672330e-151</td>
    </tr>
    <tr>
      <th>6</th>
      <td>ENSG00000102743</td>
      <td>SLC25A15</td>
      <td>rs115773631</td>
      <td>-3.519380</td>
      <td>6.031130e-123</td>
    </tr>
    <tr>
      <th>7</th>
      <td>ENSG00000125449</td>
      <td>ARMC7</td>
      <td>rs75902792</td>
      <td>-4.630780</td>
      <td>6.994600e-88</td>
    </tr>
    <tr>
      <th>8</th>
      <td>ENSG00000143183</td>
      <td>TMCO1</td>
      <td>rs114697800</td>
      <td>-2.691680</td>
      <td>4.711220e-50</td>
    </tr>
    <tr>
      <th>9</th>
      <td>ENSG00000105865</td>
      <td>DUS4L</td>
      <td>rs187529047</td>
      <td>2.818420</td>
      <td>1.040130e-44</td>
    </tr>
    <tr>
      <th>10</th>
      <td>ENSG00000120314</td>
      <td>WDR55</td>
      <td>rs11167791</td>
      <td>3.441270</td>
      <td>2.458160e-42</td>
    </tr>
    <tr>
      <th>11</th>
      <td>ENSG00000164970</td>
      <td>FAM219A</td>
      <td>rs78437190</td>
      <td>-2.935010</td>
      <td>2.794990e-37</td>
    </tr>
    <tr>
      <th>12</th>
      <td>ENSG00000162909</td>
      <td>CAPN2</td>
      <td>rs2172360</td>
      <td>3.311410</td>
      <td>9.247940e-35</td>
    </tr>
    <tr>
      <th>13</th>
      <td>ENSG00000170860</td>
      <td>LSM3</td>
      <td>rs143962723</td>
      <td>-3.295380</td>
      <td>2.297400e-33</td>
    </tr>
    <tr>
      <th>14</th>
      <td>ENSG00000244165</td>
      <td>P2RY11</td>
      <td>rs76392887</td>
      <td>-3.159010</td>
      <td>1.709660e-32</td>
    </tr>
    <tr>
      <th>15</th>
      <td>ENSG00000116898</td>
      <td>MRPS15</td>
      <td>rs144661059</td>
      <td>-2.239470</td>
      <td>1.479050e-28</td>
    </tr>
    <tr>
      <th>16</th>
      <td>ENSG00000160049</td>
      <td>DFFA</td>
      <td>rs5007607</td>
      <td>-2.658710</td>
      <td>4.592440e-26</td>
    </tr>
    <tr>
      <th>17</th>
      <td>ENSG00000171608</td>
      <td>PIK3CD</td>
      <td>rs138247028</td>
      <td>-4.697440</td>
      <td>4.235250e-25</td>
    </tr>
    <tr>
      <th>18</th>
      <td>ENSG00000117143</td>
      <td>UAP1</td>
      <td>rs12067012</td>
      <td>1.433560</td>
      <td>3.856740e-23</td>
    </tr>
    <tr>
      <th>19</th>
      <td>ENSG00000205356</td>
      <td>TECPR1</td>
      <td>rs116757663</td>
      <td>-2.618370</td>
      <td>9.657750e-18</td>
    </tr>
    <tr>
      <th>20</th>
      <td>ENSG00000114859</td>
      <td>CLCN2</td>
      <td>rs116709730</td>
      <td>4.131380</td>
      <td>1.859410e-17</td>
    </tr>
    <tr>
      <th>21</th>
      <td>ENSG00000172830</td>
      <td>SSH3</td>
      <td>rs7105623</td>
      <td>-0.638843</td>
      <td>4.644960e-17</td>
    </tr>
    <tr>
      <th>22</th>
      <td>ENSG00000183020</td>
      <td>AP2A2</td>
      <td>rs186982458</td>
      <td>2.609980</td>
      <td>2.456720e-16</td>
    </tr>
    <tr>
      <th>23</th>
      <td>ENSG00000125734</td>
      <td>GPR108</td>
      <td>rs184583582</td>
      <td>-2.822690</td>
      <td>2.470850e-15</td>
    </tr>
    <tr>
      <th>24</th>
      <td>ENSG00000161682</td>
      <td>FAM171A2</td>
      <td>rs12944040</td>
      <td>2.363410</td>
      <td>3.075220e-15</td>
    </tr>
    <tr>
      <th>25</th>
      <td>ENSG00000179409</td>
      <td>GEMIN4</td>
      <td>rs2293067</td>
      <td>1.929650</td>
      <td>8.587390e-15</td>
    </tr>
    <tr>
      <th>26</th>
      <td>ENSG00000223802</td>
      <td>CERS1</td>
      <td>rs116562609</td>
      <td>-3.084010</td>
      <td>1.026820e-13</td>
    </tr>
    <tr>
      <th>27</th>
      <td>ENSG00000172273</td>
      <td>HINFP</td>
      <td>rs115425447</td>
      <td>1.686380</td>
      <td>1.513860e-13</td>
    </tr>
    <tr>
      <th>28</th>
      <td>ENSG00000205309</td>
      <td>NT5M</td>
      <td>rs8511</td>
      <td>-2.068040</td>
      <td>2.680890e-13</td>
    </tr>
    <tr>
      <th>29</th>
      <td>ENSG00000197043</td>
      <td>ANXA6</td>
      <td>rs139372973</td>
      <td>3.311060</td>
      <td>7.743520e-13</td>
    </tr>
    <tr>
      <th>30</th>
      <td>ENSG00000221829</td>
      <td>FANCG</td>
      <td>rs10972619</td>
      <td>1.169450</td>
      <td>1.578270e-12</td>
    </tr>
    <tr>
      <th>31</th>
      <td>ENSG00000142197</td>
      <td>DOPEY2</td>
      <td>rs77145506</td>
      <td>-4.209130</td>
      <td>2.389840e-12</td>
    </tr>
    <tr>
      <th>32</th>
      <td>ENSG00000160218</td>
      <td>TRAPPC10</td>
      <td>rs111257004</td>
      <td>1.450660</td>
      <td>2.654090e-12</td>
    </tr>
    <tr>
      <th>33</th>
      <td>ENSG00000133103</td>
      <td>COG6</td>
      <td>rs147106597</td>
      <td>1.708220</td>
      <td>4.361340e-12</td>
    </tr>
    <tr>
      <th>34</th>
      <td>ENSG00000158195</td>
      <td>WASF2</td>
      <td>rs11399</td>
      <td>-2.338620</td>
      <td>1.294060e-10</td>
    </tr>
    <tr>
      <th>35</th>
      <td>ENSG00000145041</td>
      <td>VPRBP</td>
      <td>rs115267057</td>
      <td>2.494370</td>
      <td>2.579560e-09</td>
    </tr>
    <tr>
      <th>36</th>
      <td>ENSG00000110321</td>
      <td>EIF4G2</td>
      <td>rs78742382</td>
      <td>-2.541230</td>
      <td>2.636980e-09</td>
    </tr>
    <tr>
      <th>37</th>
      <td>ENSG00000110200</td>
      <td>ANAPC15</td>
      <td>rs186789865</td>
      <td>-2.565120</td>
      <td>3.441620e-09</td>
    </tr>
    <tr>
      <th>38</th>
      <td>ENSG00000149196</td>
      <td>C11orf73</td>
      <td>rs1939103</td>
      <td>2.447090</td>
      <td>4.040330e-09</td>
    </tr>
    <tr>
      <th>39</th>
      <td>ENSG00000108828</td>
      <td>VAT1</td>
      <td>rs2355375</td>
      <td>-1.710740</td>
      <td>4.750800e-09</td>
    </tr>
    <tr>
      <th>40</th>
      <td>ENSG00000179862</td>
      <td>CITED4</td>
      <td>rs115694669</td>
      <td>-5.187470</td>
      <td>8.669450e-09</td>
    </tr>
    <tr>
      <th>41</th>
      <td>ENSG00000052841</td>
      <td>TTC17</td>
      <td>rs144392024</td>
      <td>-1.613270</td>
      <td>1.177680e-08</td>
    </tr>
    <tr>
      <th>42</th>
      <td>ENSG00000166012</td>
      <td>TAF1D</td>
      <td>rs116078422</td>
      <td>-2.003780</td>
      <td>1.928780e-08</td>
    </tr>
    <tr>
      <th>43</th>
      <td>ENSG00000130287</td>
      <td>NCAN</td>
      <td>rs116770984</td>
      <td>-4.604760</td>
      <td>4.709950e-08</td>
    </tr>
    <tr>
      <th>44</th>
      <td>ENSG00000116473</td>
      <td>RAP1A</td>
      <td>rs114787626</td>
      <td>1.540230</td>
      <td>5.546490e-08</td>
    </tr>
    <tr>
      <th>45</th>
      <td>ENSG00000136861</td>
      <td>CDK5RAP2</td>
      <td>rs115356711</td>
      <td>1.754840</td>
      <td>5.685210e-08</td>
    </tr>
    <tr>
      <th>46</th>
      <td>ENSG00000154640</td>
      <td>BTG3</td>
      <td>rs2823971</td>
      <td>2.413580</td>
      <td>1.038640e-07</td>
    </tr>
    <tr>
      <th>47</th>
      <td>ENSG00000126524</td>
      <td>SBDS</td>
      <td>rs118141535</td>
      <td>2.133510</td>
      <td>1.073340e-07</td>
    </tr>
    <tr>
      <th>48</th>
      <td>ENSG00000152942</td>
      <td>RAD17</td>
      <td>rs141370675</td>
      <td>0.927315</td>
      <td>1.714180e-07</td>
    </tr>
    <tr>
      <th>49</th>
      <td>ENSG00000172893</td>
      <td>DHCR7</td>
      <td>rs115998124</td>
      <td>-2.496930</td>
      <td>1.870360e-07</td>
    </tr>
    <tr>
      <th>50</th>
      <td>ENSG00000154380</td>
      <td>ENAH</td>
      <td>rs11584969</td>
      <td>-2.854720</td>
      <td>4.772460e-07</td>
    </tr>
    <tr>
      <th>51</th>
      <td>ENSG00000154511</td>
      <td>FAM69A</td>
      <td>rs146124737</td>
      <td>-4.834600</td>
      <td>6.027520e-07</td>
    </tr>
    <tr>
      <th>52</th>
      <td>ENSG00000131238</td>
      <td>PPT1</td>
      <td>rs12090353</td>
      <td>-1.503680</td>
      <td>9.294670e-07</td>
    </tr>
    <tr>
      <th>53</th>
      <td>ENSG00000188811</td>
      <td>NHLRC3</td>
      <td>rs143196152</td>
      <td>1.264350</td>
      <td>1.191190e-06</td>
    </tr>
    <tr>
      <th>54</th>
      <td>ENSG00000122435</td>
      <td>TRMT13</td>
      <td>rs139250635</td>
      <td>1.128600</td>
      <td>1.201480e-06</td>
    </tr>
    <tr>
      <th>55</th>
      <td>ENSG00000131469</td>
      <td>RPL27</td>
      <td>rs146750609</td>
      <td>1.649960</td>
      <td>1.505230e-06</td>
    </tr>
    <tr>
      <th>56</th>
      <td>ENSG00000161558</td>
      <td>TMEM143</td>
      <td>rs142646336</td>
      <td>-4.522090</td>
      <td>1.611620e-06</td>
    </tr>
    <tr>
      <th>57</th>
      <td>ENSG00000093167</td>
      <td>LRRFIP2</td>
      <td>rs77491290</td>
      <td>2.096840</td>
      <td>2.965970e-06</td>
    </tr>
    <tr>
      <th>58</th>
      <td>ENSG00000170946</td>
      <td>DNAJC24</td>
      <td>rs75350757</td>
      <td>2.622850</td>
      <td>4.225450e-06</td>
    </tr>
    <tr>
      <th>59</th>
      <td>ENSG00000149591</td>
      <td>TAGLN</td>
      <td>rs75419894</td>
      <td>-1.560600</td>
      <td>4.950990e-06</td>
    </tr>
    <tr>
      <th>60</th>
      <td>ENSG00000131236</td>
      <td>CAP1</td>
      <td>rs150600342</td>
      <td>-1.748100</td>
      <td>7.383560e-06</td>
    </tr>
    <tr>
      <th>61</th>
      <td>ENSG00000167766</td>
      <td>ZNF83</td>
      <td>rs141851483</td>
      <td>3.292110</td>
      <td>1.065440e-05</td>
    </tr>
    <tr>
      <th>62</th>
      <td>ENSG00000179163</td>
      <td>FUCA1</td>
      <td>rs79403581</td>
      <td>-2.898940</td>
      <td>1.160650e-05</td>
    </tr>
    <tr>
      <th>63</th>
      <td>ENSG00000254585</td>
      <td>MAGEL2</td>
      <td>rs12916127</td>
      <td>-2.993250</td>
      <td>1.536980e-05</td>
    </tr>
    <tr>
      <th>64</th>
      <td>ENSG00000166436</td>
      <td>TRIM66</td>
      <td>rs143365151</td>
      <td>-4.038500</td>
      <td>1.777680e-05</td>
    </tr>
    <tr>
      <th>65</th>
      <td>ENSG00000154814</td>
      <td>OXNAD1</td>
      <td>rs186345075</td>
      <td>3.438130</td>
      <td>1.895770e-05</td>
    </tr>
    <tr>
      <th>66</th>
      <td>ENSG00000172531</td>
      <td>PPP1CA</td>
      <td>rs4930435</td>
      <td>-2.254580</td>
      <td>2.211000e-05</td>
    </tr>
    <tr>
      <th>67</th>
      <td>ENSG00000011243</td>
      <td>AKAP8L</td>
      <td>rs74393806</td>
      <td>1.997810</td>
      <td>2.666640e-05</td>
    </tr>
    <tr>
      <th>68</th>
      <td>ENSG00000248098</td>
      <td>BCKDHA</td>
      <td>rs140346884</td>
      <td>-2.913330</td>
      <td>2.763770e-05</td>
    </tr>
    <tr>
      <th>69</th>
      <td>ENSG00000143198</td>
      <td>MGST3</td>
      <td>rs144467420</td>
      <td>-2.748340</td>
      <td>4.538870e-05</td>
    </tr>
    <tr>
      <th>70</th>
      <td>ENSG00000174574</td>
      <td>AKIRIN1</td>
      <td>rs143588389</td>
      <td>-2.988750</td>
      <td>6.604600e-05</td>
    </tr>
    <tr>
      <th>71</th>
      <td>ENSG00000103707</td>
      <td>MTFMT</td>
      <td>rs137966506</td>
      <td>3.245180</td>
      <td>6.846880e-05</td>
    </tr>
    <tr>
      <th>72</th>
      <td>ENSG00000145681</td>
      <td>HAPLN1</td>
      <td>rs114887660</td>
      <td>-3.971260</td>
      <td>7.128160e-05</td>
    </tr>
    <tr>
      <th>73</th>
      <td>ENSG00000184408</td>
      <td>KCND2</td>
      <td>rs143795479</td>
      <td>-4.538730</td>
      <td>8.513240e-05</td>
    </tr>
    <tr>
      <th>74</th>
      <td>ENSG00000137841</td>
      <td>PLCB2</td>
      <td>rs62019961</td>
      <td>-4.512630</td>
      <td>8.853740e-05</td>
    </tr>
    <tr>
      <th>75</th>
      <td>ENSG00000108557</td>
      <td>RAI1</td>
      <td>rs59212772</td>
      <td>-6.439100</td>
      <td>1.211490e-04</td>
    </tr>
    <tr>
      <th>76</th>
      <td>ENSG00000006016</td>
      <td>CRLF1</td>
      <td>rs10414298</td>
      <td>-1.853580</td>
      <td>1.396770e-04</td>
    </tr>
    <tr>
      <th>77</th>
      <td>ENSG00000109084</td>
      <td>TMEM97</td>
      <td>rs144069393</td>
      <td>-2.782690</td>
      <td>1.462980e-04</td>
    </tr>
    <tr>
      <th>78</th>
      <td>ENSG00000106086</td>
      <td>PLEKHA8</td>
      <td>rs118033320</td>
      <td>2.487410</td>
      <td>1.639470e-04</td>
    </tr>
    <tr>
      <th>79</th>
      <td>ENSG00000116906</td>
      <td>GNPAT</td>
      <td>rs10779810</td>
      <td>-1.170110</td>
      <td>1.677310e-04</td>
    </tr>
    <tr>
      <th>80</th>
      <td>ENSG00000064666</td>
      <td>CNN2</td>
      <td>rs184326775</td>
      <td>-1.410370</td>
      <td>1.764350e-04</td>
    </tr>
    <tr>
      <th>81</th>
      <td>ENSG00000149256</td>
      <td>TENM4</td>
      <td>rs189331375</td>
      <td>-4.245470</td>
      <td>1.807590e-04</td>
    </tr>
    <tr>
      <th>82</th>
      <td>ENSG00000117036</td>
      <td>ETV3</td>
      <td>rs188035981</td>
      <td>3.118750</td>
      <td>1.978480e-04</td>
    </tr>
    <tr>
      <th>83</th>
      <td>ENSG00000073050</td>
      <td>XRCC1</td>
      <td>rs138843872</td>
      <td>3.159450</td>
      <td>2.268670e-04</td>
    </tr>
    <tr>
      <th>84</th>
      <td>ENSG00000185453</td>
      <td>C19orf68</td>
      <td>rs138701164</td>
      <td>4.117920</td>
      <td>2.468340e-04</td>
    </tr>
    <tr>
      <th>85</th>
      <td>ENSG00000166452</td>
      <td>AKIP1</td>
      <td>rs141004297</td>
      <td>3.508760</td>
      <td>2.546470e-04</td>
    </tr>
    <tr>
      <th>86</th>
      <td>ENSG00000092978</td>
      <td>GPATCH2</td>
      <td>rs113462534</td>
      <td>4.085070</td>
      <td>2.997230e-04</td>
    </tr>
    <tr>
      <th>87</th>
      <td>ENSG00000131473</td>
      <td>ACLY</td>
      <td>rs111713404</td>
      <td>-2.099470</td>
      <td>3.098620e-04</td>
    </tr>
    <tr>
      <th>88</th>
      <td>ENSG00000160867</td>
      <td>FGFR4</td>
      <td>rs111567185</td>
      <td>2.796980</td>
      <td>3.604680e-04</td>
    </tr>
    <tr>
      <th>89</th>
      <td>ENSG00000186472</td>
      <td>PCLO</td>
      <td>rs78573728</td>
      <td>-4.342580</td>
      <td>3.955380e-04</td>
    </tr>
    <tr>
      <th>90</th>
      <td>ENSG00000132510</td>
      <td>KDM6B</td>
      <td>rs115083985</td>
      <td>-1.067740</td>
      <td>4.668840e-04</td>
    </tr>
    <tr>
      <th>91</th>
      <td>ENSG00000070669</td>
      <td>ASNS</td>
      <td>rs79183163</td>
      <td>-2.619050</td>
      <td>4.672200e-04</td>
    </tr>
    <tr>
      <th>92</th>
      <td>ENSG00000105671</td>
      <td>DDX49</td>
      <td>rs147349438</td>
      <td>1.518330</td>
      <td>5.734510e-04</td>
    </tr>
    <tr>
      <th>93</th>
      <td>ENSG00000165895</td>
      <td>ARHGAP42</td>
      <td>rs111702886</td>
      <td>-4.671170</td>
      <td>5.748990e-04</td>
    </tr>
    <tr>
      <th>94</th>
      <td>ENSG00000166333</td>
      <td>ILK</td>
      <td>rs77801446</td>
      <td>-2.310190</td>
      <td>7.091510e-04</td>
    </tr>
    <tr>
      <th>95</th>
      <td>ENSG00000148110</td>
      <td>HIATL1</td>
      <td>rs9409727</td>
      <td>-1.103750</td>
      <td>8.033820e-04</td>
    </tr>
    <tr>
      <th>96</th>
      <td>ENSG00000117906</td>
      <td>RCN2</td>
      <td>rs8026731</td>
      <td>-1.355620</td>
      <td>8.237700e-04</td>
    </tr>
    <tr>
      <th>97</th>
      <td>ENSG00000085276</td>
      <td>MECOM</td>
      <td>rs140988682</td>
      <td>-4.053850</td>
      <td>8.395570e-04</td>
    </tr>
  </tbody>
</table>

<p>
Check the beta approximation to the permutation p-values:
</p>

<div class="org-src-container">
<pre class="src src-ipython">plt.clf()
plt.scatter(disp_qtls[<span class="org-string">'p_empirical'</span>], disp_qtls[<span class="org-string">'p_beta'</span>], s=1, c=<span class="org-string">'k'</span>)
plt.plot([0, 1], [0, 1], c=<span class="org-string">'r'</span>, ls=<span class="org-string">'--'</span>)
plt.xlabel(<span class="org-string">'Empirical p-value'</span>)
plt.ylabel(<span class="org-string">'Approximate p-value'</span>)
</pre>
</div>

<pre class="example">
Text(0,0.5,'Approximate p-value')

</pre>

<div class="figure">
<p><img src="figure/qtl-mapping.org/zi2-disp-qtl-beta.png" alt="zi2-disp-qtl-beta.png">
</p>
</div>

<p>
Plot the quantile-quantile plot of QTL test statistics:
</p>

<div class="org-src-container">
<pre class="src src-ipython">qqplot(disp_qtls[<span class="org-string">'p_beta'</span>])
</pre>
</div>


<div class="figure">
<p><img src="figure/qtl-mapping.org/zi2-disp-qtl-qq.png" alt="zi2-disp-qtl-qq.png">
</p>
</div>
</div>
</div>
</div>
<div id="postamble" class="status">
<p class="author">Author: Abhishek Sarkar</p>
<p class="date">Created: 2018-03-05 Mon 14:04</p>
<p class="validation"><a href="http://validator.w3.org/check?uri=referer">Validate</a></p>
</div>
</body>
</html>
