<!DOCTYPE html>
<html lang="en">
<head>
<!-- 2017-12-05 Tue 10:31 -->
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
<title>QTL mapping pipeline</title>
<meta name="generator" content="Org mode">
<meta name="author" content="Abhishek Sarkar">
<style type="text/css">
 <!--/*--><![CDATA[/*><!--*/
  .title  { text-align: center;
             margin-bottom: .2em; }
  .subtitle { text-align: center;
              font-size: medium;
              font-weight: bold;
              margin-top:0; }
  .todo   { font-family: monospace; color: red; }
  .done   { font-family: monospace; color: green; }
  .priority { font-family: monospace; color: orange; }
  .tag    { background-color: #eee; font-family: monospace;
            padding: 2px; font-size: 80%; font-weight: normal; }
  .timestamp { color: #bebebe; }
  .timestamp-kwd { color: #5f9ea0; }
  .org-right  { margin-left: auto; margin-right: 0px;  text-align: right; }
  .org-left   { margin-left: 0px;  margin-right: auto; text-align: left; }
  .org-center { margin-left: auto; margin-right: auto; text-align: center; }
  .underline { text-decoration: underline; }
  #postamble p, #preamble p { font-size: 90%; margin: .2em; }
  p.verse { margin-left: 3%; }
  pre {
    border: 1px solid #ccc;
    box-shadow: 3px 3px 3px #eee;
    padding: 8pt;
    font-family: monospace;
    overflow: auto;
    margin: 1.2em;
  }
  pre.src {
    position: relative;
    overflow: visible;
    padding-top: 1.2em;
  }
  pre.src:before {
    display: none;
    position: absolute;
    background-color: white;
    top: -10px;
    right: 10px;
    padding: 3px;
    border: 1px solid black;
  }
  pre.src:hover:before { display: inline;}
  /* Languages per Org manual */
  pre.src-asymptote:before { content: 'Asymptote'; }
  pre.src-awk:before { content: 'Awk'; }
  pre.src-C:before { content: 'C'; }
  /* pre.src-C++ doesn't work in CSS */
  pre.src-clojure:before { content: 'Clojure'; }
  pre.src-css:before { content: 'CSS'; }
  pre.src-D:before { content: 'D'; }
  pre.src-ditaa:before { content: 'ditaa'; }
  pre.src-dot:before { content: 'Graphviz'; }
  pre.src-calc:before { content: 'Emacs Calc'; }
  pre.src-emacs-lisp:before { content: 'Emacs Lisp'; }
  pre.src-fortran:before { content: 'Fortran'; }
  pre.src-gnuplot:before { content: 'gnuplot'; }
  pre.src-haskell:before { content: 'Haskell'; }
  pre.src-hledger:before { content: 'hledger'; }
  pre.src-java:before { content: 'Java'; }
  pre.src-js:before { content: 'Javascript'; }
  pre.src-latex:before { content: 'LaTeX'; }
  pre.src-ledger:before { content: 'Ledger'; }
  pre.src-lisp:before { content: 'Lisp'; }
  pre.src-lilypond:before { content: 'Lilypond'; }
  pre.src-lua:before { content: 'Lua'; }
  pre.src-matlab:before { content: 'MATLAB'; }
  pre.src-mscgen:before { content: 'Mscgen'; }
  pre.src-ocaml:before { content: 'Objective Caml'; }
  pre.src-octave:before { content: 'Octave'; }
  pre.src-org:before { content: 'Org mode'; }
  pre.src-oz:before { content: 'OZ'; }
  pre.src-plantuml:before { content: 'Plantuml'; }
  pre.src-processing:before { content: 'Processing.js'; }
  pre.src-python:before { content: 'Python'; }
  pre.src-R:before { content: 'R'; }
  pre.src-ruby:before { content: 'Ruby'; }
  pre.src-sass:before { content: 'Sass'; }
  pre.src-scheme:before { content: 'Scheme'; }
  pre.src-screen:before { content: 'Gnu Screen'; }
  pre.src-sed:before { content: 'Sed'; }
  pre.src-sh:before { content: 'shell'; }
  pre.src-sql:before { content: 'SQL'; }
  pre.src-sqlite:before { content: 'SQLite'; }
  /* additional languages in org.el's org-babel-load-languages alist */
  pre.src-forth:before { content: 'Forth'; }
  pre.src-io:before { content: 'IO'; }
  pre.src-J:before { content: 'J'; }
  pre.src-makefile:before { content: 'Makefile'; }
  pre.src-maxima:before { content: 'Maxima'; }
  pre.src-perl:before { content: 'Perl'; }
  pre.src-picolisp:before { content: 'Pico Lisp'; }
  pre.src-scala:before { content: 'Scala'; }
  pre.src-shell:before { content: 'Shell Script'; }
  pre.src-ebnf2ps:before { content: 'ebfn2ps'; }
  /* additional language identifiers per "defun org-babel-execute"
       in ob-*.el */
  pre.src-cpp:before  { content: 'C++'; }
  pre.src-abc:before  { content: 'ABC'; }
  pre.src-coq:before  { content: 'Coq'; }
  pre.src-groovy:before  { content: 'Groovy'; }
  /* additional language identifiers from org-babel-shell-names in
     ob-shell.el: ob-shell is the only babel language using a lambda to put
     the execution function name together. */
  pre.src-bash:before  { content: 'bash'; }
  pre.src-csh:before  { content: 'csh'; }
  pre.src-ash:before  { content: 'ash'; }
  pre.src-dash:before  { content: 'dash'; }
  pre.src-ksh:before  { content: 'ksh'; }
  pre.src-mksh:before  { content: 'mksh'; }
  pre.src-posh:before  { content: 'posh'; }
  /* Additional Emacs modes also supported by the LaTeX listings package */
  pre.src-ada:before { content: 'Ada'; }
  pre.src-asm:before { content: 'Assembler'; }
  pre.src-caml:before { content: 'Caml'; }
  pre.src-delphi:before { content: 'Delphi'; }
  pre.src-html:before { content: 'HTML'; }
  pre.src-idl:before { content: 'IDL'; }
  pre.src-mercury:before { content: 'Mercury'; }
  pre.src-metapost:before { content: 'MetaPost'; }
  pre.src-modula-2:before { content: 'Modula-2'; }
  pre.src-pascal:before { content: 'Pascal'; }
  pre.src-ps:before { content: 'PostScript'; }
  pre.src-prolog:before { content: 'Prolog'; }
  pre.src-simula:before { content: 'Simula'; }
  pre.src-tcl:before { content: 'tcl'; }
  pre.src-tex:before { content: 'TeX'; }
  pre.src-plain-tex:before { content: 'Plain TeX'; }
  pre.src-verilog:before { content: 'Verilog'; }
  pre.src-vhdl:before { content: 'VHDL'; }
  pre.src-xml:before { content: 'XML'; }
  pre.src-nxml:before { content: 'XML'; }
  /* add a generic configuration mode; LaTeX export needs an additional
     (add-to-list 'org-latex-listings-langs '(conf " ")) in .emacs */
  pre.src-conf:before { content: 'Configuration File'; }

  table { border-collapse:collapse; }
  caption.t-above { caption-side: top; }
  caption.t-bottom { caption-side: bottom; }
  td, th { vertical-align:top;  }
  th.org-right  { text-align: center;  }
  th.org-left   { text-align: center;   }
  th.org-center { text-align: center; }
  td.org-right  { text-align: right;  }
  td.org-left   { text-align: left;   }
  td.org-center { text-align: center; }
  dt { font-weight: bold; }
  .footpara { display: inline; }
  .footdef  { margin-bottom: 1em; }
  .figure { padding: 1em; }
  .figure p { text-align: center; }
  .inlinetask {
    padding: 10px;
    border: 2px solid gray;
    margin: 10px;
    background: #ffffcc;
  }
  #org-div-home-and-up
   { text-align: right; font-size: 70%; white-space: nowrap; }
  textarea { overflow-x: auto; }
  .linenr { font-size: smaller }
  .code-highlighted { background-color: #ffff00; }
  .org-info-js_info-navigation { border-style: none; }
  #org-info-js_console-label
    { font-size: 10px; font-weight: bold; white-space: nowrap; }
  .org-info-js_search-highlight
    { background-color: #ffff00; color: #000000; font-weight: bold; }
  .org-svg { width: 90%; }
  /*]]>*/-->
</style>
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<link rel="stylesheet" type="text/css" href="https://www.pirilampo.org/styles/readtheorg/css/htmlize.css"/>
<style type="text/css">body {width: 60em; margin:auto} pre.src {overflow:auto}</style>
<script type="text/javascript">
/*
@licstart  The following is the entire license notice for the
JavaScript code in this tag.

Copyright (C) 2012-2017 Free Software Foundation, Inc.

The JavaScript code in this tag is free software: you can
redistribute it and/or modify it under the terms of the GNU
General Public License (GNU GPL) as published by the Free Software
Foundation, either version 3 of the License, or (at your option)
any later version.  The code is distributed WITHOUT ANY WARRANTY;
without even the implied warranty of MERCHANTABILITY or FITNESS
FOR A PARTICULAR PURPOSE.  See the GNU GPL for more details.

As additional permission under GNU GPL version 3 section 7, you
may distribute non-source (e.g., minimized or compacted) forms of
that code without the copy of the GNU GPL normally required by
section 4, provided you include this license notice and a URL
through which recipients can access the Corresponding Source.


@licend  The above is the entire license notice
for the JavaScript code in this tag.
*/
<!--/*--><![CDATA[/*><!--*/
 function CodeHighlightOn(elem, id)
 {
   var target = document.getElementById(id);
   if(null != target) {
     elem.cacheClassElem = elem.className;
     elem.cacheClassTarget = target.className;
     target.className = "code-highlighted";
     elem.className   = "code-highlighted";
   }
 }
 function CodeHighlightOff(elem, id)
 {
   var target = document.getElementById(id);
   if(elem.cacheClassElem)
     elem.className = elem.cacheClassElem;
   if(elem.cacheClassTarget)
     target.className = elem.cacheClassTarget;
 }
/*]]>*///-->
</script>
<script type="text/x-mathjax-config">
    MathJax.Hub.Config({
        displayAlign: "center",
        displayIndent: "0em",

        "HTML-CSS": { scale: 100,
                        linebreaks: { automatic: "false" },
                        webFont: "TeX"
                       },
        SVG: {scale: 100,
              linebreaks: { automatic: "false" },
              font: "TeX"},
        NativeMML: {scale: 100},
        TeX: { equationNumbers: {autoNumber: "AMS"},
               MultLineWidth: "85%",
               TagSide: "right",
               TagIndent: ".8em"
             }
});
</script>
<script type="text/javascript"
        src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-AMS_HTML"></script>
</head>
<body>
<div id="content">
<h1 class="title">QTL mapping pipeline</h1>
<div id="table-of-contents">
<h2>Table of Contents</h2>
<div id="text-table-of-contents">
<ul>
<li><a href="#org5ed2a04">Introduction</a></li>
<li><a href="#org149d7c1">Quality control on expression</a></li>
<li><a href="#orgf315729">Mean-QTL calling</a></li>
<li><a href="#orgc8a9fe1">Variance-QTL calling</a></li>
<li><a href="#orga4d85dd">Next steps</a></li>
</ul>
</div>
</div>

<div id="outline-container-org5ed2a04" class="outline-2">
<h2 id="org5ed2a04">Introduction</h2>
<div class="outline-text-2" id="text-org5ed2a04">
<p>
Here, we set up the infrastructure to call mean and variance QTLs. 
</p>

<p>
The key issues are:
</p>

<ol class="org-ol">
<li><b>How do we account for confounders in UMI?</b> As a first pass (to test the
pipeline), use preliminary QC filters and assume no confounding.</li>
<li><b>Do we account for zeros?</b> As a first pass, we can estimate means and
variances per individual using only detected genes.</li>
<li><b>Do we jointly model mean and variance QTLs?</b> As a first pass, we can
simply include the observed mean in the variance QTL model (like <code>voom</code>,
<a href="http://genomebiology.com/2014/15/2/R29">Law et al 2014</a>).</li>
<li><b>What parameters are shared between cells/individuals/genes?</b> As a first
pass, assume no parameters are shared.</li>
</ol>
</div>
</div>

<div id="outline-container-org149d7c1" class="outline-2">
<h2 id="org149d7c1">Quality control on expression</h2>
<div class="outline-text-2" id="text-org149d7c1">
<p>
Filter out cells on percent spike-in and gene detection rate, and filter out
genes on individual detection rate.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">keep_cells</span> = functools.<span class="org-builtin">reduce</span>(
  np.logical_and,
  [
    annotations[<span class="org-string">'reads_ercc'</span>] / annotations.<span class="org-builtin">filter</span>(like=<span class="org-string">'reads_'</span>, axis=1).agg(np.<span class="org-builtin">sum</span>, axis=1) &lt; 0.5,
    annotations[<span class="org-string">'detect_hs'</span>] &gt; 4000,
  ]).values
<span class="org-variable-name">keep_genes</span> = functools.<span class="org-builtin">reduce</span>(
  np.logical_and,
  [
    (umi.groupby(annotations[<span class="org-string">'chip_id'</span>].values, axis=1).agg(np.<span class="org-builtin">sum</span>) &gt; 0).<span class="org-builtin">apply</span>(<span class="org-keyword">lambda</span> x: x.mean() &gt; 0.5, axis=1)
  ]).values
<span class="org-variable-name">umi_qc</span> = umi.loc[keep_genes, keep_cells]
<span class="org-variable-name">annotations_qc</span> = annotations.loc[keep_cells]
umi_qc.mask(umi_qc == 0)
umi_qc.shape
</pre>
</div>

<pre class="example">
(15636, 1861)

</pre>
</div>
</div>

<div id="outline-container-orgf315729" class="outline-2">
<h2 id="orgf315729">Mean-QTL calling</h2>
<div class="outline-text-2" id="text-orgf315729">
<p>
Estimate the mean count per individual, per gene using non-zero counts only,
then transform to \(log_2(CPM + 1)\). 
</p>

<p>
We still have to deal with the case where a gene is not detected in any cells
in a particular individual. As a first pass, simply treat their mean
expression as a true zero.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">vcf</span> = (pd.read_table(<span class="org-string">'/project2/gilad/singlecell-qtl/bulk/genotypes.vcf.gz'</span>, nrows=1)
       .rename(columns=<span class="org-keyword">lambda</span> x: <span class="org-string">'NA{}'</span>.<span class="org-builtin">format</span>(x) <span class="org-keyword">if</span> x.startswith(<span class="org-string">'1'</span>) <span class="org-keyword">else</span> x))
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">mean</span> = cpm(umi_qc.mask(umi_qc == 0)
           .groupby(annotations_qc[<span class="org-string">'chip_id'</span>].values, axis=1)
           .agg(np.mean)
           .<span class="org-builtin">filter</span>(items=vcf.columns)
           .rename(columns=<span class="org-keyword">lambda</span> x: x[2:])  <span class="org-comment-delimiter"># </span><span class="org-comment">match original VCF</span>
           .fillna(0),
           log2=<span class="org-constant">True</span>)
</pre>
</div>

<p>
Write out the <a href="https://qtltools.github.io/qtltools/pages/input_files.html">phenotype file</a> for <code>QTLtools</code>.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">gene_info</span> = (pd.read_table(<span class="org-string">'/home/aksarkar/projects/singlecell-qtl/data/scqtl-genes.txt.gz'</span>)
             .set_index(<span class="org-string">'gene'</span>)
             .query(<span class="org-string">'source == "H. sapiens"'</span>)
             .query(<span class="org-string">'chr != "hsX"'</span>)
             .query(<span class="org-string">'chr != "hsY"'</span>)
             .query(<span class="org-string">'chr != "hsMT"'</span>))
gene_info.head()
</pre>
</div>

<pre class="example">
                chr      start        end      name strand      source
gene                                                                    
ENSG00000000419  hs20   49551404   49575092      DPM1      -  H. sapiens
ENSG00000000457   hs1  169818772  169863408     SCYL3      -  H. sapiens
ENSG00000000460   hs1  169631245  169823221  C1orf112      +  H. sapiens
ENSG00000000938   hs1   27938575   27961788       FGR      -  H. sapiens
ENSG00000000971   hs1  196621008  196716634       CFH      +  H. sapiens
</pre>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">def</span> <span class="org-function-name">qtltools_format</span>(row):
  <span class="org-variable-name">row</span>[<span class="org-string">'#Chr'</span>] = <span class="org-string">'chr{}'</span>.<span class="org-builtin">format</span>(row[<span class="org-string">'chr'</span>][2:])
  <span class="org-variable-name">row</span>[<span class="org-string">'gid'</span>] = row.name
  <span class="org-variable-name">row</span>[<span class="org-string">'pid'</span>] = row.name
  <span class="org-keyword">return</span> row

(gene_info
 .<span class="org-builtin">apply</span>(qtltools_format, axis=1)
 .merge(mean, left_index=<span class="org-constant">True</span>, right_index=<span class="org-constant">True</span>)
 .to_csv(<span class="org-string">'mean.bed'</span>,
         sep=<span class="org-string">'\t'</span>,
         columns=[<span class="org-string">'#Chr'</span>, <span class="org-string">'start'</span>, <span class="org-string">'end'</span>, <span class="org-string">'pid'</span>, <span class="org-string">'gid'</span>, <span class="org-string">'strand'</span>] + <span class="org-builtin">list</span>(mean.columns),
         header=<span class="org-constant">True</span>,
         index=<span class="org-constant">False</span>,
         index_label=<span class="org-constant">False</span>)
)
</pre>
</div>

<p>
Index the phenotype file.
</p>

<div class="org-src-container">
<pre class="src src-sh">sbatch --partition=mstephens
<span class="org-comment-delimiter">#</span><span class="org-comment">!/bin/bash</span>
sort -k1,1 -k2,2n -k3,3n mean.bed | bgzip &gt;mean.bed.gz
tabix -p bed mean.bed.gz
</pre>
</div>

<p>
Run the QTL mapping.
</p>

<div class="org-src-container">
<pre class="src src-sh">sbatch --partition=mstephens -c4 -J mean-qtl -o mean-qtl.log
<span class="org-comment-delimiter">#</span><span class="org-comment">!/bin/bash</span>
parallel -j4 qtltools cis --vcf /project2/gilad/singlecell-qtl/bulk/genotypes.vcf.gz --bed mean.bed.gz --nominal=0.01 --chunk {#} 100 --out mean-qtl.{#}.txt ::: $(<span class="org-sh-quoted-exec">seq</span> 1 100)
</pre>
</div>

<p>
Read the results.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">mean_qtls</span> = (pd.concat([pd.read_table(<span class="org-string">'mean-qtl.{}.txt'</span>.<span class="org-builtin">format</span>(i), header=<span class="org-constant">None</span>, sep=<span class="org-string">' '</span>) <span class="org-keyword">for</span> i <span class="org-keyword">in</span> <span class="org-builtin">range</span>(1, 101) <span class="org-keyword">if</span> os.path.getsize(<span class="org-string">'mean-qtl.{}.txt'</span>.<span class="org-builtin">format</span>(i)) &gt; 0])
             .rename(columns={i: x <span class="org-keyword">for</span> i, x <span class="org-keyword">in</span> <span class="org-builtin">enumerate</span>([<span class="org-string">'gene'</span>, <span class="org-string">'chr'</span>, <span class="org-string">'start'</span>, <span class="org-string">'end'</span>, <span class="org-string">'strand'</span>, <span class="org-string">'num_vars'</span>, <span class="org-string">'distance'</span>, <span class="org-string">'id'</span>, <span class="org-string">'var_chr'</span>, <span class="org-string">'var_start'</span>, <span class="org-string">'var_end'</span>, <span class="org-string">'p'</span>, <span class="org-string">'beta'</span>, <span class="org-string">'top'</span>])})
             .set_index(<span class="org-string">'gene'</span>))
mean_qtls.head()
</pre>
</div>

<pre class="example">
                chr   start     end strand  num_vars  distance id var_chr  \
gene                                                                          
ENSG00000237683  chr1  134902  139379      -      1720    134902  T    chr1   
ENSG00000237683  chr1  134902  139379      -      1720    134902  C    chr1   
ENSG00000237683  chr1  134902  139379      -      1720    134902  G    chr1   
ENSG00000237683  chr1  134902  139379      -      1720    134902  A    chr1   
ENSG00000237683  chr1  134902  139379      -      1720    134902  T    chr1   

                 var_start  var_end         p      beta  top  
gene                                                          
ENSG00000237683     793228        0  0.004931  0.569071    0  
ENSG00000237683     796101        0  0.008686  0.527685    0  
ENSG00000237683     845275        0  0.006838  0.175679    0  
ENSG00000237683     846339        0  0.008093  0.215268    0  
ENSG00000237683     846490        0  0.008093  0.215268    0  
</pre>

<p>
<b>TODO:</b> Permutation testing within QTLtools? Multiple testing correction by
<code>ashr</code>?
</p>

<p>
Compute the total number of QTLs detected.
</p>

<div class="org-src-container">
<pre class="src src-ipython">mean_qtls.groupby(level=0)[<span class="org-string">'chr'</span>].agg(<span class="org-builtin">len</span>).<span class="org-builtin">sum</span>()
</pre>
</div>

<pre class="example">
9287

</pre>

<p>
Compare against the bulk RNA-Seq mean-QTLs.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">def</span> <span class="org-function-name">bulk_mean_qtl_format</span>(row):
  <span class="org-variable-name">row</span>[<span class="org-string">'pid'</span>] = row[<span class="org-string">'pid'</span>].split(<span class="org-string">'.'</span>)[0]
  row[<span class="org-string">'rsid'</span>], row[<span class="org-string">'var_chr'</span>], <span class="org-variable-name">row</span>[<span class="org-string">'var_start'</span>] = row[<span class="org-string">'dummy2'</span>].split(<span class="org-string">'.'</span>)
  <span class="org-keyword">del</span> row[<span class="org-string">'dummy2'</span>]
  <span class="org-keyword">return</span> row

<span class="org-variable-name">bulk_mean_qtls</span> = (pd.read_table(<span class="org-string">'/project2/gilad/singlecell-qtl/bulk/permutations.all.RNAseq_run.fixed.txt.gz.bh.txt'</span>, sep=<span class="org-string">' '</span>)
                  .<span class="org-builtin">apply</span>(bulk_mean_qtl_format, axis=1)
                  .set_index(<span class="org-string">'pid'</span>))
bulk_mean_qtls.head()
</pre>
</div>

<pre class="example">
                      bpval        rsid var_chr var_start
pid                                                        
ENSG00000047056  1.065450e-03  rs12268949   chr10   1104257
ENSG00000196139  6.354310e-03   rs1937889   chr10   5057741
ENSG00000237943  2.074140e-04   rs4515867   chr10   6672053
ENSG00000165623  1.970260e-04   rs7893239   chr10  13275510
ENSG00000152457  2.142050e-09   rs7084951   chr10  15013439
</pre>
</div>
</div>

<div id="outline-container-orgc8a9fe1" class="outline-2">
<h2 id="orgc8a9fe1">Variance-QTL calling</h2>
<div class="outline-text-2" id="text-orgc8a9fe1">
<p>
The fundamental relationship derived by <a href="http://genomebiology.com/2014/15/2/R29">Law et al 2014</a> to transform to
Gaussian phenotypes is:
</p>

<p>
\[ \mathrm{variance} = \frac{1}{\mathrm{mean}} + \mathrm{dispersion} \]
</p>

<p>
<b>TODO:</b> Do we need to worry about mean QTLs in LD with variance QTLs? Can we
just plug in the mean? Should we use predictions from multivariate mean-QTL
models? Should we fit a Gaussian process regression (per individual, or
pooling all individuals)?
</p>
</div>
</div>

<div id="outline-container-orga4d85dd" class="outline-2">
<h2 id="orga4d85dd">Next steps</h2>
<div class="outline-text-2" id="text-orga4d85dd">
<ol class="org-ol">
<li><p>
We previously used half-sibling regression (<a href="https://www.pnas.org/cgi/doi/10.1073/pnas.1511656113">Scholkopf et al 2016</a>) to correct
confounders for bulk RNA-Seq in GTEx (<a href="http://dx.doi.org/10.1101/107623">Park et al 2017</a>).
</p>

<p>
The key idea is that the only ways we could predict the expression of a
gene of interest using genes on other chromosomes are (1) true
<i>trans</i>-regulation and (2) shared confounding. For <i>cis</i>-QTL mapping, we
want to eliminate both of these, which we can do by simply regressing out
gene expression of all genes on all other chromosomes.
</p>

<p>
One open question is whether we should correct counts (which would be
required to solve issue (3)), or correct per-individual means/variances.
</p></li>

<li><p>
We previously built an <a href="https://github.com/YPARK/fqtl">inference engine which supports negative binomial
models</a> which could be easily extended to zero-inflated negative binomial
models.
</p>

<p>
In our experiments directly using the negative binomial likelihood
outperformed the Gaussian model (following the <code>voom</code>-transform, but
sharing the mean model parameters in the dispersion model).
</p>

<p>
This would allow us to solve issues (2) and (3), and potentially make (4)
easier.
</p></li>
</ol>
</div>
</div>
</div>
<div id="postamble" class="status">
<p class="author">Author: Abhishek Sarkar</p>
<p class="date">Created: 2017-12-05 Tue 10:31</p>
<p class="validation"><a href="http://validator.w3.org/check?uri=referer">Validate</a></p>
</div>
</body>
</html>
