<!DOCTYPE html>
<html lang="en">
<head>
<!-- 2018-03-02 Fri 10:23 -->
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
<title>QTL mapping pipeline</title>
<meta name="generator" content="Org mode">
<meta name="author" content="Abhishek Sarkar">
<style type="text/css">
 <!--/*--><![CDATA[/*><!--*/
  .title  { text-align: center;
             margin-bottom: .2em; }
  .subtitle { text-align: center;
              font-size: medium;
              font-weight: bold;
              margin-top:0; }
  .todo   { font-family: monospace; color: red; }
  .done   { font-family: monospace; color: green; }
  .priority { font-family: monospace; color: orange; }
  .tag    { background-color: #eee; font-family: monospace;
            padding: 2px; font-size: 80%; font-weight: normal; }
  .timestamp { color: #bebebe; }
  .timestamp-kwd { color: #5f9ea0; }
  .org-right  { margin-left: auto; margin-right: 0px;  text-align: right; }
  .org-left   { margin-left: 0px;  margin-right: auto; text-align: left; }
  .org-center { margin-left: auto; margin-right: auto; text-align: center; }
  .underline { text-decoration: underline; }
  #postamble p, #preamble p { font-size: 90%; margin: .2em; }
  p.verse { margin-left: 3%; }
  pre {
    border: 1px solid #ccc;
    box-shadow: 3px 3px 3px #eee;
    padding: 8pt;
    font-family: monospace;
    overflow: auto;
    margin: 1.2em;
  }
  pre.src {
    position: relative;
    overflow: visible;
    padding-top: 1.2em;
  }
  pre.src:before {
    display: none;
    position: absolute;
    background-color: white;
    top: -10px;
    right: 10px;
    padding: 3px;
    border: 1px solid black;
  }
  pre.src:hover:before { display: inline;}
  /* Languages per Org manual */
  pre.src-asymptote:before { content: 'Asymptote'; }
  pre.src-awk:before { content: 'Awk'; }
  pre.src-C:before { content: 'C'; }
  /* pre.src-C++ doesn't work in CSS */
  pre.src-clojure:before { content: 'Clojure'; }
  pre.src-css:before { content: 'CSS'; }
  pre.src-D:before { content: 'D'; }
  pre.src-ditaa:before { content: 'ditaa'; }
  pre.src-dot:before { content: 'Graphviz'; }
  pre.src-calc:before { content: 'Emacs Calc'; }
  pre.src-emacs-lisp:before { content: 'Emacs Lisp'; }
  pre.src-fortran:before { content: 'Fortran'; }
  pre.src-gnuplot:before { content: 'gnuplot'; }
  pre.src-haskell:before { content: 'Haskell'; }
  pre.src-hledger:before { content: 'hledger'; }
  pre.src-java:before { content: 'Java'; }
  pre.src-js:before { content: 'Javascript'; }
  pre.src-latex:before { content: 'LaTeX'; }
  pre.src-ledger:before { content: 'Ledger'; }
  pre.src-lisp:before { content: 'Lisp'; }
  pre.src-lilypond:before { content: 'Lilypond'; }
  pre.src-lua:before { content: 'Lua'; }
  pre.src-matlab:before { content: 'MATLAB'; }
  pre.src-mscgen:before { content: 'Mscgen'; }
  pre.src-ocaml:before { content: 'Objective Caml'; }
  pre.src-octave:before { content: 'Octave'; }
  pre.src-org:before { content: 'Org mode'; }
  pre.src-oz:before { content: 'OZ'; }
  pre.src-plantuml:before { content: 'Plantuml'; }
  pre.src-processing:before { content: 'Processing.js'; }
  pre.src-python:before { content: 'Python'; }
  pre.src-R:before { content: 'R'; }
  pre.src-ruby:before { content: 'Ruby'; }
  pre.src-sass:before { content: 'Sass'; }
  pre.src-scheme:before { content: 'Scheme'; }
  pre.src-screen:before { content: 'Gnu Screen'; }
  pre.src-sed:before { content: 'Sed'; }
  pre.src-sh:before { content: 'shell'; }
  pre.src-sql:before { content: 'SQL'; }
  pre.src-sqlite:before { content: 'SQLite'; }
  /* additional languages in org.el's org-babel-load-languages alist */
  pre.src-forth:before { content: 'Forth'; }
  pre.src-io:before { content: 'IO'; }
  pre.src-J:before { content: 'J'; }
  pre.src-makefile:before { content: 'Makefile'; }
  pre.src-maxima:before { content: 'Maxima'; }
  pre.src-perl:before { content: 'Perl'; }
  pre.src-picolisp:before { content: 'Pico Lisp'; }
  pre.src-scala:before { content: 'Scala'; }
  pre.src-shell:before { content: 'Shell Script'; }
  pre.src-ebnf2ps:before { content: 'ebfn2ps'; }
  /* additional language identifiers per "defun org-babel-execute"
       in ob-*.el */
  pre.src-cpp:before  { content: 'C++'; }
  pre.src-abc:before  { content: 'ABC'; }
  pre.src-coq:before  { content: 'Coq'; }
  pre.src-groovy:before  { content: 'Groovy'; }
  /* additional language identifiers from org-babel-shell-names in
     ob-shell.el: ob-shell is the only babel language using a lambda to put
     the execution function name together. */
  pre.src-bash:before  { content: 'bash'; }
  pre.src-csh:before  { content: 'csh'; }
  pre.src-ash:before  { content: 'ash'; }
  pre.src-dash:before  { content: 'dash'; }
  pre.src-ksh:before  { content: 'ksh'; }
  pre.src-mksh:before  { content: 'mksh'; }
  pre.src-posh:before  { content: 'posh'; }
  /* Additional Emacs modes also supported by the LaTeX listings package */
  pre.src-ada:before { content: 'Ada'; }
  pre.src-asm:before { content: 'Assembler'; }
  pre.src-caml:before { content: 'Caml'; }
  pre.src-delphi:before { content: 'Delphi'; }
  pre.src-html:before { content: 'HTML'; }
  pre.src-idl:before { content: 'IDL'; }
  pre.src-mercury:before { content: 'Mercury'; }
  pre.src-metapost:before { content: 'MetaPost'; }
  pre.src-modula-2:before { content: 'Modula-2'; }
  pre.src-pascal:before { content: 'Pascal'; }
  pre.src-ps:before { content: 'PostScript'; }
  pre.src-prolog:before { content: 'Prolog'; }
  pre.src-simula:before { content: 'Simula'; }
  pre.src-tcl:before { content: 'tcl'; }
  pre.src-tex:before { content: 'TeX'; }
  pre.src-plain-tex:before { content: 'Plain TeX'; }
  pre.src-verilog:before { content: 'Verilog'; }
  pre.src-vhdl:before { content: 'VHDL'; }
  pre.src-xml:before { content: 'XML'; }
  pre.src-nxml:before { content: 'XML'; }
  /* add a generic configuration mode; LaTeX export needs an additional
     (add-to-list 'org-latex-listings-langs '(conf " ")) in .emacs */
  pre.src-conf:before { content: 'Configuration File'; }

  table { border-collapse:collapse; }
  caption.t-above { caption-side: top; }
  caption.t-bottom { caption-side: bottom; }
  td, th { vertical-align:top;  }
  th.org-right  { text-align: center;  }
  th.org-left   { text-align: center;   }
  th.org-center { text-align: center; }
  td.org-right  { text-align: right;  }
  td.org-left   { text-align: left;   }
  td.org-center { text-align: center; }
  dt { font-weight: bold; }
  .footpara { display: inline; }
  .footdef  { margin-bottom: 1em; }
  .figure { padding: 1em; }
  .figure p { text-align: center; }
  .inlinetask {
    padding: 10px;
    border: 2px solid gray;
    margin: 10px;
    background: #ffffcc;
  }
  #org-div-home-and-up
   { text-align: right; font-size: 70%; white-space: nowrap; }
  textarea { overflow-x: auto; }
  .linenr { font-size: smaller }
  .code-highlighted { background-color: #ffff00; }
  .org-info-js_info-navigation { border-style: none; }
  #org-info-js_console-label
    { font-size: 10px; font-weight: bold; white-space: nowrap; }
  .org-info-js_search-highlight
    { background-color: #ffff00; color: #000000; font-weight: bold; }
  .org-svg { width: 90%; }
  /*]]>*/-->
</style>
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<link rel="stylesheet" type="text/css" href="https://www.pirilampo.org/styles/readtheorg/css/htmlize.css"/>
<style type="text/css">body {width: 60em; margin:auto} pre.src {overflow:auto}</style>
<script type="text/javascript">
/*
@licstart  The following is the entire license notice for the
JavaScript code in this tag.

Copyright (C) 2012-2017 Free Software Foundation, Inc.

The JavaScript code in this tag is free software: you can
redistribute it and/or modify it under the terms of the GNU
General Public License (GNU GPL) as published by the Free Software
Foundation, either version 3 of the License, or (at your option)
any later version.  The code is distributed WITHOUT ANY WARRANTY;
without even the implied warranty of MERCHANTABILITY or FITNESS
FOR A PARTICULAR PURPOSE.  See the GNU GPL for more details.

As additional permission under GNU GPL version 3 section 7, you
may distribute non-source (e.g., minimized or compacted) forms of
that code without the copy of the GNU GPL normally required by
section 4, provided you include this license notice and a URL
through which recipients can access the Corresponding Source.


@licend  The above is the entire license notice
for the JavaScript code in this tag.
*/
<!--/*--><![CDATA[/*><!--*/
 function CodeHighlightOn(elem, id)
 {
   var target = document.getElementById(id);
   if(null != target) {
     elem.cacheClassElem = elem.className;
     elem.cacheClassTarget = target.className;
     target.className = "code-highlighted";
     elem.className   = "code-highlighted";
   }
 }
 function CodeHighlightOff(elem, id)
 {
   var target = document.getElementById(id);
   if(elem.cacheClassElem)
     elem.className = elem.cacheClassElem;
   if(elem.cacheClassTarget)
     target.className = elem.cacheClassTarget;
 }
/*]]>*///-->
</script>
</head>
<body>
<div id="content">
<h1 class="title">QTL mapping pipeline</h1>
<div id="outline-container-org73245c3" class="outline-2">
<h2 id="org73245c3">Replicate bulk eQTLs in the single cell data</h2>
<div class="outline-text-2" id="text-org73245c3">
<p>
Read the bulk eQTLs (nominally significant).
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">def</span> <span class="org-function-name">parse_bulk_qtl_record</span>(row):
  <span class="org-variable-name">gene</span>, <span class="org-variable-name">_</span> = row[0].split(<span class="org-string">'.'</span>)
  id_, *<span class="org-variable-name">_</span> = row[5].split(<span class="org-string">'.'</span>)
  <span class="org-keyword">return</span> pd.Series({<span class="org-string">'gene'</span>: gene, <span class="org-string">'id'</span>: id_, <span class="org-string">'beta'</span>: row.iloc[-3], <span class="org-string">'p_bulk'</span>: row.iloc[-1]})
<span class="org-variable-name">bulk_qtls</span> = (pd.read_table(<span class="org-string">'/project2/gilad/singlecell-qtl/bulk/permutations.all.RNAseq_run.fixed.txt.gz'</span>, sep=<span class="org-string">' '</span>, header=<span class="org-constant">None</span>)
             .fillna(<span class="org-string">''</span>)
             .<span class="org-builtin">apply</span>(parse_bulk_qtl_record, axis=1))
</pre>
</div>

<p>
Read the bulk eQTLs (significant by permutation test).
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">def</span> <span class="org-function-name">parse_perm_bulk_qtl_record</span>(row):
  gene, *<span class="org-variable-name">_</span> = row[0].split(<span class="org-string">'.'</span>)
  <span class="org-variable-name">id_</span>, <span class="org-variable-name">chr_</span>, <span class="org-variable-name">pos</span> = row[1].split(<span class="org-string">'.'</span>)
  <span class="org-keyword">return</span> pd.Series({<span class="org-string">'gene'</span>: gene, <span class="org-string">'id'</span>: id_, <span class="org-string">'chr'</span>: chr_, <span class="org-string">'pos'</span>: pos, <span class="org-string">'p_bulk'</span>: row.iloc[-1]})
<span class="org-variable-name">perm_bulk_qtls</span> = (pd.read_table(<span class="org-string">'/project2/gilad/singlecell-qtl/bulk/permutations.all.RNAseq_run.fixed.txt.gz.bh.txt'</span>, sep=<span class="org-string">' '</span>, header=0)
                  .<span class="org-builtin">apply</span>(parse_perm_bulk_qtl_record, axis=1))
</pre>
</div>

<p>
Test each bulk QTL-gene pair in the single cell data. Mean-impute and
center genotypes.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">def</span> <span class="org-function-name">parse_vcf</span>(record):
  <span class="org-variable-name">geno</span> = [<span class="org-builtin">sum</span>(<span class="org-builtin">float</span>(h) <span class="org-keyword">for</span> h <span class="org-keyword">in</span> g.split(<span class="org-string">'/'</span>)) <span class="org-keyword">if</span> <span class="org-string">'.'</span> <span class="org-keyword">not</span> <span class="org-keyword">in</span> g <span class="org-keyword">else</span> -1 <span class="org-keyword">for</span> g <span class="org-keyword">in</span> record[9:]]
  <span class="org-keyword">return</span> pd.Series(geno)

<span class="org-keyword">def</span> <span class="org-function-name">standardize</span>(geno):
  <span class="org-comment-delimiter"># </span><span class="org-comment">Important: this has to be done after filtering to common individuals</span>
  <span class="org-variable-name">orig_geno</span> = geno.copy()
  <span class="org-comment-delimiter"># </span><span class="org-comment">Important: match the qtltools implementation</span>
  <span class="org-variable-name">geno</span> = np.ma.masked_less(geno.values, 0)
  <span class="org-variable-name">geno</span> = geno.filled(geno.mean())
  <span class="org-variable-name">geno</span> -= geno.mean()
  <span class="org-keyword">return</span> geno

<span class="org-keyword">def</span> <span class="org-function-name">extract_qtl_gene_pair</span>(bulk_qtls, means):
  <span class="org-variable-name">common_means</span>, <span class="org-variable-name">common_qtls</span> = means.align(bulk_qtls.set_index(<span class="org-string">'gene'</span>), join=<span class="org-string">'inner'</span>, axis=0)
  <span class="org-variable-name">header</span> = pd.read_table(<span class="org-string">'/project2/gilad/singlecell-qtl/bulk/genotypes.vcf.gz'</span>, skiprows=1, nrows=1, header=0)[9:]
  <span class="org-variable-name">genotypes</span> = tabix.<span class="org-builtin">open</span>(<span class="org-string">'/project2/gilad/singlecell-qtl/bulk/genotypes.vcf.gz'</span>)
  <span class="org-variable-name">X</span>, <span class="org-variable-name">Y</span> = (common_qtls
          .<span class="org-builtin">apply</span>(<span class="org-keyword">lambda</span> x: parse_vcf(<span class="org-builtin">next</span>(genotypes.query(x[<span class="org-string">'chr'</span>], <span class="org-builtin">int</span>(x[<span class="org-string">'pos'</span>]) - 1, <span class="org-builtin">int</span>(x[<span class="org-string">'pos'</span>])))), axis=1)
          .rename(columns={i: ind <span class="org-keyword">for</span> i, ind <span class="org-keyword">in</span> <span class="org-builtin">enumerate</span>(header.columns[9:])})
          .align(common_means, join=<span class="org-string">'inner'</span>, axis=<span class="org-constant">None</span>))
  <span class="org-variable-name">X</span> = X.<span class="org-builtin">apply</span>(standardize, axis=1)
  <span class="org-keyword">return</span> X, Y
</pre>
</div>

<p>
Define replication as significant at FDR 5% in both bulk and single cell.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">def</span> <span class="org-function-name">bh</span>(df, column, fdr=0.05):
  <span class="org-variable-name">num_tests</span> = df.shape[0]
  <span class="org-keyword">return</span> np.logical_and.accumulate(df.sort_values(column)[column] &lt; fdr * np.arange(1, num_tests + 1) / num_tests)
</pre>
</div>
</div>

<div id="outline-container-orga0df169" class="outline-3">
<h3 id="orga0df169">NB model</h3>
<div class="outline-text-3" id="text-orga0df169">
<p>
Estimate a lower bound on the replication rate using the following procedure:
</p>

<ol class="org-ol">
<li>For each individual, for each gene, <a href="zinb.html#orgef1a628">estimate a mean and dispersion in the
single cell data ignoring zero inflation</a> (i.e., just fit a negative
binomial model)</li>
<li>For each bulk RNA-Seq eQTL-gene pair, test the SNP against the estimated
means</li>
</ol>

<p>
Read the estimated parameters (phenotypes) then center them.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">mean</span> = (pd.read_table(<span class="org-string">'/scratch/midway2/aksarkar/singlecell/mean2.txt.gz'</span>, sep=<span class="org-string">' '</span>, index_col=0)
        .transform(<span class="org-keyword">lambda</span> x: x - x.mean(), axis=1))
</pre>
</div>

<p>
Perform the eQTL tests.
</p>

<div class="org-src-container">
<pre class="src src-ipython" id="org72705b1"><span class="org-variable-name">X</span>, <span class="org-variable-name">Y</span> = extract_qtl_gene_pair(perm_bulk_qtls, mean)
<span class="org-variable-name">result</span> = []
<span class="org-variable-name">_sf</span> = st.chi2(1).sf
<span class="org-keyword">for</span> (_, x), (name, y) <span class="org-keyword">in</span> <span class="org-builtin">zip</span>(X.iterrows(), Y.iterrows()):
  <span class="org-keyword">if</span> np.isclose(x.std(), 0):
    <span class="org-keyword">print</span>(<span class="org-string">'Skipping {}'</span>.<span class="org-builtin">format</span>(name))
    <span class="org-keyword">continue</span>
  beta, rss, *<span class="org-variable-name">_</span> = np.linalg.lstsq(x.values.reshape(-1, 1), y.values.ravel(), rcond=-1)
  <span class="org-variable-name">sigma2</span> = rss / y.shape[0]
  <span class="org-variable-name">se</span> = sigma2 / np.inner(x.values, x.values)
  <span class="org-variable-name">pval</span> = _sf(np.square(beta / se))
  result.append({<span class="org-string">'gene'</span>: name, <span class="org-string">'beta'</span>: beta[0], <span class="org-string">'p_sc'</span>: pval.ravel()[0]})
<span class="org-variable-name">merged</span> = (bulk_qtls.merge(pd.DataFrame.from_dict(result), on=<span class="org-string">'gene'</span>, suffixes=[<span class="org-string">'_bulk'</span>, <span class="org-string">'_sc'</span>])
          .set_index(<span class="org-string">'gene'</span>)
          .sort_values(<span class="org-string">'p_bulk'</span>)
          .reset_index()[[<span class="org-string">'gene'</span>, <span class="org-string">'id'</span>, <span class="org-string">'p_bulk'</span>, <span class="org-string">'beta_bulk'</span>, <span class="org-string">'p_sc'</span>, <span class="org-string">'beta_sc'</span>]])
</pre>
</div>

<p>
Write out tables for the interactive browser.
</p>

<div class="org-src-container">
<pre class="src src-ipython">merged.to_csv(<span class="org-string">'/home/aksarkar/projects/singlecell-qtl/data/nb-mean-qtls.txt.gz'</span>, compression=<span class="org-string">'gzip'</span>, sep=<span class="org-string">' '</span>)
X.to_csv(<span class="org-string">'/home/aksarkar/projects/singlecell-qtl/data/bulk-qtl-genotypes.txt.gz'</span>, compression=<span class="org-string">'gzip'</span>, sep=<span class="org-string">' '</span>) 
</pre>
</div>

<p>
Estimate the replication rate.
</p>

<div class="org-src-container">
<pre class="src src-ipython" id="org3057f9d"><span class="org-variable-name">bulk_fdr_pass</span> = bh(merged, <span class="org-string">'p_bulk'</span>)
<span class="org-variable-name">sc_fdr_pass</span> = bh(merged, <span class="org-string">'p_sc'</span>)
<span class="org-variable-name">replicated</span> = merged[np.logical_and(*bulk_fdr_pass.align(sc_fdr_pass))]
replicated.shape[0], replicated.shape[0] / merged.shape[0]
</pre>
</div>

<pre class="example">
(902, 0.9636752136752137)

</pre>

<p>
Estimate the concordance
</p>

<div class="org-src-container">
<pre class="src src-ipython" id="org6a223f0"><span class="org-variable-name">sign_pass</span> = (replicated[<span class="org-string">'beta_bulk'</span>] * replicated[<span class="org-string">'beta_sc'</span>]).values &gt; 0
replicated[sign_pass].shape[0], sign_pass.<span class="org-builtin">sum</span>() / replicated.shape[0]
</pre>
</div>

<pre class="example">
(785, 0.8702882483370288)

</pre>
</div>
</div>

<div id="outline-container-orgb7c3e0e" class="outline-3">
<h3 id="orgb7c3e0e">ZINB model</h3>
<div class="outline-text-3" id="text-orgb7c3e0e">
<p>
Estimate replication on <a href="zinb.html#orgc07a1fb">means estimated assuming a gene-specific dropout
parameters</a>.  Accounting for zero-inflation should push estimated means away
from zero, and should improve concordance for lowly expressed genes.
</p>

<p>
Read the estimated parameters (phenotypes) then center them.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">mean</span> = (pd.read_table(<span class="org-string">'/scratch/midway2/aksarkar/singlecell/zi-mean.txt.gz'</span>, sep=<span class="org-string">' '</span>, index_col=0)
        .transform(<span class="org-keyword">lambda</span> x: x - x.mean(), axis=1))
</pre>
</div>

<p>
Write out tables for the interactive browser.
</p>

<div class="org-src-container">
<pre class="src src-ipython">merged.to_csv(<span class="org-string">'/home/aksarkar/projects/singlecell-qtl/data/zinb-mean-qtls.txt.gz'</span>, compression=<span class="org-string">'gzip'</span>, sep=<span class="org-string">' '</span>)
</pre>
</div>

<p>
Estimate the replication rate.
</p>

<pre class="example">
(894, 0.9551282051282052)

</pre>

<p>
Estimate the concordance.
</p>

<pre class="example">
(752, 0.8411633109619687)

</pre>
</div>
</div>

<div id="outline-container-org513fdc9" class="outline-3">
<h3 id="org513fdc9">ZINB2 model</h3>
<div class="outline-text-3" id="text-org513fdc9">
<p>
For each individual, for each gene, estimate mean, dispersion, and dropout
parameter.
</p>

<p>
Read the estimated parameters (phenotypes) then center them.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">mean</span> = (pd.read_table(<span class="org-string">'/scratch/midway2/aksarkar/singlecell/zi2-mean.txt.gz'</span>, sep=<span class="org-string">' '</span>, index_col=0)
        .transform(<span class="org-keyword">lambda</span> x: x - x.mean(), axis=1))
</pre>
</div>

<p>
Write out tables for the interactive browser.
</p>

<div class="org-src-container">
<pre class="src src-ipython">merged.to_csv(<span class="org-string">'/home/aksarkar/projects/singlecell-qtl/data/zinb2-mean-qtls.txt.gz'</span>, compression=<span class="org-string">'gzip'</span>, sep=<span class="org-string">' '</span>)
</pre>
</div>

<p>
Estimate the replication rate.
</p>

<pre class="example">
(894, 0.9551282051282052)

</pre>

<p>
Estimate the concordance.
</p>

<pre class="example">
(752, 0.8411633109619687)

</pre>
</div>
</div>
<div id="outline-container-org651b6c4" class="outline-3">
<h3 id="org651b6c4">Look at discordance between NB/ZINB</h3>
<div class="outline-text-3" id="text-org651b6c4">
<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">nb_qtls</span> = pd.read_table(<span class="org-string">'/home/aksarkar/projects/singlecell-qtl/data/nb-mean-qtls.txt.gz'</span>, sep=<span class="org-string">' '</span>, index_col=0)
<span class="org-variable-name">zinb_qtls</span> = pd.read_table(<span class="org-string">'/home/aksarkar/projects/singlecell-qtl/data/zinb-mean-qtls.txt.gz'</span>, sep=<span class="org-string">' '</span>, index_col=0)
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">nb_qtls</span>[<span class="org-string">'pass'</span>] = bh(nb_qtls, <span class="org-string">'p_bulk'</span>) &amp; bh(nb_qtls, <span class="org-string">'p_sc'</span>)
<span class="org-variable-name">zinb_qtls</span>[<span class="org-string">'pass'</span>] = bh(zinb_qtls, <span class="org-string">'p_bulk'</span>) &amp; bh(zinb_qtls, <span class="org-string">'p_sc'</span>)
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython">nb_qtls[~nb_qtls[<span class="org-string">'pass'</span>] &amp; zinb_qtls[<span class="org-string">'pass'</span>]][[<span class="org-string">'gene'</span>]]
</pre>
</div>

<pre class="example">
gene
49   ENSG00000138107
146  ENSG00000134815
246  ENSG00000136731
271  ENSG00000143776
294  ENSG00000106341
324  ENSG00000157036
352  ENSG00000185619
455  ENSG00000160214
477  ENSG00000184014
511  ENSG00000134987
539  ENSG00000119723
548  ENSG00000224870
575  ENSG00000213347
577  ENSG00000176473
609  ENSG00000103657
611  ENSG00000106785
615  ENSG00000122490
640  ENSG00000155275
653  ENSG00000152217
687  ENSG00000160216
781  ENSG00000173214
787  ENSG00000092871
796  ENSG00000176715
802  ENSG00000128335
845  ENSG00000167625
892  ENSG00000169718
916  ENSG00000170471
922  ENSG00000089335
926  ENSG00000079432
</pre>

<div class="org-src-container">
<pre class="src src-ipython">nb_qtls[nb_qtls[<span class="org-string">'pass'</span>] &amp; ~zinb_qtls[<span class="org-string">'pass'</span>]][[<span class="org-string">'gene'</span>]]
</pre>
</div>

<pre class="example">
gene
53   ENSG00000144712
173  ENSG00000148143
208  ENSG00000181481
211  ENSG00000141295
225  ENSG00000169814
235  ENSG00000163870
270  ENSG00000039319
288  ENSG00000185418
378  ENSG00000167840
436  ENSG00000072736
459  ENSG00000198589
465  ENSG00000187741
479  ENSG00000068001
482  ENSG00000117791
491  ENSG00000163820
529  ENSG00000104228
543  ENSG00000115339
546  ENSG00000033327
566  ENSG00000256269
585  ENSG00000005059
616  ENSG00000153015
641  ENSG00000172461
645  ENSG00000205356
650  ENSG00000044090
669  ENSG00000146909
700  ENSG00000172053
769  ENSG00000154265
800  ENSG00000139629
801  ENSG00000123992
805  ENSG00000251247
806  ENSG00000103966
821  ENSG00000130695
825  ENSG00000116212
884  ENSG00000183258
885  ENSG00000112357
912  ENSG00000185963
917  ENSG00000145604
</pre>
</div>
</div>

<div id="outline-container-org401e631" class="outline-3">
<h3 id="org401e631">Conclusions</h3>
<div class="outline-text-3" id="text-org401e631">
<table class="table">


<colgroup>
<col  class="org-left">

<col  class="org-right">

<col  class="org-right">
</colgroup>
<thead>
<tr>
<th scope="col" class="org-left">Model</th>
<th scope="col" class="org-right">Replication rate</th>
<th scope="col" class="org-right">Concordance</th>
</tr>
</thead>
<tbody>
<tr>
<td class="org-left">NB</td>
<td class="org-right">96.4%</td>
<td class="org-right">87.0%</td>
</tr>

<tr>
<td class="org-left">ZINB</td>
<td class="org-right">95.5%</td>
<td class="org-right">84.1%</td>
</tr>

<tr>
<td class="org-left">ZINB2</td>
<td class="org-right">95.5%</td>
<td class="org-right">84.1%</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
</div>
<div id="postamble" class="status">
<p class="author">Author: Abhishek Sarkar</p>
<p class="date">Created: 2018-03-02 Fri 10:23</p>
<p class="validation"><a href="http://validator.w3.org/check?uri=referer">Validate</a></p>
</div>
</body>
</html>
