<!DOCTYPE html>
<html lang="en">
<head>
<!-- 2018-02-16 Fri 14:10 -->
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
<title>QTL mapping pipeline</title>
<meta name="generator" content="Org mode">
<meta name="author" content="Abhishek Sarkar">
<style type="text/css">
 <!--/*--><![CDATA[/*><!--*/
  .title  { text-align: center;
             margin-bottom: .2em; }
  .subtitle { text-align: center;
              font-size: medium;
              font-weight: bold;
              margin-top:0; }
  .todo   { font-family: monospace; color: red; }
  .done   { font-family: monospace; color: green; }
  .priority { font-family: monospace; color: orange; }
  .tag    { background-color: #eee; font-family: monospace;
            padding: 2px; font-size: 80%; font-weight: normal; }
  .timestamp { color: #bebebe; }
  .timestamp-kwd { color: #5f9ea0; }
  .org-right  { margin-left: auto; margin-right: 0px;  text-align: right; }
  .org-left   { margin-left: 0px;  margin-right: auto; text-align: left; }
  .org-center { margin-left: auto; margin-right: auto; text-align: center; }
  .underline { text-decoration: underline; }
  #postamble p, #preamble p { font-size: 90%; margin: .2em; }
  p.verse { margin-left: 3%; }
  pre {
    border: 1px solid #ccc;
    box-shadow: 3px 3px 3px #eee;
    padding: 8pt;
    font-family: monospace;
    overflow: auto;
    margin: 1.2em;
  }
  pre.src {
    position: relative;
    overflow: visible;
    padding-top: 1.2em;
  }
  pre.src:before {
    display: none;
    position: absolute;
    background-color: white;
    top: -10px;
    right: 10px;
    padding: 3px;
    border: 1px solid black;
  }
  pre.src:hover:before { display: inline;}
  /* Languages per Org manual */
  pre.src-asymptote:before { content: 'Asymptote'; }
  pre.src-awk:before { content: 'Awk'; }
  pre.src-C:before { content: 'C'; }
  /* pre.src-C++ doesn't work in CSS */
  pre.src-clojure:before { content: 'Clojure'; }
  pre.src-css:before { content: 'CSS'; }
  pre.src-D:before { content: 'D'; }
  pre.src-ditaa:before { content: 'ditaa'; }
  pre.src-dot:before { content: 'Graphviz'; }
  pre.src-calc:before { content: 'Emacs Calc'; }
  pre.src-emacs-lisp:before { content: 'Emacs Lisp'; }
  pre.src-fortran:before { content: 'Fortran'; }
  pre.src-gnuplot:before { content: 'gnuplot'; }
  pre.src-haskell:before { content: 'Haskell'; }
  pre.src-hledger:before { content: 'hledger'; }
  pre.src-java:before { content: 'Java'; }
  pre.src-js:before { content: 'Javascript'; }
  pre.src-latex:before { content: 'LaTeX'; }
  pre.src-ledger:before { content: 'Ledger'; }
  pre.src-lisp:before { content: 'Lisp'; }
  pre.src-lilypond:before { content: 'Lilypond'; }
  pre.src-lua:before { content: 'Lua'; }
  pre.src-matlab:before { content: 'MATLAB'; }
  pre.src-mscgen:before { content: 'Mscgen'; }
  pre.src-ocaml:before { content: 'Objective Caml'; }
  pre.src-octave:before { content: 'Octave'; }
  pre.src-org:before { content: 'Org mode'; }
  pre.src-oz:before { content: 'OZ'; }
  pre.src-plantuml:before { content: 'Plantuml'; }
  pre.src-processing:before { content: 'Processing.js'; }
  pre.src-python:before { content: 'Python'; }
  pre.src-R:before { content: 'R'; }
  pre.src-ruby:before { content: 'Ruby'; }
  pre.src-sass:before { content: 'Sass'; }
  pre.src-scheme:before { content: 'Scheme'; }
  pre.src-screen:before { content: 'Gnu Screen'; }
  pre.src-sed:before { content: 'Sed'; }
  pre.src-sh:before { content: 'shell'; }
  pre.src-sql:before { content: 'SQL'; }
  pre.src-sqlite:before { content: 'SQLite'; }
  /* additional languages in org.el's org-babel-load-languages alist */
  pre.src-forth:before { content: 'Forth'; }
  pre.src-io:before { content: 'IO'; }
  pre.src-J:before { content: 'J'; }
  pre.src-makefile:before { content: 'Makefile'; }
  pre.src-maxima:before { content: 'Maxima'; }
  pre.src-perl:before { content: 'Perl'; }
  pre.src-picolisp:before { content: 'Pico Lisp'; }
  pre.src-scala:before { content: 'Scala'; }
  pre.src-shell:before { content: 'Shell Script'; }
  pre.src-ebnf2ps:before { content: 'ebfn2ps'; }
  /* additional language identifiers per "defun org-babel-execute"
       in ob-*.el */
  pre.src-cpp:before  { content: 'C++'; }
  pre.src-abc:before  { content: 'ABC'; }
  pre.src-coq:before  { content: 'Coq'; }
  pre.src-groovy:before  { content: 'Groovy'; }
  /* additional language identifiers from org-babel-shell-names in
     ob-shell.el: ob-shell is the only babel language using a lambda to put
     the execution function name together. */
  pre.src-bash:before  { content: 'bash'; }
  pre.src-csh:before  { content: 'csh'; }
  pre.src-ash:before  { content: 'ash'; }
  pre.src-dash:before  { content: 'dash'; }
  pre.src-ksh:before  { content: 'ksh'; }
  pre.src-mksh:before  { content: 'mksh'; }
  pre.src-posh:before  { content: 'posh'; }
  /* Additional Emacs modes also supported by the LaTeX listings package */
  pre.src-ada:before { content: 'Ada'; }
  pre.src-asm:before { content: 'Assembler'; }
  pre.src-caml:before { content: 'Caml'; }
  pre.src-delphi:before { content: 'Delphi'; }
  pre.src-html:before { content: 'HTML'; }
  pre.src-idl:before { content: 'IDL'; }
  pre.src-mercury:before { content: 'Mercury'; }
  pre.src-metapost:before { content: 'MetaPost'; }
  pre.src-modula-2:before { content: 'Modula-2'; }
  pre.src-pascal:before { content: 'Pascal'; }
  pre.src-ps:before { content: 'PostScript'; }
  pre.src-prolog:before { content: 'Prolog'; }
  pre.src-simula:before { content: 'Simula'; }
  pre.src-tcl:before { content: 'tcl'; }
  pre.src-tex:before { content: 'TeX'; }
  pre.src-plain-tex:before { content: 'Plain TeX'; }
  pre.src-verilog:before { content: 'Verilog'; }
  pre.src-vhdl:before { content: 'VHDL'; }
  pre.src-xml:before { content: 'XML'; }
  pre.src-nxml:before { content: 'XML'; }
  /* add a generic configuration mode; LaTeX export needs an additional
     (add-to-list 'org-latex-listings-langs '(conf " ")) in .emacs */
  pre.src-conf:before { content: 'Configuration File'; }

  table { border-collapse:collapse; }
  caption.t-above { caption-side: top; }
  caption.t-bottom { caption-side: bottom; }
  td, th { vertical-align:top;  }
  th.org-right  { text-align: center;  }
  th.org-left   { text-align: center;   }
  th.org-center { text-align: center; }
  td.org-right  { text-align: right;  }
  td.org-left   { text-align: left;   }
  td.org-center { text-align: center; }
  dt { font-weight: bold; }
  .footpara { display: inline; }
  .footdef  { margin-bottom: 1em; }
  .figure { padding: 1em; }
  .figure p { text-align: center; }
  .inlinetask {
    padding: 10px;
    border: 2px solid gray;
    margin: 10px;
    background: #ffffcc;
  }
  #org-div-home-and-up
   { text-align: right; font-size: 70%; white-space: nowrap; }
  textarea { overflow-x: auto; }
  .linenr { font-size: smaller }
  .code-highlighted { background-color: #ffff00; }
  .org-info-js_info-navigation { border-style: none; }
  #org-info-js_console-label
    { font-size: 10px; font-weight: bold; white-space: nowrap; }
  .org-info-js_search-highlight
    { background-color: #ffff00; color: #000000; font-weight: bold; }
  .org-svg { width: 90%; }
  /*]]>*/-->
</style>
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<link rel="stylesheet" type="text/css" href="https://www.pirilampo.org/styles/readtheorg/css/htmlize.css"/>
<style type="text/css">body {width: 60em; margin:auto} pre.src {overflow:auto}</style>
<script type="text/javascript">
/*
@licstart  The following is the entire license notice for the
JavaScript code in this tag.

Copyright (C) 2012-2017 Free Software Foundation, Inc.

The JavaScript code in this tag is free software: you can
redistribute it and/or modify it under the terms of the GNU
General Public License (GNU GPL) as published by the Free Software
Foundation, either version 3 of the License, or (at your option)
any later version.  The code is distributed WITHOUT ANY WARRANTY;
without even the implied warranty of MERCHANTABILITY or FITNESS
FOR A PARTICULAR PURPOSE.  See the GNU GPL for more details.

As additional permission under GNU GPL version 3 section 7, you
may distribute non-source (e.g., minimized or compacted) forms of
that code without the copy of the GNU GPL normally required by
section 4, provided you include this license notice and a URL
through which recipients can access the Corresponding Source.


@licend  The above is the entire license notice
for the JavaScript code in this tag.
*/
<!--/*--><![CDATA[/*><!--*/
 function CodeHighlightOn(elem, id)
 {
   var target = document.getElementById(id);
   if(null != target) {
     elem.cacheClassElem = elem.className;
     elem.cacheClassTarget = target.className;
     target.className = "code-highlighted";
     elem.className   = "code-highlighted";
   }
 }
 function CodeHighlightOff(elem, id)
 {
   var target = document.getElementById(id);
   if(elem.cacheClassElem)
     elem.className = elem.cacheClassElem;
   if(elem.cacheClassTarget)
     target.className = elem.cacheClassTarget;
 }
/*]]>*///-->
</script>
<script type="text/x-mathjax-config">
    MathJax.Hub.Config({
        displayAlign: "center",
        displayIndent: "0em",

        "HTML-CSS": { scale: 100,
                        linebreaks: { automatic: "false" },
                        webFont: "TeX"
                       },
        SVG: {scale: 100,
              linebreaks: { automatic: "false" },
              font: "TeX"},
        NativeMML: {scale: 100},
        TeX: { equationNumbers: {autoNumber: "AMS"},
               MultLineWidth: "85%",
               TagSide: "right",
               TagIndent: ".8em"
             }
});
</script>
<script type="text/javascript"
        src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-AMS_HTML"></script>
</head>
<body>
<div id="content">
<h1 class="title">QTL mapping pipeline</h1>

<div id="outline-container-org3d0ec86" class="outline-2">
<h2 id="org3d0ec86">Introduction</h2>
<div class="outline-text-2" id="text-org3d0ec86">
<p>
Here, we set up the infrastructure to call mean and variance QTLs.
</p>

<p>
We previously <a href="zinb.html">estimated means and dispersions per individual, per gene</a>. We
now fit log linear models for each:
</p>

<p>
\[ \ln \mu_{ik} = X \beta_\mu + \epsilon_\mu \]
</p>

<p>
\[ \ln \phi_{ik} = X \beta_\phi + \epsilon_\phi \]
</p>

<p>
where \(\mu_{ik}, \phi_{ik}\) are the mean and dispersion for individual
\(i\), gene \(k\), \(X\) is the genotype matrix (\(n \times 1)\), and
\(\beta_\mu, \beta_\phi\) are the effect sizes (scalar).
</p>
</div>
</div>

<div id="outline-container-org8a59be7" class="outline-2">
<h2 id="org8a59be7">Genotype processing</h2>
<div class="outline-text-2" id="text-org8a59be7">
<p>
<code>QTLtools</code> reported 0 variants in cis for an unexpected number of genes when
using the provided genotype file, so reprocess the genotypes to fix it.
</p>

<div class="org-src-container">
<pre class="src src-sh">sbatch --partition=broadwl --job-name process-geno
<span class="org-comment-delimiter">#</span><span class="org-comment">!/bin/bash</span>
<span class="org-builtin">set</span> -e
<span class="org-builtin">source</span> activate scqtl
zcat /project/compbio/jointLCLs/genotype/hg19/YRI/vcf/chr{?,??}.hg19.vcf.gz | awk <span class="org-string">'BEGIN {print "##fileformat=VCFv4.1"} NR == 1 || ! /CHROM/ {print}'</span> | bgzip &gt;genotypes.vcf.gz
tabix genotypes.vcf.gz
</pre>
</div>
</div>
</div>

<div id="outline-container-org051ecfa" class="outline-2">
<h2 id="org051ecfa">Mean-QTL calling</h2>
<div class="outline-text-2" id="text-org051ecfa">
<p>
Read the estimated parameters (phenotypes) then center them.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">mean</span> = (pd.read_table(<span class="org-string">'/scratch/midway2/aksarkar/singlecell/zi2-mean.txt.gz'</span>, sep=<span class="org-string">' '</span>, index_col=0)
        .transform(<span class="org-keyword">lambda</span> x: x - x.mean(), axis=1))
</pre>
</div>

<p>
Write out the <a href="https://qtltools.github.io/qtltools/pages/input_files.html">phenotype file</a> for <code>qtltools</code>.
</p>

<div class="org-src-container">
<pre class="src src-ipython" id="org6264062"><span class="org-variable-name">gene_info</span> = (pd.read_table(<span class="org-string">'/home/aksarkar/projects/singlecell-qtl/data/scqtl-genes.txt.gz'</span>)
             .set_index(<span class="org-string">'gene'</span>)
             .query(<span class="org-string">'source == "H. sapiens"'</span>)
             .query(<span class="org-string">'chr != "hsX"'</span>)
             .query(<span class="org-string">'chr != "hsY"'</span>)
             .query(<span class="org-string">'chr != "hsMT"'</span>))
gene_info.head()
</pre>
</div>

<pre class="example">
chr      start        end      name strand      source
gene
ENSG00000000419  hs20   49551404   49575092      DPM1      -  H. sapiens
ENSG00000000457   hs1  169818772  169863408     SCYL3      -  H. sapiens
ENSG00000000460   hs1  169631245  169823221  C1orf112      +  H. sapiens
ENSG00000000938   hs1   27938575   27961788       FGR      -  H. sapiens
ENSG00000000971   hs1  196621008  196716634       CFH      +  H. sapiens
</pre>

<div class="org-src-container">
<pre class="src src-ipython" id="org042c79c"><span class="org-keyword">def</span> <span class="org-function-name">qtltools_format</span>(row):
  <span class="org-variable-name">row</span>[<span class="org-string">'#Chr'</span>] = <span class="org-string">'chr{}'</span>.<span class="org-builtin">format</span>(row[<span class="org-string">'chr'</span>][2:])
  <span class="org-variable-name">row</span>[<span class="org-string">'gid'</span>] = row.name
  <span class="org-variable-name">row</span>[<span class="org-string">'pid'</span>] = row.name
  <span class="org-keyword">return</span> row

<span class="org-keyword">def</span> <span class="org-function-name">write_pheno_file</span>(pheno, gene_info, output_file, holdout=<span class="org-constant">True</span>):
  <span class="org-keyword">if</span> holdout:
    <span class="org-variable-name">genes</span> = gene_info.loc[gene_info[<span class="org-string">'chr'</span>].<span class="org-builtin">apply</span>(<span class="org-keyword">lambda</span> x: <span class="org-builtin">bool</span>(<span class="org-builtin">int</span>(x[2:]) % 2)).values]
  <span class="org-keyword">else</span>:
    <span class="org-variable-name">genes</span> = gene_info
  (genes
   .<span class="org-builtin">apply</span>(qtltools_format, axis=1)
   .merge(pheno, left_index=<span class="org-constant">True</span>, right_index=<span class="org-constant">True</span>)
   .to_csv(output_file,
           sep=<span class="org-string">'\t'</span>,
           columns=[<span class="org-string">'#Chr'</span>, <span class="org-string">'start'</span>, <span class="org-string">'end'</span>, <span class="org-string">'pid'</span>, <span class="org-string">'gid'</span>, <span class="org-string">'strand'</span>] + <span class="org-builtin">list</span>(pheno.columns),
           header=<span class="org-constant">True</span>,
           index=<span class="org-constant">False</span>,
           index_label=<span class="org-constant">False</span>)
  )
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython">write_pheno_file(mean, gene_info, <span class="org-string">'/scratch/midway2/aksarkar/singlecell/mean.bed'</span>)
</pre>
</div>

<p>
Index the phenotype file.
</p>

<div class="org-src-container">
<pre class="src src-sh" id="orge80d324"><span class="org-builtin">export</span> <span class="org-variable-name">input</span>=$<span class="org-variable-name">input</span>
sbatch --partition=$<span class="org-variable-name">partition</span> --wait
<span class="org-comment-delimiter">#</span><span class="org-comment">!/bin/bash</span>
sort -k1,1 -k2,2n -k3,3n $<span class="org-variable-name">input</span> | bgzip &gt;$<span class="org-variable-name">input</span>.gz
tabix -p bed $<span class="org-variable-name">input</span>.gz
</pre>
</div>

<p>
Run the QTL mapping, using the approximate permutation test implemented in
<code>QTLtools</code>.
</p>

<div class="org-src-container">
<pre class="src src-sh" id="org5a3e21d"><span class="org-builtin">export</span> <span class="org-variable-name">pheno</span>=$<span class="org-variable-name">pheno</span>
sbatch --partition=$<span class="org-variable-name">partition</span> -N1 -c16 -J $<span class="org-variable-name">pheno</span>-qtl -o $<span class="org-variable-name">pheno</span>-qtl.log
<span class="org-comment-delimiter">#</span><span class="org-comment">!/bin/bash</span>
<span class="org-builtin">source</span> activate scqtl
module load parallel
parallel -j16 qtltools cis --vcf /project2/gilad/singlecell-qtl/bulk/genotypes.vcf.gz --bed $<span class="org-variable-name">pheno</span>.bed.gz --permute=100 --chunk {#} 100 --out $<span class="org-variable-name">pheno</span>-qtl.{#}.txt ::: $(<span class="org-sh-quoted-exec">seq</span> 1 100)
</pre>
</div>

<p>
Read the results.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">def</span> <span class="org-function-name">read_qtltools_output</span>(pheno):
  <span class="org-variable-name">file_names</span> = [<span class="org-string">'{}-qtl.{}.txt'</span>.<span class="org-builtin">format</span>(pheno, i) <span class="org-keyword">for</span> i <span class="org-keyword">in</span> <span class="org-builtin">range</span>(1, 101)]
  <span class="org-keyword">return</span> (pd.concat([pd.read_table(f, header=<span class="org-constant">None</span>, sep=<span class="org-string">' '</span>)
                     <span class="org-keyword">for</span> f <span class="org-keyword">in</span> file_names <span class="org-keyword">if</span> os.path.exists(f) <span class="org-keyword">and</span>
                     os.path.getsize(f) &gt; 0])
          .rename(columns={i: x <span class="org-keyword">for</span> i, x <span class="org-keyword">in</span> <span class="org-builtin">enumerate</span>(
            [<span class="org-string">'gene'</span>, <span class="org-string">'chr'</span>, <span class="org-string">'start'</span>, <span class="org-string">'end'</span>, <span class="org-string">'strand'</span>, <span class="org-string">'num_vars'</span>,
             <span class="org-string">'distance'</span>, <span class="org-string">'id'</span>, <span class="org-string">'var_chr'</span>, <span class="org-string">'var_start'</span>, <span class="org-string">'var_end'</span>, <span class="org-string">'df'</span>,
             <span class="org-string">'dummy'</span>, <span class="org-string">'a'</span>, <span class="org-string">'b'</span>, <span class="org-string">'p_nominal'</span>, <span class="org-string">'beta'</span>, <span class="org-string">'p_empirical'</span>, <span class="org-string">'p_beta'</span>]
          )})
          .sort_values(<span class="org-string">'p_beta'</span>))
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">mean_qtls</span> = read_qtltools_output(<span class="org-string">'mean'</span>)
</pre>
</div>

<p>
Apply the Benjamini-Hochberg procedure:
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">def</span> <span class="org-function-name">bh</span>(df):
  <span class="org-keyword">return</span> df[<span class="org-string">'p_beta'</span>] &lt; .05 * np.arange(1, df.shape[0] + 1) / df.shape[0]
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">keep_mean_qtls</span> = bh(mean_qtls)
keep_mean_qtls.<span class="org-builtin">sum</span>()
</pre>
</div>

<pre class="example">
189

</pre>

<div class="org-src-container">
<pre class="src src-ipython">mean_qtls[keep_mean_qtls].merge(gene_info, left_on=<span class="org-string">'gene'</span>, right_index=<span class="org-constant">True</span>)[[<span class="org-string">'gene'</span>, <span class="org-string">'name'</span>, <span class="org-string">'id'</span>, <span class="org-string">'beta'</span>, <span class="org-string">'p_beta'</span>]]
</pre>
</div>

<pre class="example">
gene      name           id       beta         p_beta
42  ENSG00000162517      PEF1  rs115438012  -0.614188   0.000000e+00
49  ENSG00000121486    TRMT1L   rs73059600   7.770620   0.000000e+00
34  ENSG00000151502    VPS26B   rs59014772   9.085280   0.000000e+00
26  ENSG00000088280     ASAP3  rs142168595   8.061380  2.311830e-311
60  ENSG00000154814    OXNAD1  rs186345075   6.056120  1.244260e-248
47  ENSG00000169515     CCDC8  rs115292552   8.416460  1.162980e-247
30  ENSG00000105732    ZNF574  rs138978828   6.917230  1.363250e-222
36  ENSG00000146067   FAM193B  rs191336763   6.666040  5.046890e-222
0   ENSG00000118197     DDX59  rs189972380   9.531930  5.553180e-201
26  ENSG00000086300     SNX10  rs193043296   5.148360  4.541970e-176
7   ENSG00000144909   OSBPL11  rs142359735   6.324380  4.300360e-166
32  ENSG00000198466    ZNF587  rs138158391   6.448550  1.089520e-163
43  ENSG00000142556    ZNF614   rs57469094  10.462400  2.407550e-157
22  ENSG00000116337     AMPD2  rs115795881   8.250020  1.059290e-152
24  ENSG00000152601     MBNL1  rs151091865   7.133200  1.156550e-150
31  ENSG00000028277    POU2F2  rs141805699   6.914220  2.546100e-134
27  ENSG00000105671     DDX49   rs56165767   3.878050  2.622660e-115
12  ENSG00000002822    MAD1L1   rs77575700   6.492050   8.173360e-83
49  ENSG00000116793     PHTF1  rs114564397   5.112960   9.314330e-79
6   ENSG00000162636   FAM102B     rs481848  -4.752180   8.336260e-78
0   ENSG00000171612  SLC25A33    rs2401422   4.831560   8.401550e-75
45  ENSG00000116667   C1orf21  rs190731151   5.134630   4.383330e-58
62  ENSG00000162623      TYW3   rs28602706  -0.440675   2.126720e-51
22  ENSG00000131503    ANKHD1   rs58190825   4.887060   1.848730e-48
6   ENSG00000137726     FXYD6  rs190517409   5.676260   5.540640e-47
33  ENSG00000175567      UCP2   rs80217992  -0.330121   7.408280e-47
57  ENSG00000156876     SASS6  rs149086207   0.771629   8.289340e-41
27  ENSG00000134086       VHL  rs115909610   3.021960   1.079410e-40
19  ENSG00000169136      ATF5  rs142916709   9.093010   1.342720e-38
63  ENSG00000169733      RFNG  rs116264601   7.291060   4.530920e-31
..              ...       ...          ...        ...            ...
32  ENSG00000162585   C1orf86     rs436543   0.117010   6.688970e-04
17  ENSG00000106853     PTGR1    rs7028222  -0.247891   6.997520e-04
36  ENSG00000160685    ZBTB7B   rs12134456  -7.110410   7.170970e-04
21  ENSG00000130770    ATPIF1     rs530426  -0.094513   7.544320e-04
17  ENSG00000166938     DIS3L   rs79824500   8.834890   8.095130e-04
13  ENSG00000253797    UTP14C  rs115432565   0.381140   8.122250e-04
29  ENSG00000136856    SLC2A8  rs149587918   5.056830   8.605430e-04
21  ENSG00000165140      FBP1   rs10989709  -8.628560   9.267410e-04
30  ENSG00000269343   ZNF587B  rs188747487   8.310020   9.522340e-04
5   ENSG00000085491  SLC25A24     rs607101  -0.146611   9.530610e-04
8   ENSG00000182973    CNOT10    rs6788735  -0.132575   9.827390e-04
28  ENSG00000150753      CCT5    rs2548546  -0.105278   9.926300e-04
19  ENSG00000169174     PCSK9   rs77965618   6.162400   1.024780e-03
8   ENSG00000127184     COX7C   rs12332473   0.074714   1.052670e-03
55  ENSG00000117118      SDHB   rs12066331  -0.221960   1.057660e-03
9   ENSG00000137880     GCHFR  rs190530286   4.580140   1.085450e-03
47  ENSG00000198799     LRIG2  rs144581772   0.748893   1.144320e-03
0   ENSG00000153885    KCTD15  rs146198944   2.407710   1.182870e-03
25  ENSG00000134809    TIMM10    rs2454664  -0.263685   1.202400e-03
6   ENSG00000086062   B4GALT1    rs1125479   0.229450   1.218740e-03
23  ENSG00000213523      SRA1     rs778591  -0.144387   1.301270e-03
21  ENSG00000267645   POLR2J2  rs183117024  -8.381680   1.314740e-03
42  ENSG00000116489    CAPZA1    rs1994604   0.099928   1.353860e-03
12  ENSG00000185787   MORF4L1    rs8039144  -0.083248   1.362930e-03
40  ENSG00000150401   DCUN1D2   rs74999777   6.601230   1.363650e-03
61  ENSG00000116791      CRYZ   rs28594379  -0.670421   1.408140e-03
23  ENSG00000134153      EMC7   rs17237191  -0.373564   1.419180e-03
36  ENSG00000166444       ST5  rs181100237   6.130040   1.477240e-03
47  ENSG00000121775   TMEM39B  rs111714251   7.927280   1.505640e-03
63  ENSG00000249992   TMEM158   rs73829696   9.665750   1.544280e-03

[189 rows x 5 columns]
</pre>

<p>
Check the beta approximation to the permutation p-values:
</p>

<div class="org-src-container">
<pre class="src src-ipython">plt.clf()
plt.scatter(mean_qtls[<span class="org-string">'p_empirical'</span>], mean_qtls[<span class="org-string">'p_beta'</span>], s=1, c=<span class="org-string">'k'</span>)
plt.plot([0, 1], [0, 1], c=<span class="org-string">'r'</span>, ls=<span class="org-string">'--'</span>)
plt.xlabel(<span class="org-string">'Empirical p-value'</span>)
plt.ylabel(<span class="org-string">'Approximate p-value'</span>)
</pre>
</div>

<pre class="example">
Text(0,0.5,'Approximate p-value')

</pre>

<div class="figure">
<p><img src="figure/qtl-mapping.org/zi2-mean-qtl-beta.png" alt="zi2-mean-qtl-beta.png">
</p>
</div>

<p>
Plot the quantile-quantile plot of QTL test statistics:
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">def</span> <span class="org-function-name">qqplot</span>(x):
  <span class="org-variable-name">y</span> = -np.log10(x[np.logical_and(np.isfinite(x), x &gt; 0)])
  <span class="org-variable-name">y</span> = y[np.isfinite(y)]
  <span class="org-variable-name">n</span> = y.shape[0]
  plt.clf()
  plt.scatter(np.flip(-np.log10(np.linspace(0, 1, n + 1)[1:]), -1), np.sort(y), s=1, c=<span class="org-string">'k'</span>)
  plt.plot([0, np.log10(n)], [0, np.log10(n)], color=<span class="org-string">'r'</span>, ls=<span class="org-string">'dashed'</span>)
  plt.xlabel(<span class="org-string">'Expected -$\log_{10}(p)$'</span>)
  plt.ylabel(<span class="org-string">'Observed -$\log_{10}(p)$'</span>)
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython">qqplot(mean_qtls[<span class="org-string">'p_beta'</span>])
</pre>
</div>


<div class="figure">
<p><img src="figure/qtl-mapping.org/zi2-mean-qtl-qq.png" alt="zi2-mean-qtl-qq.png">
</p>
</div>
</div>
</div>

<div id="outline-container-orgccdbbdc" class="outline-2">
<h2 id="orgccdbbdc">Dispersion-QTL calling</h2>
<div class="outline-text-2" id="text-orgccdbbdc">
<p>
Read the estimated parameters and center them.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">disp</span> = (pd.read_table(<span class="org-string">'/scratch/midway2/aksarkar/singlecell/zi2-dispersion.txt.gz'</span>, sep=<span class="org-string">' '</span>, index_col=0)
        .transform(<span class="org-keyword">lambda</span> x: x - x.mean(), axis=1))
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython">write_pheno_file(disp, gene_info, <span class="org-string">'/scratch/midway2/aksarkar/singlecell/disp.bed'</span>)
</pre>
</div>

<p>
Index the phenotype file.
</p>

<pre class="example">
Submitted batch job 43380178

</pre>

<p>
Run <code>qtltools</code>.
</p>

<pre class="example">
Submitted batch job 43380210

</pre>

<p>
Read the output.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">disp_qtls</span> = read_qtltools_output(<span class="org-string">'disp'</span>)
<span class="org-variable-name">keep_disp_qtls</span> = bh(disp_qtls)
keep_disp_qtls.<span class="org-builtin">sum</span>()
</pre>
</div>

<pre class="example">
98

</pre>

<div class="org-src-container">
<pre class="src src-ipython">disp_qtls[keep_disp_qtls].merge(gene_info, left_on=<span class="org-string">'gene'</span>, right_index=<span class="org-constant">True</span>)[[<span class="org-string">'gene'</span>, <span class="org-string">'name'</span>, <span class="org-string">'id'</span>, <span class="org-string">'beta'</span>, <span class="org-string">'p_beta'</span>]]
</pre>
</div>

<pre class="example">
gene      name           id      beta         p_beta
29  ENSG00000165283    STOML2  rs151013351 -3.309280   0.000000e+00
11  ENSG00000170085     SIMC1  rs115194011 -4.063380   0.000000e+00
18  ENSG00000197724      PHF2  rs113363990 -4.412720  7.294860e-228
23  ENSG00000174405      LIG4  rs115154048 -5.077620  2.507990e-182
51  ENSG00000176697      BDNF  rs143827698 -4.773690  2.225750e-170
21  ENSG00000123810      B9D2   rs79026999 -4.370730  1.672330e-151
17  ENSG00000102743  SLC25A15  rs115773631 -3.519380  6.031130e-123
47  ENSG00000125449     ARMC7   rs75902792 -4.630780   6.994600e-88
58  ENSG00000143183     TMCO1  rs114697800 -2.691680   4.711220e-50
40  ENSG00000105865     DUS4L  rs187529047  2.818420   1.040130e-44
28  ENSG00000120314     WDR55   rs11167791  3.441270   2.458160e-42
18  ENSG00000164970   FAM219A   rs78437190 -2.935010   2.794990e-37
28  ENSG00000162909     CAPN2    rs2172360  3.311410   9.247940e-35
46  ENSG00000170860      LSM3  rs143962723 -3.295380   2.297400e-33
24  ENSG00000244165    P2RY11   rs76392887 -3.159010   1.709660e-32
27  ENSG00000116898    MRPS15  rs144661059 -2.239470   1.479050e-28
11  ENSG00000160049      DFFA    rs5007607 -2.658710   4.592440e-26
2   ENSG00000171608    PIK3CD  rs138247028 -4.697440   4.235250e-25
50  ENSG00000117143      UAP1   rs12067012  1.433560   3.856740e-23
29  ENSG00000205356    TECPR1  rs116757663 -2.618370   9.657750e-18
33  ENSG00000114859     CLCN2  rs116709730  4.131380   1.859410e-17
21  ENSG00000172830      SSH3    rs7105623 -0.638843   4.644960e-17
29  ENSG00000183020     AP2A2  rs186982458  2.609980   2.456720e-16
26  ENSG00000125734    GPR108  rs184583582 -2.822690   2.470850e-15
36  ENSG00000161682  FAM171A2   rs12944040  2.363410   3.075220e-15
3   ENSG00000179409    GEMIN4    rs2293067  1.929650   8.587390e-15
25  ENSG00000223802     CERS1  rs116562609 -3.084010   1.026820e-13
26  ENSG00000172273     HINFP  rs115425447  1.686380   1.513860e-13
60  ENSG00000205309      NT5M       rs8511 -2.068040   2.680890e-13
10  ENSG00000197043     ANXA6  rs139372973  3.311060   7.743520e-13
..              ...       ...          ...       ...            ...
22  ENSG00000248098    BCKDHA  rs140346884 -2.913330   2.763770e-05
56  ENSG00000143198     MGST3  rs144467420 -2.748340   4.538870e-05
45  ENSG00000174574   AKIRIN1  rs143588389 -2.988750   6.604600e-05
7   ENSG00000103707     MTFMT  rs137966506  3.245180   6.846880e-05
6   ENSG00000145681    HAPLN1  rs114887660 -3.971260   7.128160e-05
65  ENSG00000184408     KCND2  rs143795479 -4.538730   8.513240e-05
43  ENSG00000137841     PLCB2   rs62019961 -4.512630   8.853740e-05
63  ENSG00000108557      RAI1   rs59212772 -6.439100   1.211490e-04
20  ENSG00000006016     CRLF1   rs10414298 -1.853580   1.396770e-04
16  ENSG00000109084    TMEM97  rs144069393 -2.782690   1.462980e-04
38  ENSG00000106086   PLEKHA8  rs118033320  2.487410   1.639470e-04
14  ENSG00000116906     GNPAT   rs10779810 -1.170110   1.677310e-04
13  ENSG00000064666      CNN2  rs184326775 -1.410370   1.764350e-04
63  ENSG00000149256     TENM4  rs189331375 -4.245470   1.807590e-04
16  ENSG00000117036      ETV3  rs188035981  3.118750   1.978480e-04
44  ENSG00000073050     XRCC1  rs138843872  3.159450   2.268670e-04
20  ENSG00000185453  C19orf68  rs138701164  4.117920   2.468340e-04
37  ENSG00000166452     AKIP1  rs141004297  3.508760   2.546470e-04
12  ENSG00000092978   GPATCH2  rs113462534  4.085070   2.997230e-04
64  ENSG00000131473      ACLY  rs111713404 -2.099470   3.098620e-04
23  ENSG00000160867     FGFR4  rs111567185  2.796980   3.604680e-04
46  ENSG00000186472      PCLO   rs78573728 -4.342580   3.955380e-04
18  ENSG00000132510     KDM6B  rs115083985 -1.067740   4.668840e-04
27  ENSG00000070669      ASNS   rs79183163 -2.619050   4.672200e-04
27  ENSG00000105671     DDX49  rs147349438  1.518330   5.734510e-04
29  ENSG00000165895  ARHGAP42  rs111702886 -4.671170   5.748990e-04
22  ENSG00000166333       ILK   rs77801446 -2.310190   7.091510e-04
20  ENSG00000148110    HIATL1    rs9409727 -1.103750   8.033820e-04
40  ENSG00000117906      RCN2    rs8026731 -1.355620   8.237700e-04
55  ENSG00000085276     MECOM  rs140988682 -4.053850   8.395570e-04

[98 rows x 5 columns]
</pre>

<p>
Check the beta approximation to the permutation p-values:
</p>

<div class="org-src-container">
<pre class="src src-ipython">plt.clf()
plt.scatter(disp_qtls[<span class="org-string">'p_empirical'</span>], disp_qtls[<span class="org-string">'p_beta'</span>], s=1, c=<span class="org-string">'k'</span>)
plt.plot([0, 1], [0, 1], c=<span class="org-string">'r'</span>, ls=<span class="org-string">'--'</span>)
plt.xlabel(<span class="org-string">'Empirical p-value'</span>)
plt.ylabel(<span class="org-string">'Approximate p-value'</span>)
</pre>
</div>

<pre class="example">
Text(0,0.5,'Approximate p-value')

</pre>

<div class="figure">
<p><img src="figure/qtl-mapping.org/zi2-disp-qtl-beta.png" alt="zi2-disp-qtl-beta.png">
</p>
</div>

<p>
Plot the quantile-quantile plot of QTL test statistics:
</p>

<div class="org-src-container">
<pre class="src src-ipython">qqplot(disp_qtls[<span class="org-string">'p_beta'</span>])
</pre>
</div>


<div class="figure">
<p><img src="figure/qtl-mapping.org/zi2-disp-qtl-qq.png" alt="zi2-disp-qtl-qq.png">
</p>
</div>
</div>
</div>
</div>
<div id="postamble" class="status">
<p class="author">Author: Abhishek Sarkar</p>
<p class="date">Created: 2018-02-16 Fri 14:10</p>
<p class="validation"><a href="http://validator.w3.org/check?uri=referer">Validate</a></p>
</div>
</body>
</html>
