<!DOCTYPE html>
<html lang="en">
<head>
<!-- 2018-01-16 Tue 14:09 -->
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
<title>QTL mapping pipeline</title>
<meta name="generator" content="Org mode">
<meta name="author" content="Abhishek Sarkar">
<style type="text/css">
 <!--/*--><![CDATA[/*><!--*/
  .title  { text-align: center;
             margin-bottom: .2em; }
  .subtitle { text-align: center;
              font-size: medium;
              font-weight: bold;
              margin-top:0; }
  .todo   { font-family: monospace; color: red; }
  .done   { font-family: monospace; color: green; }
  .priority { font-family: monospace; color: orange; }
  .tag    { background-color: #eee; font-family: monospace;
            padding: 2px; font-size: 80%; font-weight: normal; }
  .timestamp { color: #bebebe; }
  .timestamp-kwd { color: #5f9ea0; }
  .org-right  { margin-left: auto; margin-right: 0px;  text-align: right; }
  .org-left   { margin-left: 0px;  margin-right: auto; text-align: left; }
  .org-center { margin-left: auto; margin-right: auto; text-align: center; }
  .underline { text-decoration: underline; }
  #postamble p, #preamble p { font-size: 90%; margin: .2em; }
  p.verse { margin-left: 3%; }
  pre {
    border: 1px solid #ccc;
    box-shadow: 3px 3px 3px #eee;
    padding: 8pt;
    font-family: monospace;
    overflow: auto;
    margin: 1.2em;
  }
  pre.src {
    position: relative;
    overflow: visible;
    padding-top: 1.2em;
  }
  pre.src:before {
    display: none;
    position: absolute;
    background-color: white;
    top: -10px;
    right: 10px;
    padding: 3px;
    border: 1px solid black;
  }
  pre.src:hover:before { display: inline;}
  /* Languages per Org manual */
  pre.src-asymptote:before { content: 'Asymptote'; }
  pre.src-awk:before { content: 'Awk'; }
  pre.src-C:before { content: 'C'; }
  /* pre.src-C++ doesn't work in CSS */
  pre.src-clojure:before { content: 'Clojure'; }
  pre.src-css:before { content: 'CSS'; }
  pre.src-D:before { content: 'D'; }
  pre.src-ditaa:before { content: 'ditaa'; }
  pre.src-dot:before { content: 'Graphviz'; }
  pre.src-calc:before { content: 'Emacs Calc'; }
  pre.src-emacs-lisp:before { content: 'Emacs Lisp'; }
  pre.src-fortran:before { content: 'Fortran'; }
  pre.src-gnuplot:before { content: 'gnuplot'; }
  pre.src-haskell:before { content: 'Haskell'; }
  pre.src-hledger:before { content: 'hledger'; }
  pre.src-java:before { content: 'Java'; }
  pre.src-js:before { content: 'Javascript'; }
  pre.src-latex:before { content: 'LaTeX'; }
  pre.src-ledger:before { content: 'Ledger'; }
  pre.src-lisp:before { content: 'Lisp'; }
  pre.src-lilypond:before { content: 'Lilypond'; }
  pre.src-lua:before { content: 'Lua'; }
  pre.src-matlab:before { content: 'MATLAB'; }
  pre.src-mscgen:before { content: 'Mscgen'; }
  pre.src-ocaml:before { content: 'Objective Caml'; }
  pre.src-octave:before { content: 'Octave'; }
  pre.src-org:before { content: 'Org mode'; }
  pre.src-oz:before { content: 'OZ'; }
  pre.src-plantuml:before { content: 'Plantuml'; }
  pre.src-processing:before { content: 'Processing.js'; }
  pre.src-python:before { content: 'Python'; }
  pre.src-R:before { content: 'R'; }
  pre.src-ruby:before { content: 'Ruby'; }
  pre.src-sass:before { content: 'Sass'; }
  pre.src-scheme:before { content: 'Scheme'; }
  pre.src-screen:before { content: 'Gnu Screen'; }
  pre.src-sed:before { content: 'Sed'; }
  pre.src-sh:before { content: 'shell'; }
  pre.src-sql:before { content: 'SQL'; }
  pre.src-sqlite:before { content: 'SQLite'; }
  /* additional languages in org.el's org-babel-load-languages alist */
  pre.src-forth:before { content: 'Forth'; }
  pre.src-io:before { content: 'IO'; }
  pre.src-J:before { content: 'J'; }
  pre.src-makefile:before { content: 'Makefile'; }
  pre.src-maxima:before { content: 'Maxima'; }
  pre.src-perl:before { content: 'Perl'; }
  pre.src-picolisp:before { content: 'Pico Lisp'; }
  pre.src-scala:before { content: 'Scala'; }
  pre.src-shell:before { content: 'Shell Script'; }
  pre.src-ebnf2ps:before { content: 'ebfn2ps'; }
  /* additional language identifiers per "defun org-babel-execute"
       in ob-*.el */
  pre.src-cpp:before  { content: 'C++'; }
  pre.src-abc:before  { content: 'ABC'; }
  pre.src-coq:before  { content: 'Coq'; }
  pre.src-groovy:before  { content: 'Groovy'; }
  /* additional language identifiers from org-babel-shell-names in
     ob-shell.el: ob-shell is the only babel language using a lambda to put
     the execution function name together. */
  pre.src-bash:before  { content: 'bash'; }
  pre.src-csh:before  { content: 'csh'; }
  pre.src-ash:before  { content: 'ash'; }
  pre.src-dash:before  { content: 'dash'; }
  pre.src-ksh:before  { content: 'ksh'; }
  pre.src-mksh:before  { content: 'mksh'; }
  pre.src-posh:before  { content: 'posh'; }
  /* Additional Emacs modes also supported by the LaTeX listings package */
  pre.src-ada:before { content: 'Ada'; }
  pre.src-asm:before { content: 'Assembler'; }
  pre.src-caml:before { content: 'Caml'; }
  pre.src-delphi:before { content: 'Delphi'; }
  pre.src-html:before { content: 'HTML'; }
  pre.src-idl:before { content: 'IDL'; }
  pre.src-mercury:before { content: 'Mercury'; }
  pre.src-metapost:before { content: 'MetaPost'; }
  pre.src-modula-2:before { content: 'Modula-2'; }
  pre.src-pascal:before { content: 'Pascal'; }
  pre.src-ps:before { content: 'PostScript'; }
  pre.src-prolog:before { content: 'Prolog'; }
  pre.src-simula:before { content: 'Simula'; }
  pre.src-tcl:before { content: 'tcl'; }
  pre.src-tex:before { content: 'TeX'; }
  pre.src-plain-tex:before { content: 'Plain TeX'; }
  pre.src-verilog:before { content: 'Verilog'; }
  pre.src-vhdl:before { content: 'VHDL'; }
  pre.src-xml:before { content: 'XML'; }
  pre.src-nxml:before { content: 'XML'; }
  /* add a generic configuration mode; LaTeX export needs an additional
     (add-to-list 'org-latex-listings-langs '(conf " ")) in .emacs */
  pre.src-conf:before { content: 'Configuration File'; }

  table { border-collapse:collapse; }
  caption.t-above { caption-side: top; }
  caption.t-bottom { caption-side: bottom; }
  td, th { vertical-align:top;  }
  th.org-right  { text-align: center;  }
  th.org-left   { text-align: center;   }
  th.org-center { text-align: center; }
  td.org-right  { text-align: right;  }
  td.org-left   { text-align: left;   }
  td.org-center { text-align: center; }
  dt { font-weight: bold; }
  .footpara { display: inline; }
  .footdef  { margin-bottom: 1em; }
  .figure { padding: 1em; }
  .figure p { text-align: center; }
  .inlinetask {
    padding: 10px;
    border: 2px solid gray;
    margin: 10px;
    background: #ffffcc;
  }
  #org-div-home-and-up
   { text-align: right; font-size: 70%; white-space: nowrap; }
  textarea { overflow-x: auto; }
  .linenr { font-size: smaller }
  .code-highlighted { background-color: #ffff00; }
  .org-info-js_info-navigation { border-style: none; }
  #org-info-js_console-label
    { font-size: 10px; font-weight: bold; white-space: nowrap; }
  .org-info-js_search-highlight
    { background-color: #ffff00; color: #000000; font-weight: bold; }
  .org-svg { width: 90%; }
  /*]]>*/-->
</style>
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<link rel="stylesheet" type="text/css" href="https://www.pirilampo.org/styles/readtheorg/css/htmlize.css"/>
<style type="text/css">body {width: 60em; margin:auto} pre.src {overflow:auto}</style>
<script type="text/javascript">
/*
@licstart  The following is the entire license notice for the
JavaScript code in this tag.

Copyright (C) 2012-2017 Free Software Foundation, Inc.

The JavaScript code in this tag is free software: you can
redistribute it and/or modify it under the terms of the GNU
General Public License (GNU GPL) as published by the Free Software
Foundation, either version 3 of the License, or (at your option)
any later version.  The code is distributed WITHOUT ANY WARRANTY;
without even the implied warranty of MERCHANTABILITY or FITNESS
FOR A PARTICULAR PURPOSE.  See the GNU GPL for more details.

As additional permission under GNU GPL version 3 section 7, you
may distribute non-source (e.g., minimized or compacted) forms of
that code without the copy of the GNU GPL normally required by
section 4, provided you include this license notice and a URL
through which recipients can access the Corresponding Source.


@licend  The above is the entire license notice
for the JavaScript code in this tag.
*/
<!--/*--><![CDATA[/*><!--*/
 function CodeHighlightOn(elem, id)
 {
   var target = document.getElementById(id);
   if(null != target) {
     elem.cacheClassElem = elem.className;
     elem.cacheClassTarget = target.className;
     target.className = "code-highlighted";
     elem.className   = "code-highlighted";
   }
 }
 function CodeHighlightOff(elem, id)
 {
   var target = document.getElementById(id);
   if(elem.cacheClassElem)
     elem.className = elem.cacheClassElem;
   if(elem.cacheClassTarget)
     target.className = elem.cacheClassTarget;
 }
/*]]>*///-->
</script>
</head>
<body>
<div id="content">
<h1 class="title">QTL mapping pipeline</h1>
<div id="table-of-contents">
<h2>Table of Contents</h2>
<div id="text-table-of-contents">
<ul>
<li><a href="#org5ed2a04">Introduction</a></li>
<li><a href="#org051ecfa">Mean-QTL calling</a></li>
<li><a href="#org018b84e">Dispersion-QTL calling</a></li>
<li><a href="#orga4d85dd">Next steps</a></li>
</ul>
</div>
</div>

<div id="outline-container-org5ed2a04" class="outline-2">
<h2 id="org5ed2a04">Introduction</h2>
<div class="outline-text-2" id="text-org5ed2a04">
<p>
Here, we set up the infrastructure to call mean and variance QTLs, assuming
that we have <a href="zinb.html">previously estimated means and variances per individual, per
gene.</a>
</p>
</div>
</div>

<div id="outline-container-org051ecfa" class="outline-2">
<h2 id="org051ecfa">Mean-QTL calling</h2>
<div class="outline-text-2" id="text-org051ecfa">
<p>
Write out the <a href="https://qtltools.github.io/qtltools/pages/input_files.html">phenotype file</a> for <code>QTLtools</code>.
</p>

<div class="org-src-container">
<pre class="src src-ipython" id="org6264062"><span class="org-variable-name">gene_info</span> = (pd.read_table(<span class="org-string">'/home/aksarkar/projects/singlecell-qtl/data/scqtl-genes.txt.gz'</span>)
             .set_index(<span class="org-string">'gene'</span>)
             .query(<span class="org-string">'source == "H. sapiens"'</span>)
             .query(<span class="org-string">'chr != "hsX"'</span>)
             .query(<span class="org-string">'chr != "hsY"'</span>)
             .query(<span class="org-string">'chr != "hsMT"'</span>))
gene_info.head()
</pre>
</div>

<pre class="example">
chr      start        end      name strand      source
gene
ENSG00000000419  hs20   49551404   49575092      DPM1      -  H. sapiens
ENSG00000000457   hs1  169818772  169863408     SCYL3      -  H. sapiens
ENSG00000000460   hs1  169631245  169823221  C1orf112      +  H. sapiens
ENSG00000000938   hs1   27938575   27961788       FGR      -  H. sapiens
ENSG00000000971   hs1  196621008  196716634       CFH      +  H. sapiens
</pre>

<div class="org-src-container">
<pre class="src src-ipython" id="org042c79c"><span class="org-keyword">def</span> <span class="org-function-name">qtltools_format</span>(row):
  <span class="org-variable-name">row</span>[<span class="org-string">'#Chr'</span>] = <span class="org-string">'chr{}'</span>.<span class="org-builtin">format</span>(row[<span class="org-string">'chr'</span>][2:])
  <span class="org-variable-name">row</span>[<span class="org-string">'gid'</span>] = row.name
  <span class="org-variable-name">row</span>[<span class="org-string">'pid'</span>] = row.name
  <span class="org-keyword">return</span> row

<span class="org-keyword">def</span> <span class="org-function-name">write_pheno_file</span>(pheno, gene_info, output_file):
  (gene_info
   .<span class="org-builtin">apply</span>(qtltools_format, axis=1)
   .merge(pheno, left_index=<span class="org-constant">True</span>, right_index=<span class="org-constant">True</span>)
   .to_csv(output_file,
           sep=<span class="org-string">'\t'</span>,
           columns=[<span class="org-string">'#Chr'</span>, <span class="org-string">'start'</span>, <span class="org-string">'end'</span>, <span class="org-string">'pid'</span>, <span class="org-string">'gid'</span>, <span class="org-string">'strand'</span>] + <span class="org-builtin">list</span>(pheno.columns),
           header=<span class="org-constant">True</span>,
           index=<span class="org-constant">False</span>,
           index_label=<span class="org-constant">False</span>)
  )
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">mean</span> = (pd.read_table(<span class="org-string">'/scratch/midway2/aksarkar/singlecell/mean.txt.gz'</span>, sep=<span class="org-string">' '</span>, index_col=0)
        .rename(columns=<span class="org-keyword">lambda</span> x: x[2:] <span class="org-keyword">if</span> x.startswith(<span class="org-string">'NA'</span>) <span class="org-keyword">else</span> x))
mean.head()
</pre>
</div>

<pre class="example">
18489     18498     18499     18501     18502     18505  \
gene
ENSG00000000003  2.153422  1.908637  2.020811  2.235909  2.016276  1.916526
ENSG00000000005  2.717537  2.686136  1.970839  1.882466  1.934783  1.743753
ENSG00000000419  1.473960  1.381287  1.748404  1.715845  1.722835  1.655344
ENSG00000000457  2.695425  2.869927  1.661883  1.750790  1.644461  1.388998
ENSG00000000460  1.246338  1.228745  1.070099  1.072112  1.127348  0.976623

18507     18508     18519     18520    ...        19119  \
gene                                                       ...
ENSG00000000003  2.183926  2.057383  2.185234  1.972367    ...     2.228020
ENSG00000000005  1.854442  2.808901  2.835537  2.211654    ...     0.904509
ENSG00000000419  1.876334  1.778172  1.699735  1.670322    ...     1.816166
ENSG00000000457  1.649863  2.590687  2.809266  1.887073    ...     0.504516
ENSG00000000460  1.125625  1.514480  1.059698  1.170255    ...     1.119757

19128     19153     19159     19190     19193     19203  \
gene
ENSG00000000003  2.325042  2.305453  1.893718  2.172396  2.011524  1.691384
ENSG00000000005  1.855778  1.865121  2.956233  2.029287  2.142070  2.838911
ENSG00000000419  2.068414  1.947017  1.612372  1.774130  1.974698  1.299714
ENSG00000000457  1.459007  1.825049  2.791146  1.664737  1.957186  2.885063
ENSG00000000460  1.290979  1.329372  1.050951  1.334755  1.301709  2.220205

19207     19210     19257
gene
ENSG00000000003  2.275174  2.267549  2.076748
ENSG00000000005  1.951805  1.327100  1.759800
ENSG00000000419  1.876612  2.129439  2.040515
ENSG00000000457  1.824563  0.950023  1.269692
ENSG00000000460  1.437071  1.137395  1.117885

[5 rows x 32 columns]
</pre>

<p>
Visualize the data and the inferred parameters:
</p>

<div class="org-src-container">
<pre class="src src-ipython">plt.clf()

</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython">write_pheno_file(mean, gene_info, <span class="org-string">'/scratch/midway2/aksarkar/singlecell/mean.bed'</span>)
</pre>
</div>

<p>
Index the phenotype file.
</p>

<div class="org-src-container">
<pre class="src src-sh" id="orge80d324"><span class="org-builtin">export</span> <span class="org-variable-name">input</span>=$<span class="org-variable-name">input</span>
sbatch --partition=$<span class="org-variable-name">partition</span> --wait
<span class="org-comment-delimiter">#</span><span class="org-comment">!/bin/bash</span>
sort -k1,1 -k2,2n -k3,3n $<span class="org-variable-name">input</span> | bgzip &gt;$<span class="org-variable-name">input</span>.gz
tabix -p bed $<span class="org-variable-name">input</span>.gz
</pre>
</div>

<p>
Run the QTL mapping.
</p>

<div class="org-src-container">
<pre class="src src-sh" id="org5a3e21d"><span class="org-builtin">export</span> <span class="org-variable-name">pheno</span>=$<span class="org-variable-name">pheno</span>
sbatch --partition=$<span class="org-variable-name">partition</span> -N1 -c4 -J $<span class="org-variable-name">pheno</span>-qtl -o $<span class="org-variable-name">pheno</span>-qtl.log --wait
<span class="org-comment-delimiter">#</span><span class="org-comment">!/bin/bash</span>
<span class="org-builtin">source</span> activate scqtl
module load parallel
parallel -j4 qtltools cis --vcf /project2/gilad/singlecell-qtl/bulk/genotypes.vcf.gz --bed $<span class="org-variable-name">pheno</span>.bed.gz --nominal=0.01 --chunk {#} 100 --out $<span class="org-variable-name">pheno</span>-qtl.{#}.txt ::: $(<span class="org-sh-quoted-exec">seq</span> 1 100)
</pre>
</div>

<p>
Read the results.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">file_names</span> = [<span class="org-string">'mean-qtl.{}.txt'</span>.<span class="org-builtin">format</span>(i) <span class="org-keyword">for</span> i <span class="org-keyword">in</span> <span class="org-builtin">range</span>(1, 101)]
<span class="org-variable-name">mean_qtls</span> = (pd.concat([pd.read_table(f, header=<span class="org-constant">None</span>, sep=<span class="org-string">' '</span>) <span class="org-keyword">for</span> f <span class="org-keyword">in</span> file_names <span class="org-keyword">if</span> os.path.exists(f) <span class="org-keyword">and</span> os.path.getsize(f) &gt; 0])
             .rename(columns={i: x <span class="org-keyword">for</span> i, x <span class="org-keyword">in</span> <span class="org-builtin">enumerate</span>([<span class="org-string">'gene'</span>, <span class="org-string">'chr'</span>, <span class="org-string">'start'</span>, <span class="org-string">'end'</span>, <span class="org-string">'strand'</span>, <span class="org-string">'num_vars'</span>, <span class="org-string">'distance'</span>, <span class="org-string">'id'</span>, <span class="org-string">'var_chr'</span>, <span class="org-string">'var_start'</span>, <span class="org-string">'var_end'</span>, <span class="org-string">'p'</span>, <span class="org-string">'beta'</span>, <span class="org-string">'top'</span>])})
             .sort_values(<span class="org-string">'p'</span>)
             .set_index(<span class="org-string">'gene'</span>))
mean_qtls[mean_qtls[<span class="org-string">'top'</span>] == 1].head()
</pre>
</div>

<pre class="example">
chr   start     end strand  num_vars  distance id var_chr  \
gene
ENSG00000079101  chr18  596989  650334      +      5663   -596989  T   chr18
ENSG00000023191  chr11  494513  507300      -      4391    494513  G   chr11
ENSG00000177951  chr11  167785  207428      -      3537    167785  A   chr11
ENSG00000177697  chr11  832844  839831      +      5570   -832844  A   chr11
ENSG00000147364   chr8  356429  421225      +      6388   -356429  T    chr8

var_start  var_end             p      beta  top
gene
ENSG00000079101    1127932        0  1.743600e-09 -1.552050    1
ENSG00000023191     296256        0  1.174030e-08  0.589488    1
ENSG00000177951     436759        0  2.953120e-07  1.175580    1
ENSG00000177697     436759        0  7.761760e-07  1.124540    1
ENSG00000147364    1086162        0  7.801650e-07  0.816134    1
</pre>

<p>
<b>TODO:</b> Permutation testing within QTLtools? lfsr estimation by <code>ashr</code>?
</p>
</div>
</div>

<div id="outline-container-org018b84e" class="outline-2">
<h2 id="org018b84e">Dispersion-QTL calling</h2>
<div class="outline-text-2" id="text-org018b84e">
<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">disp</span> = (pd.read_table(<span class="org-string">'/scratch/midway2/aksarkar/singlecell/dispersion.txt.gz'</span>, sep=<span class="org-string">' '</span>, index_col=0)
        .rename(columns=<span class="org-keyword">lambda</span> x: x[2:] <span class="org-keyword">if</span> x.startswith(<span class="org-string">'NA'</span>) <span class="org-keyword">else</span> x))
disp.head()
</pre>
</div>

<pre class="example">
18489      18498     18499      18501     18502  \
gene
ENSG00000000003   1.167245   1.706873  1.599450   1.408319  1.711105
ENSG00000000005  10.335133  10.322262  9.905348   9.639630  9.502019
ENSG00000000419   2.515978   3.487238  2.140802   1.586118  2.195697
ENSG00000000457  10.371654  10.217912  9.560677  10.536921  9.979990
ENSG00000000460   9.115197   9.561979  9.006826   8.728274  8.700228

18505      18507      18508      18519     18520  \
gene
ENSG00000000003  2.077771   1.193537   1.548394   1.833913  1.753378
ENSG00000000005  9.582265  10.065374  10.365082  10.301587  9.595152
ENSG00000000419  3.818562   1.306487   1.470760   1.536752  2.150593
ENSG00000000457  9.621971   9.546401  10.254088  10.192816  9.819939
ENSG00000000460  8.882592   8.865636   9.414287   8.887925  9.314652

...        19119      19128     19153      19159  \
gene               ...
ENSG00000000003    ...     1.325692   0.747701  1.708594   1.236187
ENSG00000000005    ...     8.796814  10.149561  9.674639  10.224201
ENSG00000000419    ...     3.763241   1.062397  1.820775   8.463470
ENSG00000000457    ...     8.593268   9.501520  9.566214  10.446822
ENSG00000000460    ...     7.901440   8.161574  8.895837   9.729097

19190      19193      19203      19207     19210  \
gene
ENSG00000000003   1.769961   1.417471   2.819164   1.286686  1.484697
ENSG00000000005   9.807590   9.849161  10.167520  10.156749  9.744061
ENSG00000000419   1.816306   1.169699   8.482072   1.614174  1.811550
ENSG00000000457  10.278067  10.191409   9.601987   9.680202  9.291765
ENSG00000000460   9.537437   8.882127  10.174777   9.343502  8.246989

19257
gene
ENSG00000000003  1.418264
ENSG00000000005  9.928991
ENSG00000000419  1.382229
ENSG00000000457  9.178885
ENSG00000000460  8.454962

[5 rows x 32 columns]
</pre>

<div class="org-src-container">
<pre class="src src-ipython">write_pheno_file(disp, gene_info, <span class="org-string">'/scratch/midway2/aksarkar/singlecell/disp.bed'</span>)
</pre>
</div>

<p>
Index the phenotype file
</p>

<pre class="example">
Submitted batch job 42160151

</pre>

<p>
Run <code>qtltools</code>
</p>

<pre class="example">
Submitted batch job 42160171

</pre>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">file_names</span> = [<span class="org-string">'disp-qtl.{}.txt'</span>.<span class="org-builtin">format</span>(i) <span class="org-keyword">for</span> i <span class="org-keyword">in</span> <span class="org-builtin">range</span>(1, 101)]
<span class="org-variable-name">disp_qtls</span> = (pd.concat([pd.read_table(f, header=<span class="org-constant">None</span>, sep=<span class="org-string">' '</span>) <span class="org-keyword">for</span> f <span class="org-keyword">in</span> file_names <span class="org-keyword">if</span> os.path.exists(f) <span class="org-keyword">and</span> os.path.getsize(f) &gt; 0])
             .rename(columns={i: x <span class="org-keyword">for</span> i, x <span class="org-keyword">in</span> <span class="org-builtin">enumerate</span>([<span class="org-string">'gene'</span>, <span class="org-string">'chr'</span>, <span class="org-string">'start'</span>, <span class="org-string">'end'</span>, <span class="org-string">'strand'</span>, <span class="org-string">'num_vars'</span>, <span class="org-string">'distance'</span>, <span class="org-string">'id'</span>, <span class="org-string">'var_chr'</span>, <span class="org-string">'var_start'</span>, <span class="org-string">'var_end'</span>, <span class="org-string">'p'</span>, <span class="org-string">'beta'</span>, <span class="org-string">'top'</span>])})
             .sort_values(<span class="org-string">'p'</span>)
             .set_index(<span class="org-string">'gene'</span>))
disp_qtls[disp_qtls[<span class="org-string">'top'</span>] == 1].head()
</pre>
</div>

<pre class="example">
chr   start     end strand  num_vars  distance id var_chr  \
gene
ENSG00000169026   chr4  675619  683230      -      4952    675619  A    chr4
ENSG00000187583   chr1  901878  911245      +      4140   -901878  A    chr1
ENSG00000142082  chr11  215459  236931      -      3636    215459  G   chr11
ENSG00000172748   chr8  182138  197342      +      5320   -182138  G    chr8
ENSG00000255284  chr11  777579  784297      +      5412   -777579  A   chr11

var_start  var_end             p     beta  top
gene
ENSG00000169026     550716        0  2.262070e-18 -6.27400    1
ENSG00000187583    1658302        0  6.812100e-18 -7.00223    1
ENSG00000142082     865448        0  6.317700e-17  3.68775    1
ENSG00000172748     229562        0  9.070000e-16 -6.24766    1
ENSG00000255284     600513        0  2.076950e-15 -6.11760    1
</pre>

<div class="org-src-container">
<pre class="src src-ipython">disp_qtls.shape[0]
</pre>
</div>

<pre class="example">
9720

</pre>
</div>
</div>
<div id="outline-container-orga4d85dd" class="outline-2">
<h2 id="orga4d85dd">Next steps</h2>
<div class="outline-text-2" id="text-orga4d85dd">
<ol class="org-ol">
<li><p>
We previously used half-sibling regression (<a href="https://www.pnas.org/cgi/doi/10.1073/pnas.1511656113">Scholkopf et al 2016</a>) to correct
confounders for bulk RNA-Seq in GTEx (<a href="http://dx.doi.org/10.1101/107623">Park et al 2017</a>).
</p>

<p>
The key idea is that the only ways we could predict the expression of a
gene of interest using genes on other chromosomes are (1) true
<i>trans</i>-regulation and (2) shared confounding. For <i>cis</i>-QTL mapping, we
want to eliminate both of these, which we can do by simply regressing out
gene expression of all genes on all other chromosomes.
</p>

<p>
One open question is whether we should correct counts (which would be
required to solve issue (3)), or correct per-individual means/variances.
</p></li>

<li><p>
We previously built an <a href="https://github.com/YPARK/fqtl">inference engine which supports negative binomial
models</a> which could be easily extended to zero-inflated negative binomial
models.
</p>

<p>
In our experiments directly using the negative binomial likelihood
outperformed the Gaussian model (following the <code>voom</code>-transform, but
sharing the mean model parameters in the dispersion model).
</p>

<p>
This would allow us to solve issues (2) and (3), and potentially make (4)
easier.
</p></li>
</ol>
</div>
</div>
</div>
<div id="postamble" class="status">
<p class="author">Author: Abhishek Sarkar</p>
<p class="date">Created: 2018-01-16 Tue 14:09</p>
<p class="validation"><a href="http://validator.w3.org/check?uri=referer">Validate</a></p>
</div>
</body>
</html>
