<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />


<meta name="author" content="PoYuan Tung" />

<meta name="date" content="2017-09-13" />

<title>QC of single cell libraries</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-4.5.0/css/font-awesome.min.css" rel="stylesheet" />

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>


</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
button.code-folding-btn:focus {
  outline: none;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}

.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>


<div class="container-fluid main-container">

<!-- tabsets -->
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->




<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
  padding-left: 25px;
  text-indent: 0;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>

<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">singlecell-qtl</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Overview</a>
</li>
<li>
  <a href="data-collection.html">Data</a>
</li>
<li>
  <a href="analysis.html">Analysis</a>
</li>
<li>
  <a href="contributing.html">Contributing</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/jdblischak/singlecell-qtl">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">QC of single cell libraries</h1>
<h4 class="author"><em>PoYuan Tung</em></h4>
<h4 class="date"><em>2017-09-13</em></h4>

</div>


<!-- The file analysis/chunks.R contains chunks that define default settings
shared across the workflowr files. -->
<!-- Update knitr chunk options -->
<!-- Insert the date the file was last updated -->
<p><strong>Last updated:</strong> 2018-06-30</p>
<!-- Insert the code version (Git commit SHA1) if Git repository exists and R
 package git2r is installed -->
<p><strong>Code version:</strong> faf8290</p>
<p>This is for qc of the samples. Based on obsevation under the scope and the sequencing results, samples with bad quality will be removed.</p>
<div id="setup" class="section level2">
<h2>Setup</h2>
<pre class="r"><code>library(&quot;cowplot&quot;)
library(&quot;dplyr&quot;)
library(&quot;edgeR&quot;)
library(&quot;ggplot2&quot;)
library(&quot;MASS&quot;)
library(&quot;tibble&quot;)
library(&quot;tidyr&quot;)
theme_set(theme_cowplot())
source(&quot;../code/functions.R&quot;)
library(&quot;Biobase&quot;)</code></pre>
<pre class="r"><code># The palette with grey:
cbPalette &lt;- c(&quot;#999999&quot;, &quot;#E69F00&quot;, &quot;#56B4E9&quot;, &quot;#009E73&quot;, &quot;#F0E442&quot;, &quot;#0072B2&quot;, &quot;#D55E00&quot;, &quot;#CC79A7&quot;)</code></pre>
<p>Import data.</p>
<pre class="r"><code>getwd()</code></pre>
<pre><code>[1] &quot;/home/jdblischak/singlecell-qtl/analysis&quot;</code></pre>
<pre class="r"><code>eset &lt;- readRDS(&quot;../data/eset.rds&quot;)
anno &lt;- pData(eset)</code></pre>
</div>
<div id="total-mapped-reads-reads" class="section level2">
<h2>Total mapped reads reads</h2>
<pre class="r"><code>## calculate the cut-off  
cut_off_reads &lt;- quantile(anno[anno$cell_number == 0,&quot;mapped&quot;], 0.95)

cut_off_reads</code></pre>
<pre><code>    95% 
1015460 </code></pre>
<pre class="r"><code>anno$cut_off_reads &lt;- anno$mapped &gt; cut_off_reads

## numbers of cells 
sum(anno[anno$cell_number == 1, &quot;mapped&quot;] &gt; cut_off_reads)</code></pre>
<pre><code>[1] 5917</code></pre>
<pre class="r"><code>sum(anno[anno$cell_number == 1, &quot;mapped&quot;] &lt;= cut_off_reads)</code></pre>
<pre><code>[1] 1075</code></pre>
<pre class="r"><code>## density plots
plot_reads &lt;- ggplot(anno[anno$cell_number == 0 |
                          anno$cell_number == 1 , ],
       aes(x = mapped, fill = as.factor(cell_number))) + 
       geom_density(alpha = 0.5) +
       geom_vline(xintercept = cut_off_reads, colour=&quot;grey&quot;, linetype = &quot;longdash&quot;) +
       labs(x = &quot;Total mapped reads&quot;, title = &quot;Number of total mapped reads&quot;, fill = &quot;Cell number&quot;)

plot_reads</code></pre>
<p><img src="figure/sampleqc.Rmd/total-reads-1.png" width="768" style="display: block; margin: auto;" /></p>
</div>
<div id="unmapped-ratios" class="section level2">
<h2>Unmapped ratios</h2>
<p>Note: Using the 5% cutoff of samples with no cells excludes all the samples</p>
<pre class="r"><code>## calculate unmapped ratios
anno$unmapped_ratios &lt;- anno$unmapped/anno$umi

## cut off 
cut_off_unmapped &lt;- quantile(anno[anno$cell_number == 0,&quot;unmapped_ratios&quot;], 0.65)

cut_off_unmapped</code></pre>
<pre><code>      65% 
0.3994731 </code></pre>
<pre class="r"><code>anno$cut_off_unmapped &lt;- anno$unmapped_ratios &lt; cut_off_unmapped

## numbers of cells 
sum(anno[anno$cell_number == 1, &quot;unmapped_ratios&quot;] &gt;= cut_off_unmapped)</code></pre>
<pre><code>[1] 588</code></pre>
<pre class="r"><code>sum(anno[anno$cell_number == 1, &quot;unmapped_ratios&quot;] &lt; cut_off_unmapped)</code></pre>
<pre><code>[1] 6404</code></pre>
<pre class="r"><code>## density plots
plot_unmapped &lt;- ggplot(anno[anno$cell_number == 0 |
                             anno$cell_number == 1 , ],
       aes(x = unmapped_ratios *100, fill = as.factor(cell_number))) + 
       geom_density(alpha = 0.5) +
       geom_vline(xintercept = cut_off_unmapped *100, colour=&quot;grey&quot;, linetype = &quot;longdash&quot;) +
       labs(x = &quot;Unmapped reads/ total reads&quot;, title = &quot;Unmapped reads percentage&quot;)

plot_unmapped</code></pre>
<p><img src="figure/sampleqc.Rmd/unmapped-ratios-1.png" width="768" style="display: block; margin: auto;" /></p>
</div>
<div id="ercc-percentage" class="section level2">
<h2>ERCC percentage</h2>
<p>Note: Beacuse not all samples include ERCC, this is not a good cutoff.</p>
<pre class="r"><code>## calculate ercc reads percentage
anno$ercc_percentage &lt;- anno$reads_ercc / anno$mapped

## cut off 
cut_off_ercc &lt;- quantile(anno[anno$cell_number == 0,&quot;ercc_percentage&quot;], 0.25)

cut_off_ercc</code></pre>
<pre><code>      25% 
0.4768398 </code></pre>
<pre class="r"><code>anno$cut_off_ercc &lt;- anno$ercc_percentage &lt; cut_off_ercc

## numbers of cells 
sum(anno[anno$cell_number == 1, &quot;ercc_percentage&quot;] &gt;= cut_off_ercc)</code></pre>
<pre><code>[1] 542</code></pre>
<pre class="r"><code>sum(anno[anno$cell_number == 1, &quot;ercc_percentage&quot;] &lt; cut_off_ercc)</code></pre>
<pre><code>[1] 6450</code></pre>
<pre class="r"><code>## density plots
plot_ercc &lt;- ggplot(anno[anno$cell_number == 0 |
                                anno$cell_number == 1 , ],
       aes(x = ercc_percentage *100, fill = as.factor(cell_number))) + 
       geom_density(alpha = 0.5) +
       geom_vline(xintercept = cut_off_ercc *100, colour=&quot;grey&quot;, linetype = &quot;longdash&quot;) +
       labs(x = &quot;ERCC reads / total mapped reads&quot;, title = &quot;ERCC reads percentage&quot;)

plot_ercc</code></pre>
<p><img src="figure/sampleqc.Rmd/ercc-percentage-1.png" width="768" style="display: block; margin: auto;" /></p>
</div>
<div id="spike-in-percentage" class="section level2">
<h2>Spike-in percentage</h2>
<p>Note: Using the percentage of all the kinds of spike-in as the cutoff. Instead of 5%, 10% seem to be more reasonable due to different amounts of total spike-in.</p>
<pre class="r"><code>## calculate worm and fly reads percentage
anno$spike_percentage &lt;- apply(anno[,19:21],1,sum) / anno$mapped

## cut off 
cut_off_spike &lt;- quantile(anno[anno$cell_number == 0,&quot;spike_percentage&quot;], 0.10)

cut_off_spike</code></pre>
<pre><code>      10% 
0.4880592 </code></pre>
<pre class="r"><code>anno$cut_off_spike &lt;- anno$spike_percentage &lt; cut_off_spike

## numbers of cells 
sum(anno[anno$cell_number == 1, &quot;spike_percentage&quot;] &gt;= cut_off_spike)</code></pre>
<pre><code>[1] 547</code></pre>
<pre class="r"><code>sum(anno[anno$cell_number == 1, &quot;spike_percentage&quot;] &lt; cut_off_spike)</code></pre>
<pre><code>[1] 6445</code></pre>
<pre class="r"><code>## density plots
plot_spike &lt;- ggplot(anno[anno$cell_number == 0 |
                          anno$cell_number == 1 , ],
       aes(x = spike_percentage *100, fill = as.factor(cell_number))) + 
       geom_density(alpha = 0.5) +
       geom_vline(xintercept = cut_off_spike *100, colour=&quot;grey&quot;, linetype = &quot;longdash&quot;) +
       labs(x = &quot;spike-in reads / total mapped reads&quot;, title = &quot;Spike-in reads percentage&quot;)

plot_spike</code></pre>
<p><img src="figure/sampleqc.Rmd/spike-percentage-1.png" width="768" style="display: block; margin: auto;" /></p>
</div>
<div id="number-of-genes-detected" class="section level2">
<h2>Number of genes detected</h2>
<pre class="r"><code>## cut off 
cut_off_genes &lt;- quantile(anno[anno$cell_number == 0,&quot;detect_hs&quot;], 0.90)

cut_off_genes</code></pre>
<pre><code> 90% 
4730 </code></pre>
<pre class="r"><code>anno$cut_off_genes &lt;- anno$detect_hs &gt; cut_off_genes

## numbers of cells 
sum(anno[anno$cell_number == 1, &quot;detect_hs&quot;] &gt; cut_off_genes)</code></pre>
<pre><code>[1] 6192</code></pre>
<pre class="r"><code>sum(anno[anno$cell_number == 1, &quot;detect_hs&quot;] &lt;= cut_off_genes)</code></pre>
<pre><code>[1] 800</code></pre>
<pre class="r"><code>## density plots
plot_gene &lt;- ggplot(anno[anno$cell_number == 0 |
                         anno$cell_number == 1 , ],
       aes(x = detect_hs, fill = as.factor(cell_number))) + 
       geom_density(alpha = 0.5) +
       geom_vline(xintercept = cut_off_genes, colour=&quot;grey&quot;, linetype = &quot;longdash&quot;) +
       labs(x = &quot;Gene numbers&quot;, title = &quot;Numbers of detected genes&quot;)

plot_gene</code></pre>
<p><img src="figure/sampleqc.Rmd/gene-number-1.png" width="768" style="display: block; margin: auto;" /></p>
<pre class="r"><code>plot_grid(plot_reads + theme(legend.position=c(.7,.7)),
          plot_unmapped + theme(legend.position = &quot;none&quot;),
          plot_spike + theme(legend.position = &quot;none&quot;), 
          plot_gene + theme(legend.position = &quot;none&quot;),
          labels = LETTERS[1:4])</code></pre>
<p><img src="figure/sampleqc.Rmd/density%20plots-1.png" width="768" style="display: block; margin: auto;" /></p>
</div>
<div id="mitochondria" class="section level2">
<h2>Mitochondria</h2>
<pre class="r"><code>## create a list of mitochondrial genes (13 protein-coding genes)
## MT-ATP6, MT-CYB, MT-ND1, MT-ND4, MT-ND4L, MT-ND5, MT-ND6, MT-CO2, MT-CO1, MT-ND2, MT-ATP8, MT-CO3, MT-ND3
mtgene &lt;- c(&quot;ENSG00000198899&quot;, &quot;ENSG00000198727&quot;, &quot;ENSG00000198888&quot;, &quot;ENSG00000198886&quot;, &quot;ENSG00000212907&quot;, &quot;ENSG00000198786&quot;, &quot;ENSG00000198695&quot;, &quot;ENSG00000198712&quot;, &quot;ENSG00000198804&quot;, &quot;ENSG00000198763&quot;,&quot;ENSG00000228253&quot;, &quot;ENSG00000198938&quot;, &quot;ENSG00000198840&quot;)

## molecules of mt genes in single cells
eset_mt &lt;- exprs(eset)[mtgene,]
dim(eset_mt)</code></pre>
<pre><code>[1]   13 7584</code></pre>
<pre class="r"><code>## mt ratio of single cell
anno$mt_ratio &lt;- apply(eset_mt, 2, sum) / anno$mol_hs

## mt ratio vs. number of genes detected
ggplot(anno,
       aes(x = detect_hs, y = mt_ratio, 
       color = as.factor(cell_number))) +
       geom_text(aes(label = cell_number)) +
       labs(x = &quot;Number of genes&quot;, y = &quot;Mitochondrial ratio&quot;) +
       scale_fill_manual(values = cbPalette)</code></pre>
<p><img src="figure/sampleqc.Rmd/mito-1.png" width="768" style="display: block; margin: auto;" /></p>
</div>
<div id="linear-discriminat-analysis" class="section level2">
<h2>Linear Discriminat Analysis</h2>
<div id="total-molecule-vs-concentration" class="section level3">
<h3>Total molecule vs concentration</h3>
<pre class="r"><code>library(MASS)
## create 3 groups according to cell number
group_3 &lt;- rep(&quot;two&quot;,dim(anno)[1])
         group_3[grep(&quot;0&quot;, anno$cell_number)] &lt;- &quot;no&quot;
         group_3[grep(&quot;1&quot;, anno$cell_number)] &lt;- &quot;one&quot;

## create data frame
data &lt;- anno %&gt;% dplyr::select(experiment:concentration, mapped, molecules)
data &lt;- data.frame(data, group = group_3)

## perform lda
data_lda &lt;- lda(group ~ concentration + molecules, data = data)
data_lda_p &lt;- predict(data_lda, newdata = data[,c(&quot;concentration&quot;, &quot;molecules&quot;)])$class

## determine how well the model fix
table(data_lda_p, data[, &quot;group&quot;])</code></pre>
<pre><code>          
data_lda_p   no  one  two
       no     0    0    0
       one  284 6940  222
       two    2   52   84</code></pre>
<pre class="r"><code>data$data_lda_p &lt;- data_lda_p

## identify the outlier
outliers_lda &lt;- data %&gt;% rownames_to_column(&quot;sample_id&quot;) %&gt;% filter(cell_number == 1, data_lda_p == &quot;two&quot;)
outliers_lda</code></pre>
<pre><code>      sample_id experiment well batch cell_number concentration  mapped
1  02262018-E06   02262018  E06    b6           1     0.5784361 3236055
2  04052017-F03   04052017  F03    b1           1     0.1168340 1565380
3  08182017-G05   08182017  G05    b2           1     2.8030007 6133739
4  08182017-G10   08182017  G10    b2           1     1.6778928 5551708
5  08212017-B08   08212017  B08    b2           1     0.9928595 3974029
6  08212017-E02   08212017  E02    b2           1     2.0478414 4904929
7  08232017-C01   08232017  C01    b2           1     0.7138795 3386096
8  08232017-C05   08232017  C05    b2           1     1.0386415 3456184
9  08232017-E12   08232017  E12    b2           1     1.0380230 3414477
10 08232017-F01   08232017  F01    b2           1     1.7140052 4878269
11 08232017-G11   08232017  G11    b2           1     1.8179815 5005975
12 08232017-G12   08232017  G12    b2           1     1.6131800 5661160
13 08232017-H02   08232017  H02    b2           1     0.5255853 3892407
14 08242017-D06   08242017  D06    b2           1     1.6715697 4065086
15 08292017-E06   08292017  E06    b2           1     1.8983801 4757861
16 08302017-C09   08302017  C09    b2           1     0.6499613 2745495
17 08302017-D09   08302017  D09    b2           1     1.1915693 3138433
18 08302017-G05   08302017  G05    b2           1     1.6844109 4806186
19 08312017-B11   08312017  B11    b2           1     0.8779793 2706349
20 08312017-D01   08312017  D01    b2           1     0.3565580 2521219
21 08312017-G10   08312017  G10    b2           1     1.5046401 3648699
22 09262017-D03   09262017  D03    b3           1     1.8386027 3246704
23 10052017-E01   10052017  E01    b3           1     2.1426964 3904733
24 10112017-A05   10112017  A05    b3           1     2.0361340 4519609
25 10112017-B05   10112017  B05    b3           1     2.1946481 4228377
26 10112017-D04   10112017  D04    b3           1     2.0726513 4901985
27 10112017-D05   10112017  D05    b3           1     1.4777928 3244370
28 10112017-E05   10112017  E05    b3           1     0.4153881 1764184
29 10112017-F05   10112017  F05    b3           1     2.0907968 4349337
30 10112017-H05   10112017  H05    b3           1     2.5150583 4375495
31 10122017-D05   10122017  D05    b3           1     1.7121228 4257557
32 10162017-C08   10162017  C08    b3           1     1.9654983 4330386
33 10162017-E02   10162017  E02    b3           1     2.2002305 4648097
34 10162017-E05   10162017  E05    b3           1     2.7322329 5107489
35 10162017-F05   10162017  F05    b3           1     2.7496106 5095364
36 10162017-G09   10162017  G09    b3           1     2.4496919 6050854
37 10162017-G12   10162017  G12    b3           1     2.1943500 4662881
38 10162017-H08   10162017  H08    b3           1     2.2897779 4915278
39 10162017-H11   10162017  H11    b3           1     2.6873520 5448556
40 10302017-F05   10302017  F05    b4           1     1.1279243 4723648
41 11072017-C05   11072017  C05    b4           1     2.4462219 5555567
42 11072017-F05   11072017  F05    b4           1     1.9977614 5388259
43 11132017-D06   11132017  D06    b4           1     1.4905986 2920517
44 11212017-C08   11212017  C08    b4           1     1.0897800 3035924
45 11292017-C03   11292017  C03    b5           1     0.7288834 2299046
46 12052017-B12   12052017  B12    b5           1     1.2406718 3332945
47 12072017-D01   12072017  D01    b5           1     0.6185866 5162611
48 12072017-D05   12072017  D05    b5           1     0.5239694 4493960
49 12072017-D06   12072017  D06    b5           1     1.5493517 6508274
50 12072017-D08   12072017  D08    b5           1     1.7749502 7900719
51 12072017-D09   12072017  D09    b5           1     2.5319505 5988340
52 12072017-D10   12072017  D10    b5           1     2.3218666 8180072
   molecules group data_lda_p
1     282339   one        two
2     246422   one        two
3     294206   one        two
4     263658   one        two
5     317905   one        two
6     275159   one        two
7     302296   one        two
8     275188   one        two
9     303952   one        two
10    246552   one        two
11    250950   one        two
12    295622   one        two
13    368903   one        two
14    248966   one        two
15    251939   one        two
16    245208   one        two
17    243178   one        two
18    278500   one        two
19    273096   one        two
20    418731   one        two
21    270603   one        two
22    249735   one        two
23    259726   one        two
24    483817   one        two
25    373489   one        two
26    252845   one        two
27    438320   one        two
28    489482   one        two
29    543761   one        two
30    489848   one        two
31    256558   one        two
32    282929   one        two
33    261361   one        two
34    265542   one        two
35    285487   one        two
36    265230   one        two
37    271810   one        two
38    278148   one        two
39    278988   one        two
40    238854   one        two
41    274113   one        two
42    284304   one        two
43    293984   one        two
44    255066   one        two
45    233171   one        two
46    241212   one        two
47    234332   one        two
48    222125   one        two
49    289249   one        two
50    382545   one        two
51    281521   one        two
52    359559   one        two</code></pre>
<pre class="r"><code>## create filter
anno$molecule_outlier &lt;- row.names(anno) %in% outliers_lda$sample_id

## plot before and after
plot_before &lt;- ggplot(data, aes(x = concentration, y = molecules / 10^3,
               color = as.factor(group))) +
               geom_text(aes(label = cell_number, alpha = 0.5)) +
               labs(x = &quot;Concentration&quot;, y = &quot;Gene molecules (thousands)&quot;, title = &quot;Before&quot;) +
               scale_color_brewer(palette = &quot;Dark2&quot;) +
               theme(legend.position = &quot;none&quot;)


plot_after &lt;- ggplot(data, aes(x = concentration, y = molecules / 10^3,
               color = as.factor(data_lda_p))) +
               geom_text(aes(label = cell_number, alpha = 0.5)) +
               labs(x = &quot;Concentration&quot;, y = &quot;Gene molecules (thousands)&quot;, title = &quot;After&quot;) +
               scale_color_brewer(palette = &quot;Dark2&quot;) +
               theme(legend.position = &quot;none&quot;)

plot_grid(plot_before + theme(legend.position=c(.8,.85)), 
          plot_after + theme(legend.position = &quot;none&quot;),
          labels = LETTERS[1:2])</code></pre>
<p><img src="figure/sampleqc.Rmd/lda-1.png" width="1152" style="display: block; margin: auto;" /></p>
</div>
<div id="reads-to-molecule-conversion" class="section level3">
<h3>Reads to molecule conversion</h3>
<pre class="r"><code>## calculate convertion
anno$ercc_conversion &lt;- anno$mol_ercc / anno$reads_ercc

anno$conversion &lt;- anno$mol_hs / anno$reads_hs

## remove batch1 because not all sample has
anno_ercc &lt;- anno[anno$batch != &quot;b1&quot;, ]
data_ercc &lt;- data[data$batch != &quot;b1&quot;, ] 

## remove batch1 because not all sample has
ggplot(anno_ercc, aes(x = ercc_conversion, y = conversion,
  color = as.factor(cell_number))) +
  geom_text(aes(label = cell_number)) +
  labs(x = &quot;Convertion of ERCC spike-ins&quot;, y = &quot;Conversion of genes&quot;) +
  scale_color_brewer(palette = &quot;Dark2&quot;) +
  theme(legend.position = &quot;none&quot;)</code></pre>
<p><img src="figure/sampleqc.Rmd/convertion-1.png" width="1152" style="display: block; margin: auto;" /></p>
<pre class="r"><code>## try lda
data_ercc$conversion &lt;- anno_ercc$conversion
data_ercc$ercc_conversion &lt;- anno_ercc$ercc_conversion

data_ercc_lda &lt;- lda(group ~ ercc_conversion + conversion, data = data_ercc)

data_ercc_lda_p &lt;- predict(data_ercc_lda,  newdata = data_ercc[,c(&quot;ercc_conversion&quot;, &quot;conversion&quot;)])$class

## determine how well the model fix
table(data_ercc_lda_p, data_ercc[, &quot;group&quot;])</code></pre>
<pre><code>               
data_ercc_lda_p   no  one  two
            no   158  128    2
            one   60 5972  292
            two    0   12    0</code></pre>
<pre class="r"><code>data_ercc$data_ercc_lda_p &lt;- data_ercc_lda_p

## create a cutoff for outliers 
anno$conversion_outlier &lt;- anno$cell_number == 1 &amp; anno$conversion &gt; .4

## plot before and after
plot_ercc_before &lt;- ggplot(data_ercc, aes(x = ercc_conversion, y = conversion,
               color = as.factor(group))) +
               geom_text(aes(label = cell_number, alpha = 0.5)) +
               labs(x = &quot;Convertion of ERCC spike-ins&quot;, y = &quot;Conversion of genes&quot;, title = &quot;Before&quot;) +
               scale_color_brewer(palette = &quot;Dark2&quot;) +
               theme(legend.position = &quot;none&quot;)

plot_ercc_after &lt;- ggplot(data_ercc, aes(x = ercc_conversion, y = conversion,
               color = as.factor(data_ercc_lda_p))) +
               geom_text(aes(label = cell_number, alpha = 0.5)) +
               labs(x = &quot;Convertion of ERCC spike-ins&quot;, y = &quot;Conversion of genes&quot;, title = &quot;After&quot;) +
               scale_color_brewer(palette = &quot;Dark2&quot;) +
               theme(legend.position = &quot;none&quot;)

plot_grid(plot_ercc_before, 
          plot_ercc_after,
          labels = LETTERS[3:4])</code></pre>
<p><img src="figure/sampleqc.Rmd/convertion-2.png" width="1152" style="display: block; margin: auto;" /></p>
</div>
</div>
<div id="filter" class="section level2">
<h2>Filter</h2>
<div id="final-list" class="section level3">
<h3>Final list</h3>
<pre class="r"><code>## all filter
anno$filter_all &lt;- anno$cell_number == 1 &amp;
                   anno$valid_id &amp;
                   anno$cut_off_reads &amp;
                ## anno$cut_off_unmapped &amp;
                ## anno$cut_off_ercc &amp;
                   anno$cut_off_spike &amp;
                   anno$molecule_outlier != &quot;TRUE&quot; &amp;
                   anno$conversion_outlier != &quot;TRUE&quot; &amp;
                   anno$cut_off_genes 
sort(table(anno[anno$filter_all, &quot;chip_id&quot;]))</code></pre>
<pre><code>
NA18498 NA19092 NA19206 NA18856 NA18870 NA18853 NA18862 NA19127 NA19102 
     19      49      65      66      66      70      71      71      73 
NA18907 NA19114 NA19209 NA18912 NA18516 NA19140 NA18852 NA19257 NA19190 
     75      78      79      80      81      82      83      84      85 
NA18855 NA19128 NA18489 NA19108 NA19153 NA19143 NA19160 NA19225 NA19093 
     86      86      87      91      91      93      94      94      95 
NA19099 NA19101 NA18519 NA19210 NA18511 NA18522 NA19144 NA19130 NA19116 
     96      96     100     102     103     104     107     109     111 
NA19152 NA19098 NA19204 NA18517 NA19193 NA19203 NA18505 NA18913 NA18520 
    111     112     113     114     119     120     126     130     131 
NA18499 NA18858 NA19159 NA18859 NA19207 NA18508 NA18502 NA18501 NA18507 
    132     132     133     134     145     154     186     203     281 </code></pre>
<pre class="r"><code>write.table(data.frame(row.names(anno), anno[,&quot;filter_all&quot;]),
            file = &quot;../data/quality-single-cells.txt&quot;, quote = FALSE,
            sep = &quot;\t&quot;, row.names = FALSE, col.names = FALSE)</code></pre>
</div>
<div id="plots" class="section level3">
<h3>Plots</h3>
<pre class="r"><code>genes_unmapped &lt;-  ggplot(anno,
                   aes(x = detect_hs, y = unmapped_ratios * 100,
                       col = as.factor(batch), 
                       label = as.character(cell_number),
                       height = 600, width = 2000)) +
                   scale_colour_manual(values=cbPalette) +
                   geom_text(fontface = 3, alpha = 0.3) + 
                   geom_vline(xintercept = cut_off_genes, 
                              colour=&quot;grey&quot;, linetype = &quot;longdash&quot;) +
                   geom_hline(yintercept = cut_off_unmapped * 100, 
                              colour=&quot;grey&quot;, linetype = &quot;longdash&quot;) +
                   labs(x = &quot;Number of detected genes / sample&quot;, 
                        y = &quot;Percentage of unmapped reads (%)&quot;) 

genes_spike &lt;- ggplot(anno,
               aes(x = detect_hs, y = spike_percentage * 100,
                   col = as.factor(batch), 
                   label = as.character(cell_number), 
                   height = 600, width = 2000)) +
               scale_colour_manual(values=cbPalette) +
               scale_shape_manual(values=c(1:10)) +
               geom_text(fontface = 3, alpha = 0.3) + 
               geom_vline(xintercept = cut_off_genes, 
                          colour=&quot;grey&quot;, linetype = &quot;longdash&quot;) +
               geom_hline(yintercept = cut_off_spike * 100, 
                          colour=&quot;grey&quot;, linetype = &quot;longdash&quot;) +
               labs(x = &quot;Number of detected genes / samlpe&quot;, 
                    y = &quot;Percentage of spike-in reads (%)&quot;) 

reads_unmapped_num &lt;-  ggplot(anno,
                       aes(x = mapped, y = unmapped_ratios * 100,
                           col = as.factor(batch), 
                           label = as.character(cell_number), 
                           height = 600, width = 2000)) +
                       scale_colour_manual(values=cbPalette) +
                       geom_text(fontface = 3, alpha = 0.3) + 
                       geom_vline(xintercept = cut_off_reads, 
                                  colour=&quot;grey&quot;, linetype = &quot;longdash&quot;) +
                       geom_hline(yintercept = cut_off_unmapped * 100,
                                  colour=&quot;grey&quot;, linetype = &quot;longdash&quot;) +
                       labs(x = &quot;Total mapped reads / sample&quot;, 
                            y = &quot;Percentage of unmapped reads (%)&quot;) 

reads_spike_num &lt;- ggplot(anno,
                   aes(x = mapped, y = spike_percentage * 100,
                       col = as.factor(batch), 
                       label = as.character(cell_number), 
                       height = 600, width = 2000)) +
                   scale_colour_manual(values=cbPalette) +
                   geom_text(fontface = 3, alpha = 0.3) + 
                   geom_vline(xintercept = cut_off_reads, 
                              colour=&quot;grey&quot;, linetype = &quot;longdash&quot;) +
                   geom_hline(yintercept = cut_off_spike * 100, 
                              colour=&quot;grey&quot;, linetype = &quot;longdash&quot;) +
                   labs(x = &quot;Total mapped reads / sample&quot;,
                        y = &quot;Percentage of spike-in reads (%)&quot;) 

plot_grid(genes_unmapped + theme(legend.position=c(.7,.9)) + labs(col = &quot;Batch&quot;), 
          genes_spike + theme(legend.position = &quot;none&quot;),
          labels = letters[1:2])</code></pre>
<p><img src="figure/sampleqc.Rmd/plots-1.png" width="3600" style="display: block; margin: auto;" /></p>
<pre class="r"><code>plot_grid(reads_unmapped_num + theme(legend.position = &quot;none&quot;), 
          reads_spike_num + theme(legend.position = &quot;none&quot;),
          labels = letters[3:4])</code></pre>
<p><img src="figure/sampleqc.Rmd/plots-2.png" width="3600" style="display: block; margin: auto;" /></p>
<pre class="r"><code>plot_grid(ggplot(data.frame(anno[anno$filter_all,]),
       aes(x = factor(chip_id), y = conversion,
           fill = factor(batch))) +
  geom_boxplot() +
  scale_fill_manual(values = cbPalette) +
  labs(x = &quot;Individual&quot;, y = &quot;Read-to-molecule conversion efficiency&quot;) +
  theme(axis.text.x = element_text(hjust=1, angle = 90)) + 
  theme(legend.position = &quot;none&quot;),
  labels = &quot;c&quot;) </code></pre>
<p><img src="figure/sampleqc.Rmd/read-molecule%20plot-1.png" width="3600" style="display: block; margin: auto;" /></p>
<pre class="r"><code>plot_grid(reads_spike_num + theme(legend.position = &quot;none&quot;) + theme(legend.position=c(.7,.85)) + labs(col = &quot;Batch&quot;), 
          genes_unmapped + theme(legend.position = &quot;none&quot;),
          labels = letters[1:2])</code></pre>
<p><img src="figure/sampleqc.Rmd/read-molecule%20plot-2.png" width="3600" style="display: block; margin: auto;" /></p>
</div>
</div>
<div id="pca" class="section level2">
<h2>PCA</h2>
<div id="before-filter" class="section level3">
<h3>Before filter</h3>
<p>Select the most variable human genes</p>
<pre class="r"><code>## look at human genes
eset_hs &lt;- eset[fData(eset)$source == &quot;H. sapiens&quot;, ]
head(featureNames(eset_hs))</code></pre>
<pre><code>[1] &quot;ENSG00000000003&quot; &quot;ENSG00000000005&quot; &quot;ENSG00000000419&quot; &quot;ENSG00000000457&quot;
[5] &quot;ENSG00000000460&quot; &quot;ENSG00000000938&quot;</code></pre>
<pre class="r"><code>## remove genes of all 0s
eset_hs_clean &lt;- eset_hs[rowSums(exprs(eset_hs)) != 0, ]
dim(eset_hs_clean)</code></pre>
<pre><code>Features  Samples 
   19832     7584 </code></pre>
<pre class="r"><code>## convert to log2 cpm
mol_hs_cpm &lt;- cpm(exprs(eset_hs_clean), log = TRUE)
mol_hs_cpm_means &lt;- rowMeans(mol_hs_cpm)
summary(mol_hs_cpm_means)</code></pre>
<pre><code>   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
  2.473   2.527   3.103   3.811   4.602  13.049 </code></pre>
<pre class="r"><code>hist(mol_hs_cpm_means)
abline(v = median(mol_hs_cpm_means), col = &quot;red&quot;)</code></pre>
<p><img src="figure/sampleqc.Rmd/select-gene-before-1.png" width="768" style="display: block; margin: auto;" /></p>
<pre class="r"><code>mol_hs_cpm &lt;- mol_hs_cpm[mol_hs_cpm_means &gt; median(mol_hs_cpm_means), ]
dim(mol_hs_cpm)</code></pre>
<pre><code>[1] 9916 7584</code></pre>
<p>Using the genes with reasonable expression levels to perform PCA</p>
<pre class="r"><code>## pca of genes with reasonable expression levels
pca_hs &lt;- run_pca(mol_hs_cpm)

## plot 
plot_pca(pca_hs$PCs, pcx = 1, pcy = 2, explained = pca_hs$explained,
         metadata = pData(eset_hs_clean), color = &quot;batch&quot;)</code></pre>
<p><img src="figure/sampleqc.Rmd/pca-before-1.png" width="768" style="display: block; margin: auto;" /></p>
<pre class="r"><code>plot_pca(pca_hs$PCs, pcx = 1, pcy = 2, explained = pca_hs$explained,
         metadata = pData(eset_hs_clean), color = &quot;cell_number&quot;)</code></pre>
<p><img src="figure/sampleqc.Rmd/pca-before-2.png" width="768" style="display: block; margin: auto;" /></p>
<pre class="r"><code>plot_pca(pca_hs$PCs, pcx = 1, pcy = 2, explained = pca_hs$explained,
         metadata = pData(eset_hs_clean), color = &quot;chip_id&quot;)</code></pre>
<p><img src="figure/sampleqc.Rmd/pca-before-3.png" width="768" style="display: block; margin: auto;" /></p>
<pre class="r"><code>## combine to investigate the effect
pca_anno &lt;- cbind(anno, pca_hs$PCs)

## total mapped vs pc1
pc1_reads &lt;- ggplot(pca_anno, aes(x = mapped, y = PC1)) +
                     geom_text(aes(label = cell_number, 
                                   col = filter_all, alpha = 0.5)) + 
                     scale_colour_manual(values=cbPalette) +
                     geom_smooth()

## unmapped ratio vs pc1
pc1_unmapped &lt;- ggplot(pca_anno, aes(x = unmapped_ratios, y = PC1)) +
                       geom_text(aes(label = cell_number,
                                     col = filter_all, alpha = 0.5)) + 
                       scale_colour_manual(values=cbPalette) + 
                       geom_smooth()

## spike-in ratio vs pc1
pc1_spike &lt;- ggplot(pca_anno, aes(x = spike_percentage, y = PC1)) +
                     geom_text(aes(label = cell_number,
                                   col = filter_all, alpha = 0.5)) + 
                     scale_colour_manual(values=cbPalette) + 
                     geom_smooth()

## number of detected gene vs pc1
pc1_gene &lt;- ggplot(pca_anno, aes(x = detect_hs, y = PC1)) +
                     geom_text(aes(label = cell_number,
                                   col = filter_all, alpha = 0.5)) + 
                     scale_colour_manual(values=cbPalette) + 
                     geom_smooth()

plot_grid(pc1_reads + theme(legend.position=c(.7,.5)),
          pc1_unmapped + theme(legend.position = &quot;none&quot;),
          pc1_spike + theme(legend.position = &quot;none&quot;), 
          pc1_gene + theme(legend.position = &quot;none&quot;),
          labels = LETTERS[1:4])</code></pre>
<pre><code>`geom_smooth()` using method = &#39;gam&#39;
`geom_smooth()` using method = &#39;gam&#39;
`geom_smooth()` using method = &#39;gam&#39;
`geom_smooth()` using method = &#39;gam&#39;</code></pre>
<p><img src="figure/sampleqc.Rmd/pca-before-4.png" width="768" style="display: block; margin: auto;" /></p>
</div>
<div id="after-filter" class="section level3">
<h3>After filter</h3>
<pre class="r"><code>## filter bad cells
eset_hs_clean_filter &lt;- eset_hs_clean[,anno$filter_all]
dim(eset_hs_clean_filter)</code></pre>
<pre><code>Features  Samples 
   19832     5598 </code></pre>
<pre class="r"><code>## convert to log2 cpm
mol_hs_cpm_filter &lt;- cpm(exprs(eset_hs_clean_filter), log = TRUE)
stopifnot(rownames(anno[anno$filter_all,]) == colnames(mol_hs_cpm_filter))
mol_hs_cpm_filter_means &lt;- rowMeans(mol_hs_cpm_filter)
summary(mol_hs_cpm_filter_means)</code></pre>
<pre><code>   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
  2.288   2.353   3.028   3.772   4.712  12.921 </code></pre>
<pre class="r"><code>hist(mol_hs_cpm_filter_means)
abline(v = median(mol_hs_cpm_filter_means), col = &quot;red&quot;)</code></pre>
<p><img src="figure/sampleqc.Rmd/pca-after-1.png" width="768" style="display: block; margin: auto;" /></p>
<pre class="r"><code>mol_hs_cpm_filter &lt;- mol_hs_cpm_filter[mol_hs_cpm_filter_means &gt; median(mol_hs_cpm_filter_means), ]
dim(mol_hs_cpm_filter)</code></pre>
<pre><code>[1] 9916 5598</code></pre>
<pre class="r"><code>## pca of genes with reasonable expression levels
pca_hs_filter &lt;- run_pca(mol_hs_cpm_filter)

plot_pca(pca_hs_filter$PCs, pcx = 1, pcy = 2, explained = pca_hs_filter$explained,
         metadata = pData(eset_hs_clean_filter), color = &quot;batch&quot;)</code></pre>
<p><img src="figure/sampleqc.Rmd/pca-after-2.png" width="768" style="display: block; margin: auto;" /></p>
<pre class="r"><code>## combine to investigate the effect
anno_filter &lt;- anno[anno$filter_all,]
pca_anno_filter &lt;- cbind(anno_filter, pca_hs_filter$PCs)

## total mapped vs pc1
pc1_reads_filter &lt;- ggplot(pca_anno_filter, aes(x = mapped, y = PC1)) +
                     geom_text(aes(label = cell_number, 
                                   alpha = 0.5)) + 
                     scale_colour_manual(values=cbPalette) +
                     geom_smooth()

## unmapped ratio vs pc1
pc1_unmapped_filter &lt;- ggplot(pca_anno_filter, aes(x = unmapped_ratios, y = PC1)) +
                       geom_text(aes(label = cell_number,
                                     alpha = 0.5)) + 
                       scale_colour_manual(values=cbPalette) + 
                       geom_smooth()

## spike-in ratio vs pc1
pc1_spike_filter &lt;- ggplot(pca_anno_filter, aes(x = spike_percentage, y = PC1)) +
                     geom_text(aes(label = cell_number,
                                   alpha = 0.5)) + 
                     scale_colour_manual(values=cbPalette) + 
                     geom_smooth()

## number of detected gene vs pc1
pc1_gene_filter &lt;- ggplot(pca_anno_filter, aes(x = detect_hs, y = PC1)) +
                     geom_text(aes(label = cell_number,
                                   alpha = 0.5)) + 
                     scale_colour_manual(values=cbPalette) + 
                     geom_smooth()

plot_grid(pc1_reads_filter + theme(legend.position=c(.7,.5)),
          pc1_unmapped_filter + theme(legend.position = &quot;none&quot;),
          pc1_spike_filter + theme(legend.position = &quot;none&quot;), 
          pc1_gene_filter + theme(legend.position = &quot;none&quot;),
          labels = LETTERS[1:4])</code></pre>
<pre><code>`geom_smooth()` using method = &#39;gam&#39;
`geom_smooth()` using method = &#39;gam&#39;
`geom_smooth()` using method = &#39;gam&#39;
`geom_smooth()` using method = &#39;gam&#39;</code></pre>
<p><img src="figure/sampleqc.Rmd/pca-after-3.png" width="768" style="display: block; margin: auto;" /></p>
</div>
</div>
<div id="session-information" class="section level2">
<h2>Session information</h2>
<!-- Insert the session information into the document -->
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 3.4.1 (2017-06-30)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Scientific Linux 7.4 (Nitrogen)

Matrix products: default
BLAS: /project2/gilad/jdblischak/miniconda3/envs/scqtl/lib/R/lib/libRblas.so
LAPACK: /project2/gilad/jdblischak/miniconda3/envs/scqtl/lib/R/lib/libRlapack.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] parallel  methods   stats     graphics  grDevices utils     datasets 
[8] base     

other attached packages:
 [1] testit_0.6          bindrcpp_0.2        Biobase_2.38.0     
 [4] BiocGenerics_0.24.0 tidyr_0.7.1         tibble_1.3.3       
 [7] MASS_7.3-45         edgeR_3.20.1        limma_3.34.1       
[10] dplyr_0.7.4         cowplot_0.9.1       ggplot2_2.2.1      

loaded via a namespace (and not attached):
 [1] Rcpp_0.12.13       RColorBrewer_1.1-2 compiler_3.4.1    
 [4] git2r_0.19.0       plyr_1.8.4         bindr_0.1         
 [7] tools_3.4.1        digest_0.6.12      nlme_3.1-131      
[10] evaluate_0.10.1    gtable_0.2.0       lattice_0.20-34   
[13] mgcv_1.8-17        pkgconfig_2.0.1    rlang_0.1.2       
[16] Matrix_1.2-7.1     yaml_2.1.14        stringr_1.2.0     
[19] knitr_1.20         locfit_1.5-9.1     rprojroot_1.2     
[22] grid_3.4.1         glue_1.1.1         R6_2.2.0          
[25] rmarkdown_1.8      purrr_0.2.2        magrittr_1.5      
[28] backports_1.0.5    scales_0.5.0       htmltools_0.3.6   
[31] assertthat_0.1     colorspace_1.3-2   labeling_0.3      
[34] stringi_1.1.2      lazyeval_0.2.0     munsell_0.4.3     </code></pre>
</div>

<hr>
<p>
    This <a href="http://rmarkdown.rstudio.com">R Markdown</a> site was created with <a href="https://github.com/jdblischak/workflowr">workflowr</a>
</p>
<hr>

<!-- To enable disqus, uncomment the section below and provide your disqus_shortname -->

<!-- disqus
  <div id="disqus_thread"></div>
    <script type="text/javascript">
        /* * * CONFIGURATION VARIABLES: EDIT BEFORE PASTING INTO YOUR WEBPAGE * * */
        var disqus_shortname = 'rmarkdown'; // required: replace example with your forum shortname

        /* * * DON'T EDIT BELOW THIS LINE * * */
        (function() {
            var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
            dsq.src = '//' + disqus_shortname + '.disqus.com/embed.js';
            (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
        })();
    </script>
    <noscript>Please enable JavaScript to view the <a href="http://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
    <a href="http://disqus.com" class="dsq-brlink">comments powered by <span class="logo-disqus">Disqus</span></a>
-->


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
