<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />


<meta name="author" content="PoYuan Tung" />

<meta name="date" content="2017-09-13" />

<title>QC of single cell libraries</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-1.1/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-1.1/highlight.js"></script>
<link href="site_libs/font-awesome-4.5.0/css/font-awesome.min.css" rel="stylesheet" />

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs && document.readyState && document.readyState === "complete") {
   window.setTimeout(function() {
      hljs.initHighlighting();
   }, 0);
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>


</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
button.code-folding-btn:focus {
  outline: none;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}

.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>


<div class="container-fluid main-container">

<!-- tabsets -->
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->




<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
  padding-left: 25px;
  text-indent: 0;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>

<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">singlecell-qtl</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Overview</a>
</li>
<li>
  <a href="data-collection.html">Data</a>
</li>
<li>
  <a href="analysis.html">Analysis</a>
</li>
<li>
  <a href="contributing.html">Contributing</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/jdblischak/singlecell-qtl">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">QC of single cell libraries</h1>
<h4 class="author"><em>PoYuan Tung</em></h4>
<h4 class="date"><em>2017-09-13</em></h4>

</div>


<!-- The file analysis/chunks.R contains chunks that define default settings
shared across the workflowr files. -->
<!-- Update knitr chunk options -->
<!-- Insert the date the file was last updated -->
<p><strong>Last updated:</strong> 2018-02-13</p>
<!-- Insert the code version (Git commit SHA1) if Git repository exists and R
 package git2r is installed -->
<p><strong>Code version:</strong> cce60cf</p>
<p>This is for qc of the samples. Based on obsevation under the scope and the sequencing results, samples with bad quality will be removed.</p>
<section id="setup" class="level2">
<h2>Setup</h2>
<pre class="r"><code>library(&quot;cowplot&quot;)
library(&quot;dplyr&quot;)
library(&quot;ggplot2&quot;)
library(&quot;tidyr&quot;)
library(&quot;edgeR&quot;)
theme_set(theme_cowplot())
source(&quot;../code/functions.R&quot;)
library(&quot;Biobase&quot;)</code></pre>
<pre class="r"><code># The palette with grey:
cbPalette &lt;- c(&quot;#999999&quot;, &quot;#E69F00&quot;, &quot;#56B4E9&quot;, &quot;#009E73&quot;, &quot;#F0E442&quot;, &quot;#0072B2&quot;, &quot;#D55E00&quot;, &quot;#CC79A7&quot;)</code></pre>
<p>Import data.</p>
<pre class="r"><code>eset &lt;- readRDS(&quot;../data/eset.rds&quot;)
anno &lt;- pData(eset)</code></pre>
</section>
<section id="total-mapped-reads-reads" class="level2">
<h2>Total mapped reads reads</h2>
<pre class="r"><code>## calculate the cut-off  
cut_off_reads &lt;- quantile(anno[anno$cell_number == 0,&quot;mapped&quot;], 0.95)

cut_off_reads</code></pre>
<pre><code>     95% 
735553.8 </code></pre>
<pre class="r"><code>anno$cut_off_reads &lt;- anno$mapped &gt; cut_off_reads

## numbers of cells 
sum(anno[anno$cell_number == 1, &quot;mapped&quot;] &gt; cut_off_reads)</code></pre>
<pre><code>[1] 4530</code></pre>
<pre class="r"><code>sum(anno[anno$cell_number == 1, &quot;mapped&quot;] &lt;= cut_off_reads)</code></pre>
<pre><code>[1] 1389</code></pre>
<pre class="r"><code>## density plots
plot_reads &lt;- ggplot(anno[anno$cell_number == 0 |
                          anno$cell_number == 1 , ],
       aes(x = mapped, fill = as.factor(cell_number))) + 
       geom_density(alpha = 0.5) +
       geom_vline(xintercept = cut_off_reads, colour=&quot;grey&quot;, linetype = &quot;longdash&quot;) +
       labs(x = &quot;Total mapped reads&quot;, title = &quot;Number of total mapped reads&quot;, fill = &quot;Cell number&quot;)

plot_reads</code></pre>
<p><img src="figure/sampleqc.Rmd/total-reads-1.png" width="768" style="display: block; margin: auto;" /></p>
</section>
<section id="unmapped-ratios" class="level2">
<h2>Unmapped ratios</h2>
<p>Note: Using the 5% cutoff of samples with no cells excludes all the samples</p>
<pre class="r"><code>## calculate unmapped ratios
anno$unmapped_ratios &lt;- anno$unmapped/anno$umi

## cut off 
cut_off_unmapped &lt;- quantile(anno[anno$cell_number == 0,&quot;unmapped_ratios&quot;], 0.05)

cut_off_unmapped</code></pre>
<pre><code>       5% 
0.1952875 </code></pre>
<pre class="r"><code>anno$cut_off_unmapped &lt;- anno$unmapped_ratios &lt; cut_off_unmapped

## numbers of cells 
sum(anno[anno$cell_number == 1, &quot;unmapped_ratios&quot;] &gt;= cut_off_unmapped)</code></pre>
<pre><code>[1] 5891</code></pre>
<pre class="r"><code>sum(anno[anno$cell_number == 1, &quot;unmapped_ratios&quot;] &lt; cut_off_unmapped)</code></pre>
<pre><code>[1] 28</code></pre>
<pre class="r"><code>## density plots
plot_unmapped &lt;- ggplot(anno[anno$cell_number == 0 |
                             anno$cell_number == 1 , ],
       aes(x = unmapped_ratios *100, fill = as.factor(cell_number))) + 
       geom_density(alpha = 0.5) +
       geom_vline(xintercept = cut_off_unmapped *100, colour=&quot;grey&quot;, linetype = &quot;longdash&quot;) +
       labs(x = &quot;Unmapped reads/ total reads&quot;, title = &quot;Unmapped reads percentage&quot;)

plot_unmapped</code></pre>
<p><img src="figure/sampleqc.Rmd/unmapped-ratios-1.png" width="768" style="display: block; margin: auto;" /></p>
</section>
<section id="ercc-percentage" class="level2">
<h2>ERCC percentage</h2>
<p>Note: Beacuse not all samples include ERCC, this is not a good cutoff.</p>
<pre class="r"><code>## calculate ercc reads percentage
anno$ercc_percentage &lt;- anno$reads_ercc / anno$mapped

## cut off 
cut_off_ercc &lt;- quantile(anno[anno$cell_number == 0,&quot;ercc_percentage&quot;], 0.05)

cut_off_ercc</code></pre>
<pre><code>          5% 
2.582295e-06 </code></pre>
<pre class="r"><code>anno$cut_off_ercc &lt;- anno$ercc_percentage &lt; cut_off_ercc

## numbers of cells 
sum(anno[anno$cell_number == 1, &quot;ercc_percentage&quot;] &gt;= cut_off_ercc)</code></pre>
<pre><code>[1] 5772</code></pre>
<pre class="r"><code>sum(anno[anno$cell_number == 1, &quot;ercc_percentage&quot;] &lt; cut_off_ercc)</code></pre>
<pre><code>[1] 147</code></pre>
<pre class="r"><code>## density plots
plot_ercc &lt;- ggplot(anno[anno$cell_number == 0 |
                                anno$cell_number == 1 , ],
       aes(x = ercc_percentage *100, fill = as.factor(cell_number))) + 
       geom_density(alpha = 0.5) +
       geom_vline(xintercept = cut_off_ercc *100, colour=&quot;grey&quot;, linetype = &quot;longdash&quot;) +
       labs(x = &quot;ERCC reads / total mapped reads&quot;, title = &quot;ERCC reads percentage&quot;)

plot_ercc</code></pre>
<p><img src="figure/sampleqc.Rmd/ercc-percentage-1.png" width="768" style="display: block; margin: auto;" /></p>
</section>
<section id="spike-in-percentage" class="level2">
<h2>Spike-in percentage</h2>
<p>Note: Using the percentage of all the kinds of spike-in as the cutoff. Instead of 5%, 10% seem to be more reasonable due to different amounts of total spike-in.</p>
<pre class="r"><code>## calculate worm and fly reads percentage
anno$spike_percentage &lt;- apply(anno[,19:21],1,sum) / anno$mapped

## cut off 
cut_off_spike &lt;- quantile(anno[anno$cell_number == 0,&quot;spike_percentage&quot;], 0.10)

cut_off_spike</code></pre>
<pre><code>      10% 
0.4692639 </code></pre>
<pre class="r"><code>anno$cut_off_spike &lt;- anno$spike_percentage &lt; cut_off_spike

## numbers of cells 
sum(anno[anno$cell_number == 1, &quot;spike_percentage&quot;] &gt;= cut_off_spike)</code></pre>
<pre><code>[1] 467</code></pre>
<pre class="r"><code>sum(anno[anno$cell_number == 1, &quot;spike_percentage&quot;] &lt; cut_off_spike)</code></pre>
<pre><code>[1] 5452</code></pre>
<pre class="r"><code>## density plots
plot_spike &lt;- ggplot(anno[anno$cell_number == 0 |
                          anno$cell_number == 1 , ],
       aes(x = spike_percentage *100, fill = as.factor(cell_number))) + 
       geom_density(alpha = 0.5) +
       geom_vline(xintercept = cut_off_spike *100, colour=&quot;grey&quot;, linetype = &quot;longdash&quot;) +
       labs(x = &quot;spike-in reads / total mapped reads&quot;, title = &quot;Spike-in reads percentage&quot;)

plot_spike</code></pre>
<p><img src="figure/sampleqc.Rmd/spike-percentage-1.png" width="768" style="display: block; margin: auto;" /></p>
</section>
<section id="number-of-genes-detected" class="level2">
<h2>Number of genes detected</h2>
<pre class="r"><code>## cut off 
cut_off_genes &lt;- quantile(anno[anno$cell_number == 0,&quot;detect_hs&quot;], 0.90)

cut_off_genes</code></pre>
<pre><code>   90% 
4052.7 </code></pre>
<pre class="r"><code>anno$cut_off_genes &lt;- anno$detect_hs &gt; cut_off_genes

## numbers of cells 
sum(anno[anno$cell_number == 1, &quot;detect_hs&quot;] &gt; cut_off_genes)</code></pre>
<pre><code>[1] 5171</code></pre>
<pre class="r"><code>sum(anno[anno$cell_number == 1, &quot;detect_hs&quot;] &lt;= cut_off_genes)</code></pre>
<pre><code>[1] 748</code></pre>
<pre class="r"><code>## density plots
plot_gene &lt;- ggplot(anno[anno$cell_number == 0 |
                         anno$cell_number == 1 , ],
       aes(x = detect_hs, fill = as.factor(cell_number))) + 
       geom_density(alpha = 0.5) +
       geom_vline(xintercept = cut_off_genes, colour=&quot;grey&quot;, linetype = &quot;longdash&quot;) +
       labs(x = &quot;Gene numbers&quot;, title = &quot;Numbers of detected genes&quot;)

plot_gene</code></pre>
<p><img src="figure/sampleqc.Rmd/gene-number-1.png" width="768" style="display: block; margin: auto;" /></p>
<pre class="r"><code>plot_grid(plot_reads + theme(legend.position=c(.7,.7)),
          plot_unmapped + theme(legend.position = &quot;none&quot;),
          plot_spike + theme(legend.position = &quot;none&quot;), 
          plot_gene + theme(legend.position = &quot;none&quot;),
          labels = LETTERS[1:4])</code></pre>
<p><img src="figure/sampleqc.Rmd/density%20plots-1.png" width="768" style="display: block; margin: auto;" /></p>
</section>
<section id="linear-discriminat-analysis" class="level2">
<h2>Linear Discriminat Analysis</h2>
<section id="total-molecule-vs-concentration" class="level3">
<h3>Total molecule vs concentration</h3>
<pre class="r"><code>library(MASS)</code></pre>
<pre><code>
Attaching package: &#39;MASS&#39;</code></pre>
<pre><code>The following object is masked from &#39;package:dplyr&#39;:

    select</code></pre>
<pre class="r"><code>## create 3 groups according to cell number
group_3 &lt;- rep(&quot;two&quot;,dim(anno)[1])
         group_3[grep(&quot;0&quot;, anno$cell_number)] &lt;- &quot;no&quot;
         group_3[grep(&quot;1&quot;, anno$cell_number)] &lt;- &quot;one&quot;

## create data frame
data &lt;- anno %&gt;% dplyr::select(experiment:concentration, mapped, molecules)
data &lt;- data.frame(data, group = group_3)

## perform lda
data_lda &lt;- lda(group ~ concentration + molecules, data = data)
data_lda_p &lt;- predict(data_lda, newdata = data[,c(&quot;concentration&quot;, &quot;molecules&quot;)])$class

## determine how well the model fix
table(data_lda_p, data[, &quot;group&quot;])</code></pre>
<pre><code>          
data_lda_p   no  one  two
       no     0    0    0
       one  234 5882  228
       two    0   37   51</code></pre>
<pre class="r"><code>data$data_lda_p &lt;- data_lda_p


## plot before and after
plot_before &lt;- ggplot(data, aes(x = concentration, y = molecules / 10^3,
               color = as.factor(group))) +
               geom_text(aes(label = cell_number, alpha = 0.5)) +
               labs(x = &quot;Concentration&quot;, y = &quot;Gene molecules (thousands)&quot;, title = &quot;Before&quot;) +
               scale_color_brewer(palette = &quot;Dark2&quot;) +
               theme(legend.position = &quot;none&quot;)


plot_after &lt;- ggplot(data, aes(x = concentration, y = molecules / 10^3,
               color = as.factor(data_lda_p))) +
               geom_text(aes(label = cell_number, alpha = 0.5)) +
               labs(x = &quot;Concentration&quot;, y = &quot;Gene molecules (thousands)&quot;, title = &quot;After&quot;) +
               scale_color_brewer(palette = &quot;Dark2&quot;) +
               theme(legend.position = &quot;none&quot;)

plot_grid(plot_before + theme(legend.position=c(.8,.85)), 
          plot_after + theme(legend.position = &quot;none&quot;),
          labels = LETTERS[1:2])</code></pre>
<p><img src="figure/sampleqc.Rmd/lda-1.png" width="1152" style="display: block; margin: auto;" /></p>
</section>
<section id="reads-to-molecule-conversion" class="level3">
<h3>Reads to molecule conversion</h3>
<pre class="r"><code>## calculate convertion
anno$dm_conversion &lt;- anno$mol_dm / anno$reads_dm

anno$conversion &lt;- anno$mol_hs / anno$reads_hs

ggplot(anno, aes(x = dm_conversion, y = conversion,
  color = as.factor(cell_number))) +
  geom_text(aes(label = cell_number)) +
  labs(x = &quot;Convertion of Fly spike-ins&quot;, y = &quot;Conversion of genes&quot;) +
  scale_color_brewer(palette = &quot;Dark2&quot;) +
  theme(legend.position = &quot;none&quot;)</code></pre>
<pre><code>Warning: Removed 4 rows containing missing values (geom_text).</code></pre>
<p><img src="figure/sampleqc.Rmd/convertion-1.png" width="1152" style="display: block; margin: auto;" /></p>
<pre class="r"><code>## try lda
data$conversion &lt;- anno$conversion
data$dm_conversion &lt;- anno$dm_conversion

data_ercc_lda &lt;- lda(group ~ dm_conversion + conversion, data = data)

data_ercc_lda_p &lt;- predict(data_ercc_lda,  newdata = data[,c(&quot;dm_conversion&quot;, &quot;conversion&quot;)])$class</code></pre>
<pre><code>Warning in FUN(newX[, i], ...): no non-missing arguments to min; returning
Inf</code></pre>
<pre><code>Warning in FUN(newX[, i], ...): no non-missing arguments to min; returning
Inf

Warning in FUN(newX[, i], ...): no non-missing arguments to min; returning
Inf

Warning in FUN(newX[, i], ...): no non-missing arguments to min; returning
Inf</code></pre>
<pre class="r"><code>## determine how well the model fix
table(data_ercc_lda_p, data[, &quot;group&quot;])</code></pre>
<pre><code>               
data_ercc_lda_p   no  one  two
            no   137  108    5
            one   97 5807  274
            two    0    0    0</code></pre>
<pre class="r"><code>data$data_ercc_lda_p &lt;- data_ercc_lda_p


## cutoff
out_ercc_con &lt;- anno %&gt;% filter(cell_number == &quot;1&quot;, dm_conversion &gt; .094)

anno$conversion_outlier &lt;- anno$cell_number == 1 &amp; anno$dm_conversion &gt; .094

## plot before and after
plot_ercc_before &lt;- ggplot(data, aes(x = dm_conversion, y = conversion,
               color = as.factor(group))) +
               geom_text(aes(label = cell_number, alpha = 0.5)) +
               labs(x = &quot;Convertion of Fly spike-ins&quot;, y = &quot;Conversion of genes&quot;, title = &quot;Before&quot;) +
               scale_color_brewer(palette = &quot;Dark2&quot;) +
               theme(legend.position = &quot;none&quot;)

plot_ercc_after &lt;- ggplot(data, aes(x = dm_conversion, y = conversion,
               color = as.factor(data_ercc_lda_p))) +
               geom_text(aes(label = cell_number, alpha = 0.5)) +
               labs(x = &quot;Convertion of Fly spike-ins&quot;, y = &quot;Conversion of genes&quot;, title = &quot;After&quot;) +
               scale_color_brewer(palette = &quot;Dark2&quot;) +
               theme(legend.position = &quot;none&quot;)

plot_grid(plot_ercc_before, 
          plot_ercc_after,
          labels = LETTERS[3:4])</code></pre>
<pre><code>Warning: Removed 4 rows containing missing values (geom_text).</code></pre>
<pre><code>Warning: Removed 4 rows containing missing values (geom_text).</code></pre>
<p><img src="figure/sampleqc.Rmd/convertion-2.png" width="1152" style="display: block; margin: auto;" /></p>
</section>
</section>
<section id="filter" class="level2">
<h2>Filter</h2>
<section id="final-list" class="level3">
<h3>Final list</h3>
<pre class="r"><code>## all filter
anno$filter_all &lt;- anno$cell_number == 1 &amp;
                   anno$valid_id &amp;
                ##   anno$cut_off_reads &amp;
                ##   anno$cut_off_unmapped &amp;
                ##   anno$cut_off_ercc &amp;
                   anno$cut_off_spike &amp;
                   anno$cut_off_genes 
sort(table(anno[anno$filter_all, &quot;chip_id&quot;]))</code></pre>
<pre><code>
NA18498 NA19203 NA19159 NA18913 NA19204 NA19092 NA18505 NA19193 NA18519 
     19      24      31      49      49      51      53      56      62 
NA19143 NA18522 NA18859 NA19098 NA19130 NA18517 NA19210 NA19206 NA18862 
     64      65      65      65      65      67      69      72      73 
NA19102 NA19127 NA18870 NA18856 NA18907 NA18853 NA18489 NA18912 NA19114 
     73      73      74      76      77      79      82      82      82 
NA19140 NA18852 NA19209 NA19108 NA19153 NA19257 NA18855 NA19225 NA18516 
     82      85      86      88      89      89      90      91      92 
NA19101 NA19190 NA19099 NA19093 NA19116 NA19160 NA19128 NA19144 NA18511 
     94      96      98     102     106     107     108     111     112 
NA19152 NA18520 NA18499 NA19207 NA18858 NA18508 NA18502 NA18501 NA18507 
    119     136     138     150     153     168     194     198     306 </code></pre>
<pre class="r"><code>write.table(data.frame(row.names(anno), anno[,&quot;filter_all&quot;]),
            file = &quot;../data/quality-single-cells.txt&quot;, quote = FALSE,
            sep = &quot;\t&quot;, row.names = FALSE, col.names = FALSE)</code></pre>
</section>
<section id="plots" class="level3">
<h3>Plots</h3>
<pre class="r"><code>genes_unmapped &lt;-  ggplot(anno,
                   aes(x = detect_hs, y = unmapped_ratios * 100,
                       col = as.factor(valid_id), 
                       label = as.character(cell_number),
                       height = 600, width = 2000)) +
                   scale_colour_manual(values=cbPalette) +
                   geom_text(fontface = 3, alpha = 0.5) + 
                   geom_vline(xintercept = cut_off_genes, 
                              colour=&quot;grey&quot;, linetype = &quot;longdash&quot;) +
                   geom_hline(yintercept = cut_off_unmapped * 100, 
                              colour=&quot;grey&quot;, linetype = &quot;longdash&quot;) +
                   labs(x = &quot;Number of detected genes / sample&quot;, 
                        y = &quot;Percentage of unmapped reads (%)&quot;) 

genes_spike &lt;- ggplot(anno,
               aes(x = detect_hs, y = spike_percentage * 100,
                   col = as.factor(valid_id), 
                   label = as.character(cell_number), 
                   height = 600, width = 2000)) +
               scale_colour_manual(values=cbPalette) +
               scale_shape_manual(values=c(1:10)) +
               geom_text(fontface = 3, alpha = 0.5) + 
               geom_vline(xintercept = cut_off_genes, 
                          colour=&quot;grey&quot;, linetype = &quot;longdash&quot;) +
               geom_hline(yintercept = cut_off_spike * 100, 
                          colour=&quot;grey&quot;, linetype = &quot;longdash&quot;) +
               labs(x = &quot;Number of detected genes / samlpe&quot;, 
                    y = &quot;Percentage of spike-in reads (%)&quot;) 

reads_unmapped_num &lt;-  ggplot(anno,
                       aes(x = mapped, y = unmapped_ratios * 100,
                           col = as.factor(chip_id), 
                           label = as.character(cell_number), 
                           height = 600, width = 2000)) +
                       geom_text(fontface = 3, alpha = 0.5) + 
                       geom_vline(xintercept = cut_off_reads, 
                                  colour=&quot;grey&quot;, linetype = &quot;longdash&quot;) +
                       geom_hline(yintercept = cut_off_unmapped * 100,
                                  colour=&quot;grey&quot;, linetype = &quot;longdash&quot;) +
                       labs(x = &quot;Total mapped reads / sample&quot;, 
                            y = &quot;Percentage of unmapped reads (%)&quot;) 

reads_spike_num &lt;- ggplot(anno,
                   aes(x = mapped, y = spike_percentage * 100,
                       col = as.factor(chip_id), 
                       label = as.character(cell_number), 
                       height = 600, width = 2000)) +
                   geom_text(fontface = 3, alpha = 0.5) + 
                   geom_vline(xintercept = cut_off_reads, 
                              colour=&quot;grey&quot;, linetype = &quot;longdash&quot;) +
                   geom_hline(yintercept = cut_off_spike * 100, 
                              colour=&quot;grey&quot;, linetype = &quot;longdash&quot;) +
                   labs(x = &quot;Total mapped reads / sample&quot;,
                        y = &quot;Percentage of spike-in reads (%)&quot;) 

plot_grid(genes_unmapped + theme(legend.position=c(.7,.9)), 
          genes_spike + theme(legend.position = &quot;none&quot;),
          labels = letters[1:2])</code></pre>
<p><img src="figure/sampleqc.Rmd/plots-1.png" width="3600" style="display: block; margin: auto;" /></p>
<pre class="r"><code>plot_grid(reads_unmapped_num + theme(legend.position = &quot;none&quot;), 
          reads_spike_num + theme(legend.position = &quot;none&quot;),
          labels = letters[3:4])</code></pre>
<p><img src="figure/sampleqc.Rmd/plots-2.png" width="3600" style="display: block; margin: auto;" /></p>
</section>
</section>
<section id="pca" class="level2">
<h2>PCA</h2>
<section id="before-filter" class="level3">
<h3>Before filter</h3>
<p>Select the most variable human genes</p>
<pre class="r"><code>## look at human genes
eset_hs &lt;- eset[fData(eset)$source == &quot;H. sapiens&quot;, ]
head(featureNames(eset_hs))</code></pre>
<pre><code>[1] &quot;ENSG00000000003&quot; &quot;ENSG00000000005&quot; &quot;ENSG00000000419&quot; &quot;ENSG00000000457&quot;
[5] &quot;ENSG00000000460&quot; &quot;ENSG00000000938&quot;</code></pre>
<pre class="r"><code>## remove genes of all 0s
eset_hs_clean &lt;- eset_hs[rowSums(exprs(eset_hs)) != 0, ]
dim(eset_hs_clean)</code></pre>
<pre><code>Features  Samples 
   19753     6432 </code></pre>
<pre class="r"><code>## convert to log2 cpm
mol_hs_cpm &lt;- cpm(exprs(eset_hs_clean), log = TRUE)
mol_hs_cpm_means &lt;- rowMeans(mol_hs_cpm)
summary(mol_hs_cpm_means)</code></pre>
<pre><code>   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
  2.750   2.793   3.282   3.945   4.603  13.075 </code></pre>
<pre class="r"><code>hist(mol_hs_cpm_means)
abline(v = median(mol_hs_cpm_means), col = &quot;red&quot;)</code></pre>
<p><img src="figure/sampleqc.Rmd/select-gene-before-1.png" width="768" style="display: block; margin: auto;" /></p>
<pre class="r"><code>mol_hs_cpm &lt;- mol_hs_cpm[mol_hs_cpm_means &gt; median(mol_hs_cpm_means), ]
dim(mol_hs_cpm)</code></pre>
<pre><code>[1] 9876 6432</code></pre>
<p>Using the genes with reasonable expression levels to perform PCA</p>
<pre class="r"><code>## pca of genes with reasonable expression levels
pca_hs &lt;- run_pca(mol_hs_cpm)

## plot 
plot_pca(pca_hs$PCs, pcx = 1, pcy = 2, explained = pca_hs$explained,
         metadata = pData(eset_hs_clean), color = &quot;batch&quot;)</code></pre>
<p><img src="figure/sampleqc.Rmd/pca-before-1.png" width="768" style="display: block; margin: auto;" /></p>
<pre class="r"><code>plot_pca(pca_hs$PCs, pcx = 1, pcy = 2, explained = pca_hs$explained,
         metadata = pData(eset_hs_clean), color = &quot;cell_number&quot;)</code></pre>
<p><img src="figure/sampleqc.Rmd/pca-before-2.png" width="768" style="display: block; margin: auto;" /></p>
<pre class="r"><code>plot_pca(pca_hs$PCs, pcx = 1, pcy = 2, explained = pca_hs$explained,
         metadata = pData(eset_hs_clean), color = &quot;chip_id&quot;)</code></pre>
<p><img src="figure/sampleqc.Rmd/pca-before-3.png" width="768" style="display: block; margin: auto;" /></p>
<pre class="r"><code>## combine to investigate the effect
pca_anno &lt;- cbind(anno, pca_hs$PCs)

## total mapped vs pc1
pc1_reads &lt;- ggplot(pca_anno, aes(x = mapped, y = PC1)) +
                     geom_text(aes(label = cell_number, 
                                   col = filter_all, alpha = 0.5)) + 
                     scale_colour_manual(values=cbPalette) +
                     geom_smooth()

## unmapped ratio vs pc1
pc1_unmapped &lt;- ggplot(pca_anno, aes(x = unmapped_ratios, y = PC1)) +
                       geom_text(aes(label = cell_number,
                                     col = filter_all, alpha = 0.5)) + 
                       scale_colour_manual(values=cbPalette) + 
                       geom_smooth()

## spike-in ratio vs pc1
pc1_spike &lt;- ggplot(pca_anno, aes(x = spike_percentage, y = PC1)) +
                     geom_text(aes(label = cell_number,
                                   col = filter_all, alpha = 0.5)) + 
                     scale_colour_manual(values=cbPalette) + 
                     geom_smooth()

## number of detected gene vs pc1
pc1_gene &lt;- ggplot(pca_anno, aes(x = detect_hs, y = PC1)) +
                     geom_text(aes(label = cell_number,
                                   col = filter_all, alpha = 0.5)) + 
                     scale_colour_manual(values=cbPalette) + 
                     geom_smooth()

plot_grid(pc1_reads + theme(legend.position=c(.7,.5)),
          pc1_unmapped + theme(legend.position = &quot;none&quot;),
          pc1_spike + theme(legend.position = &quot;none&quot;), 
          pc1_gene + theme(legend.position = &quot;none&quot;),
          labels = LETTERS[1:4])</code></pre>
<pre><code>`geom_smooth()` using method = &#39;gam&#39;
`geom_smooth()` using method = &#39;gam&#39;
`geom_smooth()` using method = &#39;gam&#39;
`geom_smooth()` using method = &#39;gam&#39;</code></pre>
<p><img src="figure/sampleqc.Rmd/pca-before-4.png" width="768" style="display: block; margin: auto;" /></p>
</section>
<section id="after-filter" class="level3">
<h3>After filter</h3>
<pre class="r"><code>## filter bad cells
eset_hs_clean_filter &lt;- eset_hs_clean[,anno$filter_all]
dim(eset_hs_clean_filter)</code></pre>
<pre><code>Features  Samples 
   19753     4985 </code></pre>
<pre class="r"><code>## convert to log2 cpm
mol_hs_cpm_filter &lt;- cpm(exprs(eset_hs_clean_filter), log = TRUE)
stopifnot(rownames(anno[anno$filter_all,]) == colnames(mol_hs_cpm_filter))
mol_hs_cpm_filter_means &lt;- rowMeans(mol_hs_cpm_filter)
summary(mol_hs_cpm_filter_means)</code></pre>
<pre><code>   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
  2.576   2.626   3.192   3.904   4.687  12.949 </code></pre>
<pre class="r"><code>hist(mol_hs_cpm_filter_means)
abline(v = median(mol_hs_cpm_filter_means), col = &quot;red&quot;)</code></pre>
<p><img src="figure/sampleqc.Rmd/pca-after-1.png" width="768" style="display: block; margin: auto;" /></p>
<pre class="r"><code>mol_hs_cpm_filter &lt;- mol_hs_cpm_filter[mol_hs_cpm_filter_means &gt; median(mol_hs_cpm_filter_means), ]
dim(mol_hs_cpm_filter)</code></pre>
<pre><code>[1] 9876 4985</code></pre>
<pre class="r"><code>## pca of genes with reasonable expression levels
pca_hs_filter &lt;- run_pca(mol_hs_cpm_filter)

plot_pca(pca_hs_filter$PCs, pcx = 1, pcy = 2, explained = pca_hs_filter$explained,
         metadata = pData(eset_hs_clean_filter), color = &quot;batch&quot;)</code></pre>
<p><img src="figure/sampleqc.Rmd/pca-after-2.png" width="768" style="display: block; margin: auto;" /></p>
<pre class="r"><code>## combine to investigate the effect
anno_filter &lt;- anno[anno$filter_all,]
pca_anno_filter &lt;- cbind(anno_filter, pca_hs_filter$PCs)

## total mapped vs pc1
pc1_reads_filter &lt;- ggplot(pca_anno_filter, aes(x = mapped, y = PC1)) +
                     geom_text(aes(label = cell_number, 
                                   alpha = 0.5)) + 
                     scale_colour_manual(values=cbPalette) +
                     geom_smooth()

## unmapped ratio vs pc1
pc1_unmapped_filter &lt;- ggplot(pca_anno_filter, aes(x = unmapped_ratios, y = PC1)) +
                       geom_text(aes(label = cell_number,
                                     alpha = 0.5)) + 
                       scale_colour_manual(values=cbPalette) + 
                       geom_smooth()

## spike-in ratio vs pc1
pc1_spike_filter &lt;- ggplot(pca_anno_filter, aes(x = spike_percentage, y = PC1)) +
                     geom_text(aes(label = cell_number,
                                   alpha = 0.5)) + 
                     scale_colour_manual(values=cbPalette) + 
                     geom_smooth()

## number of detected gene vs pc1
pc1_gene_filter &lt;- ggplot(pca_anno_filter, aes(x = detect_hs, y = PC1)) +
                     geom_text(aes(label = cell_number,
                                   alpha = 0.5)) + 
                     scale_colour_manual(values=cbPalette) + 
                     geom_smooth()

plot_grid(pc1_reads_filter + theme(legend.position=c(.7,.5)),
          pc1_unmapped_filter + theme(legend.position = &quot;none&quot;),
          pc1_spike_filter + theme(legend.position = &quot;none&quot;), 
          pc1_gene_filter + theme(legend.position = &quot;none&quot;),
          labels = LETTERS[1:4])</code></pre>
<pre><code>`geom_smooth()` using method = &#39;gam&#39;
`geom_smooth()` using method = &#39;gam&#39;
`geom_smooth()` using method = &#39;gam&#39;
`geom_smooth()` using method = &#39;gam&#39;</code></pre>
<p><img src="figure/sampleqc.Rmd/pca-after-3.png" width="768" style="display: block; margin: auto;" /></p>
</section>
</section>
<section id="session-information" class="level2">
<h2>Session information</h2>
<!-- Insert the session information into the document -->
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 3.4.1 (2017-06-30)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Scientific Linux 7.2 (Nitrogen)

Matrix products: default
BLAS: /project2/gilad/jdblischak/miniconda3/envs/scqtl/lib/R/lib/libRblas.so
LAPACK: /project2/gilad/jdblischak/miniconda3/envs/scqtl/lib/R/lib/libRlapack.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] parallel  methods   stats     graphics  grDevices utils     datasets 
[8] base     

other attached packages:
 [1] testit_0.6          bindrcpp_0.2        MASS_7.3-45        
 [4] Biobase_2.38.0      BiocGenerics_0.24.0 edgeR_3.20.1       
 [7] limma_3.34.1        tidyr_0.7.1         dplyr_0.7.4        
[10] cowplot_0.9.1       ggplot2_2.2.1      

loaded via a namespace (and not attached):
 [1] Rcpp_0.12.13       RColorBrewer_1.1-2 compiler_3.4.1    
 [4] git2r_0.19.0       plyr_1.8.4         bindr_0.1         
 [7] tools_3.4.1        digest_0.6.12      nlme_3.1-131      
[10] evaluate_0.10.1    tibble_1.3.3       gtable_0.2.0      
[13] lattice_0.20-34    mgcv_1.8-17        pkgconfig_2.0.1   
[16] rlang_0.1.2        Matrix_1.2-7.1     yaml_2.1.14       
[19] stringr_1.2.0      knitr_1.16         locfit_1.5-9.1    
[22] rprojroot_1.2      grid_3.4.1         glue_1.1.1        
[25] R6_2.2.0           rmarkdown_1.6      purrr_0.2.2       
[28] magrittr_1.5       backports_1.0.5    scales_0.5.0      
[31] htmltools_0.3.6    assertthat_0.1     colorspace_1.3-2  
[34] labeling_0.3       stringi_1.1.2      lazyeval_0.2.0    
[37] munsell_0.4.3     </code></pre>
</section>

<hr>
<p>
    This <a href="http://rmarkdown.rstudio.com">R Markdown</a> site was created with <a href="https://github.com/jdblischak/workflowr">workflowr</a>
</p>
<hr>

<!-- To enable disqus, uncomment the section below and provide your disqus_shortname -->

<!-- disqus
  <div id="disqus_thread"></div>
    <script type="text/javascript">
        /* * * CONFIGURATION VARIABLES: EDIT BEFORE PASTING INTO YOUR WEBPAGE * * */
        var disqus_shortname = 'rmarkdown'; // required: replace example with your forum shortname

        /* * * DON'T EDIT BELOW THIS LINE * * */
        (function() {
            var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
            dsq.src = '//' + disqus_shortname + '.disqus.com/embed.js';
            (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
        })();
    </script>
    <noscript>Please enable JavaScript to view the <a href="http://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
    <a href="http://disqus.com" class="dsq-brlink">comments powered by <span class="logo-disqus">Disqus</span></a>
-->


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
