<!DOCTYPE html>
<html lang="en">
<head>
<!-- 2018-03-16 Fri 12:51 -->
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
<title>Single cell/bulk RNA-Seq concordance</title>
<meta name="generator" content="Org mode">
<meta name="author" content="Abhishek Sarkar">
<style type="text/css">
 <!--/*--><![CDATA[/*><!--*/
  .title  { text-align: center;
             margin-bottom: .2em; }
  .subtitle { text-align: center;
              font-size: medium;
              font-weight: bold;
              margin-top:0; }
  .todo   { font-family: monospace; color: red; }
  .done   { font-family: monospace; color: green; }
  .priority { font-family: monospace; color: orange; }
  .tag    { background-color: #eee; font-family: monospace;
            padding: 2px; font-size: 80%; font-weight: normal; }
  .timestamp { color: #bebebe; }
  .timestamp-kwd { color: #5f9ea0; }
  .org-right  { margin-left: auto; margin-right: 0px;  text-align: right; }
  .org-left   { margin-left: 0px;  margin-right: auto; text-align: left; }
  .org-center { margin-left: auto; margin-right: auto; text-align: center; }
  .underline { text-decoration: underline; }
  #postamble p, #preamble p { font-size: 90%; margin: .2em; }
  p.verse { margin-left: 3%; }
  pre {
    border: 1px solid #ccc;
    box-shadow: 3px 3px 3px #eee;
    padding: 8pt;
    font-family: monospace;
    overflow: auto;
    margin: 1.2em;
  }
  pre.src {
    position: relative;
    overflow: visible;
    padding-top: 1.2em;
  }
  pre.src:before {
    display: none;
    position: absolute;
    background-color: white;
    top: -10px;
    right: 10px;
    padding: 3px;
    border: 1px solid black;
  }
  pre.src:hover:before { display: inline;}
  /* Languages per Org manual */
  pre.src-asymptote:before { content: 'Asymptote'; }
  pre.src-awk:before { content: 'Awk'; }
  pre.src-C:before { content: 'C'; }
  /* pre.src-C++ doesn't work in CSS */
  pre.src-clojure:before { content: 'Clojure'; }
  pre.src-css:before { content: 'CSS'; }
  pre.src-D:before { content: 'D'; }
  pre.src-ditaa:before { content: 'ditaa'; }
  pre.src-dot:before { content: 'Graphviz'; }
  pre.src-calc:before { content: 'Emacs Calc'; }
  pre.src-emacs-lisp:before { content: 'Emacs Lisp'; }
  pre.src-fortran:before { content: 'Fortran'; }
  pre.src-gnuplot:before { content: 'gnuplot'; }
  pre.src-haskell:before { content: 'Haskell'; }
  pre.src-hledger:before { content: 'hledger'; }
  pre.src-java:before { content: 'Java'; }
  pre.src-js:before { content: 'Javascript'; }
  pre.src-latex:before { content: 'LaTeX'; }
  pre.src-ledger:before { content: 'Ledger'; }
  pre.src-lisp:before { content: 'Lisp'; }
  pre.src-lilypond:before { content: 'Lilypond'; }
  pre.src-lua:before { content: 'Lua'; }
  pre.src-matlab:before { content: 'MATLAB'; }
  pre.src-mscgen:before { content: 'Mscgen'; }
  pre.src-ocaml:before { content: 'Objective Caml'; }
  pre.src-octave:before { content: 'Octave'; }
  pre.src-org:before { content: 'Org mode'; }
  pre.src-oz:before { content: 'OZ'; }
  pre.src-plantuml:before { content: 'Plantuml'; }
  pre.src-processing:before { content: 'Processing.js'; }
  pre.src-python:before { content: 'Python'; }
  pre.src-R:before { content: 'R'; }
  pre.src-ruby:before { content: 'Ruby'; }
  pre.src-sass:before { content: 'Sass'; }
  pre.src-scheme:before { content: 'Scheme'; }
  pre.src-screen:before { content: 'Gnu Screen'; }
  pre.src-sed:before { content: 'Sed'; }
  pre.src-sh:before { content: 'shell'; }
  pre.src-sql:before { content: 'SQL'; }
  pre.src-sqlite:before { content: 'SQLite'; }
  /* additional languages in org.el's org-babel-load-languages alist */
  pre.src-forth:before { content: 'Forth'; }
  pre.src-io:before { content: 'IO'; }
  pre.src-J:before { content: 'J'; }
  pre.src-makefile:before { content: 'Makefile'; }
  pre.src-maxima:before { content: 'Maxima'; }
  pre.src-perl:before { content: 'Perl'; }
  pre.src-picolisp:before { content: 'Pico Lisp'; }
  pre.src-scala:before { content: 'Scala'; }
  pre.src-shell:before { content: 'Shell Script'; }
  pre.src-ebnf2ps:before { content: 'ebfn2ps'; }
  /* additional language identifiers per "defun org-babel-execute"
       in ob-*.el */
  pre.src-cpp:before  { content: 'C++'; }
  pre.src-abc:before  { content: 'ABC'; }
  pre.src-coq:before  { content: 'Coq'; }
  pre.src-groovy:before  { content: 'Groovy'; }
  /* additional language identifiers from org-babel-shell-names in
     ob-shell.el: ob-shell is the only babel language using a lambda to put
     the execution function name together. */
  pre.src-bash:before  { content: 'bash'; }
  pre.src-csh:before  { content: 'csh'; }
  pre.src-ash:before  { content: 'ash'; }
  pre.src-dash:before  { content: 'dash'; }
  pre.src-ksh:before  { content: 'ksh'; }
  pre.src-mksh:before  { content: 'mksh'; }
  pre.src-posh:before  { content: 'posh'; }
  /* Additional Emacs modes also supported by the LaTeX listings package */
  pre.src-ada:before { content: 'Ada'; }
  pre.src-asm:before { content: 'Assembler'; }
  pre.src-caml:before { content: 'Caml'; }
  pre.src-delphi:before { content: 'Delphi'; }
  pre.src-html:before { content: 'HTML'; }
  pre.src-idl:before { content: 'IDL'; }
  pre.src-mercury:before { content: 'Mercury'; }
  pre.src-metapost:before { content: 'MetaPost'; }
  pre.src-modula-2:before { content: 'Modula-2'; }
  pre.src-pascal:before { content: 'Pascal'; }
  pre.src-ps:before { content: 'PostScript'; }
  pre.src-prolog:before { content: 'Prolog'; }
  pre.src-simula:before { content: 'Simula'; }
  pre.src-tcl:before { content: 'tcl'; }
  pre.src-tex:before { content: 'TeX'; }
  pre.src-plain-tex:before { content: 'Plain TeX'; }
  pre.src-verilog:before { content: 'Verilog'; }
  pre.src-vhdl:before { content: 'VHDL'; }
  pre.src-xml:before { content: 'XML'; }
  pre.src-nxml:before { content: 'XML'; }
  /* add a generic configuration mode; LaTeX export needs an additional
     (add-to-list 'org-latex-listings-langs '(conf " ")) in .emacs */
  pre.src-conf:before { content: 'Configuration File'; }

  table { border-collapse:collapse; }
  caption.t-above { caption-side: top; }
  caption.t-bottom { caption-side: bottom; }
  td, th { vertical-align:top;  }
  th.org-right  { text-align: center;  }
  th.org-left   { text-align: center;   }
  th.org-center { text-align: center; }
  td.org-right  { text-align: right;  }
  td.org-left   { text-align: left;   }
  td.org-center { text-align: center; }
  dt { font-weight: bold; }
  .footpara { display: inline; }
  .footdef  { margin-bottom: 1em; }
  .figure { padding: 1em; }
  .figure p { text-align: center; }
  .inlinetask {
    padding: 10px;
    border: 2px solid gray;
    margin: 10px;
    background: #ffffcc;
  }
  #org-div-home-and-up
   { text-align: right; font-size: 70%; white-space: nowrap; }
  textarea { overflow-x: auto; }
  .linenr { font-size: smaller }
  .code-highlighted { background-color: #ffff00; }
  .org-info-js_info-navigation { border-style: none; }
  #org-info-js_console-label
    { font-size: 10px; font-weight: bold; white-space: nowrap; }
  .org-info-js_search-highlight
    { background-color: #ffff00; color: #000000; font-weight: bold; }
  .org-svg { width: 90%; }
  /*]]>*/-->
</style>
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<link rel="stylesheet" type="text/css" href="https://www.pirilampo.org/styles/readtheorg/css/htmlize.css"/>
<style type="text/css">body {width: 60em; margin:auto} pre.src {overflow:auto}</style>
<script type="text/javascript">
/*
@licstart  The following is the entire license notice for the
JavaScript code in this tag.

Copyright (C) 2012-2017 Free Software Foundation, Inc.

The JavaScript code in this tag is free software: you can
redistribute it and/or modify it under the terms of the GNU
General Public License (GNU GPL) as published by the Free Software
Foundation, either version 3 of the License, or (at your option)
any later version.  The code is distributed WITHOUT ANY WARRANTY;
without even the implied warranty of MERCHANTABILITY or FITNESS
FOR A PARTICULAR PURPOSE.  See the GNU GPL for more details.

As additional permission under GNU GPL version 3 section 7, you
may distribute non-source (e.g., minimized or compacted) forms of
that code without the copy of the GNU GPL normally required by
section 4, provided you include this license notice and a URL
through which recipients can access the Corresponding Source.


@licend  The above is the entire license notice
for the JavaScript code in this tag.
*/
<!--/*--><![CDATA[/*><!--*/
 function CodeHighlightOn(elem, id)
 {
   var target = document.getElementById(id);
   if(null != target) {
     elem.cacheClassElem = elem.className;
     elem.cacheClassTarget = target.className;
     target.className = "code-highlighted";
     elem.className   = "code-highlighted";
   }
 }
 function CodeHighlightOff(elem, id)
 {
   var target = document.getElementById(id);
   if(elem.cacheClassElem)
     elem.className = elem.cacheClassElem;
   if(elem.cacheClassTarget)
     target.className = elem.cacheClassTarget;
 }
/*]]>*///-->
</script>
</head>
<body>
<div id="content">
<h1 class="title">Single cell/bulk RNA-Seq concordance</h1>

<div id="outline-container-org52a4766" class="outline-2">
<h2 id="org52a4766">Setup</h2>
<div class="outline-text-2" id="text-org52a4766">
<div class="org-src-container">
<pre class="src src-shell" id="org29ee016">sbatch ${<span class="org-variable-name">memory</span>/#/--mem=} ${<span class="org-variable-name">runtime</span>/#/--time=} ${<span class="org-variable-name">partition</span>/#/--partition=} --job-name=ipython3 --output=ipython3.out
<span class="org-comment-delimiter">#</span><span class="org-comment">!/bin/bash</span>
<span class="org-builtin">source</span> activate scqtl
rm -f $<span class="org-variable-name">HOME</span>/.local/share/jupyter/runtime/kernel-aksarkar.json
ipython3 kernel --ip=$(<span class="org-sh-quoted-exec">hostname</span> -i) -f kernel-aksarkar.json
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython" id="orgb7a93be">%matplotlib inline

<span class="org-keyword">import</span> colorcet
<span class="org-keyword">import</span> gzip
<span class="org-keyword">import</span> io
<span class="org-keyword">import</span> matplotlib.pyplot <span class="org-keyword">as</span> plt
<span class="org-keyword">import</span> numpy <span class="org-keyword">as</span> np
<span class="org-keyword">import</span> os
<span class="org-keyword">import</span> pandas <span class="org-keyword">as</span> pd
<span class="org-keyword">import</span> re
<span class="org-keyword">import</span> requests
<span class="org-keyword">import</span> scipy.stats <span class="org-keyword">as</span> sst
</pre>
</div>
</div>
</div>

<div id="outline-container-orge2dd258" class="outline-2">
<h2 id="orge2dd258">Read the data</h2>
<div class="outline-text-2" id="text-orge2dd258">
<p>
Use <code>kallisto</code> TPM for the bulk RNA-Seq data on the iPSCs.
</p>

<p>
Our preliminary experiments normalizing gene-level counts after the fact did
not give reasonable results.
</p>

<p>
Importantly, on the bulk side we filter out genes with gene-level TPM &lt; 1.
</p>

<div class="org-src-container">
<pre class="src src-ipython" id="orgfee619f"><span class="org-keyword">def</span> <span class="org-function-name">load_umi_data</span>(min_cell_count=5):
  <span class="org-variable-name">annotations</span> = pd.read_table(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-qtl/data/scqtl-annotation.txt'</span>)
  <span class="org-variable-name">counts</span> = pd.read_table(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-qtl/data/scqtl-counts.txt.gz'</span>, index_col=0)
  <span class="org-variable-name">keep_individuals</span> = annotations[<span class="org-string">'chip_id'</span>].value_counts() &gt; min_cell_count
  <span class="org-variable-name">keep_cells</span> = annotations.<span class="org-builtin">apply</span>(<span class="org-keyword">lambda</span> x: keep_individuals.loc[x[<span class="org-string">'chip_id'</span>]], axis=1)
  <span class="org-variable-name">annotations</span> = annotations.loc[keep_cells.values]
  <span class="org-variable-name">counts</span> = counts.loc[:,keep_cells.values]
  <span class="org-keyword">return</span> counts, annotations

<span class="org-keyword">def</span> <span class="org-function-name">load_bulk_data</span>():
  <span class="org-keyword">return</span> (pd.read_table(<span class="org-string">'/project2/gilad/singlecell-qtl/bulk/kallisto_cardio.txt.gz'</span>, sep=<span class="org-string">' '</span>, index_col=<span class="org-string">'gene'</span>).
          <span class="org-comment-delimiter"># </span><span class="org-comment">ID goes "transcript|gene|..."</span>
          groupby(<span class="org-keyword">lambda</span> x: x.split(<span class="org-string">'|'</span>)[1].split(<span class="org-string">'.'</span>)[0]).
          <span class="org-comment-delimiter"># </span><span class="org-comment">Gene level TPM is sum of transcript level TPM</span>
          agg(np.<span class="org-builtin">sum</span>).
          <span class="org-comment-delimiter"># </span><span class="org-comment">IMPORTANT: Filter out TPM &lt; 1</span>
          mask(<span class="org-keyword">lambda</span> x: x &lt; 1).
          applymap(<span class="org-keyword">lambda</span> x: np.log(x + 1) / np.log(2)).
          rename(columns=<span class="org-keyword">lambda</span> x: <span class="org-string">'NA{}'</span>.<span class="org-builtin">format</span>(x[1:]) <span class="org-keyword">if</span> x[0] == <span class="org-string">'H'</span> <span class="org-keyword">else</span> x)
          .<span class="org-builtin">filter</span>(regex=<span class="org-string">'^NA'</span>))
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython" id="org1c4a4dc"><span class="org-variable-name">umi</span>, <span class="org-variable-name">annotations</span> = load_umi_data()
<span class="org-variable-name">bulk</span> = load_bulk_data()

umi.shape, bulk.shape
</pre>
</div>

<pre class="example">
((20327, 2261), (34608, 15))

</pre>

<div class="org-src-container">
<pre class="src src-ipython" id="org80f2327"><span class="org-keyword">def</span> <span class="org-function-name">plot_concordance</span>(x, y, title, filename, xlabel=<span class="org-constant">None</span>, ylabel=<span class="org-constant">None</span>, **kwargs):
  <span class="org-doc">"""Plot hexbin of concordance"""</span>
  <span class="org-variable-name">merged</span> = x.merge(y, left_index=<span class="org-constant">True</span>, right_index=<span class="org-constant">True</span>)
  <span class="org-variable-name">merged.columns</span> = [<span class="org-string">'x'</span>, <span class="org-string">'y'</span>]
  <span class="org-variable-name">lim</span> = [merged.<span class="org-builtin">min</span>().<span class="org-builtin">min</span>(), merged.<span class="org-builtin">max</span>().<span class="org-builtin">max</span>()]
  plt.clf()
  <span class="org-keyword">if</span> <span class="org-string">'gridsize'</span> <span class="org-keyword">not</span> <span class="org-keyword">in</span> kwargs:
    <span class="org-variable-name">kwargs</span>[<span class="org-string">'gridsize'</span>] = 40
  plt.hexbin(merged[<span class="org-string">'x'</span>], merged[<span class="org-string">'y'</span>], cmap=colorcet.cm[<span class="org-string">'blues'</span>], **kwargs)
  <span class="org-variable-name">cb</span> = plt.colorbar()
  cb.set_label(<span class="org-string">'Number of genes'</span>)
  plt.plot(lim, lim, color=<span class="org-string">'red'</span>)
  <span class="org-variable-name">ax</span> = plt.gca()
  ax.set_xlim([merged[<span class="org-string">'x'</span>].<span class="org-builtin">min</span>(), merged[<span class="org-string">'x'</span>].<span class="org-builtin">max</span>()])
  ax.set_ylim([merged[<span class="org-string">'y'</span>].<span class="org-builtin">min</span>(), merged[<span class="org-string">'y'</span>].<span class="org-builtin">max</span>()])
  plt.title(title)
  <span class="org-keyword">if</span> xlabel <span class="org-keyword">is</span> <span class="org-constant">None</span>:
    <span class="org-variable-name">xlabel</span> = <span class="org-string">'scRNA-Seq $\log_2(\mathrm{CPM} + 1) $'</span>
  <span class="org-keyword">if</span> ylabel <span class="org-keyword">is</span> <span class="org-constant">None</span>:
    <span class="org-variable-name">ylabel</span> = <span class="org-string">'Bulk RNA-Seq $\log_2(\mathrm{TPM} + 1)$'</span>
  plt.xlabel(xlabel)
  plt.ylabel(ylabel)
  plt.savefig(filename)

<span class="org-keyword">def</span> <span class="org-function-name">cpm</span>(counts, log2=<span class="org-constant">False</span>):
  <span class="org-variable-name">cpm</span> = counts / counts.<span class="org-builtin">sum</span>(axis=0) * 1e6
  <span class="org-keyword">if</span> log2:
    <span class="org-variable-name">cpm</span> = np.log(cpm + 1) / np.log(2)
  <span class="org-keyword">return</span> cpm
</pre>
</div>
</div>
</div>

<div id="outline-container-orgc62a880" class="outline-2">
<h2 id="orgc62a880">Plot bulk vs. pooled single cells</h2>
<div class="outline-text-2" id="text-orgc62a880">
<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">def</span> <span class="org-function-name">plot_concordance_by_individual</span>(umi, annotations, bulk, output_dir):
  <span class="org-variable-name">bulk</span>, <span class="org-variable-name">pooled_cpm</span> = bulk.align(
    cpm(umi.groupby(by=annotations[<span class="org-string">'chip_id'</span>].values, axis=1).agg(np.<span class="org-builtin">sum</span>), log2=<span class="org-constant">True</span>),
    axis=1, join=<span class="org-string">'inner'</span>)
  <span class="org-keyword">for</span> k <span class="org-keyword">in</span> bulk:
    plot_concordance(
      x=pooled_cpm[k].to_frame(),
      y=bulk[k].to_frame(),
      title=k,
      filename=<span class="org-string">'{}/{}.png'</span>.<span class="org-builtin">format</span>(output_dir, k))
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython">plot_concordance_by_individual(
  umi,
  annotations,
  bulk,
  <span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-qtl/analysis/figure/sc-vs-bulk.org/normed'</span>)
</pre>
</div>

<p>
<img src="figure/sc-vs-bulk.org/normed/NA18498.png" alt="NA18498.png">
<img src="figure/sc-vs-bulk.org/normed/NA18501.png" alt="NA18501.png">
<img src="figure/sc-vs-bulk.org/normed/NA18502.png" alt="NA18502.png">
<img src="figure/sc-vs-bulk.org/normed/NA18505.png" alt="NA18505.png">
<img src="figure/sc-vs-bulk.org/normed/NA18507.png" alt="NA18507.png">
<img src="figure/sc-vs-bulk.org/normed/NA18508.png" alt="NA18508.png">
<img src="figure/sc-vs-bulk.org/normed/NA18520.png" alt="NA18520.png">
</p>
</div>
</div>

<div id="outline-container-org50c3440" class="outline-2">
<h2 id="org50c3440">Plot bulk vs. pooled subsets</h2>
<div class="outline-text-2" id="text-org50c3440">
<p>
Plot concordance between bulk vs pools of single cells, focusing on genes
which have log-transformed expression at least 1 in both assays.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">def</span> <span class="org-function-name">plot_concordance_by_num_cells</span>(individual, umi, annotations, bulk_tpm, output_dir):
  <span class="org-variable-name">bulk_tpm</span> = bulk_tpm[individual].to_frame()
  <span class="org-variable-name">umi</span> = umi.loc[:,(annotations[<span class="org-string">'chip_id'</span>] == individual).values]
  <span class="org-keyword">for</span> num_cells <span class="org-keyword">in</span> [1, 10, 50, 100, 200]:
    <span class="org-variable-name">pooled_cpm</span> = cpm(umi.sample(n=num_cells, axis=1).<span class="org-builtin">sum</span>(axis=1).to_frame(), log2=<span class="org-constant">True</span>)
    plot_concordance(
      x=pooled_cpm,
      y=bulk_tpm,
      title=<span class="org-string">'{}, {} cell{}'</span>.<span class="org-builtin">format</span>(individual, num_cells, <span class="org-string">'s'</span> <span class="org-keyword">if</span> num_cells &gt; 1 <span class="org-keyword">else</span> <span class="org-string">''</span>),
      filename=<span class="org-string">'{}/{}-{}.png'</span>.<span class="org-builtin">format</span>(output_dir, individual, num_cells),
      gridsize=20)
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython">plot_concordance_by_num_cells(
  <span class="org-string">'NA18507'</span>,
  umi,
  annotations,
  bulk,
  <span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-qtl/analysis/figure/sc-vs-bulk.org/vs-cells/'</span>
)
</pre>
</div>

<p>
<img src="figure/sc-vs-bulk.org/vs-cells/NA18507-1.png" alt="NA18507-1.png">
<img src="figure/sc-vs-bulk.org/vs-cells/NA18507-10.png" alt="NA18507-10.png">
<img src="figure/sc-vs-bulk.org/vs-cells/NA18507-50.png" alt="NA18507-50.png">
<img src="figure/sc-vs-bulk.org/vs-cells/NA18507-100.png" alt="NA18507-100.png">
<img src="figure/sc-vs-bulk.org/vs-cells/NA18507-200.png" alt="NA18507-200.png">
</p>
</div>
</div>

<div id="outline-container-orgcdd234c" class="outline-2">
<h2 id="orgcdd234c">Plot pooled subsets vs. pooled subsets</h2>
<div class="outline-text-2" id="text-orgcdd234c">
<p>
Ensure that pools don't overlap by randomly sampling double the cells and
partitioning into two halves.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">def</span> <span class="org-function-name">plot_concordance_pooled_subsets</span>(individual, umi, annotations, output_dir):
  <span class="org-variable-name">umi</span> = umi.loc[:,(annotations[<span class="org-string">'chip_id'</span>] == individual).values]
  <span class="org-keyword">for</span> num_cells <span class="org-keyword">in</span> [1, 10, 50, 100]:
    <span class="org-variable-name">sample</span> = umi.sample(n=2 * num_cells, axis=1)
    <span class="org-variable-name">pool1</span> = cpm(sample.iloc[:,:num_cells].<span class="org-builtin">sum</span>(axis=1).to_frame(), log2=<span class="org-constant">True</span>)
    <span class="org-variable-name">pool2</span> = cpm(sample.iloc[:,num_cells:].<span class="org-builtin">sum</span>(axis=1).to_frame(), log2=<span class="org-constant">True</span>)
    plot_concordance(
      x=pool1,
      y=pool2,
      title=<span class="org-string">'{}, {} cell{}'</span>.<span class="org-builtin">format</span>(individual, num_cells, <span class="org-string">'s'</span> <span class="org-keyword">if</span> num_cells &gt; 1 <span class="org-keyword">else</span> <span class="org-string">''</span>),
      filename=<span class="org-string">'{}/{}-{}.png'</span>.<span class="org-builtin">format</span>(output_dir, individual, num_cells),
      ylabel=<span class="org-string">'scRNA-Seq $\log_2(\mathrm{CPM} + 1)$'</span>,
      gridsize=15)
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython">plot_concordance_pooled_subsets(
  <span class="org-string">'NA18507'</span>,
  umi,
  annotations,
  <span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-qtl/analysis/figure/sc-vs-bulk.org/subsets/'</span>
)
</pre>
</div>

<p>
<img src="figure/sc-vs-bulk.org/subsets/NA18507-1.png" alt="NA18507-1.png">
<img src="figure/sc-vs-bulk.org/subsets/NA18507-10.png" alt="NA18507-10.png">
<img src="figure/sc-vs-bulk.org/subsets/NA18507-50.png" alt="NA18507-50.png">
<img src="figure/sc-vs-bulk.org/subsets/NA18507-100.png" alt="NA18507-100.png">
</p>
</div>
</div>
</div>
<div id="postamble" class="status">
<p class="author">Author: Abhishek Sarkar</p>
<p class="date">Created: 2018-03-16 Fri 12:51</p>
<p class="validation"><a href="http://validator.w3.org/check?uri=referer">Validate</a></p>
</div>
</body>
</html>
