<!DOCTYPE html>
<html lang="en">
<head>
<!-- 2018-01-14 Sun 17:15 -->
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
<title>ZINB estimation</title>
<meta name="generator" content="Org mode">
<meta name="author" content="Abhishek Sarkar">
<style type="text/css">
 <!--/*--><![CDATA[/*><!--*/
  .title  { text-align: center;
             margin-bottom: .2em; }
  .subtitle { text-align: center;
              font-size: medium;
              font-weight: bold;
              margin-top:0; }
  .todo   { font-family: monospace; color: red; }
  .done   { font-family: monospace; color: green; }
  .priority { font-family: monospace; color: orange; }
  .tag    { background-color: #eee; font-family: monospace;
            padding: 2px; font-size: 80%; font-weight: normal; }
  .timestamp { color: #bebebe; }
  .timestamp-kwd { color: #5f9ea0; }
  .org-right  { margin-left: auto; margin-right: 0px;  text-align: right; }
  .org-left   { margin-left: 0px;  margin-right: auto; text-align: left; }
  .org-center { margin-left: auto; margin-right: auto; text-align: center; }
  .underline { text-decoration: underline; }
  #postamble p, #preamble p { font-size: 90%; margin: .2em; }
  p.verse { margin-left: 3%; }
  pre {
    border: 1px solid #ccc;
    box-shadow: 3px 3px 3px #eee;
    padding: 8pt;
    font-family: monospace;
    overflow: auto;
    margin: 1.2em;
  }
  pre.src {
    position: relative;
    overflow: visible;
    padding-top: 1.2em;
  }
  pre.src:before {
    display: none;
    position: absolute;
    background-color: white;
    top: -10px;
    right: 10px;
    padding: 3px;
    border: 1px solid black;
  }
  pre.src:hover:before { display: inline;}
  /* Languages per Org manual */
  pre.src-asymptote:before { content: 'Asymptote'; }
  pre.src-awk:before { content: 'Awk'; }
  pre.src-C:before { content: 'C'; }
  /* pre.src-C++ doesn't work in CSS */
  pre.src-clojure:before { content: 'Clojure'; }
  pre.src-css:before { content: 'CSS'; }
  pre.src-D:before { content: 'D'; }
  pre.src-ditaa:before { content: 'ditaa'; }
  pre.src-dot:before { content: 'Graphviz'; }
  pre.src-calc:before { content: 'Emacs Calc'; }
  pre.src-emacs-lisp:before { content: 'Emacs Lisp'; }
  pre.src-fortran:before { content: 'Fortran'; }
  pre.src-gnuplot:before { content: 'gnuplot'; }
  pre.src-haskell:before { content: 'Haskell'; }
  pre.src-hledger:before { content: 'hledger'; }
  pre.src-java:before { content: 'Java'; }
  pre.src-js:before { content: 'Javascript'; }
  pre.src-latex:before { content: 'LaTeX'; }
  pre.src-ledger:before { content: 'Ledger'; }
  pre.src-lisp:before { content: 'Lisp'; }
  pre.src-lilypond:before { content: 'Lilypond'; }
  pre.src-lua:before { content: 'Lua'; }
  pre.src-matlab:before { content: 'MATLAB'; }
  pre.src-mscgen:before { content: 'Mscgen'; }
  pre.src-ocaml:before { content: 'Objective Caml'; }
  pre.src-octave:before { content: 'Octave'; }
  pre.src-org:before { content: 'Org mode'; }
  pre.src-oz:before { content: 'OZ'; }
  pre.src-plantuml:before { content: 'Plantuml'; }
  pre.src-processing:before { content: 'Processing.js'; }
  pre.src-python:before { content: 'Python'; }
  pre.src-R:before { content: 'R'; }
  pre.src-ruby:before { content: 'Ruby'; }
  pre.src-sass:before { content: 'Sass'; }
  pre.src-scheme:before { content: 'Scheme'; }
  pre.src-screen:before { content: 'Gnu Screen'; }
  pre.src-sed:before { content: 'Sed'; }
  pre.src-sh:before { content: 'shell'; }
  pre.src-sql:before { content: 'SQL'; }
  pre.src-sqlite:before { content: 'SQLite'; }
  /* additional languages in org.el's org-babel-load-languages alist */
  pre.src-forth:before { content: 'Forth'; }
  pre.src-io:before { content: 'IO'; }
  pre.src-J:before { content: 'J'; }
  pre.src-makefile:before { content: 'Makefile'; }
  pre.src-maxima:before { content: 'Maxima'; }
  pre.src-perl:before { content: 'Perl'; }
  pre.src-picolisp:before { content: 'Pico Lisp'; }
  pre.src-scala:before { content: 'Scala'; }
  pre.src-shell:before { content: 'Shell Script'; }
  pre.src-ebnf2ps:before { content: 'ebfn2ps'; }
  /* additional language identifiers per "defun org-babel-execute"
       in ob-*.el */
  pre.src-cpp:before  { content: 'C++'; }
  pre.src-abc:before  { content: 'ABC'; }
  pre.src-coq:before  { content: 'Coq'; }
  pre.src-groovy:before  { content: 'Groovy'; }
  /* additional language identifiers from org-babel-shell-names in
     ob-shell.el: ob-shell is the only babel language using a lambda to put
     the execution function name together. */
  pre.src-bash:before  { content: 'bash'; }
  pre.src-csh:before  { content: 'csh'; }
  pre.src-ash:before  { content: 'ash'; }
  pre.src-dash:before  { content: 'dash'; }
  pre.src-ksh:before  { content: 'ksh'; }
  pre.src-mksh:before  { content: 'mksh'; }
  pre.src-posh:before  { content: 'posh'; }
  /* Additional Emacs modes also supported by the LaTeX listings package */
  pre.src-ada:before { content: 'Ada'; }
  pre.src-asm:before { content: 'Assembler'; }
  pre.src-caml:before { content: 'Caml'; }
  pre.src-delphi:before { content: 'Delphi'; }
  pre.src-html:before { content: 'HTML'; }
  pre.src-idl:before { content: 'IDL'; }
  pre.src-mercury:before { content: 'Mercury'; }
  pre.src-metapost:before { content: 'MetaPost'; }
  pre.src-modula-2:before { content: 'Modula-2'; }
  pre.src-pascal:before { content: 'Pascal'; }
  pre.src-ps:before { content: 'PostScript'; }
  pre.src-prolog:before { content: 'Prolog'; }
  pre.src-simula:before { content: 'Simula'; }
  pre.src-tcl:before { content: 'tcl'; }
  pre.src-tex:before { content: 'TeX'; }
  pre.src-plain-tex:before { content: 'Plain TeX'; }
  pre.src-verilog:before { content: 'Verilog'; }
  pre.src-vhdl:before { content: 'VHDL'; }
  pre.src-xml:before { content: 'XML'; }
  pre.src-nxml:before { content: 'XML'; }
  /* add a generic configuration mode; LaTeX export needs an additional
     (add-to-list 'org-latex-listings-langs '(conf " ")) in .emacs */
  pre.src-conf:before { content: 'Configuration File'; }

  table { border-collapse:collapse; }
  caption.t-above { caption-side: top; }
  caption.t-bottom { caption-side: bottom; }
  td, th { vertical-align:top;  }
  th.org-right  { text-align: center;  }
  th.org-left   { text-align: center;   }
  th.org-center { text-align: center; }
  td.org-right  { text-align: right;  }
  td.org-left   { text-align: left;   }
  td.org-center { text-align: center; }
  dt { font-weight: bold; }
  .footpara { display: inline; }
  .footdef  { margin-bottom: 1em; }
  .figure { padding: 1em; }
  .figure p { text-align: center; }
  .inlinetask {
    padding: 10px;
    border: 2px solid gray;
    margin: 10px;
    background: #ffffcc;
  }
  #org-div-home-and-up
   { text-align: right; font-size: 70%; white-space: nowrap; }
  textarea { overflow-x: auto; }
  .linenr { font-size: smaller }
  .code-highlighted { background-color: #ffff00; }
  .org-info-js_info-navigation { border-style: none; }
  #org-info-js_console-label
    { font-size: 10px; font-weight: bold; white-space: nowrap; }
  .org-info-js_search-highlight
    { background-color: #ffff00; color: #000000; font-weight: bold; }
  .org-svg { width: 90%; }
  /*]]>*/-->
</style>
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<link rel="stylesheet" type="text/css" href="https://www.pirilampo.org/styles/readtheorg/css/htmlize.css"/>
<style type="text/css">body {width: 60em; margin:auto} pre.src {overflow:auto}</style>
<script type="text/javascript">
/*
@licstart  The following is the entire license notice for the
JavaScript code in this tag.

Copyright (C) 2012-2017 Free Software Foundation, Inc.

The JavaScript code in this tag is free software: you can
redistribute it and/or modify it under the terms of the GNU
General Public License (GNU GPL) as published by the Free Software
Foundation, either version 3 of the License, or (at your option)
any later version.  The code is distributed WITHOUT ANY WARRANTY;
without even the implied warranty of MERCHANTABILITY or FITNESS
FOR A PARTICULAR PURPOSE.  See the GNU GPL for more details.

As additional permission under GNU GPL version 3 section 7, you
may distribute non-source (e.g., minimized or compacted) forms of
that code without the copy of the GNU GPL normally required by
section 4, provided you include this license notice and a URL
through which recipients can access the Corresponding Source.


@licend  The above is the entire license notice
for the JavaScript code in this tag.
*/
<!--/*--><![CDATA[/*><!--*/
 function CodeHighlightOn(elem, id)
 {
   var target = document.getElementById(id);
   if(null != target) {
     elem.cacheClassElem = elem.className;
     elem.cacheClassTarget = target.className;
     target.className = "code-highlighted";
     elem.className   = "code-highlighted";
   }
 }
 function CodeHighlightOff(elem, id)
 {
   var target = document.getElementById(id);
   if(elem.cacheClassElem)
     elem.className = elem.cacheClassElem;
   if(elem.cacheClassTarget)
     target.className = elem.cacheClassTarget;
 }
/*]]>*///-->
</script>
<script type="text/x-mathjax-config">
    MathJax.Hub.Config({
        displayAlign: "center",
        displayIndent: "0em",

        "HTML-CSS": { scale: 100,
                        linebreaks: { automatic: "false" },
                        webFont: "TeX"
                       },
        SVG: {scale: 100,
              linebreaks: { automatic: "false" },
              font: "TeX"},
        NativeMML: {scale: 100},
        TeX: { equationNumbers: {autoNumber: "AMS"},
               MultLineWidth: "85%",
               TagSide: "right",
               TagIndent: ".8em"
             }
});
</script>
<script type="text/javascript"
        src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-AMS_HTML"></script>
</head>
<body>
<div id="content">
<h1 class="title">ZINB estimation</h1>
<div id="table-of-contents">
<h2>Table of Contents</h2>
<div id="text-table-of-contents">
<ul>
<li><a href="#org470ab93">Introduction</a></li>
<li><a href="#org5a724da">Quality control</a></li>
<li><a href="#org8fd5c16">Read the data</a></li>
<li><a href="#orgde4b91c">Maximum likelihood estimation</a></li>
<li><a href="#org9aaf67c">Mean-QTL calling</a></li>
<li><a href="#orgad41e3e">Dispersion-QTL calling</a></li>
</ul>
</div>
</div>

<div id="outline-container-org470ab93" class="outline-2">
<h2 id="org470ab93">Introduction</h2>
<div class="outline-text-2" id="text-org470ab93">
<p>
The simplest approach to call mean/variance QTLs is to estimate a mean and a
dispersion for each individual, treat them as continuous phenotypes, and plug
into standard QTL mapping software.
</p>

<p>
Here, we find maximum likelihood estimates of a zero-inflated negative
binomial model for the dropout, mean, and dispersion per individual per gene.
</p>
</div>
</div>

<div id="outline-container-org5a724da" class="outline-2">
<h2 id="org5a724da">Quality control</h2>
<div class="outline-text-2" id="text-org5a724da">
<p>
Filter out cells on percent spike-in and gene detection rate, and filter out
genes on individual detection rate.
</p>

<div class="org-src-container">
<pre class="src src-ipython" id="org65b6f97"><span class="org-variable-name">keep_cells</span> = functools.<span class="org-builtin">reduce</span>(
  np.logical_and,
  [
    annotations[<span class="org-string">'reads_ercc'</span>] / annotations.<span class="org-builtin">filter</span>(like=<span class="org-string">'reads_'</span>, axis=1).agg(np.<span class="org-builtin">sum</span>, axis=1) &lt; 0.5,
    annotations[<span class="org-string">'detect_hs'</span>] &gt; 4000,
    annotations[<span class="org-string">'chip_id'</span>] != <span class="org-string">'NA19092'</span>,
  ]).values
<span class="org-variable-name">keep_genes</span> = functools.<span class="org-builtin">reduce</span>(
  np.logical_and,
  [
    (umi.groupby(annotations[<span class="org-string">'chip_id'</span>].values, axis=1).agg(np.<span class="org-builtin">sum</span>) &gt; 0).<span class="org-builtin">apply</span>(<span class="org-keyword">lambda</span> x: x.mean() &gt; 0.5, axis=1)
  ]).values
<span class="org-variable-name">umi_qc</span> = umi.loc[keep_genes, keep_cells]
<span class="org-variable-name">annotations_qc</span> = annotations.loc[keep_cells]
umi_qc.mask(umi_qc == 0)
umi_qc.shape
</pre>
</div>

<pre class="example">
(15636, 1810)

</pre>

<div class="org-src-container">
<pre class="src src-ipython" id="org1e8d763"><span class="org-variable-name">individuals</span> = <span class="org-builtin">sorted</span>(annotations_qc[<span class="org-string">'chip_id'</span>].unique())
<span class="org-variable-name">onehot</span> = np.zeros((umi_qc.shape[1], <span class="org-builtin">len</span>(individuals)), dtype=np.float32)
onehot[np.arange(onehot.shape[0]),annotations_qc[<span class="org-string">'chip_id'</span>].<span class="org-builtin">apply</span>(<span class="org-keyword">lambda</span> x: individuals.index(x))] = 1
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython">umi_qc.to_csv(<span class="org-string">'/scratch/midway2/aksarkar/singlecell/umi-qc.txt.gz'</span>, sep=<span class="org-string">' '</span>, compression=<span class="org-string">'gzip'</span>)
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">onehot</span> = pd.DataFrame(onehot, columns=individuals, index=umi_qc.columns)
onehot.to_csv(<span class="org-string">'/scratch/midway2/aksarkar/singlecell/onehot-qc.txt.gz'</span>, sep=<span class="org-string">' '</span>, compression=<span class="org-string">'gzip'</span>)
</pre>
</div>
</div>
</div>

<div id="outline-container-org8fd5c16" class="outline-2">
<h2 id="org8fd5c16">Read the data</h2>
<div class="outline-text-2" id="text-org8fd5c16">
<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">umi</span> = pd.read_table(<span class="org-string">'/scratch/midway2/aksarkar/singlecell/umi-qc.txt.gz'</span>, sep=<span class="org-string">' '</span>, index_col=0, nrows=4000)
<span class="org-variable-name">onehot</span> = pd.read_table(<span class="org-string">'/scratch/midway2/aksarkar/singlecell/onehot-qc.txt.gz'</span>, sep=<span class="org-string">' '</span>, index_col=0)
</pre>
</div>
</div>
</div>

<div id="outline-container-orgde4b91c" class="outline-2">
<h2 id="orgde4b91c">Maximum likelihood estimation</h2>
<div class="outline-text-2" id="text-orgde4b91c">
<p>
Use gradient descent to maximize the likelihood of the zero-inflated negative
binomial model for the data.
</p>

<p>
\[ r_{ijk} = \text{number of reads for individual \(i\), cell \(j\), gene
  \(k\)} \]
</p>

<p>
\[ r_{ijk} = \pi_j \delta_0(r_{ijk}) + (1 - \pi_j) \mathcal{NB}(r_{ijk};
  \mu_{ik}, \phi_{ik}) \]
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">def</span> <span class="org-function-name">log_zinb_positive</span>(x, mu, theta, pi, eps=1e-8):
  <span class="org-doc">"""Log likelihood of x according to ZINB model</span>

<span class="org-doc">  Lopez et al. "A deep generative model for gene expression profiles</span>
<span class="org-doc">  from single-cell RNA sequencing"</span>

<span class="org-doc">  c.f. Hilbe 2012, eq. 11.12, 11.13</span>

<span class="org-doc">  mu: mean (&gt;0)</span>
<span class="org-doc">  theta: inverse dispersion parameter (&gt;0)</span>
<span class="org-doc">  pi:logit of the dropout parameter</span>

<span class="org-doc">  """</span>
  <span class="org-variable-name">case_zero</span> = tf.nn.softplus(-pi + theta * tf.log(theta + eps) - theta * tf.log(theta + mu + eps)) \
              - tf.nn.softplus(-pi)
  <span class="org-variable-name">case_non_zero</span> = - pi - tf.nn.softplus(- pi) \
                  + theta * tf.log(theta + eps) - theta * tf.log(theta + mu + eps) \
                  + x * tf.log(mu + eps) - x * tf.log(theta + mu + eps) \
                  + tf.lgamma(x + theta) - tf.lgamma(theta) - tf.lgamma(x + 1)
  <span class="org-variable-name">mask</span> = tf.cast(tf.less(x, eps), tf.float32)
  <span class="org-keyword">return</span> tf.reduce_sum(tf.multiply(mask, case_zero) + tf.multiply(1 - mask, case_non_zero))

<span class="org-keyword">def</span> <span class="org-function-name">fit</span>(umi, onehot, learning_rate=1e-2, max_epochs=1000):
  <span class="org-doc">"""Return estimated dropout log odds, log mean, and log dispersion</span>

<span class="org-doc">  umi - count matrix (n x p; float32)</span>
<span class="org-doc">  onehot - mapping of individuals to cells (m x n; float32)</span>

<span class="org-doc">  Returns:</span>

<span class="org-doc">  llik - log likelihood</span>
<span class="org-doc">  dropout - spike log odds (n x 1)</span>
<span class="org-doc">  log_mean - log mean parameter (m x p)</span>
<span class="org-doc">  log_disp - log dispersion parameter (m x p)</span>

<span class="org-doc">  """</span>
  <span class="org-variable-name">n</span>, <span class="org-variable-name">p</span> = umi.shape
  <span class="org-variable-name">_</span>, <span class="org-variable-name">m</span> = onehot.shape

  <span class="org-variable-name">umi</span> = tf.Variable(umi, trainable=<span class="org-constant">False</span>)
  <span class="org-variable-name">onehot</span> = tf.Variable(onehot, trainable=<span class="org-constant">False</span>)

  <span class="org-variable-name">index</span> = tf.placeholder(tf.int32)

  <span class="org-variable-name">dropout</span> = tf.Variable(tf.zeros([n, 1]))
  <span class="org-variable-name">mean</span> = tf.exp(tf.Variable(tf.zeros([m, p])))
  <span class="org-variable-name">dispersion</span> = tf.exp(tf.Variable(tf.zeros([m, p])))

  <span class="org-variable-name">llik</span> = log_zinb_positive(
    umi,
    tf.matmul(onehot, mean),
    tf.matmul(onehot, dispersion),
    dropout)

  <span class="org-variable-name">train</span> = tf.train.RMSPropOptimizer(learning_rate=learning_rate).minimize(-llik)

  <span class="org-variable-name">opt</span> = [tf.log(mean), tf.log(dispersion), dropout]
  <span class="org-variable-name">curr</span> = <span class="org-builtin">float</span>(<span class="org-string">'-inf'</span>)
  <span class="org-keyword">with</span> tf.Session() <span class="org-keyword">as</span> sess:
    sess.run(tf.global_variables_initializer())
    <span class="org-keyword">for</span> i <span class="org-keyword">in</span> <span class="org-builtin">range</span>(max_epochs):
      <span class="org-variable-name">_</span>, <span class="org-variable-name">update</span> = sess.run([train, llik], feed_dict={index: i % p})
      <span class="org-keyword">if</span> <span class="org-keyword">not</span> np.isfinite(update):
        <span class="org-keyword">raise</span> tf.train.NanLossDuringTrainingError
      <span class="org-keyword">if</span> <span class="org-keyword">not</span> i % 100:
        <span class="org-keyword">print</span>(i, update)
    <span class="org-keyword">return</span> sess.run(opt)
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">mean</span>, <span class="org-variable-name">dispersion</span>, <span class="org-variable-name">dropout</span> = fit(
  umi.values.T.astype(np.float32),
  onehot.values.astype(np.float32),
  learning_rate=1e-2)
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython">pd.DataFrame(mean.T, index=umi.index, columns=onehot.columns).to_csv(<span class="org-string">'/scratch/midway2/aksarkar/singlecell/mean.txt.gz'</span>, sep=<span class="org-string">' '</span>, compression=<span class="org-string">'gzip'</span>)
pd.DataFrame(dispersion.T, index=umi.index, columns=onehot.columns).to_csv(<span class="org-string">'/scratch/midway2/aksarkar/singlecell/dispersion.txt.gz'</span>, sep=<span class="org-string">' '</span>, compression=<span class="org-string">'gzip'</span>)
pd.DataFrame(dropout, index=umi.columns).to_csv(<span class="org-string">'/scratch/midway2/aksarkar/singlecell/dropout.txt.gz'</span>, sep=<span class="org-string">' '</span>, compression=<span class="org-string">'gzip'</span>)
</pre>
</div>

<p>
This estimation problem takes too long to run interactively, so <a href="https://orgmode.org/manual/Extracting-source-code.html">tangle</a> it
from this file and run it through <code>sbatch</code>.
</p>

<div class="org-src-container">
<pre class="src src-emacs-lisp">(org-babel-tangle)
</pre>
</div>

<div class="org-src-container">
<pre class="src src-sh">sbatch --partition=broadwl --mem=8G --time=120 --job-name=zinb --output=zinb.out --error=zinb.err
<span class="org-comment-delimiter">#</span><span class="org-comment">!/bin/bash</span>
<span class="org-builtin">source</span> activate scqtl
python zinb.py
</pre>
</div>
</div>
</div>

<div id="outline-container-org9aaf67c" class="outline-2">
<h2 id="org9aaf67c">Mean-QTL calling</h2>
<div class="outline-text-2" id="text-org9aaf67c">
<p>
Write out the <a href="https://qtltools.github.io/qtltools/pages/input_files.html">phenotype file</a> for <code>QTLtools</code>.
</p>

<div class="org-src-container">
<pre class="src src-ipython" id="org389e5be"><span class="org-variable-name">gene_info</span> = (pd.read_table(<span class="org-string">'/home/aksarkar/projects/singlecell-qtl/data/scqtl-genes.txt.gz'</span>)
             .set_index(<span class="org-string">'gene'</span>)
             .query(<span class="org-string">'source == "H. sapiens"'</span>)
             .query(<span class="org-string">'chr != "hsX"'</span>)
             .query(<span class="org-string">'chr != "hsY"'</span>)
             .query(<span class="org-string">'chr != "hsMT"'</span>))
gene_info.head()
</pre>
</div>

<pre class="example">
                chr      start        end      name strand      source
gene                                                                    
ENSG00000000419  hs20   49551404   49575092      DPM1      -  H. sapiens
ENSG00000000457   hs1  169818772  169863408     SCYL3      -  H. sapiens
ENSG00000000460   hs1  169631245  169823221  C1orf112      +  H. sapiens
ENSG00000000938   hs1   27938575   27961788       FGR      -  H. sapiens
ENSG00000000971   hs1  196621008  196716634       CFH      +  H. sapiens
</pre>

<pre class="example">
                chr      start        end      name strand      source
gene                                                                    
ENSG00000000419  hs20   49551404   49575092      DPM1      -  H. sapiens
ENSG00000000457   hs1  169818772  169863408     SCYL3      -  H. sapiens
ENSG00000000460   hs1  169631245  169823221  C1orf112      +  H. sapiens
ENSG00000000938   hs1   27938575   27961788       FGR      -  H. sapiens
ENSG00000000971   hs1  196621008  196716634       CFH      +  H. sapiens
</pre>

<div class="org-src-container">
<pre class="src src-ipython" id="org5a26fff"><span class="org-keyword">def</span> <span class="org-function-name">qtltools_format</span>(row):
  <span class="org-variable-name">row</span>[<span class="org-string">'#Chr'</span>] = <span class="org-string">'chr{}'</span>.<span class="org-builtin">format</span>(row[<span class="org-string">'chr'</span>][2:])
  <span class="org-variable-name">row</span>[<span class="org-string">'gid'</span>] = row.name
  <span class="org-variable-name">row</span>[<span class="org-string">'pid'</span>] = row.name
  <span class="org-keyword">return</span> row

<span class="org-keyword">def</span> <span class="org-function-name">write_pheno_file</span>(pheno, gene_info, output_file):
  (gene_info
   .<span class="org-builtin">apply</span>(qtltools_format, axis=1)
   .merge(pheno, left_index=<span class="org-constant">True</span>, right_index=<span class="org-constant">True</span>)
   .to_csv(output_file,
           sep=<span class="org-string">'\t'</span>,
           columns=[<span class="org-string">'#Chr'</span>, <span class="org-string">'start'</span>, <span class="org-string">'end'</span>, <span class="org-string">'pid'</span>, <span class="org-string">'gid'</span>, <span class="org-string">'strand'</span>] + <span class="org-builtin">list</span>(pheno.columns),
           header=<span class="org-constant">True</span>,
           index=<span class="org-constant">False</span>,
           index_label=<span class="org-constant">False</span>)
  )
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">mean</span> = pd.read_table(<span class="org-string">'/scratch/midway2/aksarkar/singlecell/mean.txt.gz'</span>, sep=<span class="org-string">' '</span>, index_col=0)
mean.head()
</pre>
</div>

<pre class="example">
                NA18489   NA18498   NA18499   NA18501   NA18502   NA18505  \
gene                                                                          
ENSG00000000003  2.132201  1.862085  1.994689  2.203277  1.994263  1.873698   
ENSG00000000005 -1.304971 -1.137107 -2.085783 -1.440596 -2.309412 -1.127515   
ENSG00000000419  1.408769  1.332950  1.690473  1.656058  1.669314  1.605577   
ENSG00000000457 -0.993338 -2.940411 -0.788445 -0.932689 -0.906488 -0.360301   
ENSG00000000460  0.403482 -0.106141  0.520277  0.567161  0.443905  0.670071   

                  NA18507   NA18508   NA18519   NA18520    ...      NA19119  \
gene                                                       ...                
ENSG00000000003  2.185227  1.984693  2.175263  1.960210    ...     2.228055   
ENSG00000000005 -1.499075 -3.047421 -1.563732 -2.949254    ...    -2.113694   
ENSG00000000419  1.834280  1.659502  1.692899  1.579187    ...     1.806168   
ENSG00000000457 -0.763892 -0.913251 -1.235485 -0.913332    ...    -1.210785   
ENSG00000000460  0.474492  0.521287  0.653763  0.439643    ...     0.771924   

                  NA19128   NA19153   NA19159   NA19190   NA19193   NA19203  \
gene                                                                          
ENSG00000000003  2.316742  2.286229  1.866834  2.146526  2.011522  1.547072   
ENSG00000000005 -2.109344 -0.943781 -3.445312 -2.083898 -2.493103 -1.347846   
ENSG00000000419  2.015950  1.903451  1.413075  1.730209  1.968157  1.188535   
ENSG00000000457 -0.705963 -0.792029 -0.870194 -0.492704 -1.142365 -2.031550   
ENSG00000000460  0.898602  0.601947  0.173087  0.454041  0.562179  0.153599   

                  NA19207   NA19210   NA19257  
gene                                           
ENSG00000000003  2.275138  2.277549  2.068974  
ENSG00000000005 -1.550304 -1.269496 -2.731949  
ENSG00000000419  1.865725  2.116240  2.024805  
ENSG00000000457 -0.965920 -0.678640 -0.859062  
ENSG00000000460  0.797515  0.873859  0.751235  

[5 rows x 32 columns]
</pre>

<div class="org-src-container">
<pre class="src src-ipython">write_pheno_file(mean, gene_info, <span class="org-string">'/scratch/midway2/aksarkar/singlecell/mean.bed'</span>)
</pre>
</div>

<p>
Index the phenotype file.
</p>

<div class="org-src-container">
<pre class="src src-sh" id="org78fe142"><span class="org-builtin">export</span> <span class="org-variable-name">input</span>=$<span class="org-variable-name">input</span>
sbatch --partition=$<span class="org-variable-name">partition</span> --wait
<span class="org-comment-delimiter">#</span><span class="org-comment">!/bin/bash</span>
sort -k1,1 -k2,2n -k3,3n $<span class="org-variable-name">input</span> | bgzip &gt;$<span class="org-variable-name">input</span>.gz
tabix -p bed $<span class="org-variable-name">input</span>.gz
</pre>
</div>

<p>
Run the QTL mapping.
</p>

<div class="org-src-container">
<pre class="src src-sh" id="org93dc535"><span class="org-builtin">export</span> <span class="org-variable-name">pheno</span>=$<span class="org-variable-name">pheno</span>
sbatch --partition=$<span class="org-variable-name">partition</span> -N1 -c4 -J $<span class="org-variable-name">pheno</span>-qtl -o $<span class="org-variable-name">pheno</span>-qtl.log
<span class="org-comment-delimiter">#</span><span class="org-comment">!/bin/bash</span>
<span class="org-builtin">source</span> activate scqtl
module load parallel
parallel -j4 qtltools cis --vcf /project2/gilad/singlecell-qtl/bulk/genotypes.vcf.gz --bed $<span class="org-variable-name">pheno</span>.bed.gz --nominal=0.01 --chunk {#} 100 --out $<span class="org-variable-name">pheno</span>-qtl.{#}.txt ::: $(<span class="org-sh-quoted-exec">seq</span> 1 100)
</pre>
</div>

<p>
Read the results.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">file_names</span> = [<span class="org-string">'mean-qtl.{}.txt'</span>.<span class="org-builtin">format</span>(i) <span class="org-keyword">for</span> i <span class="org-keyword">in</span> <span class="org-builtin">range</span>(1, 101)]
<span class="org-variable-name">mean_qtls</span> = (pd.concat([pd.read_table(f, header=<span class="org-constant">None</span>, sep=<span class="org-string">' '</span>) <span class="org-keyword">for</span> f <span class="org-keyword">in</span> file_names <span class="org-keyword">if</span> os.path.exists(f) <span class="org-keyword">and</span> os.path.getsize(f) &gt; 0])
             .rename(columns={i: x <span class="org-keyword">for</span> i, x <span class="org-keyword">in</span> <span class="org-builtin">enumerate</span>([<span class="org-string">'gene'</span>, <span class="org-string">'chr'</span>, <span class="org-string">'start'</span>, <span class="org-string">'end'</span>, <span class="org-string">'strand'</span>, <span class="org-string">'num_vars'</span>, <span class="org-string">'distance'</span>, <span class="org-string">'id'</span>, <span class="org-string">'var_chr'</span>, <span class="org-string">'var_start'</span>, <span class="org-string">'var_end'</span>, <span class="org-string">'p'</span>, <span class="org-string">'beta'</span>, <span class="org-string">'top'</span>])})
             .sort_values(<span class="org-string">'p'</span>)
             .set_index(<span class="org-string">'gene'</span>))
mean_qtls[mean_qtls[<span class="org-string">'top'</span>] == 1].head()
</pre>
</div>

<pre class="example">
                 chr   start     end strand  num_vars  distance id var_chr  \
gene                                                                           
ENSG00000239857   chr7  916190  936073      +      7763   -916190  C    chr7   
ENSG00000169026   chr4  675619  683230      -      4952    675619  C    chr4   
ENSG00000101282  chr20  939096  982907      -      6734    939096  G   chr20   
ENSG00000186777   chr4  264465  299110      -      3531    264465  A    chr4   
ENSG00000116014  chr19  917288  921015      +      6659   -917288  G   chr19   

                 var_start  var_end             p     beta  top  
gene                                                             
ENSG00000239857     986498        0  1.273710e-16 -4.71915    1  
ENSG00000169026     810152        0  2.231630e-15 -4.74142    1  
ENSG00000101282    1025679        0  3.021520e-15 -5.27573    1  
ENSG00000186777     897103        0  4.407960e-15 -4.66579    1  
ENSG00000116014     619338        0  1.081230e-14 -5.08811    1  
</pre>

<p>
<b>TODO:</b> Permutation testing within QTLtools? lfsr estimation by <code>ashr</code>?
</p>
</div>
</div>

<div id="outline-container-orgad41e3e" class="outline-2">
<h2 id="orgad41e3e">Dispersion-QTL calling</h2>
<div class="outline-text-2" id="text-orgad41e3e">
<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">disp</span> = pd.read_table(<span class="org-string">'/scratch/midway2/aksarkar/singlecell/dispersion.txt.gz'</span>, sep=<span class="org-string">' '</span>, index_col=0)
disp.head()
</pre>
</div>

<pre class="example">
                NA18489   NA18498   NA18499   NA18501   NA18502   NA18505  \
gene                                                                          
ENSG00000000003  1.092216  1.349166  1.376199  1.151147  1.597304  1.619764   
ENSG00000000005  1.245627  7.247635 -0.062590  1.342859 -1.007912  0.411424   
ENSG00000000419  1.913491  2.318849  1.598213  1.189439  1.753222  2.410088   
ENSG00000000457 -0.699626  5.044208  0.099977 -0.107983 -0.469346  4.960834   
ENSG00000000460  1.130231  6.989872  1.317161  1.452350  0.750685  7.235482   

                  NA18507   NA18508   NA18519   NA18520    ...      NA19119  \
gene                                                       ...                
ENSG00000000003  1.129030  1.115463  1.833903  1.555660    ...     1.315700   
ENSG00000000005 -0.469625  4.921400 -0.688780  5.141915    ...     6.322920   
ENSG00000000419  1.036118  0.865517  1.408871  1.298325    ...     3.754308   
ENSG00000000457 -0.108463  0.456174 -0.609740 -0.225618    ...     0.740946   
ENSG00000000460  0.996670  0.676970  2.453142  1.251912    ...     0.717296   

                  NA19128   NA19153   NA19159   NA19190   NA19193   NA19203  \
gene                                                                          
ENSG00000000003  0.693213  1.513592  1.016362  1.539210  1.417464  1.459843   
ENSG00000000005 -0.657682 -0.084881  4.429851  0.978749  5.818314  0.995584   
ENSG00000000419  0.771990  1.453221  1.603899  1.464751  1.067739  4.966578   
ENSG00000000457 -0.268748  0.060874 -0.039906  0.009648  0.003913  6.323813   
ENSG00000000460  0.866060  0.424351  7.455638  0.493674  1.032298  0.203354   

                  NA19207   NA19210   NA19257  
gene                                           
ENSG00000000003  1.296685  1.494707  1.368400  
ENSG00000000005  0.493389 -1.198948  5.263945  
ENSG00000000419  1.536127  1.775915  1.279451  
ENSG00000000457 -0.081436  0.800806 -0.193967  
ENSG00000000460  1.145072  1.669772  1.276706  

[5 rows x 32 columns]
</pre>

<div class="org-src-container">
<pre class="src src-ipython">write_pheno_file(disp, gene_info, <span class="org-string">'/scratch/midway2/aksarkar/singlecell/disp.bed'</span>)
</pre>
</div>

<p>
Index the phenotype file
</p>

<pre class="example">
Submitted batch job 42125661

</pre>

<p>
Run <code>qtltools</code>
</p>

<pre class="example">
Submitted batch job 42125670

</pre>

<p>
There are no detected dispersion QTLs, so the following code raises an exception.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">file_names</span> = [<span class="org-string">'disp-qtl.{}.txt'</span>.<span class="org-builtin">format</span>(i) <span class="org-keyword">for</span> i <span class="org-keyword">in</span> <span class="org-builtin">range</span>(1, 101)]
<span class="org-variable-name">disp_qtls</span> = (pd.concat([pd.read_table(f, header=<span class="org-constant">None</span>, sep=<span class="org-string">' '</span>) <span class="org-keyword">for</span> f <span class="org-keyword">in</span> file_names <span class="org-keyword">if</span> os.path.exists(f) <span class="org-keyword">and</span> os.path.getsize(f) &gt; 0])
             .rename(columns={i: x <span class="org-keyword">for</span> i, x <span class="org-keyword">in</span> <span class="org-builtin">enumerate</span>([<span class="org-string">'gene'</span>, <span class="org-string">'chr'</span>, <span class="org-string">'start'</span>, <span class="org-string">'end'</span>, <span class="org-string">'strand'</span>, <span class="org-string">'num_vars'</span>, <span class="org-string">'distance'</span>, <span class="org-string">'id'</span>, <span class="org-string">'var_chr'</span>, <span class="org-string">'var_start'</span>, <span class="org-string">'var_end'</span>, <span class="org-string">'p'</span>, <span class="org-string">'beta'</span>, <span class="org-string">'top'</span>])})
             .sort_values(<span class="org-string">'p'</span>)
             .set_index(<span class="org-string">'gene'</span>))
disp_qtls[disp_qtls[<span class="org-string">'top'</span>] == 1].head()
</pre>
</div>
</div>
</div>
</div>
<div id="postamble" class="status">
<p class="author">Author: Abhishek Sarkar</p>
<p class="date">Created: 2018-01-14 Sun 17:15</p>
<p class="validation"><a href="http://validator.w3.org/check?uri=referer">Validate</a></p>
</div>
</body>
</html>
