#+TITLE: Hurdle model estimation
#+SETUPFILE: setup.org

* Introduction

  The key idea of ~voom~ ([[http://genomebiology.com/2014/15/2/R29][Law et al 2014]]) is that the distribution of \(Y =
  \log_2(\mathrm{CPM} + 1)\) is approximately Gaussian. For simplicity,
  consider one gene in one cell.

  \[ R = \mbox{number of reads} \]

  Assuming \(R\) follows the NB2 negative binomial model (Hilbe 2012):

  \[ E[R] = \lambda \]
  \[ V[R] = \lambda + \phi \lambda^2 \]

  By the definition of CPM:

  \[ Y = \log_2(R) + \mathrm{const} \]

  By first order Taylor expansion:

  \[ E[Y] = \mu \approx \log_2 \lambda + \mathrm{const} \]

  By the delta method:

  \[ V[Y] = \sigma^2 \approx V[R] / \lambda^2 = 1/\lambda + \phi \]

  The key idea of ~mast~ ([[https://dx.doi.org/10.1186/s13059-015-0844-5][Finak et al 2015]]) is to model non-zero \(R\) using a
  Gaussian distribution, and model zero \(R\) using logistic regression. The
  key distinction between this approach and the [[file:zinb.org][zero-inflated
  negative binomial model]] is that zeroes are assumed to arise from only one
  process.

  This naturally suggests a Bayesian model which simultaneously calls mean and
  variance QTLs based on the likelihood:

  \[ Y \mid Y = 0 \sim \mathrm{Bernoulli}(1 - \mathrm{sigmoid}(AX\theta_m + \delta)) \]

  \[ Y \mid Y > 0 \sim N(AX\theta_m + \delta, \exp(-AX\theta_m - \delta) +
  \exp(AX\theta_v) + \sigma^2) \]

  where \(\delta\) is a known constant that depends on library size, \(A\) maps
  from cells to individuals, and \(X\) is the genotype matrix.

  The key issues are:

  1. *Is it possible to develop an individual-level model?* An alternative
     model would be to estimate a variance for each individual, and then plug
     into existing QTL mapping software. This approach is naturally suggested
     by the ~voom~ approximation, but it is not obvious that the approach will
     generalize to zero-inflated data.

     Fitting the proposed model will require writing new software as opposed to
     plugging into existing software. We previously built an efficient
     algorithm and inference engine to build these kind of models ([[https://www.biorxiv.org/content/early/2017/02/10/107623][Park et al
     2017]]), but it has one major issue: it is computationally expensive to
     perform statistical tests on the fitted models because the approximate
     Bayesian inferences underestimate the uncertainty in the estimated values.

  2. *Do we need to actually fit the dropout model?* ~mast~ conditions on the
     observed \(Y = 0\), not on a latent \(Z = 0\), which means we could simply
     do the same and ignore zeros.

     If we did so, then mean/variance QTL effect size estimation will be less
     robust for genes with high dropout. This might not be a problem depending
     on the stringency of gene filtering.

  3. *For a single gene, do we need to worry about mean QTLs in LD with
     variance QTLs?* We previously [[https://github.com/YPARK/fqtl][built multivariate mean/variance QTL models]]
     which could account for LD, and could share information between the mean
     and variance models.

     The fundamental problem is that if we assume that the mean and dispersion
     both have genetic components, then the mean is no longer independent of
     the dispersion.

     This actually could be derived without using the fact that both depend on
     the same genotypes if we use second-order Taylor expansion:

     \[ \mu \approx \log_2 \lambda + \frac{V[R]}{2 \lambda^2} \]

  4. *Do we need to share parameters between genes?* ~mast~ assumes genes are
     conditionally independent. But this is no longer true when nearby genes
     can be driven by overlapping (or correlated) /cis/-genotypes.

     We previously developed multiresponse QTL models which learned the target
     genes of causal variants, allowing the true target gene to explain away
     nearby correlated genes ([[https://www.biorxiv.org/content/early/2017/11/14/219428][Park et al 2017]]).

* Setup                                                            :noexport:

  #+BEGIN_SRC emacs-lisp
    (org-babel-lob-ingest "/home/aksarkar/projects/singlecell-qtl/analysis/sc-vs-bulk.org")
    (org-babel-lob-ingest "/home/aksarkar/projects/singlecell-qtl/analysis/qtl-mapping.org")
    (org-babel-lob-ingest "/home/aksarkar/.emacs.d/org-templates/library.org")
  #+END_SRC

  #+RESULTS:
  : 1

  #+CALL: ipython3(memory="16G", venv="scqtl", partition="gpu2") :dir /scratch/midway2/aksarkar/singlecell

  #+RESULTS:
  : Submitted batch job 41289837

  #+BEGIN_SRC ipython
    import functools
    import gzip
    import os.path
    import numpy as np
    import pandas as pd
    import pickle
    import tabix
  #+END_SRC

  #+RESULTS:
  :RESULTS:
  :END:

  #+CALL: concordance-def()

  #+RESULTS:
  :RESULTS:
  :END:

  #+CALL: load-data-defs()

  #+RESULTS:
  :RESULTS:
  :END:

  #+CALL: load-data()

  #+RESULTS:
  :RESULTS:
  : ((20327, 2261), (34608, 15))
  :END:

  #+CALL: umi-qc()

  #+RESULTS:
  :RESULTS:
  : (15636, 1810)
  :END:

  #+CALL: onehot-qc()

  #+RESULTS:
  :RESULTS:
  :END:

* Test case

  #+CALL: get-gene-info() :exports both

  #+RESULTS:
  :RESULTS:
  #+BEGIN_EXAMPLE
                    chr      start        end      name strand      source
    gene                                                                    
    ENSG00000000419  hs20   49551404   49575092      DPM1      -  H. sapiens
    ENSG00000000457   hs1  169818772  169863408     SCYL3      -  H. sapiens
    ENSG00000000460   hs1  169631245  169823221  C1orf112      +  H. sapiens
    ENSG00000000938   hs1   27938575   27961788       FGR      -  H. sapiens
    ENSG00000000971   hs1  196621008  196716634       CFH      +  H. sapiens
  #+END_EXAMPLE
  :END:

  #+BEGIN_SRC ipython
    gene_info.query('name == "ZSWIM7"')
  #+END_SRC

  #+RESULTS:
  :RESULTS:
  #+BEGIN_EXAMPLE
                    chr     start       end    name strand      source
    gene                                                                
    ENSG00000214941  hs17  15879874  15903031  ZSWIM7      -  H. sapiens
  #+END_EXAMPLE
  :END:

  #+BEGIN_SRC ipython
    with gzip.open('/project2/gilad/singlecell-qtl/bulk/genotypes.vcf.gz', 'rt') as f:
      header = next(f).split()
    vcf = tabix.open('/project2/gilad/singlecell-qtl/bulk/genotypes.vcf.gz')
  #+END_SRC

  #+RESULTS:
  :RESULTS:
  :END:

  #+BEGIN_SRC ipython
    def _extract_genotypes(row, vcf, header, window):
      if row['strand'] == '+':
        start = row['start'] - window
        end = row['start']
      else:
        start = row['end']
        end = row['end'] + window
      result = pd.DataFrame(list(vcf.query('chr{}'.format(row['chr'][2:]), start, end)), columns=header).iloc[:,9:].astype(np.float32)
      return result
  #+END_SRC

  #+RESULTS:
  :RESULTS:
  :END:

  #+BEGIN_SRC ipython
    genotypes = _extract_genotypes(gene_info.loc['ENSG00000214941'], vcf=vcf, header=header, window=int(1e6)).rename(columns=lambda x: 'NA{}'.format(x))[sorted(annotations_qc['chip_id'].unique())]
    genotypes -= genotypes.agg(np.mean, axis=1)
    genotypes.shape
  #+END_SRC

  #+RESULTS:
  :RESULTS:
  : (2506, 21)
  :END:

  #+BEGIN_SRC ipython
    with open('test_data.pkl', 'wb') as f:
      pickle.dump({'onehot': onehot,
                   'genotypes': genotypes.values.T,
                   'counts': umi_qc.loc['ENSG00000214941'].values.astype(np.float32).reshape(-1, 1),
                   'normalizers': -np.log(umi_qc.agg(np.sum)).values.reshape(-1, 1)}, f)
  #+END_SRC

  #+RESULTS:
  :RESULTS:
  :END:
