---
title: "Identifying single cells"
author: "John Blischak"
date: 2017-08-23
output: html_document
---

<!-- The file analysis/chunks.R contains chunks that define default settings
shared across the workflowr files. -->
```{r read-chunk, include=FALSE, cache=FALSE}
knitr::read_chunk("chunks.R")
```

<!-- Update knitr chunk options -->
```{r knitr-opts-chunk, include=FALSE}
```

<!-- Insert the date the file was last updated -->
```{r last-updated, echo=FALSE, results='asis'}
```

<!-- Insert the code version (Git commit SHA1) if Git repository exists and R
 package git2r is installed -->
```{r code-version, echo=FALSE, results='asis'}
```

[Link to Yoruba samples on Coriell][coriell]

[coriell]: https://catalog.coriell.org/0/Sections/Collections/NHGRI/1000IndSamples.aspx?PgId=665&did=MGENOME27

## Setup

```{r packages, message=FALSE}
library("cowplot")
library("dplyr")
library("ggplot2")
library("stringr")
library("tidyr")
theme_set(theme_cowplot())
library("Biobase")
```

Import data.

```{r import}
fname <- Sys.glob("../data/eset/*.rds")
eset <- Reduce(combine, Map(readRDS, fname))
eset
```

## How many of the identities are plausible?

```{r plausible}
table(eset$valid_id)
mean(eset$valid_id)
```

## Plot per C1 chip

```{r ind-per-chip}
anno <- pData(eset)
df_e <- anno %>%
  group_by(experiment) %>%
  summarize(i1 = unique(individual.1),
            i2 = unique(individual.2),
            i3 = unique(individual.3),
            i4 = unique(individual.4))
op <- par(cex = 2, las = 2, mfrow = c(2, 5))
for (i in 1:nrow(df_e)) {
  e <- df_e$experiment[i]
  n1 <- sum(anno$chip_id[anno$experiment == e] == df_e$i1[i])
  n2 <- sum(anno$chip_id[anno$experiment == e] == df_e$i2[i])
  n3 <- sum(anno$chip_id[anno$experiment == e] == df_e$i3[i])
  n4 <- sum(anno$chip_id[anno$experiment == e] == df_e$i4[i])
  n_other <- 96 - n1 - n2 -n3 - n4
  barplot(c(n1, n2, n3, n4, n_other), main = sprintf("C1 chip %s", e),
          names.arg = c(df_e$i1[i], df_e$i2[i], df_e$i3[i],df_e$i4[i], "Other"),
          ylab = "Number of single cells")
}
par(op)
```

## Expected versus observed cells

Each individual is expected to get 25% of the wells (i.e. 24) for each C1 chip
it is on.

```{r obs-v-exp}
expected <- df_e %>% select(starts_with("i")) %>% unlist %>% table * 96 / 4
expected <- as.data.frame(expected)
colnames(expected) <- c("ind", "count")
observed <- table(anno$chip_id)
observed <- as.data.frame(observed)
colnames(observed) <- c("ind", "count")
df_obs <- merge(expected, observed, by = "ind", suffixes = c(".exp", ".obs"),
                all.x = TRUE, sort = TRUE)
# Some individuals were assigned zero cells
df_obs$count.obs[is.na(df_obs$count.obs)] <- 0
df_obs
# The max observed number of cells for any indvidual
max_cells <- max(df_obs$count.exp, df_obs$count.obs)
ggplot(df_obs, aes(x = count.exp, y = count.obs)) +
  geom_text(aes(label = ind)) +
  geom_abline(intercept = 0, slope = 1, color = "red", linetype = "dashed") +
  xlim(0, max_cells) + ylim(0, max_cells) +
  labs(x = "Expected number of cells per individual",
       y = "Observed number of cells per individual",
       title = "Observed versus expected number of cells")
```

## Unexpectedly high counts

```{r unexpectedly-high-counts}
observed %>% filter(count > 10, !(ind %in% expected$ind)) %>%
  arrange(desc(count))
```

## Potential reasons for failure

Sequencing depth

```{r reason-raw}
reason_raw <- ggplot(anno, aes(x = valid_id, y = raw)) +
  geom_boxplot() +
  scale_x_discrete(labels = c("Wrong", "Correct")) +
  labs(x = "Predicted individual of origin",
       y = "Raw number of reads",
       title = "Raw number of reads")
reason_raw
```

Percent mapped

```{r reason-mapped}
reason_raw %+% aes(y = mapped / raw * 100) +
  labs(y = "Percent of reads mapped",
       title = "Percent of reads mapped")
```

Cell number

```{r reason-cell-number}
barplot(with(anno, table(valid_id, cell_number)), beside = TRUE,
        col = c("red", "black"),
        xlab = "Number of cells in well", main = "Number of cells in well",
        ylab = "Number of cells correct (black) versus wrong (red)")
```

Concentration

```{r reason-concentation}
reason_raw %+% aes(y = concentration) +
  labs(y = "Concentration",
       title = "Concentration")
```

TRA-1-60

```{r reason-tra-1-60}
barplot(with(anno, table(valid_id, tra1.60)), beside = TRUE,
        col = c("red", "black"),
        xlab = "TRA-1-60", main = "TRA-1-60",
        ylab = "Number of cells correct (black) versus wrong (red)")
```

SNPs with at least one overlapping read

```{r reason-snps-w-min}
reason_raw %+% aes(y = snps_w_min) +
  labs(y = "SNPs with at least one overlapping read",
       title = "SNPs with at least one overlapping read")
```

chipmix

```{r reason-chipmix}
reason_raw %+% aes(y = chipmix) +
  labs(y = "chipmix", title = "chipmix")
```

freemix

```{r reason-freemix}
reason_raw %+% aes(y = freemix) +
  labs(y = "freemix", title = "freemix")
```

chipmix versus freemix

```{r chipmix-v-freemix}
ggplot(anno, aes(x = freemix, y = chipmix, color = valid_id)) +
  geom_point() +
  facet_wrap(~valid_id) +
  labs(title = "chipmix versus freemix")
```

chipmix versus SNPs with overlapping reads

```{r chipmix-v-snps-w-min}
ggplot(anno, aes(x = snps_w_min, y = chipmix, color = valid_id)) +
  geom_point() +
  facet_wrap(~valid_id) +
  labs(title = "chipmix versus SNPs with at least one overlapping read",
       x = "SNPs with at least one overlapping read")
```

chipmix versus raw number of reads

```{r chipmix-v-raw}
ggplot(anno, aes(x = raw, y = chipmix, color = valid_id)) +
  geom_point() +
  facet_wrap(~valid_id) +
  labs(title = "chipmix versus number of raw reads",
       x = "Number of raw reads")
```

## Session information

<!-- Insert the session information into the document -->
```{r session-info}
```
