#+TITLE: QTL mapping pipeline
#+SETUPFILE: setup.org

* Introduction

  Here, we set up the infrastructure to call mean and variance QTLs. 

  The key issues are:

  1. *How do we account for confounders in UMI?* As a first pass (to test the
     pipeline), use preliminary QC filters and assume no confounding.
  2. *Do we account for zeros?* As a first pass, we can estimate means and
     variances per individual using only detected genes.
  3. *Do we jointly model mean and variance QTLs?* As a first pass, we can
     simply include the observed mean in the variance QTL model (like ~voom~,
     [[http://genomebiology.com/2014/15/2/R29][Law et al 2014]]).
  4. *What parameters are shared between cells/individuals/genes?* As a first
     pass, assume no parameters are shared.

* Setup                                                            :noexport:

  #+BEGIN_SRC emacs-lisp
    (org-babel-lob-ingest "/home/aksarkar/projects/singlecell-qtl/analysis/sc-vs-bulk.org")
    (org-babel-lob-ingest "/home/aksarkar/.emacs.d/org-templates/library.org")
  #+END_SRC

  #+RESULTS:
  : 1

  #+CALL: ipython3(memory="16G", venv="scqtl") :dir /scratch/midway2/aksarkar/singlecell

  #+RESULTS:
  : Submitted batch job 41610560

  #+BEGIN_SRC ipython
    import functools
    import os.path
    import numpy as np
    import pandas as pd
  #+END_SRC

  #+RESULTS:
  :RESULTS:
  :END:

  #+CALL: concordance-def()

  #+RESULTS:
  :RESULTS:
  :END:

  #+CALL: load-data-defs()

  #+RESULTS:
  :RESULTS:
  :END:

  #+CALL: load-data()

  #+RESULTS:
  :RESULTS:
  : ((20327, 2261), (34608, 15))
  :END:

* Next steps
   
  1. We previously used half-sibling regression ([[https://www.pnas.org/cgi/doi/10.1073/pnas.1511656113][Scholkopf et al 2016]]) to correct
     confounders for bulk RNA-Seq in GTEx ([[http://dx.doi.org/10.1101/107623][Park et al 2017]]).

     The key idea is that the only ways we could predict the expression of a
     gene of interest using genes on other chromosomes are (1) true
     /trans/-regulation and (2) shared confounding. For /cis/-QTL mapping, we
     want to eliminate both of these, which we can do by simply regressing out
     gene expression of all genes on all other chromosomes.

     One open question is whether we should correct counts (which would be
     required to solve issue (3)), or correct per-individual means/variances.

  2. We previously built an [[https://github.com/YPARK/fqtl][inference engine which supports negative binomial
     models]] which could be easily extended to zero-inflated negative binomial
     models.

     In our experiments directly using the negative binomial likelihood
     outperformed the Gaussian model (following the ~voom~-transform, but
     sharing the mean model parameters in the dispersion model).

     This would allow us to solve issues (2) and (3), and potentially make (4)
     easier.
